// @cdktf/provider-google/lib/identity-platform-project-default-config/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentityPlatformProjectDefaultConfig = exports.IdentityPlatformProjectDefaultConfigTimeoutsOutputReference = exports.identityPlatformProjectDefaultConfigTimeoutsToTerraform = exports.IdentityPlatformProjectDefaultConfigSignInOutputReference = exports.identityPlatformProjectDefaultConfigSignInToTerraform = exports.IdentityPlatformProjectDefaultConfigSignInPhoneNumberOutputReference = exports.identityPlatformProjectDefaultConfigSignInPhoneNumberToTerraform = exports.IdentityPlatformProjectDefaultConfigSignInEmailOutputReference = exports.identityPlatformProjectDefaultConfigSignInEmailToTerraform = exports.IdentityPlatformProjectDefaultConfigSignInAnonymousOutputReference = exports.identityPlatformProjectDefaultConfigSignInAnonymousToTerraform = exports.IdentityPlatformProjectDefaultConfigSignInHashConfigList = exports.IdentityPlatformProjectDefaultConfigSignInHashConfigOutputReference = exports.identityPlatformProjectDefaultConfigSignInHashConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function identityPlatformProjectDefaultConfigSignInHashConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.identityPlatformProjectDefaultConfigSignInHashConfigToTerraform = identityPlatformProjectDefaultConfigSignInHashConfigToTerraform;
class IdentityPlatformProjectDefaultConfigSignInHashConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // algorithm - computed: true, optional: false, required: false
    get algorithm() {
        return this.getStringAttribute('algorithm');
    }
    // memory_cost - computed: true, optional: false, required: false
    get memoryCost() {
        return this.getNumberAttribute('memory_cost');
    }
    // rounds - computed: true, optional: false, required: false
    get rounds() {
        return this.getNumberAttribute('rounds');
    }
    // salt_separator - computed: true, optional: false, required: false
    get saltSeparator() {
        return this.getStringAttribute('salt_separator');
    }
    // signer_key - computed: true, optional: false, required: false
    get signerKey() {
        return this.getStringAttribute('signer_key');
    }
}
exports.IdentityPlatformProjectDefaultConfigSignInHashConfigOutputReference = IdentityPlatformProjectDefaultConfigSignInHashConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IdentityPlatformProjectDefaultConfigSignInHashConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.identityPlatformProjectDefaultConfig.IdentityPlatformProjectDefaultConfigSignInHashConfigOutputReference", version: "5.0.1" };
class IdentityPlatformProjectDefaultConfigSignInHashConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IdentityPlatformProjectDefaultConfigSignInHashConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IdentityPlatformProjectDefaultConfigSignInHashConfigList = IdentityPlatformProjectDefaultConfigSignInHashConfigList;
_b = JSII_RTTI_SYMBOL_1;
IdentityPlatformProjectDefaultConfigSignInHashConfigList[_b] = { fqn: "@cdktf/provider-google.identityPlatformProjectDefaultConfig.IdentityPlatformProjectDefaultConfigSignInHashConfigList", version: "5.0.1" };
function identityPlatformProjectDefaultConfigSignInAnonymousToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.identityPlatformProjectDefaultConfigSignInAnonymousToTerraform = identityPlatformProjectDefaultConfigSignInAnonymousToTerraform;
class IdentityPlatformProjectDefaultConfigSignInAnonymousOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.IdentityPlatformProjectDefaultConfigSignInAnonymousOutputReference = IdentityPlatformProjectDefaultConfigSignInAnonymousOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IdentityPlatformProjectDefaultConfigSignInAnonymousOutputReference[_c] = { fqn: "@cdktf/provider-google.identityPlatformProjectDefaultConfig.IdentityPlatformProjectDefaultConfigSignInAnonymousOutputReference", version: "5.0.1" };
function identityPlatformProjectDefaultConfigSignInEmailToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        password_required: cdktf.booleanToTerraform(struct.passwordRequired),
    };
}
exports.identityPlatformProjectDefaultConfigSignInEmailToTerraform = identityPlatformProjectDefaultConfigSignInEmailToTerraform;
class IdentityPlatformProjectDefaultConfigSignInEmailOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._passwordRequired !== undefined) {
            hasAnyValues = true;
            internalValueResult.passwordRequired = this._passwordRequired;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._passwordRequired = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._passwordRequired = value.passwordRequired;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get passwordRequired() {
        return this.getBooleanAttribute('password_required');
    }
    set passwordRequired(value) {
        this._passwordRequired = value;
    }
    resetPasswordRequired() {
        this._passwordRequired = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passwordRequiredInput() {
        return this._passwordRequired;
    }
}
exports.IdentityPlatformProjectDefaultConfigSignInEmailOutputReference = IdentityPlatformProjectDefaultConfigSignInEmailOutputReference;
_d = JSII_RTTI_SYMBOL_1;
IdentityPlatformProjectDefaultConfigSignInEmailOutputReference[_d] = { fqn: "@cdktf/provider-google.identityPlatformProjectDefaultConfig.IdentityPlatformProjectDefaultConfigSignInEmailOutputReference", version: "5.0.1" };
function identityPlatformProjectDefaultConfigSignInPhoneNumberToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        test_phone_numbers: cdktf.hashMapper(cdktf.stringToTerraform)(struct.testPhoneNumbers),
    };
}
exports.identityPlatformProjectDefaultConfigSignInPhoneNumberToTerraform = identityPlatformProjectDefaultConfigSignInPhoneNumberToTerraform;
class IdentityPlatformProjectDefaultConfigSignInPhoneNumberOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._testPhoneNumbers !== undefined) {
            hasAnyValues = true;
            internalValueResult.testPhoneNumbers = this._testPhoneNumbers;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._testPhoneNumbers = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._testPhoneNumbers = value.testPhoneNumbers;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get testPhoneNumbers() {
        return this.getStringMapAttribute('test_phone_numbers');
    }
    set testPhoneNumbers(value) {
        this._testPhoneNumbers = value;
    }
    resetTestPhoneNumbers() {
        this._testPhoneNumbers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get testPhoneNumbersInput() {
        return this._testPhoneNumbers;
    }
}
exports.IdentityPlatformProjectDefaultConfigSignInPhoneNumberOutputReference = IdentityPlatformProjectDefaultConfigSignInPhoneNumberOutputReference;
_e = JSII_RTTI_SYMBOL_1;
IdentityPlatformProjectDefaultConfigSignInPhoneNumberOutputReference[_e] = { fqn: "@cdktf/provider-google.identityPlatformProjectDefaultConfig.IdentityPlatformProjectDefaultConfigSignInPhoneNumberOutputReference", version: "5.0.1" };
function identityPlatformProjectDefaultConfigSignInToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allow_duplicate_emails: cdktf.booleanToTerraform(struct.allowDuplicateEmails),
        anonymous: identityPlatformProjectDefaultConfigSignInAnonymousToTerraform(struct.anonymous),
        email: identityPlatformProjectDefaultConfigSignInEmailToTerraform(struct.email),
        phone_number: identityPlatformProjectDefaultConfigSignInPhoneNumberToTerraform(struct.phoneNumber),
    };
}
exports.identityPlatformProjectDefaultConfigSignInToTerraform = identityPlatformProjectDefaultConfigSignInToTerraform;
class IdentityPlatformProjectDefaultConfigSignInOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // hash_config - computed: true, optional: false, required: false
        this._hashConfig = new IdentityPlatformProjectDefaultConfigSignInHashConfigList(this, "hash_config", false);
        // anonymous - computed: false, optional: true, required: false
        this._anonymous = new IdentityPlatformProjectDefaultConfigSignInAnonymousOutputReference(this, "anonymous");
        // email - computed: false, optional: true, required: false
        this._email = new IdentityPlatformProjectDefaultConfigSignInEmailOutputReference(this, "email");
        // phone_number - computed: false, optional: true, required: false
        this._phoneNumber = new IdentityPlatformProjectDefaultConfigSignInPhoneNumberOutputReference(this, "phone_number");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allowDuplicateEmails !== undefined) {
            hasAnyValues = true;
            internalValueResult.allowDuplicateEmails = this._allowDuplicateEmails;
        }
        if (this._anonymous?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.anonymous = this._anonymous?.internalValue;
        }
        if (this._email?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.email = this._email?.internalValue;
        }
        if (this._phoneNumber?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.phoneNumber = this._phoneNumber?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allowDuplicateEmails = undefined;
            this._anonymous.internalValue = undefined;
            this._email.internalValue = undefined;
            this._phoneNumber.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allowDuplicateEmails = value.allowDuplicateEmails;
            this._anonymous.internalValue = value.anonymous;
            this._email.internalValue = value.email;
            this._phoneNumber.internalValue = value.phoneNumber;
        }
    }
    get allowDuplicateEmails() {
        return this.getBooleanAttribute('allow_duplicate_emails');
    }
    set allowDuplicateEmails(value) {
        this._allowDuplicateEmails = value;
    }
    resetAllowDuplicateEmails() {
        this._allowDuplicateEmails = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowDuplicateEmailsInput() {
        return this._allowDuplicateEmails;
    }
    get hashConfig() {
        return this._hashConfig;
    }
    get anonymous() {
        return this._anonymous;
    }
    putAnonymous(value) {
        this._anonymous.internalValue = value;
    }
    resetAnonymous() {
        this._anonymous.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get anonymousInput() {
        return this._anonymous.internalValue;
    }
    get email() {
        return this._email;
    }
    putEmail(value) {
        this._email.internalValue = value;
    }
    resetEmail() {
        this._email.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email.internalValue;
    }
    get phoneNumber() {
        return this._phoneNumber;
    }
    putPhoneNumber(value) {
        this._phoneNumber.internalValue = value;
    }
    resetPhoneNumber() {
        this._phoneNumber.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumberInput() {
        return this._phoneNumber.internalValue;
    }
}
exports.IdentityPlatformProjectDefaultConfigSignInOutputReference = IdentityPlatformProjectDefaultConfigSignInOutputReference;
_f = JSII_RTTI_SYMBOL_1;
IdentityPlatformProjectDefaultConfigSignInOutputReference[_f] = { fqn: "@cdktf/provider-google.identityPlatformProjectDefaultConfig.IdentityPlatformProjectDefaultConfigSignInOutputReference", version: "5.0.1" };
function identityPlatformProjectDefaultConfigTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.identityPlatformProjectDefaultConfigTimeoutsToTerraform = identityPlatformProjectDefaultConfigTimeoutsToTerraform;
class IdentityPlatformProjectDefaultConfigTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.IdentityPlatformProjectDefaultConfigTimeoutsOutputReference = IdentityPlatformProjectDefaultConfigTimeoutsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
IdentityPlatformProjectDefaultConfigTimeoutsOutputReference[_g] = { fqn: "@cdktf/provider-google.identityPlatformProjectDefaultConfig.IdentityPlatformProjectDefaultConfigTimeoutsOutputReference", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/identity_platform_project_default_config google_identity_platform_project_default_config}
*/
class IdentityPlatformProjectDefaultConfig extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/identity_platform_project_default_config google_identity_platform_project_default_config} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IdentityPlatformProjectDefaultConfigConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_identity_platform_project_default_config',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // sign_in - computed: false, optional: true, required: false
        this._signIn = new IdentityPlatformProjectDefaultConfigSignInOutputReference(this, "sign_in");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new IdentityPlatformProjectDefaultConfigTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._project = config.project;
        this._signIn.internalValue = config.signIn;
        this._timeouts.internalValue = config.timeouts;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get signIn() {
        return this._signIn;
    }
    putSignIn(value) {
        this._signIn.internalValue = value;
    }
    resetSignIn() {
        this._signIn.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get signInInput() {
        return this._signIn.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
            sign_in: identityPlatformProjectDefaultConfigSignInToTerraform(this._signIn.internalValue),
            timeouts: identityPlatformProjectDefaultConfigTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.IdentityPlatformProjectDefaultConfig = IdentityPlatformProjectDefaultConfig;
_h = JSII_RTTI_SYMBOL_1;
IdentityPlatformProjectDefaultConfig[_h] = { fqn: "@cdktf/provider-google.identityPlatformProjectDefaultConfig.IdentityPlatformProjectDefaultConfig", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
IdentityPlatformProjectDefaultConfig.tfResourceType = "google_identity_platform_project_default_config";
