// @cdktf/provider-google/lib/compute-autoscaler/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComputeAutoscaler = exports.ComputeAutoscalerTimeoutsOutputReference = exports.computeAutoscalerTimeoutsToTerraform = exports.ComputeAutoscalerAutoscalingPolicyOutputReference = exports.computeAutoscalerAutoscalingPolicyToTerraform = exports.ComputeAutoscalerAutoscalingPolicyScalingSchedulesList = exports.ComputeAutoscalerAutoscalingPolicyScalingSchedulesOutputReference = exports.computeAutoscalerAutoscalingPolicyScalingSchedulesToTerraform = exports.ComputeAutoscalerAutoscalingPolicyScaleInControlOutputReference = exports.computeAutoscalerAutoscalingPolicyScaleInControlToTerraform = exports.ComputeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasOutputReference = exports.computeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasToTerraform = exports.ComputeAutoscalerAutoscalingPolicyMetricList = exports.ComputeAutoscalerAutoscalingPolicyMetricOutputReference = exports.computeAutoscalerAutoscalingPolicyMetricToTerraform = exports.ComputeAutoscalerAutoscalingPolicyLoadBalancingUtilizationOutputReference = exports.computeAutoscalerAutoscalingPolicyLoadBalancingUtilizationToTerraform = exports.ComputeAutoscalerAutoscalingPolicyCpuUtilizationOutputReference = exports.computeAutoscalerAutoscalingPolicyCpuUtilizationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function computeAutoscalerAutoscalingPolicyCpuUtilizationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        predictive_method: cdktf.stringToTerraform(struct.predictiveMethod),
        target: cdktf.numberToTerraform(struct.target),
    };
}
exports.computeAutoscalerAutoscalingPolicyCpuUtilizationToTerraform = computeAutoscalerAutoscalingPolicyCpuUtilizationToTerraform;
class ComputeAutoscalerAutoscalingPolicyCpuUtilizationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._predictiveMethod !== undefined) {
            hasAnyValues = true;
            internalValueResult.predictiveMethod = this._predictiveMethod;
        }
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._predictiveMethod = undefined;
            this._target = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._predictiveMethod = value.predictiveMethod;
            this._target = value.target;
        }
    }
    get predictiveMethod() {
        return this.getStringAttribute('predictive_method');
    }
    set predictiveMethod(value) {
        this._predictiveMethod = value;
    }
    resetPredictiveMethod() {
        this._predictiveMethod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get predictiveMethodInput() {
        return this._predictiveMethod;
    }
    get target() {
        return this.getNumberAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
}
exports.ComputeAutoscalerAutoscalingPolicyCpuUtilizationOutputReference = ComputeAutoscalerAutoscalingPolicyCpuUtilizationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerAutoscalingPolicyCpuUtilizationOutputReference[_a] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerAutoscalingPolicyCpuUtilizationOutputReference", version: "5.0.1" };
function computeAutoscalerAutoscalingPolicyLoadBalancingUtilizationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target: cdktf.numberToTerraform(struct.target),
    };
}
exports.computeAutoscalerAutoscalingPolicyLoadBalancingUtilizationToTerraform = computeAutoscalerAutoscalingPolicyLoadBalancingUtilizationToTerraform;
class ComputeAutoscalerAutoscalingPolicyLoadBalancingUtilizationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._target = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._target = value.target;
        }
    }
    get target() {
        return this.getNumberAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
}
exports.ComputeAutoscalerAutoscalingPolicyLoadBalancingUtilizationOutputReference = ComputeAutoscalerAutoscalingPolicyLoadBalancingUtilizationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerAutoscalingPolicyLoadBalancingUtilizationOutputReference[_b] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerAutoscalingPolicyLoadBalancingUtilizationOutputReference", version: "5.0.1" };
function computeAutoscalerAutoscalingPolicyMetricToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        target: cdktf.numberToTerraform(struct.target),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.computeAutoscalerAutoscalingPolicyMetricToTerraform = computeAutoscalerAutoscalingPolicyMetricToTerraform;
class ComputeAutoscalerAutoscalingPolicyMetricOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._target = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._target = value.target;
            this._type = value.type;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get target() {
        return this.getNumberAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ComputeAutoscalerAutoscalingPolicyMetricOutputReference = ComputeAutoscalerAutoscalingPolicyMetricOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerAutoscalingPolicyMetricOutputReference[_c] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerAutoscalingPolicyMetricOutputReference", version: "5.0.1" };
class ComputeAutoscalerAutoscalingPolicyMetricList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeAutoscalerAutoscalingPolicyMetricOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeAutoscalerAutoscalingPolicyMetricList = ComputeAutoscalerAutoscalingPolicyMetricList;
_d = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerAutoscalingPolicyMetricList[_d] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerAutoscalingPolicyMetricList", version: "5.0.1" };
function computeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        fixed: cdktf.numberToTerraform(struct.fixed),
        percent: cdktf.numberToTerraform(struct.percent),
    };
}
exports.computeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasToTerraform = computeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasToTerraform;
class ComputeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fixed !== undefined) {
            hasAnyValues = true;
            internalValueResult.fixed = this._fixed;
        }
        if (this._percent !== undefined) {
            hasAnyValues = true;
            internalValueResult.percent = this._percent;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fixed = undefined;
            this._percent = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fixed = value.fixed;
            this._percent = value.percent;
        }
    }
    get fixed() {
        return this.getNumberAttribute('fixed');
    }
    set fixed(value) {
        this._fixed = value;
    }
    resetFixed() {
        this._fixed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fixedInput() {
        return this._fixed;
    }
    get percent() {
        return this.getNumberAttribute('percent');
    }
    set percent(value) {
        this._percent = value;
    }
    resetPercent() {
        this._percent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get percentInput() {
        return this._percent;
    }
}
exports.ComputeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasOutputReference = ComputeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasOutputReference[_e] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasOutputReference", version: "5.0.1" };
function computeAutoscalerAutoscalingPolicyScaleInControlToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        time_window_sec: cdktf.numberToTerraform(struct.timeWindowSec),
        max_scaled_in_replicas: computeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasToTerraform(struct.maxScaledInReplicas),
    };
}
exports.computeAutoscalerAutoscalingPolicyScaleInControlToTerraform = computeAutoscalerAutoscalingPolicyScaleInControlToTerraform;
class ComputeAutoscalerAutoscalingPolicyScaleInControlOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // max_scaled_in_replicas - computed: false, optional: true, required: false
        this._maxScaledInReplicas = new ComputeAutoscalerAutoscalingPolicyScaleInControlMaxScaledInReplicasOutputReference(this, "max_scaled_in_replicas");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._timeWindowSec !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeWindowSec = this._timeWindowSec;
        }
        if (this._maxScaledInReplicas?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxScaledInReplicas = this._maxScaledInReplicas?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._timeWindowSec = undefined;
            this._maxScaledInReplicas.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._timeWindowSec = value.timeWindowSec;
            this._maxScaledInReplicas.internalValue = value.maxScaledInReplicas;
        }
    }
    get timeWindowSec() {
        return this.getNumberAttribute('time_window_sec');
    }
    set timeWindowSec(value) {
        this._timeWindowSec = value;
    }
    resetTimeWindowSec() {
        this._timeWindowSec = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeWindowSecInput() {
        return this._timeWindowSec;
    }
    get maxScaledInReplicas() {
        return this._maxScaledInReplicas;
    }
    putMaxScaledInReplicas(value) {
        this._maxScaledInReplicas.internalValue = value;
    }
    resetMaxScaledInReplicas() {
        this._maxScaledInReplicas.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxScaledInReplicasInput() {
        return this._maxScaledInReplicas.internalValue;
    }
}
exports.ComputeAutoscalerAutoscalingPolicyScaleInControlOutputReference = ComputeAutoscalerAutoscalingPolicyScaleInControlOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerAutoscalingPolicyScaleInControlOutputReference[_f] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerAutoscalingPolicyScaleInControlOutputReference", version: "5.0.1" };
function computeAutoscalerAutoscalingPolicyScalingSchedulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        disabled: cdktf.booleanToTerraform(struct.disabled),
        duration_sec: cdktf.numberToTerraform(struct.durationSec),
        min_required_replicas: cdktf.numberToTerraform(struct.minRequiredReplicas),
        name: cdktf.stringToTerraform(struct.name),
        schedule: cdktf.stringToTerraform(struct.schedule),
        time_zone: cdktf.stringToTerraform(struct.timeZone),
    };
}
exports.computeAutoscalerAutoscalingPolicyScalingSchedulesToTerraform = computeAutoscalerAutoscalingPolicyScalingSchedulesToTerraform;
class ComputeAutoscalerAutoscalingPolicyScalingSchedulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._disabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.disabled = this._disabled;
        }
        if (this._durationSec !== undefined) {
            hasAnyValues = true;
            internalValueResult.durationSec = this._durationSec;
        }
        if (this._minRequiredReplicas !== undefined) {
            hasAnyValues = true;
            internalValueResult.minRequiredReplicas = this._minRequiredReplicas;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._schedule !== undefined) {
            hasAnyValues = true;
            internalValueResult.schedule = this._schedule;
        }
        if (this._timeZone !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeZone = this._timeZone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._description = undefined;
            this._disabled = undefined;
            this._durationSec = undefined;
            this._minRequiredReplicas = undefined;
            this._name = undefined;
            this._schedule = undefined;
            this._timeZone = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._description = value.description;
            this._disabled = value.disabled;
            this._durationSec = value.durationSec;
            this._minRequiredReplicas = value.minRequiredReplicas;
            this._name = value.name;
            this._schedule = value.schedule;
            this._timeZone = value.timeZone;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    set disabled(value) {
        this._disabled = value;
    }
    resetDisabled() {
        this._disabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disabledInput() {
        return this._disabled;
    }
    get durationSec() {
        return this.getNumberAttribute('duration_sec');
    }
    set durationSec(value) {
        this._durationSec = value;
    }
    // Temporarily expose input value. Use with caution.
    get durationSecInput() {
        return this._durationSec;
    }
    get minRequiredReplicas() {
        return this.getNumberAttribute('min_required_replicas');
    }
    set minRequiredReplicas(value) {
        this._minRequiredReplicas = value;
    }
    // Temporarily expose input value. Use with caution.
    get minRequiredReplicasInput() {
        return this._minRequiredReplicas;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    set schedule(value) {
        this._schedule = value;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule;
    }
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
    set timeZone(value) {
        this._timeZone = value;
    }
    resetTimeZone() {
        this._timeZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeZoneInput() {
        return this._timeZone;
    }
}
exports.ComputeAutoscalerAutoscalingPolicyScalingSchedulesOutputReference = ComputeAutoscalerAutoscalingPolicyScalingSchedulesOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerAutoscalingPolicyScalingSchedulesOutputReference[_g] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerAutoscalingPolicyScalingSchedulesOutputReference", version: "5.0.1" };
class ComputeAutoscalerAutoscalingPolicyScalingSchedulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeAutoscalerAutoscalingPolicyScalingSchedulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeAutoscalerAutoscalingPolicyScalingSchedulesList = ComputeAutoscalerAutoscalingPolicyScalingSchedulesList;
_h = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerAutoscalingPolicyScalingSchedulesList[_h] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerAutoscalingPolicyScalingSchedulesList", version: "5.0.1" };
function computeAutoscalerAutoscalingPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cooldown_period: cdktf.numberToTerraform(struct.cooldownPeriod),
        max_replicas: cdktf.numberToTerraform(struct.maxReplicas),
        min_replicas: cdktf.numberToTerraform(struct.minReplicas),
        mode: cdktf.stringToTerraform(struct.mode),
        cpu_utilization: computeAutoscalerAutoscalingPolicyCpuUtilizationToTerraform(struct.cpuUtilization),
        load_balancing_utilization: computeAutoscalerAutoscalingPolicyLoadBalancingUtilizationToTerraform(struct.loadBalancingUtilization),
        metric: cdktf.listMapper(computeAutoscalerAutoscalingPolicyMetricToTerraform, true)(struct.metric),
        scale_in_control: computeAutoscalerAutoscalingPolicyScaleInControlToTerraform(struct.scaleInControl),
        scaling_schedules: cdktf.listMapper(computeAutoscalerAutoscalingPolicyScalingSchedulesToTerraform, true)(struct.scalingSchedules),
    };
}
exports.computeAutoscalerAutoscalingPolicyToTerraform = computeAutoscalerAutoscalingPolicyToTerraform;
class ComputeAutoscalerAutoscalingPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cpu_utilization - computed: false, optional: true, required: false
        this._cpuUtilization = new ComputeAutoscalerAutoscalingPolicyCpuUtilizationOutputReference(this, "cpu_utilization");
        // load_balancing_utilization - computed: false, optional: true, required: false
        this._loadBalancingUtilization = new ComputeAutoscalerAutoscalingPolicyLoadBalancingUtilizationOutputReference(this, "load_balancing_utilization");
        // metric - computed: false, optional: true, required: false
        this._metric = new ComputeAutoscalerAutoscalingPolicyMetricList(this, "metric", false);
        // scale_in_control - computed: false, optional: true, required: false
        this._scaleInControl = new ComputeAutoscalerAutoscalingPolicyScaleInControlOutputReference(this, "scale_in_control");
        // scaling_schedules - computed: false, optional: true, required: false
        this._scalingSchedules = new ComputeAutoscalerAutoscalingPolicyScalingSchedulesList(this, "scaling_schedules", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cooldownPeriod !== undefined) {
            hasAnyValues = true;
            internalValueResult.cooldownPeriod = this._cooldownPeriod;
        }
        if (this._maxReplicas !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxReplicas = this._maxReplicas;
        }
        if (this._minReplicas !== undefined) {
            hasAnyValues = true;
            internalValueResult.minReplicas = this._minReplicas;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._cpuUtilization?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuUtilization = this._cpuUtilization?.internalValue;
        }
        if (this._loadBalancingUtilization?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.loadBalancingUtilization = this._loadBalancingUtilization?.internalValue;
        }
        if (this._metric?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.metric = this._metric?.internalValue;
        }
        if (this._scaleInControl?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.scaleInControl = this._scaleInControl?.internalValue;
        }
        if (this._scalingSchedules?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.scalingSchedules = this._scalingSchedules?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cooldownPeriod = undefined;
            this._maxReplicas = undefined;
            this._minReplicas = undefined;
            this._mode = undefined;
            this._cpuUtilization.internalValue = undefined;
            this._loadBalancingUtilization.internalValue = undefined;
            this._metric.internalValue = undefined;
            this._scaleInControl.internalValue = undefined;
            this._scalingSchedules.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cooldownPeriod = value.cooldownPeriod;
            this._maxReplicas = value.maxReplicas;
            this._minReplicas = value.minReplicas;
            this._mode = value.mode;
            this._cpuUtilization.internalValue = value.cpuUtilization;
            this._loadBalancingUtilization.internalValue = value.loadBalancingUtilization;
            this._metric.internalValue = value.metric;
            this._scaleInControl.internalValue = value.scaleInControl;
            this._scalingSchedules.internalValue = value.scalingSchedules;
        }
    }
    get cooldownPeriod() {
        return this.getNumberAttribute('cooldown_period');
    }
    set cooldownPeriod(value) {
        this._cooldownPeriod = value;
    }
    resetCooldownPeriod() {
        this._cooldownPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cooldownPeriodInput() {
        return this._cooldownPeriod;
    }
    get maxReplicas() {
        return this.getNumberAttribute('max_replicas');
    }
    set maxReplicas(value) {
        this._maxReplicas = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxReplicasInput() {
        return this._maxReplicas;
    }
    get minReplicas() {
        return this.getNumberAttribute('min_replicas');
    }
    set minReplicas(value) {
        this._minReplicas = value;
    }
    // Temporarily expose input value. Use with caution.
    get minReplicasInput() {
        return this._minReplicas;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get cpuUtilization() {
        return this._cpuUtilization;
    }
    putCpuUtilization(value) {
        this._cpuUtilization.internalValue = value;
    }
    resetCpuUtilization() {
        this._cpuUtilization.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuUtilizationInput() {
        return this._cpuUtilization.internalValue;
    }
    get loadBalancingUtilization() {
        return this._loadBalancingUtilization;
    }
    putLoadBalancingUtilization(value) {
        this._loadBalancingUtilization.internalValue = value;
    }
    resetLoadBalancingUtilization() {
        this._loadBalancingUtilization.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancingUtilizationInput() {
        return this._loadBalancingUtilization.internalValue;
    }
    get metric() {
        return this._metric;
    }
    putMetric(value) {
        this._metric.internalValue = value;
    }
    resetMetric() {
        this._metric.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metricInput() {
        return this._metric.internalValue;
    }
    get scaleInControl() {
        return this._scaleInControl;
    }
    putScaleInControl(value) {
        this._scaleInControl.internalValue = value;
    }
    resetScaleInControl() {
        this._scaleInControl.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scaleInControlInput() {
        return this._scaleInControl.internalValue;
    }
    get scalingSchedules() {
        return this._scalingSchedules;
    }
    putScalingSchedules(value) {
        this._scalingSchedules.internalValue = value;
    }
    resetScalingSchedules() {
        this._scalingSchedules.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scalingSchedulesInput() {
        return this._scalingSchedules.internalValue;
    }
}
exports.ComputeAutoscalerAutoscalingPolicyOutputReference = ComputeAutoscalerAutoscalingPolicyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerAutoscalingPolicyOutputReference[_j] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerAutoscalingPolicyOutputReference", version: "5.0.1" };
function computeAutoscalerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.computeAutoscalerTimeoutsToTerraform = computeAutoscalerTimeoutsToTerraform;
class ComputeAutoscalerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ComputeAutoscalerTimeoutsOutputReference = ComputeAutoscalerTimeoutsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ComputeAutoscalerTimeoutsOutputReference[_k] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscalerTimeoutsOutputReference", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/compute_autoscaler google_compute_autoscaler}
*/
class ComputeAutoscaler extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/compute_autoscaler google_compute_autoscaler} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComputeAutoscalerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_autoscaler',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // autoscaling_policy - computed: false, optional: false, required: true
        this._autoscalingPolicy = new ComputeAutoscalerAutoscalingPolicyOutputReference(this, "autoscaling_policy");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComputeAutoscalerTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._target = config.target;
        this._zone = config.zone;
        this._autoscalingPolicy.internalValue = config.autoscalingPolicy;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get target() {
        return this.getStringAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    get autoscalingPolicy() {
        return this._autoscalingPolicy;
    }
    putAutoscalingPolicy(value) {
        this._autoscalingPolicy.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingPolicyInput() {
        return this._autoscalingPolicy.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            target: cdktf.stringToTerraform(this._target),
            zone: cdktf.stringToTerraform(this._zone),
            autoscaling_policy: computeAutoscalerAutoscalingPolicyToTerraform(this._autoscalingPolicy.internalValue),
            timeouts: computeAutoscalerTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.ComputeAutoscaler = ComputeAutoscaler;
_l = JSII_RTTI_SYMBOL_1;
ComputeAutoscaler[_l] = { fqn: "@cdktf/provider-google.computeAutoscaler.ComputeAutoscaler", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
ComputeAutoscaler.tfResourceType = "google_compute_autoscaler";
