// @cdktf/provider-google/lib/data-google-beyondcorp-app-connector/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleBeyondcorpAppConnector = exports.DataGoogleBeyondcorpAppConnectorPrincipalInfoList = exports.DataGoogleBeyondcorpAppConnectorPrincipalInfoOutputReference = exports.dataGoogleBeyondcorpAppConnectorPrincipalInfoToTerraform = exports.DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountList = exports.DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountOutputReference = exports.dataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountToTerraform = dataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountToTerraform;
class DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
}
exports.DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountOutputReference = DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnector.DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountOutputReference", version: "5.0.1" };
class DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountList = DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountList[_b] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnector.DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountList", version: "5.0.1" };
function dataGoogleBeyondcorpAppConnectorPrincipalInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleBeyondcorpAppConnectorPrincipalInfoToTerraform = dataGoogleBeyondcorpAppConnectorPrincipalInfoToTerraform;
class DataGoogleBeyondcorpAppConnectorPrincipalInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // service_account - computed: true, optional: false, required: false
        this._serviceAccount = new DataGoogleBeyondcorpAppConnectorPrincipalInfoServiceAccountList(this, "service_account", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get serviceAccount() {
        return this._serviceAccount;
    }
}
exports.DataGoogleBeyondcorpAppConnectorPrincipalInfoOutputReference = DataGoogleBeyondcorpAppConnectorPrincipalInfoOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnectorPrincipalInfoOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnector.DataGoogleBeyondcorpAppConnectorPrincipalInfoOutputReference", version: "5.0.1" };
class DataGoogleBeyondcorpAppConnectorPrincipalInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleBeyondcorpAppConnectorPrincipalInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleBeyondcorpAppConnectorPrincipalInfoList = DataGoogleBeyondcorpAppConnectorPrincipalInfoList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnectorPrincipalInfoList[_d] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnector.DataGoogleBeyondcorpAppConnectorPrincipalInfoList", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/beyondcorp_app_connector google_beyondcorp_app_connector}
*/
class DataGoogleBeyondcorpAppConnector extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/beyondcorp_app_connector google_beyondcorp_app_connector} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleBeyondcorpAppConnectorConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_beyondcorp_app_connector',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // principal_info - computed: true, optional: false, required: false
        this._principalInfo = new DataGoogleBeyondcorpAppConnectorPrincipalInfoList(this, "principal_info", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get principalInfo() {
        return this._principalInfo;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
        };
    }
}
exports.DataGoogleBeyondcorpAppConnector = DataGoogleBeyondcorpAppConnector;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnector[_e] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnector.DataGoogleBeyondcorpAppConnector", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleBeyondcorpAppConnector.tfResourceType = "google_beyondcorp_app_connector";
