// @cdktf/provider-google/lib/container-azure-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerAzureCluster = exports.ContainerAzureClusterTimeoutsOutputReference = exports.containerAzureClusterTimeoutsToTerraform = exports.ContainerAzureClusterNetworkingOutputReference = exports.containerAzureClusterNetworkingToTerraform = exports.ContainerAzureClusterFleetOutputReference = exports.containerAzureClusterFleetToTerraform = exports.ContainerAzureClusterControlPlaneOutputReference = exports.containerAzureClusterControlPlaneToTerraform = exports.ContainerAzureClusterControlPlaneSshConfigOutputReference = exports.containerAzureClusterControlPlaneSshConfigToTerraform = exports.ContainerAzureClusterControlPlaneRootVolumeOutputReference = exports.containerAzureClusterControlPlaneRootVolumeToTerraform = exports.ContainerAzureClusterControlPlaneReplicaPlacementsList = exports.ContainerAzureClusterControlPlaneReplicaPlacementsOutputReference = exports.containerAzureClusterControlPlaneReplicaPlacementsToTerraform = exports.ContainerAzureClusterControlPlaneProxyConfigOutputReference = exports.containerAzureClusterControlPlaneProxyConfigToTerraform = exports.ContainerAzureClusterControlPlaneMainVolumeOutputReference = exports.containerAzureClusterControlPlaneMainVolumeToTerraform = exports.ContainerAzureClusterControlPlaneDatabaseEncryptionOutputReference = exports.containerAzureClusterControlPlaneDatabaseEncryptionToTerraform = exports.ContainerAzureClusterAuthorizationOutputReference = exports.containerAzureClusterAuthorizationToTerraform = exports.ContainerAzureClusterAuthorizationAdminUsersList = exports.ContainerAzureClusterAuthorizationAdminUsersOutputReference = exports.containerAzureClusterAuthorizationAdminUsersToTerraform = exports.ContainerAzureClusterWorkloadIdentityConfigList = exports.ContainerAzureClusterWorkloadIdentityConfigOutputReference = exports.containerAzureClusterWorkloadIdentityConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function containerAzureClusterWorkloadIdentityConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.containerAzureClusterWorkloadIdentityConfigToTerraform = containerAzureClusterWorkloadIdentityConfigToTerraform;
class ContainerAzureClusterWorkloadIdentityConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // identity_provider - computed: true, optional: false, required: false
    get identityProvider() {
        return this.getStringAttribute('identity_provider');
    }
    // issuer_uri - computed: true, optional: false, required: false
    get issuerUri() {
        return this.getStringAttribute('issuer_uri');
    }
    // workload_pool - computed: true, optional: false, required: false
    get workloadPool() {
        return this.getStringAttribute('workload_pool');
    }
}
exports.ContainerAzureClusterWorkloadIdentityConfigOutputReference = ContainerAzureClusterWorkloadIdentityConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterWorkloadIdentityConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterWorkloadIdentityConfigOutputReference", version: "5.0.1" };
class ContainerAzureClusterWorkloadIdentityConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ContainerAzureClusterWorkloadIdentityConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ContainerAzureClusterWorkloadIdentityConfigList = ContainerAzureClusterWorkloadIdentityConfigList;
_b = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterWorkloadIdentityConfigList[_b] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterWorkloadIdentityConfigList", version: "5.0.1" };
function containerAzureClusterAuthorizationAdminUsersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        username: cdktf.stringToTerraform(struct.username),
    };
}
exports.containerAzureClusterAuthorizationAdminUsersToTerraform = containerAzureClusterAuthorizationAdminUsersToTerraform;
class ContainerAzureClusterAuthorizationAdminUsersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._username !== undefined) {
            hasAnyValues = true;
            internalValueResult.username = this._username;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._username = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._username = value.username;
        }
    }
    get username() {
        return this.getStringAttribute('username');
    }
    set username(value) {
        this._username = value;
    }
    // Temporarily expose input value. Use with caution.
    get usernameInput() {
        return this._username;
    }
}
exports.ContainerAzureClusterAuthorizationAdminUsersOutputReference = ContainerAzureClusterAuthorizationAdminUsersOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterAuthorizationAdminUsersOutputReference[_c] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterAuthorizationAdminUsersOutputReference", version: "5.0.1" };
class ContainerAzureClusterAuthorizationAdminUsersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ContainerAzureClusterAuthorizationAdminUsersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ContainerAzureClusterAuthorizationAdminUsersList = ContainerAzureClusterAuthorizationAdminUsersList;
_d = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterAuthorizationAdminUsersList[_d] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterAuthorizationAdminUsersList", version: "5.0.1" };
function containerAzureClusterAuthorizationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        admin_users: cdktf.listMapper(containerAzureClusterAuthorizationAdminUsersToTerraform, true)(struct.adminUsers),
    };
}
exports.containerAzureClusterAuthorizationToTerraform = containerAzureClusterAuthorizationToTerraform;
class ContainerAzureClusterAuthorizationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // admin_users - computed: false, optional: false, required: true
        this._adminUsers = new ContainerAzureClusterAuthorizationAdminUsersList(this, "admin_users", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._adminUsers?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.adminUsers = this._adminUsers?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._adminUsers.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._adminUsers.internalValue = value.adminUsers;
        }
    }
    get adminUsers() {
        return this._adminUsers;
    }
    putAdminUsers(value) {
        this._adminUsers.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get adminUsersInput() {
        return this._adminUsers.internalValue;
    }
}
exports.ContainerAzureClusterAuthorizationOutputReference = ContainerAzureClusterAuthorizationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterAuthorizationOutputReference[_e] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterAuthorizationOutputReference", version: "5.0.1" };
function containerAzureClusterControlPlaneDatabaseEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key_id: cdktf.stringToTerraform(struct.keyId),
    };
}
exports.containerAzureClusterControlPlaneDatabaseEncryptionToTerraform = containerAzureClusterControlPlaneDatabaseEncryptionToTerraform;
class ContainerAzureClusterControlPlaneDatabaseEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._keyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.keyId = this._keyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._keyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._keyId = value.keyId;
        }
    }
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    set keyId(value) {
        this._keyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyIdInput() {
        return this._keyId;
    }
}
exports.ContainerAzureClusterControlPlaneDatabaseEncryptionOutputReference = ContainerAzureClusterControlPlaneDatabaseEncryptionOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterControlPlaneDatabaseEncryptionOutputReference[_f] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterControlPlaneDatabaseEncryptionOutputReference", version: "5.0.1" };
function containerAzureClusterControlPlaneMainVolumeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        size_gib: cdktf.numberToTerraform(struct.sizeGib),
    };
}
exports.containerAzureClusterControlPlaneMainVolumeToTerraform = containerAzureClusterControlPlaneMainVolumeToTerraform;
class ContainerAzureClusterControlPlaneMainVolumeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sizeGib !== undefined) {
            hasAnyValues = true;
            internalValueResult.sizeGib = this._sizeGib;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sizeGib = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sizeGib = value.sizeGib;
        }
    }
    get sizeGib() {
        return this.getNumberAttribute('size_gib');
    }
    set sizeGib(value) {
        this._sizeGib = value;
    }
    resetSizeGib() {
        this._sizeGib = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sizeGibInput() {
        return this._sizeGib;
    }
}
exports.ContainerAzureClusterControlPlaneMainVolumeOutputReference = ContainerAzureClusterControlPlaneMainVolumeOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterControlPlaneMainVolumeOutputReference[_g] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterControlPlaneMainVolumeOutputReference", version: "5.0.1" };
function containerAzureClusterControlPlaneProxyConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_group_id: cdktf.stringToTerraform(struct.resourceGroupId),
        secret_id: cdktf.stringToTerraform(struct.secretId),
    };
}
exports.containerAzureClusterControlPlaneProxyConfigToTerraform = containerAzureClusterControlPlaneProxyConfigToTerraform;
class ContainerAzureClusterControlPlaneProxyConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceGroupId = this._resourceGroupId;
        }
        if (this._secretId !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretId = this._secretId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceGroupId = undefined;
            this._secretId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceGroupId = value.resourceGroupId;
            this._secretId = value.secretId;
        }
    }
    get resourceGroupId() {
        return this.getStringAttribute('resource_group_id');
    }
    set resourceGroupId(value) {
        this._resourceGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceGroupIdInput() {
        return this._resourceGroupId;
    }
    get secretId() {
        return this.getStringAttribute('secret_id');
    }
    set secretId(value) {
        this._secretId = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretIdInput() {
        return this._secretId;
    }
}
exports.ContainerAzureClusterControlPlaneProxyConfigOutputReference = ContainerAzureClusterControlPlaneProxyConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterControlPlaneProxyConfigOutputReference[_h] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterControlPlaneProxyConfigOutputReference", version: "5.0.1" };
function containerAzureClusterControlPlaneReplicaPlacementsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        azure_availability_zone: cdktf.stringToTerraform(struct.azureAvailabilityZone),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
    };
}
exports.containerAzureClusterControlPlaneReplicaPlacementsToTerraform = containerAzureClusterControlPlaneReplicaPlacementsToTerraform;
class ContainerAzureClusterControlPlaneReplicaPlacementsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._azureAvailabilityZone !== undefined) {
            hasAnyValues = true;
            internalValueResult.azureAvailabilityZone = this._azureAvailabilityZone;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._azureAvailabilityZone = undefined;
            this._subnetId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._azureAvailabilityZone = value.azureAvailabilityZone;
            this._subnetId = value.subnetId;
        }
    }
    get azureAvailabilityZone() {
        return this.getStringAttribute('azure_availability_zone');
    }
    set azureAvailabilityZone(value) {
        this._azureAvailabilityZone = value;
    }
    // Temporarily expose input value. Use with caution.
    get azureAvailabilityZoneInput() {
        return this._azureAvailabilityZone;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
}
exports.ContainerAzureClusterControlPlaneReplicaPlacementsOutputReference = ContainerAzureClusterControlPlaneReplicaPlacementsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterControlPlaneReplicaPlacementsOutputReference[_j] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterControlPlaneReplicaPlacementsOutputReference", version: "5.0.1" };
class ContainerAzureClusterControlPlaneReplicaPlacementsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ContainerAzureClusterControlPlaneReplicaPlacementsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ContainerAzureClusterControlPlaneReplicaPlacementsList = ContainerAzureClusterControlPlaneReplicaPlacementsList;
_k = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterControlPlaneReplicaPlacementsList[_k] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterControlPlaneReplicaPlacementsList", version: "5.0.1" };
function containerAzureClusterControlPlaneRootVolumeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        size_gib: cdktf.numberToTerraform(struct.sizeGib),
    };
}
exports.containerAzureClusterControlPlaneRootVolumeToTerraform = containerAzureClusterControlPlaneRootVolumeToTerraform;
class ContainerAzureClusterControlPlaneRootVolumeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sizeGib !== undefined) {
            hasAnyValues = true;
            internalValueResult.sizeGib = this._sizeGib;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sizeGib = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sizeGib = value.sizeGib;
        }
    }
    get sizeGib() {
        return this.getNumberAttribute('size_gib');
    }
    set sizeGib(value) {
        this._sizeGib = value;
    }
    resetSizeGib() {
        this._sizeGib = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sizeGibInput() {
        return this._sizeGib;
    }
}
exports.ContainerAzureClusterControlPlaneRootVolumeOutputReference = ContainerAzureClusterControlPlaneRootVolumeOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterControlPlaneRootVolumeOutputReference[_l] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterControlPlaneRootVolumeOutputReference", version: "5.0.1" };
function containerAzureClusterControlPlaneSshConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        authorized_key: cdktf.stringToTerraform(struct.authorizedKey),
    };
}
exports.containerAzureClusterControlPlaneSshConfigToTerraform = containerAzureClusterControlPlaneSshConfigToTerraform;
class ContainerAzureClusterControlPlaneSshConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._authorizedKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.authorizedKey = this._authorizedKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._authorizedKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._authorizedKey = value.authorizedKey;
        }
    }
    get authorizedKey() {
        return this.getStringAttribute('authorized_key');
    }
    set authorizedKey(value) {
        this._authorizedKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get authorizedKeyInput() {
        return this._authorizedKey;
    }
}
exports.ContainerAzureClusterControlPlaneSshConfigOutputReference = ContainerAzureClusterControlPlaneSshConfigOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterControlPlaneSshConfigOutputReference[_m] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterControlPlaneSshConfigOutputReference", version: "5.0.1" };
function containerAzureClusterControlPlaneToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        tags: cdktf.hashMapper(cdktf.stringToTerraform)(struct.tags),
        version: cdktf.stringToTerraform(struct.version),
        vm_size: cdktf.stringToTerraform(struct.vmSize),
        database_encryption: containerAzureClusterControlPlaneDatabaseEncryptionToTerraform(struct.databaseEncryption),
        main_volume: containerAzureClusterControlPlaneMainVolumeToTerraform(struct.mainVolume),
        proxy_config: containerAzureClusterControlPlaneProxyConfigToTerraform(struct.proxyConfig),
        replica_placements: cdktf.listMapper(containerAzureClusterControlPlaneReplicaPlacementsToTerraform, true)(struct.replicaPlacements),
        root_volume: containerAzureClusterControlPlaneRootVolumeToTerraform(struct.rootVolume),
        ssh_config: containerAzureClusterControlPlaneSshConfigToTerraform(struct.sshConfig),
    };
}
exports.containerAzureClusterControlPlaneToTerraform = containerAzureClusterControlPlaneToTerraform;
class ContainerAzureClusterControlPlaneOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // database_encryption - computed: false, optional: true, required: false
        this._databaseEncryption = new ContainerAzureClusterControlPlaneDatabaseEncryptionOutputReference(this, "database_encryption");
        // main_volume - computed: false, optional: true, required: false
        this._mainVolume = new ContainerAzureClusterControlPlaneMainVolumeOutputReference(this, "main_volume");
        // proxy_config - computed: false, optional: true, required: false
        this._proxyConfig = new ContainerAzureClusterControlPlaneProxyConfigOutputReference(this, "proxy_config");
        // replica_placements - computed: false, optional: true, required: false
        this._replicaPlacements = new ContainerAzureClusterControlPlaneReplicaPlacementsList(this, "replica_placements", false);
        // root_volume - computed: false, optional: true, required: false
        this._rootVolume = new ContainerAzureClusterControlPlaneRootVolumeOutputReference(this, "root_volume");
        // ssh_config - computed: false, optional: false, required: true
        this._sshConfig = new ContainerAzureClusterControlPlaneSshConfigOutputReference(this, "ssh_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        if (this._vmSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.vmSize = this._vmSize;
        }
        if (this._databaseEncryption?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseEncryption = this._databaseEncryption?.internalValue;
        }
        if (this._mainVolume?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.mainVolume = this._mainVolume?.internalValue;
        }
        if (this._proxyConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.proxyConfig = this._proxyConfig?.internalValue;
        }
        if (this._replicaPlacements?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.replicaPlacements = this._replicaPlacements?.internalValue;
        }
        if (this._rootVolume?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rootVolume = this._rootVolume?.internalValue;
        }
        if (this._sshConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sshConfig = this._sshConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._subnetId = undefined;
            this._tags = undefined;
            this._version = undefined;
            this._vmSize = undefined;
            this._databaseEncryption.internalValue = undefined;
            this._mainVolume.internalValue = undefined;
            this._proxyConfig.internalValue = undefined;
            this._replicaPlacements.internalValue = undefined;
            this._rootVolume.internalValue = undefined;
            this._sshConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._subnetId = value.subnetId;
            this._tags = value.tags;
            this._version = value.version;
            this._vmSize = value.vmSize;
            this._databaseEncryption.internalValue = value.databaseEncryption;
            this._mainVolume.internalValue = value.mainVolume;
            this._proxyConfig.internalValue = value.proxyConfig;
            this._replicaPlacements.internalValue = value.replicaPlacements;
            this._rootVolume.internalValue = value.rootVolume;
            this._sshConfig.internalValue = value.sshConfig;
        }
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get vmSize() {
        return this.getStringAttribute('vm_size');
    }
    set vmSize(value) {
        this._vmSize = value;
    }
    resetVmSize() {
        this._vmSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vmSizeInput() {
        return this._vmSize;
    }
    get databaseEncryption() {
        return this._databaseEncryption;
    }
    putDatabaseEncryption(value) {
        this._databaseEncryption.internalValue = value;
    }
    resetDatabaseEncryption() {
        this._databaseEncryption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseEncryptionInput() {
        return this._databaseEncryption.internalValue;
    }
    get mainVolume() {
        return this._mainVolume;
    }
    putMainVolume(value) {
        this._mainVolume.internalValue = value;
    }
    resetMainVolume() {
        this._mainVolume.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mainVolumeInput() {
        return this._mainVolume.internalValue;
    }
    get proxyConfig() {
        return this._proxyConfig;
    }
    putProxyConfig(value) {
        this._proxyConfig.internalValue = value;
    }
    resetProxyConfig() {
        this._proxyConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get proxyConfigInput() {
        return this._proxyConfig.internalValue;
    }
    get replicaPlacements() {
        return this._replicaPlacements;
    }
    putReplicaPlacements(value) {
        this._replicaPlacements.internalValue = value;
    }
    resetReplicaPlacements() {
        this._replicaPlacements.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicaPlacementsInput() {
        return this._replicaPlacements.internalValue;
    }
    get rootVolume() {
        return this._rootVolume;
    }
    putRootVolume(value) {
        this._rootVolume.internalValue = value;
    }
    resetRootVolume() {
        this._rootVolume.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rootVolumeInput() {
        return this._rootVolume.internalValue;
    }
    get sshConfig() {
        return this._sshConfig;
    }
    putSshConfig(value) {
        this._sshConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get sshConfigInput() {
        return this._sshConfig.internalValue;
    }
}
exports.ContainerAzureClusterControlPlaneOutputReference = ContainerAzureClusterControlPlaneOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterControlPlaneOutputReference[_o] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterControlPlaneOutputReference", version: "5.0.1" };
function containerAzureClusterFleetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        project: cdktf.stringToTerraform(struct.project),
    };
}
exports.containerAzureClusterFleetToTerraform = containerAzureClusterFleetToTerraform;
class ContainerAzureClusterFleetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._project !== undefined) {
            hasAnyValues = true;
            internalValueResult.project = this._project;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._project = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._project = value.project;
        }
    }
    // membership - computed: true, optional: false, required: false
    get membership() {
        return this.getStringAttribute('membership');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
}
exports.ContainerAzureClusterFleetOutputReference = ContainerAzureClusterFleetOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterFleetOutputReference[_p] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterFleetOutputReference", version: "5.0.1" };
function containerAzureClusterNetworkingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        pod_address_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.podAddressCidrBlocks),
        service_address_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.serviceAddressCidrBlocks),
        virtual_network_id: cdktf.stringToTerraform(struct.virtualNetworkId),
    };
}
exports.containerAzureClusterNetworkingToTerraform = containerAzureClusterNetworkingToTerraform;
class ContainerAzureClusterNetworkingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._podAddressCidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.podAddressCidrBlocks = this._podAddressCidrBlocks;
        }
        if (this._serviceAddressCidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAddressCidrBlocks = this._serviceAddressCidrBlocks;
        }
        if (this._virtualNetworkId !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualNetworkId = this._virtualNetworkId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._podAddressCidrBlocks = undefined;
            this._serviceAddressCidrBlocks = undefined;
            this._virtualNetworkId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._podAddressCidrBlocks = value.podAddressCidrBlocks;
            this._serviceAddressCidrBlocks = value.serviceAddressCidrBlocks;
            this._virtualNetworkId = value.virtualNetworkId;
        }
    }
    get podAddressCidrBlocks() {
        return this.getListAttribute('pod_address_cidr_blocks');
    }
    set podAddressCidrBlocks(value) {
        this._podAddressCidrBlocks = value;
    }
    // Temporarily expose input value. Use with caution.
    get podAddressCidrBlocksInput() {
        return this._podAddressCidrBlocks;
    }
    get serviceAddressCidrBlocks() {
        return this.getListAttribute('service_address_cidr_blocks');
    }
    set serviceAddressCidrBlocks(value) {
        this._serviceAddressCidrBlocks = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAddressCidrBlocksInput() {
        return this._serviceAddressCidrBlocks;
    }
    get virtualNetworkId() {
        return this.getStringAttribute('virtual_network_id');
    }
    set virtualNetworkId(value) {
        this._virtualNetworkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualNetworkIdInput() {
        return this._virtualNetworkId;
    }
}
exports.ContainerAzureClusterNetworkingOutputReference = ContainerAzureClusterNetworkingOutputReference;
_q = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterNetworkingOutputReference[_q] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterNetworkingOutputReference", version: "5.0.1" };
function containerAzureClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.containerAzureClusterTimeoutsToTerraform = containerAzureClusterTimeoutsToTerraform;
class ContainerAzureClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ContainerAzureClusterTimeoutsOutputReference = ContainerAzureClusterTimeoutsOutputReference;
_r = JSII_RTTI_SYMBOL_1;
ContainerAzureClusterTimeoutsOutputReference[_r] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureClusterTimeoutsOutputReference", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/container_azure_cluster google_container_azure_cluster}
*/
class ContainerAzureCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/container_azure_cluster google_container_azure_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ContainerAzureClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_container_azure_cluster',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // workload_identity_config - computed: true, optional: false, required: false
        this._workloadIdentityConfig = new ContainerAzureClusterWorkloadIdentityConfigList(this, "workload_identity_config", false);
        // authorization - computed: false, optional: false, required: true
        this._authorization = new ContainerAzureClusterAuthorizationOutputReference(this, "authorization");
        // control_plane - computed: false, optional: false, required: true
        this._controlPlane = new ContainerAzureClusterControlPlaneOutputReference(this, "control_plane");
        // fleet - computed: false, optional: false, required: true
        this._fleet = new ContainerAzureClusterFleetOutputReference(this, "fleet");
        // networking - computed: false, optional: false, required: true
        this._networking = new ContainerAzureClusterNetworkingOutputReference(this, "networking");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ContainerAzureClusterTimeoutsOutputReference(this, "timeouts");
        this._annotations = config.annotations;
        this._azureRegion = config.azureRegion;
        this._client = config.client;
        this._description = config.description;
        this._id = config.id;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._resourceGroupId = config.resourceGroupId;
        this._authorization.internalValue = config.authorization;
        this._controlPlane.internalValue = config.controlPlane;
        this._fleet.internalValue = config.fleet;
        this._networking.internalValue = config.networking;
        this._timeouts.internalValue = config.timeouts;
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    get azureRegion() {
        return this.getStringAttribute('azure_region');
    }
    set azureRegion(value) {
        this._azureRegion = value;
    }
    // Temporarily expose input value. Use with caution.
    get azureRegionInput() {
        return this._azureRegion;
    }
    get client() {
        return this.getStringAttribute('client');
    }
    set client(value) {
        this._client = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientInput() {
        return this._client;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // reconciling - computed: true, optional: false, required: false
    get reconciling() {
        return this.getBooleanAttribute('reconciling');
    }
    get resourceGroupId() {
        return this.getStringAttribute('resource_group_id');
    }
    set resourceGroupId(value) {
        this._resourceGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceGroupIdInput() {
        return this._resourceGroupId;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get workloadIdentityConfig() {
        return this._workloadIdentityConfig;
    }
    get authorization() {
        return this._authorization;
    }
    putAuthorization(value) {
        this._authorization.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationInput() {
        return this._authorization.internalValue;
    }
    get controlPlane() {
        return this._controlPlane;
    }
    putControlPlane(value) {
        this._controlPlane.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get controlPlaneInput() {
        return this._controlPlane.internalValue;
    }
    get fleet() {
        return this._fleet;
    }
    putFleet(value) {
        this._fleet.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get fleetInput() {
        return this._fleet.internalValue;
    }
    get networking() {
        return this._networking;
    }
    putNetworking(value) {
        this._networking.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkingInput() {
        return this._networking.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            annotations: cdktf.hashMapper(cdktf.stringToTerraform)(this._annotations),
            azure_region: cdktf.stringToTerraform(this._azureRegion),
            client: cdktf.stringToTerraform(this._client),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            resource_group_id: cdktf.stringToTerraform(this._resourceGroupId),
            authorization: containerAzureClusterAuthorizationToTerraform(this._authorization.internalValue),
            control_plane: containerAzureClusterControlPlaneToTerraform(this._controlPlane.internalValue),
            fleet: containerAzureClusterFleetToTerraform(this._fleet.internalValue),
            networking: containerAzureClusterNetworkingToTerraform(this._networking.internalValue),
            timeouts: containerAzureClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.ContainerAzureCluster = ContainerAzureCluster;
_s = JSII_RTTI_SYMBOL_1;
ContainerAzureCluster[_s] = { fqn: "@cdktf/provider-google.containerAzureCluster.ContainerAzureCluster", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
ContainerAzureCluster.tfResourceType = "google_container_azure_cluster";
