// @cdktf/provider-google/lib/cloudbuild-trigger/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudbuildTrigger = exports.CloudbuildTriggerWebhookConfigOutputReference = exports.cloudbuildTriggerWebhookConfigToTerraform = exports.CloudbuildTriggerTriggerTemplateOutputReference = exports.cloudbuildTriggerTriggerTemplateToTerraform = exports.CloudbuildTriggerTimeoutsOutputReference = exports.cloudbuildTriggerTimeoutsToTerraform = exports.CloudbuildTriggerSourceToBuildOutputReference = exports.cloudbuildTriggerSourceToBuildToTerraform = exports.CloudbuildTriggerPubsubConfigOutputReference = exports.cloudbuildTriggerPubsubConfigToTerraform = exports.CloudbuildTriggerGithubOutputReference = exports.cloudbuildTriggerGithubToTerraform = exports.CloudbuildTriggerGithubPushOutputReference = exports.cloudbuildTriggerGithubPushToTerraform = exports.CloudbuildTriggerGithubPullRequestOutputReference = exports.cloudbuildTriggerGithubPullRequestToTerraform = exports.CloudbuildTriggerGitFileSourceOutputReference = exports.cloudbuildTriggerGitFileSourceToTerraform = exports.CloudbuildTriggerBuildOutputReference = exports.cloudbuildTriggerBuildToTerraform = exports.CloudbuildTriggerBuildStepList = exports.CloudbuildTriggerBuildStepOutputReference = exports.cloudbuildTriggerBuildStepToTerraform = exports.CloudbuildTriggerBuildStepVolumesList = exports.CloudbuildTriggerBuildStepVolumesOutputReference = exports.cloudbuildTriggerBuildStepVolumesToTerraform = exports.CloudbuildTriggerBuildSourceOutputReference = exports.cloudbuildTriggerBuildSourceToTerraform = exports.CloudbuildTriggerBuildSourceStorageSourceOutputReference = exports.cloudbuildTriggerBuildSourceStorageSourceToTerraform = exports.CloudbuildTriggerBuildSourceRepoSourceOutputReference = exports.cloudbuildTriggerBuildSourceRepoSourceToTerraform = exports.CloudbuildTriggerBuildSecretList = exports.CloudbuildTriggerBuildSecretOutputReference = exports.cloudbuildTriggerBuildSecretToTerraform = exports.CloudbuildTriggerBuildOptionsOutputReference = exports.cloudbuildTriggerBuildOptionsToTerraform = exports.CloudbuildTriggerBuildOptionsVolumesList = exports.CloudbuildTriggerBuildOptionsVolumesOutputReference = exports.cloudbuildTriggerBuildOptionsVolumesToTerraform = exports.CloudbuildTriggerBuildAvailableSecretsOutputReference = exports.cloudbuildTriggerBuildAvailableSecretsToTerraform = exports.CloudbuildTriggerBuildAvailableSecretsSecretManagerList = exports.CloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference = exports.cloudbuildTriggerBuildAvailableSecretsSecretManagerToTerraform = exports.CloudbuildTriggerBuildArtifactsOutputReference = exports.cloudbuildTriggerBuildArtifactsToTerraform = exports.CloudbuildTriggerBuildArtifactsObjectsOutputReference = exports.cloudbuildTriggerBuildArtifactsObjectsToTerraform = exports.CloudbuildTriggerBuildArtifactsObjectsTimingList = exports.CloudbuildTriggerBuildArtifactsObjectsTimingOutputReference = exports.cloudbuildTriggerBuildArtifactsObjectsTimingToTerraform = exports.CloudbuildTriggerApprovalConfigOutputReference = exports.cloudbuildTriggerApprovalConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudbuildTriggerApprovalConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        approval_required: cdktf.booleanToTerraform(struct.approvalRequired),
    };
}
exports.cloudbuildTriggerApprovalConfigToTerraform = cloudbuildTriggerApprovalConfigToTerraform;
class CloudbuildTriggerApprovalConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._approvalRequired !== undefined) {
            hasAnyValues = true;
            internalValueResult.approvalRequired = this._approvalRequired;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._approvalRequired = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._approvalRequired = value.approvalRequired;
        }
    }
    get approvalRequired() {
        return this.getBooleanAttribute('approval_required');
    }
    set approvalRequired(value) {
        this._approvalRequired = value;
    }
    resetApprovalRequired() {
        this._approvalRequired = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get approvalRequiredInput() {
        return this._approvalRequired;
    }
}
exports.CloudbuildTriggerApprovalConfigOutputReference = CloudbuildTriggerApprovalConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerApprovalConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerApprovalConfigOutputReference", version: "4.0.3" };
function cloudbuildTriggerBuildArtifactsObjectsTimingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudbuildTriggerBuildArtifactsObjectsTimingToTerraform = cloudbuildTriggerBuildArtifactsObjectsTimingToTerraform;
class CloudbuildTriggerBuildArtifactsObjectsTimingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // end_time - computed: true, optional: false, required: false
    get endTime() {
        return this.getStringAttribute('end_time');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
}
exports.CloudbuildTriggerBuildArtifactsObjectsTimingOutputReference = CloudbuildTriggerBuildArtifactsObjectsTimingOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildArtifactsObjectsTimingOutputReference[_b] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildArtifactsObjectsTimingOutputReference", version: "4.0.3" };
class CloudbuildTriggerBuildArtifactsObjectsTimingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudbuildTriggerBuildArtifactsObjectsTimingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudbuildTriggerBuildArtifactsObjectsTimingList = CloudbuildTriggerBuildArtifactsObjectsTimingList;
_c = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildArtifactsObjectsTimingList[_c] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildArtifactsObjectsTimingList", version: "4.0.3" };
function cloudbuildTriggerBuildArtifactsObjectsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        location: cdktf.stringToTerraform(struct.location),
        paths: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.paths),
    };
}
exports.cloudbuildTriggerBuildArtifactsObjectsToTerraform = cloudbuildTriggerBuildArtifactsObjectsToTerraform;
class CloudbuildTriggerBuildArtifactsObjectsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // timing - computed: true, optional: false, required: false
        this._timing = new CloudbuildTriggerBuildArtifactsObjectsTimingList(this, "timing", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._location !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._paths !== undefined) {
            hasAnyValues = true;
            internalValueResult.paths = this._paths;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._location = undefined;
            this._paths = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._location = value.location;
            this._paths = value.paths;
        }
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get paths() {
        return this.getListAttribute('paths');
    }
    set paths(value) {
        this._paths = value;
    }
    resetPaths() {
        this._paths = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathsInput() {
        return this._paths;
    }
    get timing() {
        return this._timing;
    }
}
exports.CloudbuildTriggerBuildArtifactsObjectsOutputReference = CloudbuildTriggerBuildArtifactsObjectsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildArtifactsObjectsOutputReference[_d] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildArtifactsObjectsOutputReference", version: "4.0.3" };
function cloudbuildTriggerBuildArtifactsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        images: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.images),
        objects: cloudbuildTriggerBuildArtifactsObjectsToTerraform(struct.objects),
    };
}
exports.cloudbuildTriggerBuildArtifactsToTerraform = cloudbuildTriggerBuildArtifactsToTerraform;
class CloudbuildTriggerBuildArtifactsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // objects - computed: false, optional: true, required: false
        this._objects = new CloudbuildTriggerBuildArtifactsObjectsOutputReference(this, "objects");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._images !== undefined) {
            hasAnyValues = true;
            internalValueResult.images = this._images;
        }
        if (this._objects?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.objects = this._objects?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._images = undefined;
            this._objects.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._images = value.images;
            this._objects.internalValue = value.objects;
        }
    }
    get images() {
        return this.getListAttribute('images');
    }
    set images(value) {
        this._images = value;
    }
    resetImages() {
        this._images = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imagesInput() {
        return this._images;
    }
    get objects() {
        return this._objects;
    }
    putObjects(value) {
        this._objects.internalValue = value;
    }
    resetObjects() {
        this._objects.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get objectsInput() {
        return this._objects.internalValue;
    }
}
exports.CloudbuildTriggerBuildArtifactsOutputReference = CloudbuildTriggerBuildArtifactsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildArtifactsOutputReference[_e] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildArtifactsOutputReference", version: "4.0.3" };
function cloudbuildTriggerBuildAvailableSecretsSecretManagerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        env: cdktf.stringToTerraform(struct.env),
        version_name: cdktf.stringToTerraform(struct.versionName),
    };
}
exports.cloudbuildTriggerBuildAvailableSecretsSecretManagerToTerraform = cloudbuildTriggerBuildAvailableSecretsSecretManagerToTerraform;
class CloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._env !== undefined) {
            hasAnyValues = true;
            internalValueResult.env = this._env;
        }
        if (this._versionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.versionName = this._versionName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._env = undefined;
            this._versionName = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._env = value.env;
            this._versionName = value.versionName;
        }
    }
    get env() {
        return this.getStringAttribute('env');
    }
    set env(value) {
        this._env = value;
    }
    // Temporarily expose input value. Use with caution.
    get envInput() {
        return this._env;
    }
    get versionName() {
        return this.getStringAttribute('version_name');
    }
    set versionName(value) {
        this._versionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionNameInput() {
        return this._versionName;
    }
}
exports.CloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference = CloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference;
_f = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference[_f] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference", version: "4.0.3" };
class CloudbuildTriggerBuildAvailableSecretsSecretManagerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudbuildTriggerBuildAvailableSecretsSecretManagerList = CloudbuildTriggerBuildAvailableSecretsSecretManagerList;
_g = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildAvailableSecretsSecretManagerList[_g] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildAvailableSecretsSecretManagerList", version: "4.0.3" };
function cloudbuildTriggerBuildAvailableSecretsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_manager: cdktf.listMapper(cloudbuildTriggerBuildAvailableSecretsSecretManagerToTerraform, true)(struct.secretManager),
    };
}
exports.cloudbuildTriggerBuildAvailableSecretsToTerraform = cloudbuildTriggerBuildAvailableSecretsToTerraform;
class CloudbuildTriggerBuildAvailableSecretsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // secret_manager - computed: false, optional: false, required: true
        this._secretManager = new CloudbuildTriggerBuildAvailableSecretsSecretManagerList(this, "secret_manager", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretManager?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretManager = this._secretManager?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretManager.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretManager.internalValue = value.secretManager;
        }
    }
    get secretManager() {
        return this._secretManager;
    }
    putSecretManager(value) {
        this._secretManager.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretManagerInput() {
        return this._secretManager.internalValue;
    }
}
exports.CloudbuildTriggerBuildAvailableSecretsOutputReference = CloudbuildTriggerBuildAvailableSecretsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildAvailableSecretsOutputReference[_h] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildAvailableSecretsOutputReference", version: "4.0.3" };
function cloudbuildTriggerBuildOptionsVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.cloudbuildTriggerBuildOptionsVolumesToTerraform = cloudbuildTriggerBuildOptionsVolumesToTerraform;
class CloudbuildTriggerBuildOptionsVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._path = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._path = value.path;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.CloudbuildTriggerBuildOptionsVolumesOutputReference = CloudbuildTriggerBuildOptionsVolumesOutputReference;
_j = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildOptionsVolumesOutputReference[_j] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildOptionsVolumesOutputReference", version: "4.0.3" };
class CloudbuildTriggerBuildOptionsVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudbuildTriggerBuildOptionsVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudbuildTriggerBuildOptionsVolumesList = CloudbuildTriggerBuildOptionsVolumesList;
_k = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildOptionsVolumesList[_k] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildOptionsVolumesList", version: "4.0.3" };
function cloudbuildTriggerBuildOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_size_gb: cdktf.numberToTerraform(struct.diskSizeGb),
        dynamic_substitutions: cdktf.booleanToTerraform(struct.dynamicSubstitutions),
        env: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.env),
        log_streaming_option: cdktf.stringToTerraform(struct.logStreamingOption),
        logging: cdktf.stringToTerraform(struct.logging),
        machine_type: cdktf.stringToTerraform(struct.machineType),
        requested_verify_option: cdktf.stringToTerraform(struct.requestedVerifyOption),
        secret_env: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.secretEnv),
        source_provenance_hash: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.sourceProvenanceHash),
        substitution_option: cdktf.stringToTerraform(struct.substitutionOption),
        worker_pool: cdktf.stringToTerraform(struct.workerPool),
        volumes: cdktf.listMapper(cloudbuildTriggerBuildOptionsVolumesToTerraform, true)(struct.volumes),
    };
}
exports.cloudbuildTriggerBuildOptionsToTerraform = cloudbuildTriggerBuildOptionsToTerraform;
class CloudbuildTriggerBuildOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // volumes - computed: false, optional: true, required: false
        this._volumes = new CloudbuildTriggerBuildOptionsVolumesList(this, "volumes", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskSizeGb !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSizeGb = this._diskSizeGb;
        }
        if (this._dynamicSubstitutions !== undefined) {
            hasAnyValues = true;
            internalValueResult.dynamicSubstitutions = this._dynamicSubstitutions;
        }
        if (this._env !== undefined) {
            hasAnyValues = true;
            internalValueResult.env = this._env;
        }
        if (this._logStreamingOption !== undefined) {
            hasAnyValues = true;
            internalValueResult.logStreamingOption = this._logStreamingOption;
        }
        if (this._logging !== undefined) {
            hasAnyValues = true;
            internalValueResult.logging = this._logging;
        }
        if (this._machineType !== undefined) {
            hasAnyValues = true;
            internalValueResult.machineType = this._machineType;
        }
        if (this._requestedVerifyOption !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestedVerifyOption = this._requestedVerifyOption;
        }
        if (this._secretEnv !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretEnv = this._secretEnv;
        }
        if (this._sourceProvenanceHash !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceProvenanceHash = this._sourceProvenanceHash;
        }
        if (this._substitutionOption !== undefined) {
            hasAnyValues = true;
            internalValueResult.substitutionOption = this._substitutionOption;
        }
        if (this._workerPool !== undefined) {
            hasAnyValues = true;
            internalValueResult.workerPool = this._workerPool;
        }
        if (this._volumes?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumes = this._volumes?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskSizeGb = undefined;
            this._dynamicSubstitutions = undefined;
            this._env = undefined;
            this._logStreamingOption = undefined;
            this._logging = undefined;
            this._machineType = undefined;
            this._requestedVerifyOption = undefined;
            this._secretEnv = undefined;
            this._sourceProvenanceHash = undefined;
            this._substitutionOption = undefined;
            this._workerPool = undefined;
            this._volumes.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskSizeGb = value.diskSizeGb;
            this._dynamicSubstitutions = value.dynamicSubstitutions;
            this._env = value.env;
            this._logStreamingOption = value.logStreamingOption;
            this._logging = value.logging;
            this._machineType = value.machineType;
            this._requestedVerifyOption = value.requestedVerifyOption;
            this._secretEnv = value.secretEnv;
            this._sourceProvenanceHash = value.sourceProvenanceHash;
            this._substitutionOption = value.substitutionOption;
            this._workerPool = value.workerPool;
            this._volumes.internalValue = value.volumes;
        }
    }
    get diskSizeGb() {
        return this.getNumberAttribute('disk_size_gb');
    }
    set diskSizeGb(value) {
        this._diskSizeGb = value;
    }
    resetDiskSizeGb() {
        this._diskSizeGb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeGbInput() {
        return this._diskSizeGb;
    }
    get dynamicSubstitutions() {
        return this.getBooleanAttribute('dynamic_substitutions');
    }
    set dynamicSubstitutions(value) {
        this._dynamicSubstitutions = value;
    }
    resetDynamicSubstitutions() {
        this._dynamicSubstitutions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dynamicSubstitutionsInput() {
        return this._dynamicSubstitutions;
    }
    get env() {
        return this.getListAttribute('env');
    }
    set env(value) {
        this._env = value;
    }
    resetEnv() {
        this._env = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get envInput() {
        return this._env;
    }
    get logStreamingOption() {
        return this.getStringAttribute('log_streaming_option');
    }
    set logStreamingOption(value) {
        this._logStreamingOption = value;
    }
    resetLogStreamingOption() {
        this._logStreamingOption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logStreamingOptionInput() {
        return this._logStreamingOption;
    }
    get logging() {
        return this.getStringAttribute('logging');
    }
    set logging(value) {
        this._logging = value;
    }
    resetLogging() {
        this._logging = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loggingInput() {
        return this._logging;
    }
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    set machineType(value) {
        this._machineType = value;
    }
    resetMachineType() {
        this._machineType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get machineTypeInput() {
        return this._machineType;
    }
    get requestedVerifyOption() {
        return this.getStringAttribute('requested_verify_option');
    }
    set requestedVerifyOption(value) {
        this._requestedVerifyOption = value;
    }
    resetRequestedVerifyOption() {
        this._requestedVerifyOption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestedVerifyOptionInput() {
        return this._requestedVerifyOption;
    }
    get secretEnv() {
        return this.getListAttribute('secret_env');
    }
    set secretEnv(value) {
        this._secretEnv = value;
    }
    resetSecretEnv() {
        this._secretEnv = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretEnvInput() {
        return this._secretEnv;
    }
    get sourceProvenanceHash() {
        return this.getListAttribute('source_provenance_hash');
    }
    set sourceProvenanceHash(value) {
        this._sourceProvenanceHash = value;
    }
    resetSourceProvenanceHash() {
        this._sourceProvenanceHash = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceProvenanceHashInput() {
        return this._sourceProvenanceHash;
    }
    get substitutionOption() {
        return this.getStringAttribute('substitution_option');
    }
    set substitutionOption(value) {
        this._substitutionOption = value;
    }
    resetSubstitutionOption() {
        this._substitutionOption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get substitutionOptionInput() {
        return this._substitutionOption;
    }
    get workerPool() {
        return this.getStringAttribute('worker_pool');
    }
    set workerPool(value) {
        this._workerPool = value;
    }
    resetWorkerPool() {
        this._workerPool = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workerPoolInput() {
        return this._workerPool;
    }
    get volumes() {
        return this._volumes;
    }
    putVolumes(value) {
        this._volumes.internalValue = value;
    }
    resetVolumes() {
        this._volumes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesInput() {
        return this._volumes.internalValue;
    }
}
exports.CloudbuildTriggerBuildOptionsOutputReference = CloudbuildTriggerBuildOptionsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildOptionsOutputReference[_l] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildOptionsOutputReference", version: "4.0.3" };
function cloudbuildTriggerBuildSecretToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_name: cdktf.stringToTerraform(struct.kmsKeyName),
        secret_env: cdktf.hashMapper(cdktf.stringToTerraform)(struct.secretEnv),
    };
}
exports.cloudbuildTriggerBuildSecretToTerraform = cloudbuildTriggerBuildSecretToTerraform;
class CloudbuildTriggerBuildSecretOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyName = this._kmsKeyName;
        }
        if (this._secretEnv !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretEnv = this._secretEnv;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._kmsKeyName = undefined;
            this._secretEnv = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._kmsKeyName = value.kmsKeyName;
            this._secretEnv = value.secretEnv;
        }
    }
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
    set kmsKeyName(value) {
        this._kmsKeyName = value;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyNameInput() {
        return this._kmsKeyName;
    }
    get secretEnv() {
        return this.getStringMapAttribute('secret_env');
    }
    set secretEnv(value) {
        this._secretEnv = value;
    }
    resetSecretEnv() {
        this._secretEnv = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretEnvInput() {
        return this._secretEnv;
    }
}
exports.CloudbuildTriggerBuildSecretOutputReference = CloudbuildTriggerBuildSecretOutputReference;
_m = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildSecretOutputReference[_m] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildSecretOutputReference", version: "4.0.3" };
class CloudbuildTriggerBuildSecretList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudbuildTriggerBuildSecretOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudbuildTriggerBuildSecretList = CloudbuildTriggerBuildSecretList;
_o = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildSecretList[_o] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildSecretList", version: "4.0.3" };
function cloudbuildTriggerBuildSourceRepoSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        branch_name: cdktf.stringToTerraform(struct.branchName),
        commit_sha: cdktf.stringToTerraform(struct.commitSha),
        dir: cdktf.stringToTerraform(struct.dir),
        invert_regex: cdktf.booleanToTerraform(struct.invertRegex),
        project_id: cdktf.stringToTerraform(struct.projectId),
        repo_name: cdktf.stringToTerraform(struct.repoName),
        substitutions: cdktf.hashMapper(cdktf.stringToTerraform)(struct.substitutions),
        tag_name: cdktf.stringToTerraform(struct.tagName),
    };
}
exports.cloudbuildTriggerBuildSourceRepoSourceToTerraform = cloudbuildTriggerBuildSourceRepoSourceToTerraform;
class CloudbuildTriggerBuildSourceRepoSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._branchName !== undefined) {
            hasAnyValues = true;
            internalValueResult.branchName = this._branchName;
        }
        if (this._commitSha !== undefined) {
            hasAnyValues = true;
            internalValueResult.commitSha = this._commitSha;
        }
        if (this._dir !== undefined) {
            hasAnyValues = true;
            internalValueResult.dir = this._dir;
        }
        if (this._invertRegex !== undefined) {
            hasAnyValues = true;
            internalValueResult.invertRegex = this._invertRegex;
        }
        if (this._projectId !== undefined) {
            hasAnyValues = true;
            internalValueResult.projectId = this._projectId;
        }
        if (this._repoName !== undefined) {
            hasAnyValues = true;
            internalValueResult.repoName = this._repoName;
        }
        if (this._substitutions !== undefined) {
            hasAnyValues = true;
            internalValueResult.substitutions = this._substitutions;
        }
        if (this._tagName !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagName = this._tagName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._branchName = undefined;
            this._commitSha = undefined;
            this._dir = undefined;
            this._invertRegex = undefined;
            this._projectId = undefined;
            this._repoName = undefined;
            this._substitutions = undefined;
            this._tagName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._branchName = value.branchName;
            this._commitSha = value.commitSha;
            this._dir = value.dir;
            this._invertRegex = value.invertRegex;
            this._projectId = value.projectId;
            this._repoName = value.repoName;
            this._substitutions = value.substitutions;
            this._tagName = value.tagName;
        }
    }
    get branchName() {
        return this.getStringAttribute('branch_name');
    }
    set branchName(value) {
        this._branchName = value;
    }
    resetBranchName() {
        this._branchName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get branchNameInput() {
        return this._branchName;
    }
    get commitSha() {
        return this.getStringAttribute('commit_sha');
    }
    set commitSha(value) {
        this._commitSha = value;
    }
    resetCommitSha() {
        this._commitSha = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commitShaInput() {
        return this._commitSha;
    }
    get dir() {
        return this.getStringAttribute('dir');
    }
    set dir(value) {
        this._dir = value;
    }
    resetDir() {
        this._dir = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dirInput() {
        return this._dir;
    }
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    set invertRegex(value) {
        this._invertRegex = value;
    }
    resetInvertRegex() {
        this._invertRegex = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invertRegexInput() {
        return this._invertRegex;
    }
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    set projectId(value) {
        this._projectId = value;
    }
    resetProjectId() {
        this._projectId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectIdInput() {
        return this._projectId;
    }
    get repoName() {
        return this.getStringAttribute('repo_name');
    }
    set repoName(value) {
        this._repoName = value;
    }
    // Temporarily expose input value. Use with caution.
    get repoNameInput() {
        return this._repoName;
    }
    get substitutions() {
        return this.getStringMapAttribute('substitutions');
    }
    set substitutions(value) {
        this._substitutions = value;
    }
    resetSubstitutions() {
        this._substitutions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get substitutionsInput() {
        return this._substitutions;
    }
    get tagName() {
        return this.getStringAttribute('tag_name');
    }
    set tagName(value) {
        this._tagName = value;
    }
    resetTagName() {
        this._tagName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagNameInput() {
        return this._tagName;
    }
}
exports.CloudbuildTriggerBuildSourceRepoSourceOutputReference = CloudbuildTriggerBuildSourceRepoSourceOutputReference;
_p = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildSourceRepoSourceOutputReference[_p] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildSourceRepoSourceOutputReference", version: "4.0.3" };
function cloudbuildTriggerBuildSourceStorageSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        generation: cdktf.stringToTerraform(struct.generation),
        object: cdktf.stringToTerraform(struct.object),
    };
}
exports.cloudbuildTriggerBuildSourceStorageSourceToTerraform = cloudbuildTriggerBuildSourceStorageSourceToTerraform;
class CloudbuildTriggerBuildSourceStorageSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._generation !== undefined) {
            hasAnyValues = true;
            internalValueResult.generation = this._generation;
        }
        if (this._object !== undefined) {
            hasAnyValues = true;
            internalValueResult.object = this._object;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._generation = undefined;
            this._object = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._generation = value.generation;
            this._object = value.object;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get generation() {
        return this.getStringAttribute('generation');
    }
    set generation(value) {
        this._generation = value;
    }
    resetGeneration() {
        this._generation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get generationInput() {
        return this._generation;
    }
    get object() {
        return this.getStringAttribute('object');
    }
    set object(value) {
        this._object = value;
    }
    // Temporarily expose input value. Use with caution.
    get objectInput() {
        return this._object;
    }
}
exports.CloudbuildTriggerBuildSourceStorageSourceOutputReference = CloudbuildTriggerBuildSourceStorageSourceOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildSourceStorageSourceOutputReference[_q] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildSourceStorageSourceOutputReference", version: "4.0.3" };
function cloudbuildTriggerBuildSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repo_source: cloudbuildTriggerBuildSourceRepoSourceToTerraform(struct.repoSource),
        storage_source: cloudbuildTriggerBuildSourceStorageSourceToTerraform(struct.storageSource),
    };
}
exports.cloudbuildTriggerBuildSourceToTerraform = cloudbuildTriggerBuildSourceToTerraform;
class CloudbuildTriggerBuildSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // repo_source - computed: false, optional: true, required: false
        this._repoSource = new CloudbuildTriggerBuildSourceRepoSourceOutputReference(this, "repo_source");
        // storage_source - computed: false, optional: true, required: false
        this._storageSource = new CloudbuildTriggerBuildSourceStorageSourceOutputReference(this, "storage_source");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repoSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.repoSource = this._repoSource?.internalValue;
        }
        if (this._storageSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageSource = this._storageSource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repoSource.internalValue = undefined;
            this._storageSource.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repoSource.internalValue = value.repoSource;
            this._storageSource.internalValue = value.storageSource;
        }
    }
    get repoSource() {
        return this._repoSource;
    }
    putRepoSource(value) {
        this._repoSource.internalValue = value;
    }
    resetRepoSource() {
        this._repoSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repoSourceInput() {
        return this._repoSource.internalValue;
    }
    get storageSource() {
        return this._storageSource;
    }
    putStorageSource(value) {
        this._storageSource.internalValue = value;
    }
    resetStorageSource() {
        this._storageSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageSourceInput() {
        return this._storageSource.internalValue;
    }
}
exports.CloudbuildTriggerBuildSourceOutputReference = CloudbuildTriggerBuildSourceOutputReference;
_r = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildSourceOutputReference[_r] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildSourceOutputReference", version: "4.0.3" };
function cloudbuildTriggerBuildStepVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.cloudbuildTriggerBuildStepVolumesToTerraform = cloudbuildTriggerBuildStepVolumesToTerraform;
class CloudbuildTriggerBuildStepVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._path = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._path = value.path;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.CloudbuildTriggerBuildStepVolumesOutputReference = CloudbuildTriggerBuildStepVolumesOutputReference;
_s = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildStepVolumesOutputReference[_s] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildStepVolumesOutputReference", version: "4.0.3" };
class CloudbuildTriggerBuildStepVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudbuildTriggerBuildStepVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudbuildTriggerBuildStepVolumesList = CloudbuildTriggerBuildStepVolumesList;
_t = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildStepVolumesList[_t] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildStepVolumesList", version: "4.0.3" };
function cloudbuildTriggerBuildStepToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        args: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.args),
        dir: cdktf.stringToTerraform(struct.dir),
        entrypoint: cdktf.stringToTerraform(struct.entrypoint),
        env: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.env),
        id: cdktf.stringToTerraform(struct.id),
        name: cdktf.stringToTerraform(struct.name),
        script: cdktf.stringToTerraform(struct.script),
        secret_env: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.secretEnv),
        timeout: cdktf.stringToTerraform(struct.timeout),
        timing: cdktf.stringToTerraform(struct.timing),
        wait_for: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.waitFor),
        volumes: cdktf.listMapper(cloudbuildTriggerBuildStepVolumesToTerraform, true)(struct.volumes),
    };
}
exports.cloudbuildTriggerBuildStepToTerraform = cloudbuildTriggerBuildStepToTerraform;
class CloudbuildTriggerBuildStepOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // volumes - computed: false, optional: true, required: false
        this._volumes = new CloudbuildTriggerBuildStepVolumesList(this, "volumes", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._args !== undefined) {
            hasAnyValues = true;
            internalValueResult.args = this._args;
        }
        if (this._dir !== undefined) {
            hasAnyValues = true;
            internalValueResult.dir = this._dir;
        }
        if (this._entrypoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.entrypoint = this._entrypoint;
        }
        if (this._env !== undefined) {
            hasAnyValues = true;
            internalValueResult.env = this._env;
        }
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._script !== undefined) {
            hasAnyValues = true;
            internalValueResult.script = this._script;
        }
        if (this._secretEnv !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretEnv = this._secretEnv;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        if (this._timing !== undefined) {
            hasAnyValues = true;
            internalValueResult.timing = this._timing;
        }
        if (this._waitFor !== undefined) {
            hasAnyValues = true;
            internalValueResult.waitFor = this._waitFor;
        }
        if (this._volumes?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumes = this._volumes?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._args = undefined;
            this._dir = undefined;
            this._entrypoint = undefined;
            this._env = undefined;
            this._id = undefined;
            this._name = undefined;
            this._script = undefined;
            this._secretEnv = undefined;
            this._timeout = undefined;
            this._timing = undefined;
            this._waitFor = undefined;
            this._volumes.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._args = value.args;
            this._dir = value.dir;
            this._entrypoint = value.entrypoint;
            this._env = value.env;
            this._id = value.id;
            this._name = value.name;
            this._script = value.script;
            this._secretEnv = value.secretEnv;
            this._timeout = value.timeout;
            this._timing = value.timing;
            this._waitFor = value.waitFor;
            this._volumes.internalValue = value.volumes;
        }
    }
    get args() {
        return this.getListAttribute('args');
    }
    set args(value) {
        this._args = value;
    }
    resetArgs() {
        this._args = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argsInput() {
        return this._args;
    }
    get dir() {
        return this.getStringAttribute('dir');
    }
    set dir(value) {
        this._dir = value;
    }
    resetDir() {
        this._dir = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dirInput() {
        return this._dir;
    }
    get entrypoint() {
        return this.getStringAttribute('entrypoint');
    }
    set entrypoint(value) {
        this._entrypoint = value;
    }
    resetEntrypoint() {
        this._entrypoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get entrypointInput() {
        return this._entrypoint;
    }
    get env() {
        return this.getListAttribute('env');
    }
    set env(value) {
        this._env = value;
    }
    resetEnv() {
        this._env = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get envInput() {
        return this._env;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get script() {
        return this.getStringAttribute('script');
    }
    set script(value) {
        this._script = value;
    }
    resetScript() {
        this._script = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scriptInput() {
        return this._script;
    }
    get secretEnv() {
        return this.getListAttribute('secret_env');
    }
    set secretEnv(value) {
        this._secretEnv = value;
    }
    resetSecretEnv() {
        this._secretEnv = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretEnvInput() {
        return this._secretEnv;
    }
    get timeout() {
        return this.getStringAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get timing() {
        return this.getStringAttribute('timing');
    }
    set timing(value) {
        this._timing = value;
    }
    resetTiming() {
        this._timing = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timingInput() {
        return this._timing;
    }
    get waitFor() {
        return this.getListAttribute('wait_for');
    }
    set waitFor(value) {
        this._waitFor = value;
    }
    resetWaitFor() {
        this._waitFor = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForInput() {
        return this._waitFor;
    }
    get volumes() {
        return this._volumes;
    }
    putVolumes(value) {
        this._volumes.internalValue = value;
    }
    resetVolumes() {
        this._volumes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesInput() {
        return this._volumes.internalValue;
    }
}
exports.CloudbuildTriggerBuildStepOutputReference = CloudbuildTriggerBuildStepOutputReference;
_u = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildStepOutputReference[_u] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildStepOutputReference", version: "4.0.3" };
class CloudbuildTriggerBuildStepList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudbuildTriggerBuildStepOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudbuildTriggerBuildStepList = CloudbuildTriggerBuildStepList;
_v = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildStepList[_v] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildStepList", version: "4.0.3" };
function cloudbuildTriggerBuildToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        images: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.images),
        logs_bucket: cdktf.stringToTerraform(struct.logsBucket),
        queue_ttl: cdktf.stringToTerraform(struct.queueTtl),
        substitutions: cdktf.hashMapper(cdktf.stringToTerraform)(struct.substitutions),
        tags: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tags),
        timeout: cdktf.stringToTerraform(struct.timeout),
        artifacts: cloudbuildTriggerBuildArtifactsToTerraform(struct.artifacts),
        available_secrets: cloudbuildTriggerBuildAvailableSecretsToTerraform(struct.availableSecrets),
        options: cloudbuildTriggerBuildOptionsToTerraform(struct.options),
        secret: cdktf.listMapper(cloudbuildTriggerBuildSecretToTerraform, true)(struct.secret),
        source: cloudbuildTriggerBuildSourceToTerraform(struct.source),
        step: cdktf.listMapper(cloudbuildTriggerBuildStepToTerraform, true)(struct.step),
    };
}
exports.cloudbuildTriggerBuildToTerraform = cloudbuildTriggerBuildToTerraform;
class CloudbuildTriggerBuildOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // artifacts - computed: false, optional: true, required: false
        this._artifacts = new CloudbuildTriggerBuildArtifactsOutputReference(this, "artifacts");
        // available_secrets - computed: false, optional: true, required: false
        this._availableSecrets = new CloudbuildTriggerBuildAvailableSecretsOutputReference(this, "available_secrets");
        // options - computed: false, optional: true, required: false
        this._options = new CloudbuildTriggerBuildOptionsOutputReference(this, "options");
        // secret - computed: false, optional: true, required: false
        this._secret = new CloudbuildTriggerBuildSecretList(this, "secret", false);
        // source - computed: false, optional: true, required: false
        this._source = new CloudbuildTriggerBuildSourceOutputReference(this, "source");
        // step - computed: false, optional: false, required: true
        this._step = new CloudbuildTriggerBuildStepList(this, "step", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._images !== undefined) {
            hasAnyValues = true;
            internalValueResult.images = this._images;
        }
        if (this._logsBucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.logsBucket = this._logsBucket;
        }
        if (this._queueTtl !== undefined) {
            hasAnyValues = true;
            internalValueResult.queueTtl = this._queueTtl;
        }
        if (this._substitutions !== undefined) {
            hasAnyValues = true;
            internalValueResult.substitutions = this._substitutions;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        if (this._artifacts?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.artifacts = this._artifacts?.internalValue;
        }
        if (this._availableSecrets?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.availableSecrets = this._availableSecrets?.internalValue;
        }
        if (this._options?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.options = this._options?.internalValue;
        }
        if (this._secret?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret?.internalValue;
        }
        if (this._source?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.source = this._source?.internalValue;
        }
        if (this._step?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.step = this._step?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._images = undefined;
            this._logsBucket = undefined;
            this._queueTtl = undefined;
            this._substitutions = undefined;
            this._tags = undefined;
            this._timeout = undefined;
            this._artifacts.internalValue = undefined;
            this._availableSecrets.internalValue = undefined;
            this._options.internalValue = undefined;
            this._secret.internalValue = undefined;
            this._source.internalValue = undefined;
            this._step.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._images = value.images;
            this._logsBucket = value.logsBucket;
            this._queueTtl = value.queueTtl;
            this._substitutions = value.substitutions;
            this._tags = value.tags;
            this._timeout = value.timeout;
            this._artifacts.internalValue = value.artifacts;
            this._availableSecrets.internalValue = value.availableSecrets;
            this._options.internalValue = value.options;
            this._secret.internalValue = value.secret;
            this._source.internalValue = value.source;
            this._step.internalValue = value.step;
        }
    }
    get images() {
        return this.getListAttribute('images');
    }
    set images(value) {
        this._images = value;
    }
    resetImages() {
        this._images = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imagesInput() {
        return this._images;
    }
    get logsBucket() {
        return this.getStringAttribute('logs_bucket');
    }
    set logsBucket(value) {
        this._logsBucket = value;
    }
    resetLogsBucket() {
        this._logsBucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logsBucketInput() {
        return this._logsBucket;
    }
    get queueTtl() {
        return this.getStringAttribute('queue_ttl');
    }
    set queueTtl(value) {
        this._queueTtl = value;
    }
    resetQueueTtl() {
        this._queueTtl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queueTtlInput() {
        return this._queueTtl;
    }
    get substitutions() {
        return this.getStringMapAttribute('substitutions');
    }
    set substitutions(value) {
        this._substitutions = value;
    }
    resetSubstitutions() {
        this._substitutions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get substitutionsInput() {
        return this._substitutions;
    }
    get tags() {
        return this.getListAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get timeout() {
        return this.getStringAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get artifacts() {
        return this._artifacts;
    }
    putArtifacts(value) {
        this._artifacts.internalValue = value;
    }
    resetArtifacts() {
        this._artifacts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get artifactsInput() {
        return this._artifacts.internalValue;
    }
    get availableSecrets() {
        return this._availableSecrets;
    }
    putAvailableSecrets(value) {
        this._availableSecrets.internalValue = value;
    }
    resetAvailableSecrets() {
        this._availableSecrets.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availableSecretsInput() {
        return this._availableSecrets.internalValue;
    }
    get options() {
        return this._options;
    }
    putOptions(value) {
        this._options.internalValue = value;
    }
    resetOptions() {
        this._options.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get optionsInput() {
        return this._options.internalValue;
    }
    get secret() {
        return this._secret;
    }
    putSecret(value) {
        this._secret.internalValue = value;
    }
    resetSecret() {
        this._secret.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret.internalValue;
    }
    get source() {
        return this._source;
    }
    putSource(value) {
        this._source.internalValue = value;
    }
    resetSource() {
        this._source.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source.internalValue;
    }
    get step() {
        return this._step;
    }
    putStep(value) {
        this._step.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get stepInput() {
        return this._step.internalValue;
    }
}
exports.CloudbuildTriggerBuildOutputReference = CloudbuildTriggerBuildOutputReference;
_w = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerBuildOutputReference[_w] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerBuildOutputReference", version: "4.0.3" };
function cloudbuildTriggerGitFileSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        repo_type: cdktf.stringToTerraform(struct.repoType),
        revision: cdktf.stringToTerraform(struct.revision),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.cloudbuildTriggerGitFileSourceToTerraform = cloudbuildTriggerGitFileSourceToTerraform;
class CloudbuildTriggerGitFileSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._repoType !== undefined) {
            hasAnyValues = true;
            internalValueResult.repoType = this._repoType;
        }
        if (this._revision !== undefined) {
            hasAnyValues = true;
            internalValueResult.revision = this._revision;
        }
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
            this._repoType = undefined;
            this._revision = undefined;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
            this._repoType = value.repoType;
            this._revision = value.revision;
            this._uri = value.uri;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get repoType() {
        return this.getStringAttribute('repo_type');
    }
    set repoType(value) {
        this._repoType = value;
    }
    // Temporarily expose input value. Use with caution.
    get repoTypeInput() {
        return this._repoType;
    }
    get revision() {
        return this.getStringAttribute('revision');
    }
    set revision(value) {
        this._revision = value;
    }
    resetRevision() {
        this._revision = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get revisionInput() {
        return this._revision;
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    resetUri() {
        this._uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.CloudbuildTriggerGitFileSourceOutputReference = CloudbuildTriggerGitFileSourceOutputReference;
_x = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerGitFileSourceOutputReference[_x] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerGitFileSourceOutputReference", version: "4.0.3" };
function cloudbuildTriggerGithubPullRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        branch: cdktf.stringToTerraform(struct.branch),
        comment_control: cdktf.stringToTerraform(struct.commentControl),
        invert_regex: cdktf.booleanToTerraform(struct.invertRegex),
    };
}
exports.cloudbuildTriggerGithubPullRequestToTerraform = cloudbuildTriggerGithubPullRequestToTerraform;
class CloudbuildTriggerGithubPullRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._branch !== undefined) {
            hasAnyValues = true;
            internalValueResult.branch = this._branch;
        }
        if (this._commentControl !== undefined) {
            hasAnyValues = true;
            internalValueResult.commentControl = this._commentControl;
        }
        if (this._invertRegex !== undefined) {
            hasAnyValues = true;
            internalValueResult.invertRegex = this._invertRegex;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._branch = undefined;
            this._commentControl = undefined;
            this._invertRegex = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._branch = value.branch;
            this._commentControl = value.commentControl;
            this._invertRegex = value.invertRegex;
        }
    }
    get branch() {
        return this.getStringAttribute('branch');
    }
    set branch(value) {
        this._branch = value;
    }
    // Temporarily expose input value. Use with caution.
    get branchInput() {
        return this._branch;
    }
    get commentControl() {
        return this.getStringAttribute('comment_control');
    }
    set commentControl(value) {
        this._commentControl = value;
    }
    resetCommentControl() {
        this._commentControl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commentControlInput() {
        return this._commentControl;
    }
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    set invertRegex(value) {
        this._invertRegex = value;
    }
    resetInvertRegex() {
        this._invertRegex = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invertRegexInput() {
        return this._invertRegex;
    }
}
exports.CloudbuildTriggerGithubPullRequestOutputReference = CloudbuildTriggerGithubPullRequestOutputReference;
_y = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerGithubPullRequestOutputReference[_y] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerGithubPullRequestOutputReference", version: "4.0.3" };
function cloudbuildTriggerGithubPushToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        branch: cdktf.stringToTerraform(struct.branch),
        invert_regex: cdktf.booleanToTerraform(struct.invertRegex),
        tag: cdktf.stringToTerraform(struct.tag),
    };
}
exports.cloudbuildTriggerGithubPushToTerraform = cloudbuildTriggerGithubPushToTerraform;
class CloudbuildTriggerGithubPushOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._branch !== undefined) {
            hasAnyValues = true;
            internalValueResult.branch = this._branch;
        }
        if (this._invertRegex !== undefined) {
            hasAnyValues = true;
            internalValueResult.invertRegex = this._invertRegex;
        }
        if (this._tag !== undefined) {
            hasAnyValues = true;
            internalValueResult.tag = this._tag;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._branch = undefined;
            this._invertRegex = undefined;
            this._tag = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._branch = value.branch;
            this._invertRegex = value.invertRegex;
            this._tag = value.tag;
        }
    }
    get branch() {
        return this.getStringAttribute('branch');
    }
    set branch(value) {
        this._branch = value;
    }
    resetBranch() {
        this._branch = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get branchInput() {
        return this._branch;
    }
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    set invertRegex(value) {
        this._invertRegex = value;
    }
    resetInvertRegex() {
        this._invertRegex = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invertRegexInput() {
        return this._invertRegex;
    }
    get tag() {
        return this.getStringAttribute('tag');
    }
    set tag(value) {
        this._tag = value;
    }
    resetTag() {
        this._tag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagInput() {
        return this._tag;
    }
}
exports.CloudbuildTriggerGithubPushOutputReference = CloudbuildTriggerGithubPushOutputReference;
_z = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerGithubPushOutputReference[_z] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerGithubPushOutputReference", version: "4.0.3" };
function cloudbuildTriggerGithubToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        owner: cdktf.stringToTerraform(struct.owner),
        pull_request: cloudbuildTriggerGithubPullRequestToTerraform(struct.pullRequest),
        push: cloudbuildTriggerGithubPushToTerraform(struct.push),
    };
}
exports.cloudbuildTriggerGithubToTerraform = cloudbuildTriggerGithubToTerraform;
class CloudbuildTriggerGithubOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // pull_request - computed: false, optional: true, required: false
        this._pullRequest = new CloudbuildTriggerGithubPullRequestOutputReference(this, "pull_request");
        // push - computed: false, optional: true, required: false
        this._push = new CloudbuildTriggerGithubPushOutputReference(this, "push");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._owner !== undefined) {
            hasAnyValues = true;
            internalValueResult.owner = this._owner;
        }
        if (this._pullRequest?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.pullRequest = this._pullRequest?.internalValue;
        }
        if (this._push?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.push = this._push?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._owner = undefined;
            this._pullRequest.internalValue = undefined;
            this._push.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._owner = value.owner;
            this._pullRequest.internalValue = value.pullRequest;
            this._push.internalValue = value.push;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get owner() {
        return this.getStringAttribute('owner');
    }
    set owner(value) {
        this._owner = value;
    }
    resetOwner() {
        this._owner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ownerInput() {
        return this._owner;
    }
    get pullRequest() {
        return this._pullRequest;
    }
    putPullRequest(value) {
        this._pullRequest.internalValue = value;
    }
    resetPullRequest() {
        this._pullRequest.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pullRequestInput() {
        return this._pullRequest.internalValue;
    }
    get push() {
        return this._push;
    }
    putPush(value) {
        this._push.internalValue = value;
    }
    resetPush() {
        this._push.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pushInput() {
        return this._push.internalValue;
    }
}
exports.CloudbuildTriggerGithubOutputReference = CloudbuildTriggerGithubOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerGithubOutputReference[_0] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerGithubOutputReference", version: "4.0.3" };
function cloudbuildTriggerPubsubConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        service_account_email: cdktf.stringToTerraform(struct.serviceAccountEmail),
        topic: cdktf.stringToTerraform(struct.topic),
    };
}
exports.cloudbuildTriggerPubsubConfigToTerraform = cloudbuildTriggerPubsubConfigToTerraform;
class CloudbuildTriggerPubsubConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._serviceAccountEmail !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccountEmail = this._serviceAccountEmail;
        }
        if (this._topic !== undefined) {
            hasAnyValues = true;
            internalValueResult.topic = this._topic;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._serviceAccountEmail = undefined;
            this._topic = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._serviceAccountEmail = value.serviceAccountEmail;
            this._topic = value.topic;
        }
    }
    get serviceAccountEmail() {
        return this.getStringAttribute('service_account_email');
    }
    set serviceAccountEmail(value) {
        this._serviceAccountEmail = value;
    }
    resetServiceAccountEmail() {
        this._serviceAccountEmail = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountEmailInput() {
        return this._serviceAccountEmail;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // subscription - computed: true, optional: false, required: false
    get subscription() {
        return this.getStringAttribute('subscription');
    }
    get topic() {
        return this.getStringAttribute('topic');
    }
    set topic(value) {
        this._topic = value;
    }
    // Temporarily expose input value. Use with caution.
    get topicInput() {
        return this._topic;
    }
}
exports.CloudbuildTriggerPubsubConfigOutputReference = CloudbuildTriggerPubsubConfigOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerPubsubConfigOutputReference[_1] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerPubsubConfigOutputReference", version: "4.0.3" };
function cloudbuildTriggerSourceToBuildToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ref: cdktf.stringToTerraform(struct.ref),
        repo_type: cdktf.stringToTerraform(struct.repoType),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.cloudbuildTriggerSourceToBuildToTerraform = cloudbuildTriggerSourceToBuildToTerraform;
class CloudbuildTriggerSourceToBuildOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ref !== undefined) {
            hasAnyValues = true;
            internalValueResult.ref = this._ref;
        }
        if (this._repoType !== undefined) {
            hasAnyValues = true;
            internalValueResult.repoType = this._repoType;
        }
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ref = undefined;
            this._repoType = undefined;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ref = value.ref;
            this._repoType = value.repoType;
            this._uri = value.uri;
        }
    }
    get ref() {
        return this.getStringAttribute('ref');
    }
    set ref(value) {
        this._ref = value;
    }
    // Temporarily expose input value. Use with caution.
    get refInput() {
        return this._ref;
    }
    get repoType() {
        return this.getStringAttribute('repo_type');
    }
    set repoType(value) {
        this._repoType = value;
    }
    // Temporarily expose input value. Use with caution.
    get repoTypeInput() {
        return this._repoType;
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.CloudbuildTriggerSourceToBuildOutputReference = CloudbuildTriggerSourceToBuildOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerSourceToBuildOutputReference[_2] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerSourceToBuildOutputReference", version: "4.0.3" };
function cloudbuildTriggerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudbuildTriggerTimeoutsToTerraform = cloudbuildTriggerTimeoutsToTerraform;
class CloudbuildTriggerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CloudbuildTriggerTimeoutsOutputReference = CloudbuildTriggerTimeoutsOutputReference;
_3 = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerTimeoutsOutputReference[_3] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerTimeoutsOutputReference", version: "4.0.3" };
function cloudbuildTriggerTriggerTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        branch_name: cdktf.stringToTerraform(struct.branchName),
        commit_sha: cdktf.stringToTerraform(struct.commitSha),
        dir: cdktf.stringToTerraform(struct.dir),
        invert_regex: cdktf.booleanToTerraform(struct.invertRegex),
        project_id: cdktf.stringToTerraform(struct.projectId),
        repo_name: cdktf.stringToTerraform(struct.repoName),
        tag_name: cdktf.stringToTerraform(struct.tagName),
    };
}
exports.cloudbuildTriggerTriggerTemplateToTerraform = cloudbuildTriggerTriggerTemplateToTerraform;
class CloudbuildTriggerTriggerTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._branchName !== undefined) {
            hasAnyValues = true;
            internalValueResult.branchName = this._branchName;
        }
        if (this._commitSha !== undefined) {
            hasAnyValues = true;
            internalValueResult.commitSha = this._commitSha;
        }
        if (this._dir !== undefined) {
            hasAnyValues = true;
            internalValueResult.dir = this._dir;
        }
        if (this._invertRegex !== undefined) {
            hasAnyValues = true;
            internalValueResult.invertRegex = this._invertRegex;
        }
        if (this._projectId !== undefined) {
            hasAnyValues = true;
            internalValueResult.projectId = this._projectId;
        }
        if (this._repoName !== undefined) {
            hasAnyValues = true;
            internalValueResult.repoName = this._repoName;
        }
        if (this._tagName !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagName = this._tagName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._branchName = undefined;
            this._commitSha = undefined;
            this._dir = undefined;
            this._invertRegex = undefined;
            this._projectId = undefined;
            this._repoName = undefined;
            this._tagName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._branchName = value.branchName;
            this._commitSha = value.commitSha;
            this._dir = value.dir;
            this._invertRegex = value.invertRegex;
            this._projectId = value.projectId;
            this._repoName = value.repoName;
            this._tagName = value.tagName;
        }
    }
    get branchName() {
        return this.getStringAttribute('branch_name');
    }
    set branchName(value) {
        this._branchName = value;
    }
    resetBranchName() {
        this._branchName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get branchNameInput() {
        return this._branchName;
    }
    get commitSha() {
        return this.getStringAttribute('commit_sha');
    }
    set commitSha(value) {
        this._commitSha = value;
    }
    resetCommitSha() {
        this._commitSha = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commitShaInput() {
        return this._commitSha;
    }
    get dir() {
        return this.getStringAttribute('dir');
    }
    set dir(value) {
        this._dir = value;
    }
    resetDir() {
        this._dir = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dirInput() {
        return this._dir;
    }
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    set invertRegex(value) {
        this._invertRegex = value;
    }
    resetInvertRegex() {
        this._invertRegex = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invertRegexInput() {
        return this._invertRegex;
    }
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    set projectId(value) {
        this._projectId = value;
    }
    resetProjectId() {
        this._projectId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectIdInput() {
        return this._projectId;
    }
    get repoName() {
        return this.getStringAttribute('repo_name');
    }
    set repoName(value) {
        this._repoName = value;
    }
    resetRepoName() {
        this._repoName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repoNameInput() {
        return this._repoName;
    }
    get tagName() {
        return this.getStringAttribute('tag_name');
    }
    set tagName(value) {
        this._tagName = value;
    }
    resetTagName() {
        this._tagName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagNameInput() {
        return this._tagName;
    }
}
exports.CloudbuildTriggerTriggerTemplateOutputReference = CloudbuildTriggerTriggerTemplateOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerTriggerTemplateOutputReference[_4] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerTriggerTemplateOutputReference", version: "4.0.3" };
function cloudbuildTriggerWebhookConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret: cdktf.stringToTerraform(struct.secret),
    };
}
exports.cloudbuildTriggerWebhookConfigToTerraform = cloudbuildTriggerWebhookConfigToTerraform;
class CloudbuildTriggerWebhookConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secret !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secret = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secret = value.secret;
        }
    }
    get secret() {
        return this.getStringAttribute('secret');
    }
    set secret(value) {
        this._secret = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.CloudbuildTriggerWebhookConfigOutputReference = CloudbuildTriggerWebhookConfigOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
CloudbuildTriggerWebhookConfigOutputReference[_5] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTriggerWebhookConfigOutputReference", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/cloudbuild_trigger google_cloudbuild_trigger}
*/
class CloudbuildTrigger extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/cloudbuild_trigger google_cloudbuild_trigger} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudbuildTriggerConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_cloudbuild_trigger',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // approval_config - computed: false, optional: true, required: false
        this._approvalConfig = new CloudbuildTriggerApprovalConfigOutputReference(this, "approval_config");
        // build - computed: false, optional: true, required: false
        this._build = new CloudbuildTriggerBuildOutputReference(this, "build");
        // git_file_source - computed: false, optional: true, required: false
        this._gitFileSource = new CloudbuildTriggerGitFileSourceOutputReference(this, "git_file_source");
        // github - computed: false, optional: true, required: false
        this._github = new CloudbuildTriggerGithubOutputReference(this, "github");
        // pubsub_config - computed: false, optional: true, required: false
        this._pubsubConfig = new CloudbuildTriggerPubsubConfigOutputReference(this, "pubsub_config");
        // source_to_build - computed: false, optional: true, required: false
        this._sourceToBuild = new CloudbuildTriggerSourceToBuildOutputReference(this, "source_to_build");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudbuildTriggerTimeoutsOutputReference(this, "timeouts");
        // trigger_template - computed: false, optional: true, required: false
        this._triggerTemplate = new CloudbuildTriggerTriggerTemplateOutputReference(this, "trigger_template");
        // webhook_config - computed: false, optional: true, required: false
        this._webhookConfig = new CloudbuildTriggerWebhookConfigOutputReference(this, "webhook_config");
        this._description = config.description;
        this._disabled = config.disabled;
        this._filename = config.filename;
        this._filter = config.filter;
        this._id = config.id;
        this._ignoredFiles = config.ignoredFiles;
        this._includeBuildLogs = config.includeBuildLogs;
        this._includedFiles = config.includedFiles;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._serviceAccount = config.serviceAccount;
        this._substitutions = config.substitutions;
        this._tags = config.tags;
        this._approvalConfig.internalValue = config.approvalConfig;
        this._build.internalValue = config.buildAttribute;
        this._gitFileSource.internalValue = config.gitFileSource;
        this._github.internalValue = config.github;
        this._pubsubConfig.internalValue = config.pubsubConfig;
        this._sourceToBuild.internalValue = config.sourceToBuild;
        this._timeouts.internalValue = config.timeouts;
        this._triggerTemplate.internalValue = config.triggerTemplate;
        this._webhookConfig.internalValue = config.webhookConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    set disabled(value) {
        this._disabled = value;
    }
    resetDisabled() {
        this._disabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disabledInput() {
        return this._disabled;
    }
    get filename() {
        return this.getStringAttribute('filename');
    }
    set filename(value) {
        this._filename = value;
    }
    resetFilename() {
        this._filename = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filenameInput() {
        return this._filename;
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ignoredFiles() {
        return this.getListAttribute('ignored_files');
    }
    set ignoredFiles(value) {
        this._ignoredFiles = value;
    }
    resetIgnoredFiles() {
        this._ignoredFiles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoredFilesInput() {
        return this._ignoredFiles;
    }
    get includeBuildLogs() {
        return this.getStringAttribute('include_build_logs');
    }
    set includeBuildLogs(value) {
        this._includeBuildLogs = value;
    }
    resetIncludeBuildLogs() {
        this._includeBuildLogs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeBuildLogsInput() {
        return this._includeBuildLogs;
    }
    get includedFiles() {
        return this.getListAttribute('included_files');
    }
    set includedFiles(value) {
        this._includedFiles = value;
    }
    resetIncludedFiles() {
        this._includedFiles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includedFilesInput() {
        return this._includedFiles;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    set serviceAccount(value) {
        this._serviceAccount = value;
    }
    resetServiceAccount() {
        this._serviceAccount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountInput() {
        return this._serviceAccount;
    }
    get substitutions() {
        return this.getStringMapAttribute('substitutions');
    }
    set substitutions(value) {
        this._substitutions = value;
    }
    resetSubstitutions() {
        this._substitutions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get substitutionsInput() {
        return this._substitutions;
    }
    get tags() {
        return this.getListAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // trigger_id - computed: true, optional: false, required: false
    get triggerId() {
        return this.getStringAttribute('trigger_id');
    }
    get approvalConfig() {
        return this._approvalConfig;
    }
    putApprovalConfig(value) {
        this._approvalConfig.internalValue = value;
    }
    resetApprovalConfig() {
        this._approvalConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get approvalConfigInput() {
        return this._approvalConfig.internalValue;
    }
    get buildAttribute() {
        return this._build;
    }
    putBuildAttribute(value) {
        this._build.internalValue = value;
    }
    resetBuildAttribute() {
        this._build.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get buildAttributeInput() {
        return this._build.internalValue;
    }
    get gitFileSource() {
        return this._gitFileSource;
    }
    putGitFileSource(value) {
        this._gitFileSource.internalValue = value;
    }
    resetGitFileSource() {
        this._gitFileSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gitFileSourceInput() {
        return this._gitFileSource.internalValue;
    }
    get github() {
        return this._github;
    }
    putGithub(value) {
        this._github.internalValue = value;
    }
    resetGithub() {
        this._github.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get githubInput() {
        return this._github.internalValue;
    }
    get pubsubConfig() {
        return this._pubsubConfig;
    }
    putPubsubConfig(value) {
        this._pubsubConfig.internalValue = value;
    }
    resetPubsubConfig() {
        this._pubsubConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pubsubConfigInput() {
        return this._pubsubConfig.internalValue;
    }
    get sourceToBuild() {
        return this._sourceToBuild;
    }
    putSourceToBuild(value) {
        this._sourceToBuild.internalValue = value;
    }
    resetSourceToBuild() {
        this._sourceToBuild.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceToBuildInput() {
        return this._sourceToBuild.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get triggerTemplate() {
        return this._triggerTemplate;
    }
    putTriggerTemplate(value) {
        this._triggerTemplate.internalValue = value;
    }
    resetTriggerTemplate() {
        this._triggerTemplate.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get triggerTemplateInput() {
        return this._triggerTemplate.internalValue;
    }
    get webhookConfig() {
        return this._webhookConfig;
    }
    putWebhookConfig(value) {
        this._webhookConfig.internalValue = value;
    }
    resetWebhookConfig() {
        this._webhookConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get webhookConfigInput() {
        return this._webhookConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            disabled: cdktf.booleanToTerraform(this._disabled),
            filename: cdktf.stringToTerraform(this._filename),
            filter: cdktf.stringToTerraform(this._filter),
            id: cdktf.stringToTerraform(this._id),
            ignored_files: cdktf.listMapper(cdktf.stringToTerraform, false)(this._ignoredFiles),
            include_build_logs: cdktf.stringToTerraform(this._includeBuildLogs),
            included_files: cdktf.listMapper(cdktf.stringToTerraform, false)(this._includedFiles),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            service_account: cdktf.stringToTerraform(this._serviceAccount),
            substitutions: cdktf.hashMapper(cdktf.stringToTerraform)(this._substitutions),
            tags: cdktf.listMapper(cdktf.stringToTerraform, false)(this._tags),
            approval_config: cloudbuildTriggerApprovalConfigToTerraform(this._approvalConfig.internalValue),
            build: cloudbuildTriggerBuildToTerraform(this._build.internalValue),
            git_file_source: cloudbuildTriggerGitFileSourceToTerraform(this._gitFileSource.internalValue),
            github: cloudbuildTriggerGithubToTerraform(this._github.internalValue),
            pubsub_config: cloudbuildTriggerPubsubConfigToTerraform(this._pubsubConfig.internalValue),
            source_to_build: cloudbuildTriggerSourceToBuildToTerraform(this._sourceToBuild.internalValue),
            timeouts: cloudbuildTriggerTimeoutsToTerraform(this._timeouts.internalValue),
            trigger_template: cloudbuildTriggerTriggerTemplateToTerraform(this._triggerTemplate.internalValue),
            webhook_config: cloudbuildTriggerWebhookConfigToTerraform(this._webhookConfig.internalValue),
        };
    }
}
exports.CloudbuildTrigger = CloudbuildTrigger;
_6 = JSII_RTTI_SYMBOL_1;
CloudbuildTrigger[_6] = { fqn: "@cdktf/provider-google.cloudbuildTrigger.CloudbuildTrigger", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
CloudbuildTrigger.tfResourceType = "google_cloudbuild_trigger";
