// @cdktf/provider-google/lib/kms-crypto-key-version/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KmsCryptoKeyVersion = exports.KmsCryptoKeyVersionTimeoutsOutputReference = exports.kmsCryptoKeyVersionTimeoutsToTerraform = exports.KmsCryptoKeyVersionAttestationList = exports.KmsCryptoKeyVersionAttestationOutputReference = exports.kmsCryptoKeyVersionAttestationToTerraform = exports.KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsList = exports.KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsOutputReference = exports.kmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsToTerraform = exports.KmsCryptoKeyVersionAttestationCertChainsList = exports.KmsCryptoKeyVersionAttestationCertChainsOutputReference = exports.kmsCryptoKeyVersionAttestationCertChainsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function kmsCryptoKeyVersionAttestationCertChainsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.kmsCryptoKeyVersionAttestationCertChainsToTerraform = kmsCryptoKeyVersionAttestationCertChainsToTerraform;
class KmsCryptoKeyVersionAttestationCertChainsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cavium_certs - computed: true, optional: false, required: false
    get caviumCerts() {
        return this.getStringAttribute('cavium_certs');
    }
    // google_card_certs - computed: true, optional: false, required: false
    get googleCardCerts() {
        return this.getStringAttribute('google_card_certs');
    }
    // google_partition_certs - computed: true, optional: false, required: false
    get googlePartitionCerts() {
        return this.getStringAttribute('google_partition_certs');
    }
}
exports.KmsCryptoKeyVersionAttestationCertChainsOutputReference = KmsCryptoKeyVersionAttestationCertChainsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
KmsCryptoKeyVersionAttestationCertChainsOutputReference[_a] = { fqn: "@cdktf/provider-google.kmsCryptoKeyVersion.KmsCryptoKeyVersionAttestationCertChainsOutputReference", version: "5.0.1" };
class KmsCryptoKeyVersionAttestationCertChainsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KmsCryptoKeyVersionAttestationCertChainsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KmsCryptoKeyVersionAttestationCertChainsList = KmsCryptoKeyVersionAttestationCertChainsList;
_b = JSII_RTTI_SYMBOL_1;
KmsCryptoKeyVersionAttestationCertChainsList[_b] = { fqn: "@cdktf/provider-google.kmsCryptoKeyVersion.KmsCryptoKeyVersionAttestationCertChainsList", version: "5.0.1" };
function kmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.kmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsToTerraform = kmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsToTerraform;
class KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ekm_connection_key_path - computed: true, optional: false, required: false
    get ekmConnectionKeyPath() {
        return this.getStringAttribute('ekm_connection_key_path');
    }
    // external_key_uri - computed: true, optional: false, required: false
    get externalKeyUri() {
        return this.getStringAttribute('external_key_uri');
    }
}
exports.KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsOutputReference = KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsOutputReference[_c] = { fqn: "@cdktf/provider-google.kmsCryptoKeyVersion.KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsOutputReference", version: "5.0.1" };
class KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsList = KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsList;
_d = JSII_RTTI_SYMBOL_1;
KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsList[_d] = { fqn: "@cdktf/provider-google.kmsCryptoKeyVersion.KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsList", version: "5.0.1" };
function kmsCryptoKeyVersionAttestationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.kmsCryptoKeyVersionAttestationToTerraform = kmsCryptoKeyVersionAttestationToTerraform;
class KmsCryptoKeyVersionAttestationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cert_chains - computed: true, optional: false, required: false
        this._certChains = new KmsCryptoKeyVersionAttestationCertChainsList(this, "cert_chains", false);
        // external_protection_level_options - computed: true, optional: false, required: false
        this._externalProtectionLevelOptions = new KmsCryptoKeyVersionAttestationExternalProtectionLevelOptionsList(this, "external_protection_level_options", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get certChains() {
        return this._certChains;
    }
    // content - computed: true, optional: false, required: false
    get content() {
        return this.getStringAttribute('content');
    }
    get externalProtectionLevelOptions() {
        return this._externalProtectionLevelOptions;
    }
    // format - computed: true, optional: false, required: false
    get format() {
        return this.getStringAttribute('format');
    }
}
exports.KmsCryptoKeyVersionAttestationOutputReference = KmsCryptoKeyVersionAttestationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
KmsCryptoKeyVersionAttestationOutputReference[_e] = { fqn: "@cdktf/provider-google.kmsCryptoKeyVersion.KmsCryptoKeyVersionAttestationOutputReference", version: "5.0.1" };
class KmsCryptoKeyVersionAttestationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KmsCryptoKeyVersionAttestationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KmsCryptoKeyVersionAttestationList = KmsCryptoKeyVersionAttestationList;
_f = JSII_RTTI_SYMBOL_1;
KmsCryptoKeyVersionAttestationList[_f] = { fqn: "@cdktf/provider-google.kmsCryptoKeyVersion.KmsCryptoKeyVersionAttestationList", version: "5.0.1" };
function kmsCryptoKeyVersionTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.kmsCryptoKeyVersionTimeoutsToTerraform = kmsCryptoKeyVersionTimeoutsToTerraform;
class KmsCryptoKeyVersionTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.KmsCryptoKeyVersionTimeoutsOutputReference = KmsCryptoKeyVersionTimeoutsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
KmsCryptoKeyVersionTimeoutsOutputReference[_g] = { fqn: "@cdktf/provider-google.kmsCryptoKeyVersion.KmsCryptoKeyVersionTimeoutsOutputReference", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/kms_crypto_key_version google_kms_crypto_key_version}
*/
class KmsCryptoKeyVersion extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/kms_crypto_key_version google_kms_crypto_key_version} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options KmsCryptoKeyVersionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_kms_crypto_key_version',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // attestation - computed: true, optional: false, required: false
        this._attestation = new KmsCryptoKeyVersionAttestationList(this, "attestation", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new KmsCryptoKeyVersionTimeoutsOutputReference(this, "timeouts");
        this._cryptoKey = config.cryptoKey;
        this._id = config.id;
        this._state = config.state;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // algorithm - computed: true, optional: false, required: false
    get algorithm() {
        return this.getStringAttribute('algorithm');
    }
    get attestation() {
        return this._attestation;
    }
    get cryptoKey() {
        return this.getStringAttribute('crypto_key');
    }
    set cryptoKey(value) {
        this._cryptoKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get cryptoKeyInput() {
        return this._cryptoKey;
    }
    // generate_time - computed: true, optional: false, required: false
    get generateTime() {
        return this.getStringAttribute('generate_time');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // protection_level - computed: true, optional: false, required: false
    get protectionLevel() {
        return this.getStringAttribute('protection_level');
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            crypto_key: cdktf.stringToTerraform(this._cryptoKey),
            id: cdktf.stringToTerraform(this._id),
            state: cdktf.stringToTerraform(this._state),
            timeouts: kmsCryptoKeyVersionTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.KmsCryptoKeyVersion = KmsCryptoKeyVersion;
_h = JSII_RTTI_SYMBOL_1;
KmsCryptoKeyVersion[_h] = { fqn: "@cdktf/provider-google.kmsCryptoKeyVersion.KmsCryptoKeyVersion", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
KmsCryptoKeyVersion.tfResourceType = "google_kms_crypto_key_version";
