// @cdktf/provider-google/lib/data-google-cloud-run-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleCloudRunService = exports.DataGoogleCloudRunServiceTrafficList = exports.DataGoogleCloudRunServiceTrafficOutputReference = exports.dataGoogleCloudRunServiceTrafficToTerraform = exports.DataGoogleCloudRunServiceTemplateList = exports.DataGoogleCloudRunServiceTemplateOutputReference = exports.dataGoogleCloudRunServiceTemplateToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecList = exports.DataGoogleCloudRunServiceTemplateSpecOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecVolumesList = exports.DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecVolumesToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretList = exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList = exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersList = exports.DataGoogleCloudRunServiceTemplateSpecContainersOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList = exports.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersResourcesList = exports.DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersResourcesToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersPortsList = exports.DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersPortsToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToTerraform = exports.DataGoogleCloudRunServiceTemplateMetadataList = exports.DataGoogleCloudRunServiceTemplateMetadataOutputReference = exports.dataGoogleCloudRunServiceTemplateMetadataToTerraform = exports.DataGoogleCloudRunServiceStatusList = exports.DataGoogleCloudRunServiceStatusOutputReference = exports.dataGoogleCloudRunServiceStatusToTerraform = exports.DataGoogleCloudRunServiceStatusConditionsList = exports.DataGoogleCloudRunServiceStatusConditionsOutputReference = exports.dataGoogleCloudRunServiceStatusConditionsToTerraform = exports.DataGoogleCloudRunServiceMetadataList = exports.DataGoogleCloudRunServiceMetadataOutputReference = exports.dataGoogleCloudRunServiceMetadataToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleCloudRunServiceMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceMetadataToTerraform = dataGoogleCloudRunServiceMetadataToTerraform;
class DataGoogleCloudRunServiceMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // annotations - computed: true, optional: false, required: false
        this._annotations = new cdktf.StringMap(this, "annotations");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get annotations() {
        return this._annotations;
    }
    // generation - computed: true, optional: false, required: false
    get generation() {
        return this.getNumberAttribute('generation');
    }
    get labels() {
        return this._labels;
    }
    // namespace - computed: true, optional: false, required: false
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    // resource_version - computed: true, optional: false, required: false
    get resourceVersion() {
        return this.getStringAttribute('resource_version');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
}
exports.DataGoogleCloudRunServiceMetadataOutputReference = DataGoogleCloudRunServiceMetadataOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceMetadataOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceMetadataOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceMetadataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceMetadataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceMetadataList = DataGoogleCloudRunServiceMetadataList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceMetadataList[_b] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceMetadataList", version: "5.0.1" };
function dataGoogleCloudRunServiceStatusConditionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceStatusConditionsToTerraform = dataGoogleCloudRunServiceStatusConditionsToTerraform;
class DataGoogleCloudRunServiceStatusConditionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleCloudRunServiceStatusConditionsOutputReference = DataGoogleCloudRunServiceStatusConditionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusConditionsOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusConditionsOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceStatusConditionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceStatusConditionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceStatusConditionsList = DataGoogleCloudRunServiceStatusConditionsList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusConditionsList[_d] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusConditionsList", version: "5.0.1" };
function dataGoogleCloudRunServiceStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceStatusToTerraform = dataGoogleCloudRunServiceStatusToTerraform;
class DataGoogleCloudRunServiceStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // conditions - computed: true, optional: false, required: false
        this._conditions = new DataGoogleCloudRunServiceStatusConditionsList(this, "conditions", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get conditions() {
        return this._conditions;
    }
    // latest_created_revision_name - computed: true, optional: false, required: false
    get latestCreatedRevisionName() {
        return this.getStringAttribute('latest_created_revision_name');
    }
    // latest_ready_revision_name - computed: true, optional: false, required: false
    get latestReadyRevisionName() {
        return this.getStringAttribute('latest_ready_revision_name');
    }
    // observed_generation - computed: true, optional: false, required: false
    get observedGeneration() {
        return this.getNumberAttribute('observed_generation');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
}
exports.DataGoogleCloudRunServiceStatusOutputReference = DataGoogleCloudRunServiceStatusOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceStatusList = DataGoogleCloudRunServiceStatusList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusList[_f] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateMetadataToTerraform = dataGoogleCloudRunServiceTemplateMetadataToTerraform;
class DataGoogleCloudRunServiceTemplateMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // annotations - computed: true, optional: false, required: false
        this._annotations = new cdktf.StringMap(this, "annotations");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get annotations() {
        return this._annotations;
    }
    // generation - computed: true, optional: false, required: false
    get generation() {
        return this.getNumberAttribute('generation');
    }
    get labels() {
        return this._labels;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // namespace - computed: true, optional: false, required: false
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    // resource_version - computed: true, optional: false, required: false
    get resourceVersion() {
        return this.getStringAttribute('resource_version');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
}
exports.DataGoogleCloudRunServiceTemplateMetadataOutputReference = DataGoogleCloudRunServiceTemplateMetadataOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateMetadataOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateMetadataOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateMetadataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateMetadataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateMetadataList = DataGoogleCloudRunServiceTemplateMetadataList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateMetadataList[_h] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateMetadataList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList = DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList[_k] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // secret_key_ref - computed: true, optional: false, required: false
        this._secretKeyRef = new DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList(this, "secret_key_ref", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get secretKeyRef() {
        return this._secretKeyRef;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList = DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList[_m] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // value_from - computed: true, optional: false, required: false
        this._valueFrom = new DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList(this, "value_from", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
    get valueFrom() {
        return this._valueFrom;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvList = DataGoogleCloudRunServiceTemplateSpecContainersEnvList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvList[_p] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList[_r] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // local_object_reference - computed: true, optional: false, required: false
        this._localObjectReference = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList(this, "local_object_reference", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get localObjectReference() {
        return this._localObjectReference;
    }
    // optional - computed: true, optional: false, required: false
    get optional() {
        return this.getBooleanAttribute('optional');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList[_t] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList[_v] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // local_object_reference - computed: true, optional: false, required: false
        this._localObjectReference = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList(this, "local_object_reference", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get localObjectReference() {
        return this._localObjectReference;
    }
    // optional - computed: true, optional: false, required: false
    get optional() {
        return this.getBooleanAttribute('optional');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList[_x] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // config_map_ref - computed: true, optional: false, required: false
        this._configMapRef = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList(this, "config_map_ref", false);
        // secret_ref - computed: true, optional: false, required: false
        this._secretRef = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList(this, "secret_ref", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get configMapRef() {
        return this._configMapRef;
    }
    // prefix - computed: true, optional: false, required: false
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    get secretRef() {
        return this._secretRef;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList[_z] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersPortsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersPortsToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersPortsToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // container_port - computed: true, optional: false, required: false
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersPortsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersPortsList = DataGoogleCloudRunServiceTemplateSpecContainersPortsList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersPortsList[_1] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersPortsList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersResourcesToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersResourcesToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // limits - computed: true, optional: false, required: false
        this._limits = new cdktf.StringMap(this, "limits");
        // requests - computed: true, optional: false, required: false
        this._requests = new cdktf.StringMap(this, "requests");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get limits() {
        return this._limits;
    }
    get requests() {
        return this._requests;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersResourcesList = DataGoogleCloudRunServiceTemplateSpecContainersResourcesList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersResourcesList[_3] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersResourcesList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // mount_path - computed: true, optional: false, required: false
    get mountPath() {
        return this.getStringAttribute('mount_path');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList = DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList[_5] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersToTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // env - computed: true, optional: false, required: false
        this._env = new DataGoogleCloudRunServiceTemplateSpecContainersEnvList(this, "env", true);
        // env_from - computed: true, optional: false, required: false
        this._envFrom = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList(this, "env_from", false);
        // ports - computed: true, optional: false, required: false
        this._ports = new DataGoogleCloudRunServiceTemplateSpecContainersPortsList(this, "ports", false);
        // resources - computed: true, optional: false, required: false
        this._resources = new DataGoogleCloudRunServiceTemplateSpecContainersResourcesList(this, "resources", false);
        // volume_mounts - computed: true, optional: false, required: false
        this._volumeMounts = new DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList(this, "volume_mounts", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // args - computed: true, optional: false, required: false
    get args() {
        return this.getListAttribute('args');
    }
    // command - computed: true, optional: false, required: false
    get command() {
        return this.getListAttribute('command');
    }
    get env() {
        return this._env;
    }
    get envFrom() {
        return this._envFrom;
    }
    // image - computed: true, optional: false, required: false
    get image() {
        return this.getStringAttribute('image');
    }
    get ports() {
        return this._ports;
    }
    get resources() {
        return this._resources;
    }
    get volumeMounts() {
        return this._volumeMounts;
    }
    // working_dir - computed: true, optional: false, required: false
    get workingDir() {
        return this.getStringAttribute('working_dir');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersList = DataGoogleCloudRunServiceTemplateSpecContainersList;
_7 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersList[_7] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToTerraform;
class DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getNumberAttribute('mode');
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference = DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList = DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList;
_9 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList[_9] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecVolumesSecretToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretToTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesSecretToTerraform;
class DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // items - computed: true, optional: false, required: false
        this._items = new DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList(this, "items", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // default_mode - computed: true, optional: false, required: false
    get defaultMode() {
        return this.getNumberAttribute('default_mode');
    }
    get items() {
        return this._items;
    }
    // secret_name - computed: true, optional: false, required: false
    get secretName() {
        return this.getStringAttribute('secret_name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference = DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference[_10] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecVolumesSecretList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretList = DataGoogleCloudRunServiceTemplateSpecVolumesSecretList;
_11 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesSecretList[_11] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesSecretList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesToTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesToTerraform;
class DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // secret - computed: true, optional: false, required: false
        this._secret = new DataGoogleCloudRunServiceTemplateSpecVolumesSecretList(this, "secret", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get secret() {
        return this._secret;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference = DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference[_12] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesList = DataGoogleCloudRunServiceTemplateSpecVolumesList;
_13 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesList[_13] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecToTerraform = dataGoogleCloudRunServiceTemplateSpecToTerraform;
class DataGoogleCloudRunServiceTemplateSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // containers - computed: true, optional: false, required: false
        this._containers = new DataGoogleCloudRunServiceTemplateSpecContainersList(this, "containers", false);
        // volumes - computed: true, optional: false, required: false
        this._volumes = new DataGoogleCloudRunServiceTemplateSpecVolumesList(this, "volumes", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // container_concurrency - computed: true, optional: false, required: false
    get containerConcurrency() {
        return this.getNumberAttribute('container_concurrency');
    }
    get containers() {
        return this._containers;
    }
    // service_account_name - computed: true, optional: false, required: false
    get serviceAccountName() {
        return this.getStringAttribute('service_account_name');
    }
    // serving_state - computed: true, optional: false, required: false
    get servingState() {
        return this.getStringAttribute('serving_state');
    }
    // timeout_seconds - computed: true, optional: false, required: false
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    get volumes() {
        return this._volumes;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecOutputReference = DataGoogleCloudRunServiceTemplateSpecOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecOutputReference[_14] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecList = DataGoogleCloudRunServiceTemplateSpecList;
_15 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecList[_15] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecList", version: "5.0.1" };
function dataGoogleCloudRunServiceTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateToTerraform = dataGoogleCloudRunServiceTemplateToTerraform;
class DataGoogleCloudRunServiceTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // metadata - computed: true, optional: false, required: false
        this._metadata = new DataGoogleCloudRunServiceTemplateMetadataList(this, "metadata", false);
        // spec - computed: true, optional: false, required: false
        this._spec = new DataGoogleCloudRunServiceTemplateSpecList(this, "spec", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get metadata() {
        return this._metadata;
    }
    get spec() {
        return this._spec;
    }
}
exports.DataGoogleCloudRunServiceTemplateOutputReference = DataGoogleCloudRunServiceTemplateOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateOutputReference[_16] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTemplateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateList = DataGoogleCloudRunServiceTemplateList;
_17 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateList[_17] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateList", version: "5.0.1" };
function dataGoogleCloudRunServiceTrafficToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTrafficToTerraform = dataGoogleCloudRunServiceTrafficToTerraform;
class DataGoogleCloudRunServiceTrafficOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // latest_revision - computed: true, optional: false, required: false
    get latestRevision() {
        return this.getBooleanAttribute('latest_revision');
    }
    // percent - computed: true, optional: false, required: false
    get percent() {
        return this.getNumberAttribute('percent');
    }
    // revision_name - computed: true, optional: false, required: false
    get revisionName() {
        return this.getStringAttribute('revision_name');
    }
    // tag - computed: true, optional: false, required: false
    get tag() {
        return this.getStringAttribute('tag');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
}
exports.DataGoogleCloudRunServiceTrafficOutputReference = DataGoogleCloudRunServiceTrafficOutputReference;
_18 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTrafficOutputReference[_18] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTrafficOutputReference", version: "5.0.1" };
class DataGoogleCloudRunServiceTrafficList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTrafficOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTrafficList = DataGoogleCloudRunServiceTrafficList;
_19 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTrafficList[_19] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTrafficList", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/cloud_run_service google_cloud_run_service}
*/
class DataGoogleCloudRunService extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/cloud_run_service google_cloud_run_service} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleCloudRunServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloud_run_service',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // metadata - computed: true, optional: false, required: false
        this._metadata = new DataGoogleCloudRunServiceMetadataList(this, "metadata", false);
        // status - computed: true, optional: false, required: false
        this._status = new DataGoogleCloudRunServiceStatusList(this, "status", false);
        // template - computed: true, optional: false, required: false
        this._template = new DataGoogleCloudRunServiceTemplateList(this, "template", false);
        // traffic - computed: true, optional: false, required: false
        this._traffic = new DataGoogleCloudRunServiceTrafficList(this, "traffic", false);
        this._id = config.id;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // autogenerate_revision_name - computed: true, optional: false, required: false
    get autogenerateRevisionName() {
        return this.getBooleanAttribute('autogenerate_revision_name');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get metadata() {
        return this._metadata;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get status() {
        return this._status;
    }
    get template() {
        return this._template;
    }
    get traffic() {
        return this._traffic;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
}
exports.DataGoogleCloudRunService = DataGoogleCloudRunService;
_20 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunService[_20] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunService", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleCloudRunService.tfResourceType = "google_cloud_run_service";
