// @cdktf/provider-google/lib/cloud-run-v2-job/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudRunV2Job = exports.CloudRunV2JobTimeoutsOutputReference = exports.cloudRunV2JobTimeoutsToTerraform = exports.CloudRunV2JobTemplateOutputReference = exports.cloudRunV2JobTemplateToTerraform = exports.CloudRunV2JobTemplateTemplateOutputReference = exports.cloudRunV2JobTemplateTemplateToTerraform = exports.CloudRunV2JobTemplateTemplateVpcAccessOutputReference = exports.cloudRunV2JobTemplateTemplateVpcAccessToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesList = exports.CloudRunV2JobTemplateTemplateVolumesOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesSecretOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesSecretToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesSecretItemsList = exports.CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform = exports.CloudRunV2JobTemplateTemplateContainersList = exports.CloudRunV2JobTemplateTemplateContainersOutputReference = exports.cloudRunV2JobTemplateTemplateContainersToTerraform = exports.CloudRunV2JobTemplateTemplateContainersVolumeMountsList = exports.CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference = exports.cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform = exports.CloudRunV2JobTemplateTemplateContainersStartupProbeOutputReference = exports.cloudRunV2JobTemplateTemplateContainersStartupProbeToTerraform = exports.CloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketOutputReference = exports.cloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketToTerraform = exports.CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetOutputReference = exports.cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetToTerraform = exports.CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersList = exports.CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference = exports.cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform = exports.CloudRunV2JobTemplateTemplateContainersResourcesOutputReference = exports.cloudRunV2JobTemplateTemplateContainersResourcesToTerraform = exports.CloudRunV2JobTemplateTemplateContainersPortsList = exports.CloudRunV2JobTemplateTemplateContainersPortsOutputReference = exports.cloudRunV2JobTemplateTemplateContainersPortsToTerraform = exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeOutputReference = exports.cloudRunV2JobTemplateTemplateContainersLivenessProbeToTerraform = exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketOutputReference = exports.cloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketToTerraform = exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetOutputReference = exports.cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetToTerraform = exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersList = exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference = exports.cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform = exports.CloudRunV2JobTemplateTemplateContainersEnvList = exports.CloudRunV2JobTemplateTemplateContainersEnvOutputReference = exports.cloudRunV2JobTemplateTemplateContainersEnvToTerraform = exports.CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference = exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform = exports.CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference = exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform = exports.CloudRunV2JobBinaryAuthorizationOutputReference = exports.cloudRunV2JobBinaryAuthorizationToTerraform = exports.CloudRunV2JobTerminalConditionList = exports.CloudRunV2JobTerminalConditionOutputReference = exports.cloudRunV2JobTerminalConditionToTerraform = exports.CloudRunV2JobLatestCreatedExecutionList = exports.CloudRunV2JobLatestCreatedExecutionOutputReference = exports.cloudRunV2JobLatestCreatedExecutionToTerraform = exports.CloudRunV2JobConditionsList = exports.CloudRunV2JobConditionsOutputReference = exports.cloudRunV2JobConditionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudRunV2JobConditionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudRunV2JobConditionsToTerraform = cloudRunV2JobConditionsToTerraform;
class CloudRunV2JobConditionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // execution_reason - computed: true, optional: false, required: false
    get executionReason() {
        return this.getStringAttribute('execution_reason');
    }
    // last_transition_time - computed: true, optional: false, required: false
    get lastTransitionTime() {
        return this.getStringAttribute('last_transition_time');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
    // revision_reason - computed: true, optional: false, required: false
    get revisionReason() {
        return this.getStringAttribute('revision_reason');
    }
    // severity - computed: true, optional: false, required: false
    get severity() {
        return this.getStringAttribute('severity');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.CloudRunV2JobConditionsOutputReference = CloudRunV2JobConditionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudRunV2JobConditionsOutputReference[_a] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobConditionsOutputReference", version: "4.0.3" };
class CloudRunV2JobConditionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobConditionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobConditionsList = CloudRunV2JobConditionsList;
_b = JSII_RTTI_SYMBOL_1;
CloudRunV2JobConditionsList[_b] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobConditionsList", version: "4.0.3" };
function cloudRunV2JobLatestCreatedExecutionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudRunV2JobLatestCreatedExecutionToTerraform = cloudRunV2JobLatestCreatedExecutionToTerraform;
class CloudRunV2JobLatestCreatedExecutionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // completion_time - computed: true, optional: false, required: false
    get completionTime() {
        return this.getStringAttribute('completion_time');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.CloudRunV2JobLatestCreatedExecutionOutputReference = CloudRunV2JobLatestCreatedExecutionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudRunV2JobLatestCreatedExecutionOutputReference[_c] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobLatestCreatedExecutionOutputReference", version: "4.0.3" };
class CloudRunV2JobLatestCreatedExecutionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobLatestCreatedExecutionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobLatestCreatedExecutionList = CloudRunV2JobLatestCreatedExecutionList;
_d = JSII_RTTI_SYMBOL_1;
CloudRunV2JobLatestCreatedExecutionList[_d] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobLatestCreatedExecutionList", version: "4.0.3" };
function cloudRunV2JobTerminalConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudRunV2JobTerminalConditionToTerraform = cloudRunV2JobTerminalConditionToTerraform;
class CloudRunV2JobTerminalConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // execution_reason - computed: true, optional: false, required: false
    get executionReason() {
        return this.getStringAttribute('execution_reason');
    }
    // last_transition_time - computed: true, optional: false, required: false
    get lastTransitionTime() {
        return this.getStringAttribute('last_transition_time');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
    // revision_reason - computed: true, optional: false, required: false
    get revisionReason() {
        return this.getStringAttribute('revision_reason');
    }
    // severity - computed: true, optional: false, required: false
    get severity() {
        return this.getStringAttribute('severity');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.CloudRunV2JobTerminalConditionOutputReference = CloudRunV2JobTerminalConditionOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTerminalConditionOutputReference[_e] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTerminalConditionOutputReference", version: "4.0.3" };
class CloudRunV2JobTerminalConditionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTerminalConditionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTerminalConditionList = CloudRunV2JobTerminalConditionList;
_f = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTerminalConditionList[_f] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTerminalConditionList", version: "4.0.3" };
function cloudRunV2JobBinaryAuthorizationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        breakglass_justification: cdktf.stringToTerraform(struct.breakglassJustification),
        use_default: cdktf.booleanToTerraform(struct.useDefault),
    };
}
exports.cloudRunV2JobBinaryAuthorizationToTerraform = cloudRunV2JobBinaryAuthorizationToTerraform;
class CloudRunV2JobBinaryAuthorizationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._breakglassJustification !== undefined) {
            hasAnyValues = true;
            internalValueResult.breakglassJustification = this._breakglassJustification;
        }
        if (this._useDefault !== undefined) {
            hasAnyValues = true;
            internalValueResult.useDefault = this._useDefault;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._breakglassJustification = undefined;
            this._useDefault = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._breakglassJustification = value.breakglassJustification;
            this._useDefault = value.useDefault;
        }
    }
    get breakglassJustification() {
        return this.getStringAttribute('breakglass_justification');
    }
    set breakglassJustification(value) {
        this._breakglassJustification = value;
    }
    resetBreakglassJustification() {
        this._breakglassJustification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get breakglassJustificationInput() {
        return this._breakglassJustification;
    }
    get useDefault() {
        return this.getBooleanAttribute('use_default');
    }
    set useDefault(value) {
        this._useDefault = value;
    }
    resetUseDefault() {
        this._useDefault = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useDefaultInput() {
        return this._useDefault;
    }
}
exports.CloudRunV2JobBinaryAuthorizationOutputReference = CloudRunV2JobBinaryAuthorizationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CloudRunV2JobBinaryAuthorizationOutputReference[_g] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobBinaryAuthorizationOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret: cdktf.stringToTerraform(struct.secret),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform = cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform;
class CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secret !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secret = undefined;
            this._version = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secret = value.secret;
            this._version = value.version;
        }
    }
    get secret() {
        return this.getStringAttribute('secret');
    }
    set secret(value) {
        this._secret = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference = CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference;
_h = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference[_h] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_key_ref: cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform(struct.secretKeyRef),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform = cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform;
class CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // secret_key_ref - computed: false, optional: true, required: false
        this._secretKeyRef = new CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference(this, "secret_key_ref");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretKeyRef?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretKeyRef = this._secretKeyRef?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretKeyRef.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretKeyRef.internalValue = value.secretKeyRef;
        }
    }
    get secretKeyRef() {
        return this._secretKeyRef;
    }
    putSecretKeyRef(value) {
        this._secretKeyRef.internalValue = value;
    }
    resetSecretKeyRef() {
        this._secretKeyRef.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretKeyRefInput() {
        return this._secretKeyRef.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference = CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference;
_j = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference[_j] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersEnvToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
        value_source: cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform(struct.valueSource),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersEnvToTerraform = cloudRunV2JobTemplateTemplateContainersEnvToTerraform;
class CloudRunV2JobTemplateTemplateContainersEnvOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // value_source - computed: false, optional: true, required: false
        this._valueSource = new CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference(this, "value_source");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        if (this._valueSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.valueSource = this._valueSource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
            this._valueSource.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
            this._valueSource.internalValue = value.valueSource;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
    get valueSource() {
        return this._valueSource;
    }
    putValueSource(value) {
        this._valueSource.internalValue = value;
    }
    resetValueSource() {
        this._valueSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueSourceInput() {
        return this._valueSource.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersEnvOutputReference = CloudRunV2JobTemplateTemplateContainersEnvOutputReference;
_k = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersEnvOutputReference[_k] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersEnvOutputReference", version: "4.0.3" };
class CloudRunV2JobTemplateTemplateContainersEnvList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersEnvOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersEnvList = CloudRunV2JobTemplateTemplateContainersEnvList;
_l = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersEnvList[_l] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersEnvList", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform = cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform;
class CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference = CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference;
_m = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference[_m] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference", version: "4.0.3" };
class CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersList = CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersList;
_o = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersList[_o] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersList", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        http_headers: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform, true)(struct.httpHeaders),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetToTerraform = cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetToTerraform;
class CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // http_headers - computed: false, optional: true, required: false
        this._httpHeaders = new CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetHttpHeadersList(this, "http_headers", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._httpHeaders?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpHeaders = this._httpHeaders?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
            this._httpHeaders.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
            this._httpHeaders.internalValue = value.httpHeaders;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get httpHeaders() {
        return this._httpHeaders;
    }
    putHttpHeaders(value) {
        this._httpHeaders.internalValue = value;
    }
    resetHttpHeaders() {
        this._httpHeaders.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpHeadersInput() {
        return this._httpHeaders.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetOutputReference = CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetOutputReference;
_p = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetOutputReference[_p] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketToTerraform = cloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketToTerraform;
class CloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._port = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._port = value.port;
        }
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketOutputReference = CloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketOutputReference[_q] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersLivenessProbeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        failure_threshold: cdktf.numberToTerraform(struct.failureThreshold),
        initial_delay_seconds: cdktf.numberToTerraform(struct.initialDelaySeconds),
        period_seconds: cdktf.numberToTerraform(struct.periodSeconds),
        timeout_seconds: cdktf.numberToTerraform(struct.timeoutSeconds),
        http_get: cloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetToTerraform(struct.httpGet),
        tcp_socket: cloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketToTerraform(struct.tcpSocket),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersLivenessProbeToTerraform = cloudRunV2JobTemplateTemplateContainersLivenessProbeToTerraform;
class CloudRunV2JobTemplateTemplateContainersLivenessProbeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // http_get - computed: false, optional: true, required: false
        this._httpGet = new CloudRunV2JobTemplateTemplateContainersLivenessProbeHttpGetOutputReference(this, "http_get");
        // tcp_socket - computed: false, optional: true, required: false
        this._tcpSocket = new CloudRunV2JobTemplateTemplateContainersLivenessProbeTcpSocketOutputReference(this, "tcp_socket");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._failureThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.failureThreshold = this._failureThreshold;
        }
        if (this._initialDelaySeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialDelaySeconds = this._initialDelaySeconds;
        }
        if (this._periodSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.periodSeconds = this._periodSeconds;
        }
        if (this._timeoutSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutSeconds = this._timeoutSeconds;
        }
        if (this._httpGet?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpGet = this._httpGet?.internalValue;
        }
        if (this._tcpSocket?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tcpSocket = this._tcpSocket?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._failureThreshold = undefined;
            this._initialDelaySeconds = undefined;
            this._periodSeconds = undefined;
            this._timeoutSeconds = undefined;
            this._httpGet.internalValue = undefined;
            this._tcpSocket.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._failureThreshold = value.failureThreshold;
            this._initialDelaySeconds = value.initialDelaySeconds;
            this._periodSeconds = value.periodSeconds;
            this._timeoutSeconds = value.timeoutSeconds;
            this._httpGet.internalValue = value.httpGet;
            this._tcpSocket.internalValue = value.tcpSocket;
        }
    }
    get failureThreshold() {
        return this.getNumberAttribute('failure_threshold');
    }
    set failureThreshold(value) {
        this._failureThreshold = value;
    }
    resetFailureThreshold() {
        this._failureThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failureThresholdInput() {
        return this._failureThreshold;
    }
    get initialDelaySeconds() {
        return this.getNumberAttribute('initial_delay_seconds');
    }
    set initialDelaySeconds(value) {
        this._initialDelaySeconds = value;
    }
    resetInitialDelaySeconds() {
        this._initialDelaySeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialDelaySecondsInput() {
        return this._initialDelaySeconds;
    }
    get periodSeconds() {
        return this.getNumberAttribute('period_seconds');
    }
    set periodSeconds(value) {
        this._periodSeconds = value;
    }
    resetPeriodSeconds() {
        this._periodSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get periodSecondsInput() {
        return this._periodSeconds;
    }
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    set timeoutSeconds(value) {
        this._timeoutSeconds = value;
    }
    resetTimeoutSeconds() {
        this._timeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutSecondsInput() {
        return this._timeoutSeconds;
    }
    get httpGet() {
        return this._httpGet;
    }
    putHttpGet(value) {
        this._httpGet.internalValue = value;
    }
    resetHttpGet() {
        this._httpGet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpGetInput() {
        return this._httpGet.internalValue;
    }
    get tcpSocket() {
        return this._tcpSocket;
    }
    putTcpSocket(value) {
        this._tcpSocket.internalValue = value;
    }
    resetTcpSocket() {
        this._tcpSocket.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tcpSocketInput() {
        return this._tcpSocket.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersLivenessProbeOutputReference = CloudRunV2JobTemplateTemplateContainersLivenessProbeOutputReference;
_r = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersLivenessProbeOutputReference[_r] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersLivenessProbeOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersPortsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_port: cdktf.numberToTerraform(struct.containerPort),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersPortsToTerraform = cloudRunV2JobTemplateTemplateContainersPortsToTerraform;
class CloudRunV2JobTemplateTemplateContainersPortsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerPort = this._containerPort;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._containerPort = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._containerPort = value.containerPort;
            this._name = value.name;
        }
    }
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    set containerPort(value) {
        this._containerPort = value;
    }
    resetContainerPort() {
        this._containerPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerPortInput() {
        return this._containerPort;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersPortsOutputReference = CloudRunV2JobTemplateTemplateContainersPortsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersPortsOutputReference[_s] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersPortsOutputReference", version: "4.0.3" };
class CloudRunV2JobTemplateTemplateContainersPortsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersPortsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersPortsList = CloudRunV2JobTemplateTemplateContainersPortsList;
_t = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersPortsList[_t] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersPortsList", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        limits: cdktf.hashMapper(cdktf.stringToTerraform)(struct.limits),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersResourcesToTerraform = cloudRunV2JobTemplateTemplateContainersResourcesToTerraform;
class CloudRunV2JobTemplateTemplateContainersResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._limits !== undefined) {
            hasAnyValues = true;
            internalValueResult.limits = this._limits;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._limits = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._limits = value.limits;
        }
    }
    get limits() {
        return this.getStringMapAttribute('limits');
    }
    set limits(value) {
        this._limits = value;
    }
    resetLimits() {
        this._limits = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get limitsInput() {
        return this._limits;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersResourcesOutputReference = CloudRunV2JobTemplateTemplateContainersResourcesOutputReference;
_u = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersResourcesOutputReference[_u] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersResourcesOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform = cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform;
class CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference = CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference;
_v = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference[_v] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference", version: "4.0.3" };
class CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersList = CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersList;
_w = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersList[_w] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersList", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        http_headers: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform, true)(struct.httpHeaders),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetToTerraform = cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetToTerraform;
class CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // http_headers - computed: false, optional: true, required: false
        this._httpHeaders = new CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetHttpHeadersList(this, "http_headers", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._httpHeaders?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpHeaders = this._httpHeaders?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
            this._httpHeaders.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
            this._httpHeaders.internalValue = value.httpHeaders;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get httpHeaders() {
        return this._httpHeaders;
    }
    putHttpHeaders(value) {
        this._httpHeaders.internalValue = value;
    }
    resetHttpHeaders() {
        this._httpHeaders.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpHeadersInput() {
        return this._httpHeaders.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetOutputReference = CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetOutputReference;
_x = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetOutputReference[_x] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketToTerraform = cloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketToTerraform;
class CloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._port = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._port = value.port;
        }
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketOutputReference = CloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketOutputReference;
_y = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketOutputReference[_y] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersStartupProbeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        failure_threshold: cdktf.numberToTerraform(struct.failureThreshold),
        initial_delay_seconds: cdktf.numberToTerraform(struct.initialDelaySeconds),
        period_seconds: cdktf.numberToTerraform(struct.periodSeconds),
        timeout_seconds: cdktf.numberToTerraform(struct.timeoutSeconds),
        http_get: cloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetToTerraform(struct.httpGet),
        tcp_socket: cloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketToTerraform(struct.tcpSocket),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersStartupProbeToTerraform = cloudRunV2JobTemplateTemplateContainersStartupProbeToTerraform;
class CloudRunV2JobTemplateTemplateContainersStartupProbeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // http_get - computed: false, optional: true, required: false
        this._httpGet = new CloudRunV2JobTemplateTemplateContainersStartupProbeHttpGetOutputReference(this, "http_get");
        // tcp_socket - computed: false, optional: true, required: false
        this._tcpSocket = new CloudRunV2JobTemplateTemplateContainersStartupProbeTcpSocketOutputReference(this, "tcp_socket");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._failureThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.failureThreshold = this._failureThreshold;
        }
        if (this._initialDelaySeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialDelaySeconds = this._initialDelaySeconds;
        }
        if (this._periodSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.periodSeconds = this._periodSeconds;
        }
        if (this._timeoutSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutSeconds = this._timeoutSeconds;
        }
        if (this._httpGet?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpGet = this._httpGet?.internalValue;
        }
        if (this._tcpSocket?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tcpSocket = this._tcpSocket?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._failureThreshold = undefined;
            this._initialDelaySeconds = undefined;
            this._periodSeconds = undefined;
            this._timeoutSeconds = undefined;
            this._httpGet.internalValue = undefined;
            this._tcpSocket.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._failureThreshold = value.failureThreshold;
            this._initialDelaySeconds = value.initialDelaySeconds;
            this._periodSeconds = value.periodSeconds;
            this._timeoutSeconds = value.timeoutSeconds;
            this._httpGet.internalValue = value.httpGet;
            this._tcpSocket.internalValue = value.tcpSocket;
        }
    }
    get failureThreshold() {
        return this.getNumberAttribute('failure_threshold');
    }
    set failureThreshold(value) {
        this._failureThreshold = value;
    }
    resetFailureThreshold() {
        this._failureThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failureThresholdInput() {
        return this._failureThreshold;
    }
    get initialDelaySeconds() {
        return this.getNumberAttribute('initial_delay_seconds');
    }
    set initialDelaySeconds(value) {
        this._initialDelaySeconds = value;
    }
    resetInitialDelaySeconds() {
        this._initialDelaySeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialDelaySecondsInput() {
        return this._initialDelaySeconds;
    }
    get periodSeconds() {
        return this.getNumberAttribute('period_seconds');
    }
    set periodSeconds(value) {
        this._periodSeconds = value;
    }
    resetPeriodSeconds() {
        this._periodSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get periodSecondsInput() {
        return this._periodSeconds;
    }
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    set timeoutSeconds(value) {
        this._timeoutSeconds = value;
    }
    resetTimeoutSeconds() {
        this._timeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutSecondsInput() {
        return this._timeoutSeconds;
    }
    get httpGet() {
        return this._httpGet;
    }
    putHttpGet(value) {
        this._httpGet.internalValue = value;
    }
    resetHttpGet() {
        this._httpGet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpGetInput() {
        return this._httpGet.internalValue;
    }
    get tcpSocket() {
        return this._tcpSocket;
    }
    putTcpSocket(value) {
        this._tcpSocket.internalValue = value;
    }
    resetTcpSocket() {
        this._tcpSocket.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tcpSocketInput() {
        return this._tcpSocket.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersStartupProbeOutputReference = CloudRunV2JobTemplateTemplateContainersStartupProbeOutputReference;
_z = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersStartupProbeOutputReference[_z] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersStartupProbeOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mount_path: cdktf.stringToTerraform(struct.mountPath),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform = cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform;
class CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mountPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.mountPath = this._mountPath;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._mountPath = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._mountPath = value.mountPath;
            this._name = value.name;
        }
    }
    get mountPath() {
        return this.getStringAttribute('mount_path');
    }
    set mountPath(value) {
        this._mountPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get mountPathInput() {
        return this._mountPath;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference = CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference[_0] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference", version: "4.0.3" };
class CloudRunV2JobTemplateTemplateContainersVolumeMountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersVolumeMountsList = CloudRunV2JobTemplateTemplateContainersVolumeMountsList;
_1 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersVolumeMountsList[_1] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersVolumeMountsList", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateContainersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        args: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.args),
        command: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.command),
        image: cdktf.stringToTerraform(struct.image),
        name: cdktf.stringToTerraform(struct.name),
        working_dir: cdktf.stringToTerraform(struct.workingDir),
        env: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersEnvToTerraform, true)(struct.env),
        liveness_probe: cloudRunV2JobTemplateTemplateContainersLivenessProbeToTerraform(struct.livenessProbe),
        ports: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersPortsToTerraform, true)(struct.ports),
        resources: cloudRunV2JobTemplateTemplateContainersResourcesToTerraform(struct.resources),
        startup_probe: cloudRunV2JobTemplateTemplateContainersStartupProbeToTerraform(struct.startupProbe),
        volume_mounts: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform, true)(struct.volumeMounts),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersToTerraform = cloudRunV2JobTemplateTemplateContainersToTerraform;
class CloudRunV2JobTemplateTemplateContainersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // env - computed: false, optional: true, required: false
        this._env = new CloudRunV2JobTemplateTemplateContainersEnvList(this, "env", false);
        // liveness_probe - computed: false, optional: true, required: false
        this._livenessProbe = new CloudRunV2JobTemplateTemplateContainersLivenessProbeOutputReference(this, "liveness_probe");
        // ports - computed: false, optional: true, required: false
        this._ports = new CloudRunV2JobTemplateTemplateContainersPortsList(this, "ports", false);
        // resources - computed: false, optional: true, required: false
        this._resources = new CloudRunV2JobTemplateTemplateContainersResourcesOutputReference(this, "resources");
        // startup_probe - computed: false, optional: true, required: false
        this._startupProbe = new CloudRunV2JobTemplateTemplateContainersStartupProbeOutputReference(this, "startup_probe");
        // volume_mounts - computed: false, optional: true, required: false
        this._volumeMounts = new CloudRunV2JobTemplateTemplateContainersVolumeMountsList(this, "volume_mounts", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._args !== undefined) {
            hasAnyValues = true;
            internalValueResult.args = this._args;
        }
        if (this._command !== undefined) {
            hasAnyValues = true;
            internalValueResult.command = this._command;
        }
        if (this._image !== undefined) {
            hasAnyValues = true;
            internalValueResult.image = this._image;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._workingDir !== undefined) {
            hasAnyValues = true;
            internalValueResult.workingDir = this._workingDir;
        }
        if (this._env?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.env = this._env?.internalValue;
        }
        if (this._livenessProbe?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.livenessProbe = this._livenessProbe?.internalValue;
        }
        if (this._ports?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ports = this._ports?.internalValue;
        }
        if (this._resources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources?.internalValue;
        }
        if (this._startupProbe?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.startupProbe = this._startupProbe?.internalValue;
        }
        if (this._volumeMounts?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeMounts = this._volumeMounts?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._args = undefined;
            this._command = undefined;
            this._image = undefined;
            this._name = undefined;
            this._workingDir = undefined;
            this._env.internalValue = undefined;
            this._livenessProbe.internalValue = undefined;
            this._ports.internalValue = undefined;
            this._resources.internalValue = undefined;
            this._startupProbe.internalValue = undefined;
            this._volumeMounts.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._args = value.args;
            this._command = value.command;
            this._image = value.image;
            this._name = value.name;
            this._workingDir = value.workingDir;
            this._env.internalValue = value.env;
            this._livenessProbe.internalValue = value.livenessProbe;
            this._ports.internalValue = value.ports;
            this._resources.internalValue = value.resources;
            this._startupProbe.internalValue = value.startupProbe;
            this._volumeMounts.internalValue = value.volumeMounts;
        }
    }
    get args() {
        return this.getListAttribute('args');
    }
    set args(value) {
        this._args = value;
    }
    resetArgs() {
        this._args = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argsInput() {
        return this._args;
    }
    get command() {
        return this.getListAttribute('command');
    }
    set command(value) {
        this._command = value;
    }
    resetCommand() {
        this._command = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commandInput() {
        return this._command;
    }
    get image() {
        return this.getStringAttribute('image');
    }
    set image(value) {
        this._image = value;
    }
    // Temporarily expose input value. Use with caution.
    get imageInput() {
        return this._image;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get workingDir() {
        return this.getStringAttribute('working_dir');
    }
    set workingDir(value) {
        this._workingDir = value;
    }
    resetWorkingDir() {
        this._workingDir = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workingDirInput() {
        return this._workingDir;
    }
    get env() {
        return this._env;
    }
    putEnv(value) {
        this._env.internalValue = value;
    }
    resetEnv() {
        this._env.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get envInput() {
        return this._env.internalValue;
    }
    get livenessProbe() {
        return this._livenessProbe;
    }
    putLivenessProbe(value) {
        this._livenessProbe.internalValue = value;
    }
    resetLivenessProbe() {
        this._livenessProbe.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get livenessProbeInput() {
        return this._livenessProbe.internalValue;
    }
    get ports() {
        return this._ports;
    }
    putPorts(value) {
        this._ports.internalValue = value;
    }
    resetPorts() {
        this._ports.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portsInput() {
        return this._ports.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    resetResources() {
        this._resources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
    get startupProbe() {
        return this._startupProbe;
    }
    putStartupProbe(value) {
        this._startupProbe.internalValue = value;
    }
    resetStartupProbe() {
        this._startupProbe.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startupProbeInput() {
        return this._startupProbe.internalValue;
    }
    get volumeMounts() {
        return this._volumeMounts;
    }
    putVolumeMounts(value) {
        this._volumeMounts.internalValue = value;
    }
    resetVolumeMounts() {
        this._volumeMounts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeMountsInput() {
        return this._volumeMounts.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersOutputReference = CloudRunV2JobTemplateTemplateContainersOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersOutputReference[_2] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersOutputReference", version: "4.0.3" };
class CloudRunV2JobTemplateTemplateContainersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersList = CloudRunV2JobTemplateTemplateContainersList;
_3 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersList[_3] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersList", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instances: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.instances),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform = cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform;
class CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instances !== undefined) {
            hasAnyValues = true;
            internalValueResult.instances = this._instances;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._instances = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._instances = value.instances;
        }
    }
    get instances() {
        return this.getListAttribute('instances');
    }
    set instances(value) {
        this._instances = value;
    }
    resetInstances() {
        this._instances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instancesInput() {
        return this._instances;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference = CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference[_4] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.numberToTerraform(struct.mode),
        path: cdktf.stringToTerraform(struct.path),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform = cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform;
class CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._mode = undefined;
            this._path = undefined;
            this._version = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._mode = value.mode;
            this._path = value.path;
            this._version = value.version;
        }
    }
    get mode() {
        return this.getNumberAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference = CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference[_5] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference", version: "4.0.3" };
class CloudRunV2JobTemplateTemplateVolumesSecretItemsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesSecretItemsList = CloudRunV2JobTemplateTemplateVolumesSecretItemsList;
_6 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesSecretItemsList[_6] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesSecretItemsList", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateVolumesSecretToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_mode: cdktf.numberToTerraform(struct.defaultMode),
        secret: cdktf.stringToTerraform(struct.secret),
        items: cdktf.listMapper(cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform, true)(struct.items),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesSecretToTerraform = cloudRunV2JobTemplateTemplateVolumesSecretToTerraform;
class CloudRunV2JobTemplateTemplateVolumesSecretOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // items - computed: false, optional: true, required: false
        this._items = new CloudRunV2JobTemplateTemplateVolumesSecretItemsList(this, "items", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultMode = this._defaultMode;
        }
        if (this._secret !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret;
        }
        if (this._items?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.items = this._items?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultMode = undefined;
            this._secret = undefined;
            this._items.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultMode = value.defaultMode;
            this._secret = value.secret;
            this._items.internalValue = value.items;
        }
    }
    get defaultMode() {
        return this.getNumberAttribute('default_mode');
    }
    set defaultMode(value) {
        this._defaultMode = value;
    }
    resetDefaultMode() {
        this._defaultMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultModeInput() {
        return this._defaultMode;
    }
    get secret() {
        return this.getStringAttribute('secret');
    }
    set secret(value) {
        this._secret = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret;
    }
    get items() {
        return this._items;
    }
    putItems(value) {
        this._items.internalValue = value;
    }
    resetItems() {
        this._items.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get itemsInput() {
        return this._items.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesSecretOutputReference = CloudRunV2JobTemplateTemplateVolumesSecretOutputReference;
_7 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesSecretOutputReference[_7] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesSecretOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        cloud_sql_instance: cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform(struct.cloudSqlInstance),
        secret: cloudRunV2JobTemplateTemplateVolumesSecretToTerraform(struct.secret),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesToTerraform = cloudRunV2JobTemplateTemplateVolumesToTerraform;
class CloudRunV2JobTemplateTemplateVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cloud_sql_instance - computed: false, optional: true, required: false
        this._cloudSqlInstance = new CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference(this, "cloud_sql_instance");
        // secret - computed: false, optional: true, required: false
        this._secret = new CloudRunV2JobTemplateTemplateVolumesSecretOutputReference(this, "secret");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._cloudSqlInstance?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudSqlInstance = this._cloudSqlInstance?.internalValue;
        }
        if (this._secret?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._cloudSqlInstance.internalValue = undefined;
            this._secret.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._cloudSqlInstance.internalValue = value.cloudSqlInstance;
            this._secret.internalValue = value.secret;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get cloudSqlInstance() {
        return this._cloudSqlInstance;
    }
    putCloudSqlInstance(value) {
        this._cloudSqlInstance.internalValue = value;
    }
    resetCloudSqlInstance() {
        this._cloudSqlInstance.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudSqlInstanceInput() {
        return this._cloudSqlInstance.internalValue;
    }
    get secret() {
        return this._secret;
    }
    putSecret(value) {
        this._secret.internalValue = value;
    }
    resetSecret() {
        this._secret.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesOutputReference = CloudRunV2JobTemplateTemplateVolumesOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesOutputReference[_8] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesOutputReference", version: "4.0.3" };
class CloudRunV2JobTemplateTemplateVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesList = CloudRunV2JobTemplateTemplateVolumesList;
_9 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesList[_9] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesList", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateVpcAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connector: cdktf.stringToTerraform(struct.connector),
        egress: cdktf.stringToTerraform(struct.egress),
    };
}
exports.cloudRunV2JobTemplateTemplateVpcAccessToTerraform = cloudRunV2JobTemplateTemplateVpcAccessToTerraform;
class CloudRunV2JobTemplateTemplateVpcAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connector !== undefined) {
            hasAnyValues = true;
            internalValueResult.connector = this._connector;
        }
        if (this._egress !== undefined) {
            hasAnyValues = true;
            internalValueResult.egress = this._egress;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._connector = undefined;
            this._egress = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._connector = value.connector;
            this._egress = value.egress;
        }
    }
    get connector() {
        return this.getStringAttribute('connector');
    }
    set connector(value) {
        this._connector = value;
    }
    resetConnector() {
        this._connector = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectorInput() {
        return this._connector;
    }
    get egress() {
        return this.getStringAttribute('egress');
    }
    set egress(value) {
        this._egress = value;
    }
    resetEgress() {
        this._egress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get egressInput() {
        return this._egress;
    }
}
exports.CloudRunV2JobTemplateTemplateVpcAccessOutputReference = CloudRunV2JobTemplateTemplateVpcAccessOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVpcAccessOutputReference[_10] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVpcAccessOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_key: cdktf.stringToTerraform(struct.encryptionKey),
        execution_environment: cdktf.stringToTerraform(struct.executionEnvironment),
        max_retries: cdktf.numberToTerraform(struct.maxRetries),
        service_account: cdktf.stringToTerraform(struct.serviceAccount),
        timeout: cdktf.stringToTerraform(struct.timeout),
        containers: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersToTerraform, true)(struct.containers),
        volumes: cdktf.listMapper(cloudRunV2JobTemplateTemplateVolumesToTerraform, true)(struct.volumes),
        vpc_access: cloudRunV2JobTemplateTemplateVpcAccessToTerraform(struct.vpcAccess),
    };
}
exports.cloudRunV2JobTemplateTemplateToTerraform = cloudRunV2JobTemplateTemplateToTerraform;
class CloudRunV2JobTemplateTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // containers - computed: false, optional: true, required: false
        this._containers = new CloudRunV2JobTemplateTemplateContainersList(this, "containers", false);
        // volumes - computed: false, optional: true, required: false
        this._volumes = new CloudRunV2JobTemplateTemplateVolumesList(this, "volumes", false);
        // vpc_access - computed: false, optional: true, required: false
        this._vpcAccess = new CloudRunV2JobTemplateTemplateVpcAccessOutputReference(this, "vpc_access");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionKey = this._encryptionKey;
        }
        if (this._executionEnvironment !== undefined) {
            hasAnyValues = true;
            internalValueResult.executionEnvironment = this._executionEnvironment;
        }
        if (this._maxRetries !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxRetries = this._maxRetries;
        }
        if (this._serviceAccount !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccount = this._serviceAccount;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        if (this._containers?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.containers = this._containers?.internalValue;
        }
        if (this._volumes?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumes = this._volumes?.internalValue;
        }
        if (this._vpcAccess?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpcAccess = this._vpcAccess?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._encryptionKey = undefined;
            this._executionEnvironment = undefined;
            this._maxRetries = undefined;
            this._serviceAccount = undefined;
            this._timeout = undefined;
            this._containers.internalValue = undefined;
            this._volumes.internalValue = undefined;
            this._vpcAccess.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._encryptionKey = value.encryptionKey;
            this._executionEnvironment = value.executionEnvironment;
            this._maxRetries = value.maxRetries;
            this._serviceAccount = value.serviceAccount;
            this._timeout = value.timeout;
            this._containers.internalValue = value.containers;
            this._volumes.internalValue = value.volumes;
            this._vpcAccess.internalValue = value.vpcAccess;
        }
    }
    get encryptionKey() {
        return this.getStringAttribute('encryption_key');
    }
    set encryptionKey(value) {
        this._encryptionKey = value;
    }
    resetEncryptionKey() {
        this._encryptionKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionKeyInput() {
        return this._encryptionKey;
    }
    get executionEnvironment() {
        return this.getStringAttribute('execution_environment');
    }
    set executionEnvironment(value) {
        this._executionEnvironment = value;
    }
    resetExecutionEnvironment() {
        this._executionEnvironment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get executionEnvironmentInput() {
        return this._executionEnvironment;
    }
    get maxRetries() {
        return this.getNumberAttribute('max_retries');
    }
    set maxRetries(value) {
        this._maxRetries = value;
    }
    resetMaxRetries() {
        this._maxRetries = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxRetriesInput() {
        return this._maxRetries;
    }
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    set serviceAccount(value) {
        this._serviceAccount = value;
    }
    resetServiceAccount() {
        this._serviceAccount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountInput() {
        return this._serviceAccount;
    }
    get timeout() {
        return this.getStringAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get containers() {
        return this._containers;
    }
    putContainers(value) {
        this._containers.internalValue = value;
    }
    resetContainers() {
        this._containers.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containersInput() {
        return this._containers.internalValue;
    }
    get volumes() {
        return this._volumes;
    }
    putVolumes(value) {
        this._volumes.internalValue = value;
    }
    resetVolumes() {
        this._volumes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesInput() {
        return this._volumes.internalValue;
    }
    get vpcAccess() {
        return this._vpcAccess;
    }
    putVpcAccess(value) {
        this._vpcAccess.internalValue = value;
    }
    resetVpcAccess() {
        this._vpcAccess.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcAccessInput() {
        return this._vpcAccess.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateOutputReference = CloudRunV2JobTemplateTemplateOutputReference;
_11 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateOutputReference[_11] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateOutputReference", version: "4.0.3" };
function cloudRunV2JobTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        parallelism: cdktf.numberToTerraform(struct.parallelism),
        task_count: cdktf.numberToTerraform(struct.taskCount),
        template: cloudRunV2JobTemplateTemplateToTerraform(struct.template),
    };
}
exports.cloudRunV2JobTemplateToTerraform = cloudRunV2JobTemplateToTerraform;
class CloudRunV2JobTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // template - computed: false, optional: false, required: true
        this._template = new CloudRunV2JobTemplateTemplateOutputReference(this, "template");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._parallelism !== undefined) {
            hasAnyValues = true;
            internalValueResult.parallelism = this._parallelism;
        }
        if (this._taskCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.taskCount = this._taskCount;
        }
        if (this._template?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.template = this._template?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._labels = undefined;
            this._parallelism = undefined;
            this._taskCount = undefined;
            this._template.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._labels = value.labels;
            this._parallelism = value.parallelism;
            this._taskCount = value.taskCount;
            this._template.internalValue = value.template;
        }
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get parallelism() {
        return this.getNumberAttribute('parallelism');
    }
    set parallelism(value) {
        this._parallelism = value;
    }
    resetParallelism() {
        this._parallelism = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parallelismInput() {
        return this._parallelism;
    }
    get taskCount() {
        return this.getNumberAttribute('task_count');
    }
    set taskCount(value) {
        this._taskCount = value;
    }
    resetTaskCount() {
        this._taskCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taskCountInput() {
        return this._taskCount;
    }
    get template() {
        return this._template;
    }
    putTemplate(value) {
        this._template.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get templateInput() {
        return this._template.internalValue;
    }
}
exports.CloudRunV2JobTemplateOutputReference = CloudRunV2JobTemplateOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateOutputReference[_12] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateOutputReference", version: "4.0.3" };
function cloudRunV2JobTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudRunV2JobTimeoutsToTerraform = cloudRunV2JobTimeoutsToTerraform;
class CloudRunV2JobTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CloudRunV2JobTimeoutsOutputReference = CloudRunV2JobTimeoutsOutputReference;
_13 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTimeoutsOutputReference[_13] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTimeoutsOutputReference", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/cloud_run_v2_job google_cloud_run_v2_job}
*/
class CloudRunV2Job extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/cloud_run_v2_job google_cloud_run_v2_job} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudRunV2JobConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloud_run_v2_job',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // conditions - computed: true, optional: false, required: false
        this._conditions = new CloudRunV2JobConditionsList(this, "conditions", false);
        // latest_created_execution - computed: true, optional: false, required: false
        this._latestCreatedExecution = new CloudRunV2JobLatestCreatedExecutionList(this, "latest_created_execution", false);
        // terminal_condition - computed: true, optional: false, required: false
        this._terminalCondition = new CloudRunV2JobTerminalConditionList(this, "terminal_condition", false);
        // binary_authorization - computed: false, optional: true, required: false
        this._binaryAuthorization = new CloudRunV2JobBinaryAuthorizationOutputReference(this, "binary_authorization");
        // template - computed: false, optional: false, required: true
        this._template = new CloudRunV2JobTemplateOutputReference(this, "template");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudRunV2JobTimeoutsOutputReference(this, "timeouts");
        this._client = config.client;
        this._clientVersion = config.clientVersion;
        this._id = config.id;
        this._labels = config.labels;
        this._launchStage = config.launchStage;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._binaryAuthorization.internalValue = config.binaryAuthorization;
        this._template.internalValue = config.template;
        this._timeouts.internalValue = config.timeouts;
    }
    get client() {
        return this.getStringAttribute('client');
    }
    set client(value) {
        this._client = value;
    }
    resetClient() {
        this._client = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientInput() {
        return this._client;
    }
    get clientVersion() {
        return this.getStringAttribute('client_version');
    }
    set clientVersion(value) {
        this._clientVersion = value;
    }
    resetClientVersion() {
        this._clientVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientVersionInput() {
        return this._clientVersion;
    }
    get conditions() {
        return this._conditions;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    // execution_count - computed: true, optional: false, required: false
    get executionCount() {
        return this.getNumberAttribute('execution_count');
    }
    // generation - computed: true, optional: false, required: false
    get generation() {
        return this.getStringAttribute('generation');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get latestCreatedExecution() {
        return this._latestCreatedExecution;
    }
    get launchStage() {
        return this.getStringAttribute('launch_stage');
    }
    set launchStage(value) {
        this._launchStage = value;
    }
    resetLaunchStage() {
        this._launchStage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchStageInput() {
        return this._launchStage;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // observed_generation - computed: true, optional: false, required: false
    get observedGeneration() {
        return this.getStringAttribute('observed_generation');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // reconciling - computed: true, optional: false, required: false
    get reconciling() {
        return this.getBooleanAttribute('reconciling');
    }
    get terminalCondition() {
        return this._terminalCondition;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    get binaryAuthorization() {
        return this._binaryAuthorization;
    }
    putBinaryAuthorization(value) {
        this._binaryAuthorization.internalValue = value;
    }
    resetBinaryAuthorization() {
        this._binaryAuthorization.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get binaryAuthorizationInput() {
        return this._binaryAuthorization.internalValue;
    }
    get template() {
        return this._template;
    }
    putTemplate(value) {
        this._template.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get templateInput() {
        return this._template.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            client: cdktf.stringToTerraform(this._client),
            client_version: cdktf.stringToTerraform(this._clientVersion),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            launch_stage: cdktf.stringToTerraform(this._launchStage),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            binary_authorization: cloudRunV2JobBinaryAuthorizationToTerraform(this._binaryAuthorization.internalValue),
            template: cloudRunV2JobTemplateToTerraform(this._template.internalValue),
            timeouts: cloudRunV2JobTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.CloudRunV2Job = CloudRunV2Job;
_14 = JSII_RTTI_SYMBOL_1;
CloudRunV2Job[_14] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2Job", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
CloudRunV2Job.tfResourceType = "google_cloud_run_v2_job";
