// @cdktf/provider-google/lib/compute-resource-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComputeResourcePolicy = exports.ComputeResourcePolicyTimeoutsOutputReference = exports.computeResourcePolicyTimeoutsToTerraform = exports.ComputeResourcePolicySnapshotSchedulePolicyOutputReference = exports.computeResourcePolicySnapshotSchedulePolicyToTerraform = exports.ComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference = exports.computeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToTerraform = exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference = exports.computeResourcePolicySnapshotSchedulePolicyScheduleToTerraform = exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference = exports.computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToTerraform = exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList = exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference = exports.computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToTerraform = exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference = exports.computeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToTerraform = exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference = exports.computeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToTerraform = exports.ComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference = exports.computeResourcePolicySnapshotSchedulePolicyRetentionPolicyToTerraform = exports.ComputeResourcePolicyInstanceSchedulePolicyOutputReference = exports.computeResourcePolicyInstanceSchedulePolicyToTerraform = exports.ComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference = exports.computeResourcePolicyInstanceSchedulePolicyVmStopScheduleToTerraform = exports.ComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference = exports.computeResourcePolicyInstanceSchedulePolicyVmStartScheduleToTerraform = exports.ComputeResourcePolicyGroupPlacementPolicyOutputReference = exports.computeResourcePolicyGroupPlacementPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function computeResourcePolicyGroupPlacementPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        availability_domain_count: cdktf.numberToTerraform(struct.availabilityDomainCount),
        collocation: cdktf.stringToTerraform(struct.collocation),
        vm_count: cdktf.numberToTerraform(struct.vmCount),
    };
}
exports.computeResourcePolicyGroupPlacementPolicyToTerraform = computeResourcePolicyGroupPlacementPolicyToTerraform;
class ComputeResourcePolicyGroupPlacementPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._availabilityDomainCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.availabilityDomainCount = this._availabilityDomainCount;
        }
        if (this._collocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.collocation = this._collocation;
        }
        if (this._vmCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.vmCount = this._vmCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._availabilityDomainCount = undefined;
            this._collocation = undefined;
            this._vmCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._availabilityDomainCount = value.availabilityDomainCount;
            this._collocation = value.collocation;
            this._vmCount = value.vmCount;
        }
    }
    get availabilityDomainCount() {
        return this.getNumberAttribute('availability_domain_count');
    }
    set availabilityDomainCount(value) {
        this._availabilityDomainCount = value;
    }
    resetAvailabilityDomainCount() {
        this._availabilityDomainCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityDomainCountInput() {
        return this._availabilityDomainCount;
    }
    get collocation() {
        return this.getStringAttribute('collocation');
    }
    set collocation(value) {
        this._collocation = value;
    }
    resetCollocation() {
        this._collocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get collocationInput() {
        return this._collocation;
    }
    get vmCount() {
        return this.getNumberAttribute('vm_count');
    }
    set vmCount(value) {
        this._vmCount = value;
    }
    resetVmCount() {
        this._vmCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vmCountInput() {
        return this._vmCount;
    }
}
exports.ComputeResourcePolicyGroupPlacementPolicyOutputReference = ComputeResourcePolicyGroupPlacementPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicyGroupPlacementPolicyOutputReference[_a] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicyGroupPlacementPolicyOutputReference", version: "4.0.3" };
function computeResourcePolicyInstanceSchedulePolicyVmStartScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        schedule: cdktf.stringToTerraform(struct.schedule),
    };
}
exports.computeResourcePolicyInstanceSchedulePolicyVmStartScheduleToTerraform = computeResourcePolicyInstanceSchedulePolicyVmStartScheduleToTerraform;
class ComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._schedule !== undefined) {
            hasAnyValues = true;
            internalValueResult.schedule = this._schedule;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._schedule = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._schedule = value.schedule;
        }
    }
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    set schedule(value) {
        this._schedule = value;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule;
    }
}
exports.ComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference = ComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference[_b] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference", version: "4.0.3" };
function computeResourcePolicyInstanceSchedulePolicyVmStopScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        schedule: cdktf.stringToTerraform(struct.schedule),
    };
}
exports.computeResourcePolicyInstanceSchedulePolicyVmStopScheduleToTerraform = computeResourcePolicyInstanceSchedulePolicyVmStopScheduleToTerraform;
class ComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._schedule !== undefined) {
            hasAnyValues = true;
            internalValueResult.schedule = this._schedule;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._schedule = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._schedule = value.schedule;
        }
    }
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    set schedule(value) {
        this._schedule = value;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule;
    }
}
exports.ComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference = ComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference[_c] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference", version: "4.0.3" };
function computeResourcePolicyInstanceSchedulePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expiration_time: cdktf.stringToTerraform(struct.expirationTime),
        start_time: cdktf.stringToTerraform(struct.startTime),
        time_zone: cdktf.stringToTerraform(struct.timeZone),
        vm_start_schedule: computeResourcePolicyInstanceSchedulePolicyVmStartScheduleToTerraform(struct.vmStartSchedule),
        vm_stop_schedule: computeResourcePolicyInstanceSchedulePolicyVmStopScheduleToTerraform(struct.vmStopSchedule),
    };
}
exports.computeResourcePolicyInstanceSchedulePolicyToTerraform = computeResourcePolicyInstanceSchedulePolicyToTerraform;
class ComputeResourcePolicyInstanceSchedulePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // vm_start_schedule - computed: false, optional: true, required: false
        this._vmStartSchedule = new ComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference(this, "vm_start_schedule");
        // vm_stop_schedule - computed: false, optional: true, required: false
        this._vmStopSchedule = new ComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference(this, "vm_stop_schedule");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._expirationTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.expirationTime = this._expirationTime;
        }
        if (this._startTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime;
        }
        if (this._timeZone !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeZone = this._timeZone;
        }
        if (this._vmStartSchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.vmStartSchedule = this._vmStartSchedule?.internalValue;
        }
        if (this._vmStopSchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.vmStopSchedule = this._vmStopSchedule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._expirationTime = undefined;
            this._startTime = undefined;
            this._timeZone = undefined;
            this._vmStartSchedule.internalValue = undefined;
            this._vmStopSchedule.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._expirationTime = value.expirationTime;
            this._startTime = value.startTime;
            this._timeZone = value.timeZone;
            this._vmStartSchedule.internalValue = value.vmStartSchedule;
            this._vmStopSchedule.internalValue = value.vmStopSchedule;
        }
    }
    get expirationTime() {
        return this.getStringAttribute('expiration_time');
    }
    set expirationTime(value) {
        this._expirationTime = value;
    }
    resetExpirationTime() {
        this._expirationTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expirationTimeInput() {
        return this._expirationTime;
    }
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    set startTime(value) {
        this._startTime = value;
    }
    resetStartTime() {
        this._startTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime;
    }
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
    set timeZone(value) {
        this._timeZone = value;
    }
    // Temporarily expose input value. Use with caution.
    get timeZoneInput() {
        return this._timeZone;
    }
    get vmStartSchedule() {
        return this._vmStartSchedule;
    }
    putVmStartSchedule(value) {
        this._vmStartSchedule.internalValue = value;
    }
    resetVmStartSchedule() {
        this._vmStartSchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vmStartScheduleInput() {
        return this._vmStartSchedule.internalValue;
    }
    get vmStopSchedule() {
        return this._vmStopSchedule;
    }
    putVmStopSchedule(value) {
        this._vmStopSchedule.internalValue = value;
    }
    resetVmStopSchedule() {
        this._vmStopSchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vmStopScheduleInput() {
        return this._vmStopSchedule.internalValue;
    }
}
exports.ComputeResourcePolicyInstanceSchedulePolicyOutputReference = ComputeResourcePolicyInstanceSchedulePolicyOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicyInstanceSchedulePolicyOutputReference[_d] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicyInstanceSchedulePolicyOutputReference", version: "4.0.3" };
function computeResourcePolicySnapshotSchedulePolicyRetentionPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_retention_days: cdktf.numberToTerraform(struct.maxRetentionDays),
        on_source_disk_delete: cdktf.stringToTerraform(struct.onSourceDiskDelete),
    };
}
exports.computeResourcePolicySnapshotSchedulePolicyRetentionPolicyToTerraform = computeResourcePolicySnapshotSchedulePolicyRetentionPolicyToTerraform;
class ComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxRetentionDays !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxRetentionDays = this._maxRetentionDays;
        }
        if (this._onSourceDiskDelete !== undefined) {
            hasAnyValues = true;
            internalValueResult.onSourceDiskDelete = this._onSourceDiskDelete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxRetentionDays = undefined;
            this._onSourceDiskDelete = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxRetentionDays = value.maxRetentionDays;
            this._onSourceDiskDelete = value.onSourceDiskDelete;
        }
    }
    get maxRetentionDays() {
        return this.getNumberAttribute('max_retention_days');
    }
    set maxRetentionDays(value) {
        this._maxRetentionDays = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxRetentionDaysInput() {
        return this._maxRetentionDays;
    }
    get onSourceDiskDelete() {
        return this.getStringAttribute('on_source_disk_delete');
    }
    set onSourceDiskDelete(value) {
        this._onSourceDiskDelete = value;
    }
    resetOnSourceDiskDelete() {
        this._onSourceDiskDelete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onSourceDiskDeleteInput() {
        return this._onSourceDiskDelete;
    }
}
exports.ComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference = ComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference[_e] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference", version: "4.0.3" };
function computeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        days_in_cycle: cdktf.numberToTerraform(struct.daysInCycle),
        start_time: cdktf.stringToTerraform(struct.startTime),
    };
}
exports.computeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToTerraform = computeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToTerraform;
class ComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._daysInCycle !== undefined) {
            hasAnyValues = true;
            internalValueResult.daysInCycle = this._daysInCycle;
        }
        if (this._startTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._daysInCycle = undefined;
            this._startTime = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._daysInCycle = value.daysInCycle;
            this._startTime = value.startTime;
        }
    }
    get daysInCycle() {
        return this.getNumberAttribute('days_in_cycle');
    }
    set daysInCycle(value) {
        this._daysInCycle = value;
    }
    // Temporarily expose input value. Use with caution.
    get daysInCycleInput() {
        return this._daysInCycle;
    }
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    set startTime(value) {
        this._startTime = value;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime;
    }
}
exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference = ComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference[_f] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference", version: "4.0.3" };
function computeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hours_in_cycle: cdktf.numberToTerraform(struct.hoursInCycle),
        start_time: cdktf.stringToTerraform(struct.startTime),
    };
}
exports.computeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToTerraform = computeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToTerraform;
class ComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hoursInCycle !== undefined) {
            hasAnyValues = true;
            internalValueResult.hoursInCycle = this._hoursInCycle;
        }
        if (this._startTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hoursInCycle = undefined;
            this._startTime = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hoursInCycle = value.hoursInCycle;
            this._startTime = value.startTime;
        }
    }
    get hoursInCycle() {
        return this.getNumberAttribute('hours_in_cycle');
    }
    set hoursInCycle(value) {
        this._hoursInCycle = value;
    }
    // Temporarily expose input value. Use with caution.
    get hoursInCycleInput() {
        return this._hoursInCycle;
    }
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    set startTime(value) {
        this._startTime = value;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime;
    }
}
exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference = ComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference[_g] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference", version: "4.0.3" };
function computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.stringToTerraform(struct.day),
        start_time: cdktf.stringToTerraform(struct.startTime),
    };
}
exports.computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToTerraform = computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToTerraform;
class ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._startTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._day = undefined;
            this._startTime = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._day = value.day;
            this._startTime = value.startTime;
        }
    }
    get day() {
        return this.getStringAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    set startTime(value) {
        this._startTime = value;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime;
    }
}
exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference = ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference[_h] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference", version: "4.0.3" };
class ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList = ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList;
_j = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList[_j] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList", version: "4.0.3" };
function computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day_of_weeks: cdktf.listMapper(computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToTerraform, true)(struct.dayOfWeeks),
    };
}
exports.computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToTerraform = computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToTerraform;
class ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // day_of_weeks - computed: false, optional: false, required: true
        this._dayOfWeeks = new ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList(this, "day_of_weeks", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dayOfWeeks?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dayOfWeeks = this._dayOfWeeks?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dayOfWeeks.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dayOfWeeks.internalValue = value.dayOfWeeks;
        }
    }
    get dayOfWeeks() {
        return this._dayOfWeeks;
    }
    putDayOfWeeks(value) {
        this._dayOfWeeks.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get dayOfWeeksInput() {
        return this._dayOfWeeks.internalValue;
    }
}
exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference = ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference[_k] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference", version: "4.0.3" };
function computeResourcePolicySnapshotSchedulePolicyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        daily_schedule: computeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToTerraform(struct.dailySchedule),
        hourly_schedule: computeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToTerraform(struct.hourlySchedule),
        weekly_schedule: computeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToTerraform(struct.weeklySchedule),
    };
}
exports.computeResourcePolicySnapshotSchedulePolicyScheduleToTerraform = computeResourcePolicySnapshotSchedulePolicyScheduleToTerraform;
class ComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // daily_schedule - computed: false, optional: true, required: false
        this._dailySchedule = new ComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference(this, "daily_schedule");
        // hourly_schedule - computed: false, optional: true, required: false
        this._hourlySchedule = new ComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference(this, "hourly_schedule");
        // weekly_schedule - computed: false, optional: true, required: false
        this._weeklySchedule = new ComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference(this, "weekly_schedule");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dailySchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dailySchedule = this._dailySchedule?.internalValue;
        }
        if (this._hourlySchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hourlySchedule = this._hourlySchedule?.internalValue;
        }
        if (this._weeklySchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.weeklySchedule = this._weeklySchedule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dailySchedule.internalValue = undefined;
            this._hourlySchedule.internalValue = undefined;
            this._weeklySchedule.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dailySchedule.internalValue = value.dailySchedule;
            this._hourlySchedule.internalValue = value.hourlySchedule;
            this._weeklySchedule.internalValue = value.weeklySchedule;
        }
    }
    get dailySchedule() {
        return this._dailySchedule;
    }
    putDailySchedule(value) {
        this._dailySchedule.internalValue = value;
    }
    resetDailySchedule() {
        this._dailySchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dailyScheduleInput() {
        return this._dailySchedule.internalValue;
    }
    get hourlySchedule() {
        return this._hourlySchedule;
    }
    putHourlySchedule(value) {
        this._hourlySchedule.internalValue = value;
    }
    resetHourlySchedule() {
        this._hourlySchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourlyScheduleInput() {
        return this._hourlySchedule.internalValue;
    }
    get weeklySchedule() {
        return this._weeklySchedule;
    }
    putWeeklySchedule(value) {
        this._weeklySchedule.internalValue = value;
    }
    resetWeeklySchedule() {
        this._weeklySchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weeklyScheduleInput() {
        return this._weeklySchedule.internalValue;
    }
}
exports.ComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference = ComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference[_l] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference", version: "4.0.3" };
function computeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        chain_name: cdktf.stringToTerraform(struct.chainName),
        guest_flush: cdktf.booleanToTerraform(struct.guestFlush),
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        storage_locations: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.storageLocations),
    };
}
exports.computeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToTerraform = computeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToTerraform;
class ComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._chainName !== undefined) {
            hasAnyValues = true;
            internalValueResult.chainName = this._chainName;
        }
        if (this._guestFlush !== undefined) {
            hasAnyValues = true;
            internalValueResult.guestFlush = this._guestFlush;
        }
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._storageLocations !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageLocations = this._storageLocations;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._chainName = undefined;
            this._guestFlush = undefined;
            this._labels = undefined;
            this._storageLocations = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._chainName = value.chainName;
            this._guestFlush = value.guestFlush;
            this._labels = value.labels;
            this._storageLocations = value.storageLocations;
        }
    }
    get chainName() {
        return this.getStringAttribute('chain_name');
    }
    set chainName(value) {
        this._chainName = value;
    }
    resetChainName() {
        this._chainName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get chainNameInput() {
        return this._chainName;
    }
    get guestFlush() {
        return this.getBooleanAttribute('guest_flush');
    }
    set guestFlush(value) {
        this._guestFlush = value;
    }
    resetGuestFlush() {
        this._guestFlush = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get guestFlushInput() {
        return this._guestFlush;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get storageLocations() {
        return cdktf.Fn.tolist(this.getListAttribute('storage_locations'));
    }
    set storageLocations(value) {
        this._storageLocations = value;
    }
    resetStorageLocations() {
        this._storageLocations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageLocationsInput() {
        return this._storageLocations;
    }
}
exports.ComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference = ComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference[_m] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference", version: "4.0.3" };
function computeResourcePolicySnapshotSchedulePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        retention_policy: computeResourcePolicySnapshotSchedulePolicyRetentionPolicyToTerraform(struct.retentionPolicy),
        schedule: computeResourcePolicySnapshotSchedulePolicyScheduleToTerraform(struct.schedule),
        snapshot_properties: computeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToTerraform(struct.snapshotProperties),
    };
}
exports.computeResourcePolicySnapshotSchedulePolicyToTerraform = computeResourcePolicySnapshotSchedulePolicyToTerraform;
class ComputeResourcePolicySnapshotSchedulePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // retention_policy - computed: false, optional: true, required: false
        this._retentionPolicy = new ComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference(this, "retention_policy");
        // schedule - computed: false, optional: false, required: true
        this._schedule = new ComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference(this, "schedule");
        // snapshot_properties - computed: false, optional: true, required: false
        this._snapshotProperties = new ComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference(this, "snapshot_properties");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._retentionPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.retentionPolicy = this._retentionPolicy?.internalValue;
        }
        if (this._schedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.schedule = this._schedule?.internalValue;
        }
        if (this._snapshotProperties?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.snapshotProperties = this._snapshotProperties?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._retentionPolicy.internalValue = undefined;
            this._schedule.internalValue = undefined;
            this._snapshotProperties.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._retentionPolicy.internalValue = value.retentionPolicy;
            this._schedule.internalValue = value.schedule;
            this._snapshotProperties.internalValue = value.snapshotProperties;
        }
    }
    get retentionPolicy() {
        return this._retentionPolicy;
    }
    putRetentionPolicy(value) {
        this._retentionPolicy.internalValue = value;
    }
    resetRetentionPolicy() {
        this._retentionPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPolicyInput() {
        return this._retentionPolicy.internalValue;
    }
    get schedule() {
        return this._schedule;
    }
    putSchedule(value) {
        this._schedule.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule.internalValue;
    }
    get snapshotProperties() {
        return this._snapshotProperties;
    }
    putSnapshotProperties(value) {
        this._snapshotProperties.internalValue = value;
    }
    resetSnapshotProperties() {
        this._snapshotProperties.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotPropertiesInput() {
        return this._snapshotProperties.internalValue;
    }
}
exports.ComputeResourcePolicySnapshotSchedulePolicyOutputReference = ComputeResourcePolicySnapshotSchedulePolicyOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicySnapshotSchedulePolicyOutputReference[_o] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicySnapshotSchedulePolicyOutputReference", version: "4.0.3" };
function computeResourcePolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.computeResourcePolicyTimeoutsToTerraform = computeResourcePolicyTimeoutsToTerraform;
class ComputeResourcePolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.ComputeResourcePolicyTimeoutsOutputReference = ComputeResourcePolicyTimeoutsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicyTimeoutsOutputReference[_p] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicyTimeoutsOutputReference", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/compute_resource_policy google_compute_resource_policy}
*/
class ComputeResourcePolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/compute_resource_policy google_compute_resource_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComputeResourcePolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_resource_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // group_placement_policy - computed: false, optional: true, required: false
        this._groupPlacementPolicy = new ComputeResourcePolicyGroupPlacementPolicyOutputReference(this, "group_placement_policy");
        // instance_schedule_policy - computed: false, optional: true, required: false
        this._instanceSchedulePolicy = new ComputeResourcePolicyInstanceSchedulePolicyOutputReference(this, "instance_schedule_policy");
        // snapshot_schedule_policy - computed: false, optional: true, required: false
        this._snapshotSchedulePolicy = new ComputeResourcePolicySnapshotSchedulePolicyOutputReference(this, "snapshot_schedule_policy");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComputeResourcePolicyTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
        this._groupPlacementPolicy.internalValue = config.groupPlacementPolicy;
        this._instanceSchedulePolicy.internalValue = config.instanceSchedulePolicy;
        this._snapshotSchedulePolicy.internalValue = config.snapshotSchedulePolicy;
        this._timeouts.internalValue = config.timeouts;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get groupPlacementPolicy() {
        return this._groupPlacementPolicy;
    }
    putGroupPlacementPolicy(value) {
        this._groupPlacementPolicy.internalValue = value;
    }
    resetGroupPlacementPolicy() {
        this._groupPlacementPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupPlacementPolicyInput() {
        return this._groupPlacementPolicy.internalValue;
    }
    get instanceSchedulePolicy() {
        return this._instanceSchedulePolicy;
    }
    putInstanceSchedulePolicy(value) {
        this._instanceSchedulePolicy.internalValue = value;
    }
    resetInstanceSchedulePolicy() {
        this._instanceSchedulePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceSchedulePolicyInput() {
        return this._instanceSchedulePolicy.internalValue;
    }
    get snapshotSchedulePolicy() {
        return this._snapshotSchedulePolicy;
    }
    putSnapshotSchedulePolicy(value) {
        this._snapshotSchedulePolicy.internalValue = value;
    }
    resetSnapshotSchedulePolicy() {
        this._snapshotSchedulePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotSchedulePolicyInput() {
        return this._snapshotSchedulePolicy.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
            group_placement_policy: computeResourcePolicyGroupPlacementPolicyToTerraform(this._groupPlacementPolicy.internalValue),
            instance_schedule_policy: computeResourcePolicyInstanceSchedulePolicyToTerraform(this._instanceSchedulePolicy.internalValue),
            snapshot_schedule_policy: computeResourcePolicySnapshotSchedulePolicyToTerraform(this._snapshotSchedulePolicy.internalValue),
            timeouts: computeResourcePolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.ComputeResourcePolicy = ComputeResourcePolicy;
_q = JSII_RTTI_SYMBOL_1;
ComputeResourcePolicy[_q] = { fqn: "@cdktf/provider-google.computeResourcePolicy.ComputeResourcePolicy", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
ComputeResourcePolicy.tfResourceType = "google_compute_resource_policy";
