// @cdktf/provider-google/lib/data-google-compute-instance-template/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeInstanceTemplate = exports.DataGoogleComputeInstanceTemplateShieldedInstanceConfigList = exports.DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference = exports.dataGoogleComputeInstanceTemplateShieldedInstanceConfigToTerraform = exports.DataGoogleComputeInstanceTemplateServiceAccountList = exports.DataGoogleComputeInstanceTemplateServiceAccountOutputReference = exports.dataGoogleComputeInstanceTemplateServiceAccountToTerraform = exports.DataGoogleComputeInstanceTemplateSchedulingList = exports.DataGoogleComputeInstanceTemplateSchedulingOutputReference = exports.dataGoogleComputeInstanceTemplateSchedulingToTerraform = exports.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList = exports.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference = exports.dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToTerraform = exports.DataGoogleComputeInstanceTemplateReservationAffinityList = exports.DataGoogleComputeInstanceTemplateReservationAffinityOutputReference = exports.dataGoogleComputeInstanceTemplateReservationAffinityToTerraform = exports.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList = exports.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference = exports.dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToTerraform = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceList = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceToTerraform = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToTerraform = exports.DataGoogleComputeInstanceTemplateGuestAcceleratorList = exports.DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference = exports.dataGoogleComputeInstanceTemplateGuestAcceleratorToTerraform = exports.DataGoogleComputeInstanceTemplateDiskList = exports.DataGoogleComputeInstanceTemplateDiskOutputReference = exports.dataGoogleComputeInstanceTemplateDiskToTerraform = exports.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList = exports.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference = exports.dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToTerraform = exports.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList = exports.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference = exports.dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToTerraform = exports.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList = exports.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference = exports.dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToTerraform = dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToTerraform;
class DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_nested_virtualization - computed: true, optional: false, required: false
    get enableNestedVirtualization() {
        return this.getBooleanAttribute('enable_nested_virtualization');
    }
    // threads_per_core - computed: true, optional: false, required: false
    get threadsPerCore() {
        return this.getNumberAttribute('threads_per_core');
    }
    // visible_core_count - computed: true, optional: false, required: false
    get visibleCoreCount() {
        return this.getNumberAttribute('visible_core_count');
    }
}
exports.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference = DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList = DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToTerraform = dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToTerraform;
class DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_confidential_compute - computed: true, optional: false, required: false
    get enableConfidentialCompute() {
        return this.getBooleanAttribute('enable_confidential_compute');
    }
}
exports.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference = DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList = DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToTerraform = dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToTerraform;
class DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
}
exports.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference = DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList = DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateDiskToTerraform = dataGoogleComputeInstanceTemplateDiskToTerraform;
class DataGoogleComputeInstanceTemplateDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // disk_encryption_key - computed: true, optional: false, required: false
        this._diskEncryptionKey = new DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList(this, "disk_encryption_key", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // auto_delete - computed: true, optional: false, required: false
    get autoDelete() {
        return this.getBooleanAttribute('auto_delete');
    }
    // boot - computed: true, optional: false, required: false
    get boot() {
        return this.getBooleanAttribute('boot');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    get diskEncryptionKey() {
        return this._diskEncryptionKey;
    }
    // disk_name - computed: true, optional: false, required: false
    get diskName() {
        return this.getStringAttribute('disk_name');
    }
    // disk_size_gb - computed: true, optional: false, required: false
    get diskSizeGb() {
        return this.getNumberAttribute('disk_size_gb');
    }
    // disk_type - computed: true, optional: false, required: false
    get diskType() {
        return this.getStringAttribute('disk_type');
    }
    // interface - computed: true, optional: false, required: false
    get interface() {
        return this.getStringAttribute('interface');
    }
    get labels() {
        return this._labels;
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
    // resource_policies - computed: true, optional: false, required: false
    get resourcePolicies() {
        return this.getListAttribute('resource_policies');
    }
    // source - computed: true, optional: false, required: false
    get source() {
        return this.getStringAttribute('source');
    }
    // source_image - computed: true, optional: false, required: false
    get sourceImage() {
        return this.getStringAttribute('source_image');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceTemplateDiskOutputReference = DataGoogleComputeInstanceTemplateDiskOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateDiskList = DataGoogleComputeInstanceTemplateDiskList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateGuestAcceleratorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateGuestAcceleratorToTerraform = dataGoogleComputeInstanceTemplateGuestAcceleratorToTerraform;
class DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // count - computed: true, optional: false, required: false
    get count() {
        return this.getNumberAttribute('count');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference = DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateGuestAcceleratorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateGuestAcceleratorList = DataGoogleComputeInstanceTemplateGuestAcceleratorList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateGuestAcceleratorList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateGuestAcceleratorList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToTerraform;
class DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // nat_ip - computed: true, optional: false, required: false
    get natIp() {
        return this.getStringAttribute('nat_ip');
    }
    // network_tier - computed: true, optional: false, required: false
    get networkTier() {
        return this.getStringAttribute('network_tier');
    }
    // public_ptr_domain_name - computed: true, optional: false, required: false
    get publicPtrDomainName() {
        return this.getStringAttribute('public_ptr_domain_name');
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference = DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList = DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList[_m] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform;
class DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_cidr_range - computed: true, optional: false, required: false
    get ipCidrRange() {
        return this.getStringAttribute('ip_cidr_range');
    }
    // subnetwork_range_name - computed: true, optional: false, required: false
    get subnetworkRangeName() {
        return this.getStringAttribute('subnetwork_range_name');
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference = DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList = DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList[_p] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform;
class DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // external_ipv6 - computed: true, optional: false, required: false
    get externalIpv6() {
        return this.getStringAttribute('external_ipv6');
    }
    // external_ipv6_prefix_length - computed: true, optional: false, required: false
    get externalIpv6PrefixLength() {
        return this.getStringAttribute('external_ipv6_prefix_length');
    }
    // network_tier - computed: true, optional: false, required: false
    get networkTier() {
        return this.getStringAttribute('network_tier');
    }
    // public_ptr_domain_name - computed: true, optional: false, required: false
    get publicPtrDomainName() {
        return this.getStringAttribute('public_ptr_domain_name');
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference = DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList = DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList[_r] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateNetworkInterfaceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceToTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceToTerraform;
class DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // access_config - computed: true, optional: false, required: false
        this._accessConfig = new DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList(this, "access_config", false);
        // alias_ip_range - computed: true, optional: false, required: false
        this._aliasIpRange = new DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList(this, "alias_ip_range", false);
        // ipv6_access_config - computed: true, optional: false, required: false
        this._ipv6AccessConfig = new DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList(this, "ipv6_access_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get accessConfig() {
        return this._accessConfig;
    }
    get aliasIpRange() {
        return this._aliasIpRange;
    }
    get ipv6AccessConfig() {
        return this._ipv6AccessConfig;
    }
    // ipv6_access_type - computed: true, optional: false, required: false
    get ipv6AccessType() {
        return this.getStringAttribute('ipv6_access_type');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    // network_ip - computed: true, optional: false, required: false
    get networkIp() {
        return this.getStringAttribute('network_ip');
    }
    // nic_type - computed: true, optional: false, required: false
    get nicType() {
        return this.getStringAttribute('nic_type');
    }
    // queue_count - computed: true, optional: false, required: false
    get queueCount() {
        return this.getNumberAttribute('queue_count');
    }
    // stack_type - computed: true, optional: false, required: false
    get stackType() {
        return this.getStringAttribute('stack_type');
    }
    // subnetwork - computed: true, optional: false, required: false
    get subnetwork() {
        return this.getStringAttribute('subnetwork');
    }
    // subnetwork_project - computed: true, optional: false, required: false
    get subnetworkProject() {
        return this.getStringAttribute('subnetwork_project');
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference = DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateNetworkInterfaceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceList = DataGoogleComputeInstanceTemplateNetworkInterfaceList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceList[_t] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToTerraform = dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToTerraform;
class DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return this.getListAttribute('values');
    }
}
exports.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference = DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList = DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList[_v] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateReservationAffinityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateReservationAffinityToTerraform = dataGoogleComputeInstanceTemplateReservationAffinityToTerraform;
class DataGoogleComputeInstanceTemplateReservationAffinityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // specific_reservation - computed: true, optional: false, required: false
        this._specificReservation = new DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList(this, "specific_reservation", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get specificReservation() {
        return this._specificReservation;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceTemplateReservationAffinityOutputReference = DataGoogleComputeInstanceTemplateReservationAffinityOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateReservationAffinityOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateReservationAffinityOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateReservationAffinityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateReservationAffinityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateReservationAffinityList = DataGoogleComputeInstanceTemplateReservationAffinityList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateReservationAffinityList[_x] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateReservationAffinityList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToTerraform = dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToTerraform;
class DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // operator - computed: true, optional: false, required: false
    get operator() {
        return this.getStringAttribute('operator');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference = DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList = DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList[_z] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateSchedulingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateSchedulingToTerraform = dataGoogleComputeInstanceTemplateSchedulingToTerraform;
class DataGoogleComputeInstanceTemplateSchedulingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // node_affinities - computed: true, optional: false, required: false
        this._nodeAffinities = new DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList(this, "node_affinities", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // automatic_restart - computed: true, optional: false, required: false
    get automaticRestart() {
        return this.getBooleanAttribute('automatic_restart');
    }
    // instance_termination_action - computed: true, optional: false, required: false
    get instanceTerminationAction() {
        return this.getStringAttribute('instance_termination_action');
    }
    // min_node_cpus - computed: true, optional: false, required: false
    get minNodeCpus() {
        return this.getNumberAttribute('min_node_cpus');
    }
    get nodeAffinities() {
        return this._nodeAffinities;
    }
    // on_host_maintenance - computed: true, optional: false, required: false
    get onHostMaintenance() {
        return this.getStringAttribute('on_host_maintenance');
    }
    // preemptible - computed: true, optional: false, required: false
    get preemptible() {
        return this.getBooleanAttribute('preemptible');
    }
    // provisioning_model - computed: true, optional: false, required: false
    get provisioningModel() {
        return this.getStringAttribute('provisioning_model');
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingOutputReference = DataGoogleComputeInstanceTemplateSchedulingOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateSchedulingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateSchedulingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingList = DataGoogleComputeInstanceTemplateSchedulingList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingList[_1] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateServiceAccountToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateServiceAccountToTerraform = dataGoogleComputeInstanceTemplateServiceAccountToTerraform;
class DataGoogleComputeInstanceTemplateServiceAccountOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
    // scopes - computed: true, optional: false, required: false
    get scopes() {
        return cdktf.Fn.tolist(this.getListAttribute('scopes'));
    }
}
exports.DataGoogleComputeInstanceTemplateServiceAccountOutputReference = DataGoogleComputeInstanceTemplateServiceAccountOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateServiceAccountOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateServiceAccountOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateServiceAccountList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateServiceAccountOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateServiceAccountList = DataGoogleComputeInstanceTemplateServiceAccountList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateServiceAccountList[_3] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateServiceAccountList", version: "4.0.3" };
function dataGoogleComputeInstanceTemplateShieldedInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateShieldedInstanceConfigToTerraform = dataGoogleComputeInstanceTemplateShieldedInstanceConfigToTerraform;
class DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_integrity_monitoring - computed: true, optional: false, required: false
    get enableIntegrityMonitoring() {
        return this.getBooleanAttribute('enable_integrity_monitoring');
    }
    // enable_secure_boot - computed: true, optional: false, required: false
    get enableSecureBoot() {
        return this.getBooleanAttribute('enable_secure_boot');
    }
    // enable_vtpm - computed: true, optional: false, required: false
    get enableVtpm() {
        return this.getBooleanAttribute('enable_vtpm');
    }
}
exports.DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference = DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference", version: "4.0.3" };
class DataGoogleComputeInstanceTemplateShieldedInstanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateShieldedInstanceConfigList = DataGoogleComputeInstanceTemplateShieldedInstanceConfigList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateShieldedInstanceConfigList[_5] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateShieldedInstanceConfigList", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/compute_instance_template google_compute_instance_template}
*/
class DataGoogleComputeInstanceTemplate extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/compute_instance_template google_compute_instance_template} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeInstanceTemplateConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_compute_instance_template',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // advanced_machine_features - computed: true, optional: false, required: false
        this._advancedMachineFeatures = new DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList(this, "advanced_machine_features", false);
        // confidential_instance_config - computed: true, optional: false, required: false
        this._confidentialInstanceConfig = new DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList(this, "confidential_instance_config", false);
        // disk - computed: true, optional: false, required: false
        this._disk = new DataGoogleComputeInstanceTemplateDiskList(this, "disk", false);
        // guest_accelerator - computed: true, optional: false, required: false
        this._guestAccelerator = new DataGoogleComputeInstanceTemplateGuestAcceleratorList(this, "guest_accelerator", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // metadata - computed: true, optional: false, required: false
        this._metadata = new cdktf.StringMap(this, "metadata");
        // network_interface - computed: true, optional: false, required: false
        this._networkInterface = new DataGoogleComputeInstanceTemplateNetworkInterfaceList(this, "network_interface", false);
        // reservation_affinity - computed: true, optional: false, required: false
        this._reservationAffinity = new DataGoogleComputeInstanceTemplateReservationAffinityList(this, "reservation_affinity", false);
        // scheduling - computed: true, optional: false, required: false
        this._scheduling = new DataGoogleComputeInstanceTemplateSchedulingList(this, "scheduling", false);
        // service_account - computed: true, optional: false, required: false
        this._serviceAccount = new DataGoogleComputeInstanceTemplateServiceAccountList(this, "service_account", false);
        // shielded_instance_config - computed: true, optional: false, required: false
        this._shieldedInstanceConfig = new DataGoogleComputeInstanceTemplateShieldedInstanceConfigList(this, "shielded_instance_config", false);
        this._filter = config.filter;
        this._id = config.id;
        this._mostRecent = config.mostRecent;
        this._name = config.name;
        this._project = config.project;
    }
    get advancedMachineFeatures() {
        return this._advancedMachineFeatures;
    }
    // can_ip_forward - computed: true, optional: false, required: false
    get canIpForward() {
        return this.getBooleanAttribute('can_ip_forward');
    }
    get confidentialInstanceConfig() {
        return this._confidentialInstanceConfig;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get disk() {
        return this._disk;
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get guestAccelerator() {
        return this._guestAccelerator;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_description - computed: true, optional: false, required: false
    get instanceDescription() {
        return this.getStringAttribute('instance_description');
    }
    get labels() {
        return this._labels;
    }
    // machine_type - computed: true, optional: false, required: false
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    get metadata() {
        return this._metadata;
    }
    // metadata_fingerprint - computed: true, optional: false, required: false
    get metadataFingerprint() {
        return this.getStringAttribute('metadata_fingerprint');
    }
    // metadata_startup_script - computed: true, optional: false, required: false
    get metadataStartupScript() {
        return this.getStringAttribute('metadata_startup_script');
    }
    // min_cpu_platform - computed: true, optional: false, required: false
    get minCpuPlatform() {
        return this.getStringAttribute('min_cpu_platform');
    }
    get mostRecent() {
        return this.getBooleanAttribute('most_recent');
    }
    set mostRecent(value) {
        this._mostRecent = value;
    }
    resetMostRecent() {
        this._mostRecent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mostRecentInput() {
        return this._mostRecent;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // name_prefix - computed: true, optional: false, required: false
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    get networkInterface() {
        return this._networkInterface;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // region - computed: true, optional: false, required: false
    get region() {
        return this.getStringAttribute('region');
    }
    get reservationAffinity() {
        return this._reservationAffinity;
    }
    get scheduling() {
        return this._scheduling;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get serviceAccount() {
        return this._serviceAccount;
    }
    get shieldedInstanceConfig() {
        return this._shieldedInstanceConfig;
    }
    // tags - computed: true, optional: false, required: false
    get tags() {
        return cdktf.Fn.tolist(this.getListAttribute('tags'));
    }
    // tags_fingerprint - computed: true, optional: false, required: false
    get tagsFingerprint() {
        return this.getStringAttribute('tags_fingerprint');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            filter: cdktf.stringToTerraform(this._filter),
            id: cdktf.stringToTerraform(this._id),
            most_recent: cdktf.booleanToTerraform(this._mostRecent),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
}
exports.DataGoogleComputeInstanceTemplate = DataGoogleComputeInstanceTemplate;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplate[_6] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplate", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeInstanceTemplate.tfResourceType = "google_compute_instance_template";
