// @cdktf/provider-google/lib/data-google-compute-instance-group-manager/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeInstanceGroupManager = exports.DataGoogleComputeInstanceGroupManagerVersionList = exports.DataGoogleComputeInstanceGroupManagerVersionOutputReference = exports.dataGoogleComputeInstanceGroupManagerVersionToTerraform = exports.DataGoogleComputeInstanceGroupManagerVersionTargetSizeList = exports.DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference = exports.dataGoogleComputeInstanceGroupManagerVersionTargetSizeToTerraform = exports.DataGoogleComputeInstanceGroupManagerUpdatePolicyList = exports.DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference = exports.dataGoogleComputeInstanceGroupManagerUpdatePolicyToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatusList = exports.DataGoogleComputeInstanceGroupManagerStatusOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatusToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatusVersionTargetList = exports.DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatusVersionTargetToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatusStatefulList = exports.DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatusStatefulToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList = exports.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatefulDiskList = exports.DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatefulDiskToTerraform = exports.DataGoogleComputeInstanceGroupManagerNamedPortList = exports.DataGoogleComputeInstanceGroupManagerNamedPortOutputReference = exports.dataGoogleComputeInstanceGroupManagerNamedPortToTerraform = exports.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList = exports.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference = exports.dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToTerraform = dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToTerraform;
class DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // health_check - computed: true, optional: false, required: false
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    // initial_delay_sec - computed: true, optional: false, required: false
    get initialDelaySec() {
        return this.getNumberAttribute('initial_delay_sec');
    }
}
exports.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference = DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList = DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList", version: "5.0.1" };
function dataGoogleComputeInstanceGroupManagerNamedPortToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerNamedPortToTerraform = dataGoogleComputeInstanceGroupManagerNamedPortToTerraform;
class DataGoogleComputeInstanceGroupManagerNamedPortOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataGoogleComputeInstanceGroupManagerNamedPortOutputReference = DataGoogleComputeInstanceGroupManagerNamedPortOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerNamedPortOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerNamedPortOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerNamedPortList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerNamedPortOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerNamedPortList = DataGoogleComputeInstanceGroupManagerNamedPortList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerNamedPortList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerNamedPortList", version: "5.0.1" };
function dataGoogleComputeInstanceGroupManagerStatefulDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatefulDiskToTerraform = dataGoogleComputeInstanceGroupManagerStatefulDiskToTerraform;
class DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_rule - computed: true, optional: false, required: false
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference = DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerStatefulDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatefulDiskList = DataGoogleComputeInstanceGroupManagerStatefulDiskList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatefulDiskList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatefulDiskList", version: "5.0.1" };
function dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform;
class DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all_effective - computed: true, optional: false, required: false
    get allEffective() {
        return this.getBooleanAttribute('all_effective');
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList = DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList", version: "5.0.1" };
function dataGoogleComputeInstanceGroupManagerStatusStatefulToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatusStatefulToTerraform = dataGoogleComputeInstanceGroupManagerStatusStatefulToTerraform;
class DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // per_instance_configs - computed: true, optional: false, required: false
        this._perInstanceConfigs = new DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList(this, "per_instance_configs", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // has_stateful_config - computed: true, optional: false, required: false
    get hasStatefulConfig() {
        return this.getBooleanAttribute('has_stateful_config');
    }
    get perInstanceConfigs() {
        return this._perInstanceConfigs;
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference = DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerStatusStatefulList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusStatefulList = DataGoogleComputeInstanceGroupManagerStatusStatefulList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusStatefulList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusStatefulList", version: "5.0.1" };
function dataGoogleComputeInstanceGroupManagerStatusVersionTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatusVersionTargetToTerraform = dataGoogleComputeInstanceGroupManagerStatusVersionTargetToTerraform;
class DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_reached - computed: true, optional: false, required: false
    get isReached() {
        return this.getBooleanAttribute('is_reached');
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference = DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerStatusVersionTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusVersionTargetList = DataGoogleComputeInstanceGroupManagerStatusVersionTargetList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusVersionTargetList[_m] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusVersionTargetList", version: "5.0.1" };
function dataGoogleComputeInstanceGroupManagerStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatusToTerraform = dataGoogleComputeInstanceGroupManagerStatusToTerraform;
class DataGoogleComputeInstanceGroupManagerStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // stateful - computed: true, optional: false, required: false
        this._stateful = new DataGoogleComputeInstanceGroupManagerStatusStatefulList(this, "stateful", false);
        // version_target - computed: true, optional: false, required: false
        this._versionTarget = new DataGoogleComputeInstanceGroupManagerStatusVersionTargetList(this, "version_target", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_stable - computed: true, optional: false, required: false
    get isStable() {
        return this.getBooleanAttribute('is_stable');
    }
    get stateful() {
        return this._stateful;
    }
    get versionTarget() {
        return this._versionTarget;
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusOutputReference = DataGoogleComputeInstanceGroupManagerStatusOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusList = DataGoogleComputeInstanceGroupManagerStatusList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusList[_p] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusList", version: "5.0.1" };
function dataGoogleComputeInstanceGroupManagerUpdatePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerUpdatePolicyToTerraform = dataGoogleComputeInstanceGroupManagerUpdatePolicyToTerraform;
class DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_surge_fixed - computed: true, optional: false, required: false
    get maxSurgeFixed() {
        return this.getNumberAttribute('max_surge_fixed');
    }
    // max_surge_percent - computed: true, optional: false, required: false
    get maxSurgePercent() {
        return this.getNumberAttribute('max_surge_percent');
    }
    // max_unavailable_fixed - computed: true, optional: false, required: false
    get maxUnavailableFixed() {
        return this.getNumberAttribute('max_unavailable_fixed');
    }
    // max_unavailable_percent - computed: true, optional: false, required: false
    get maxUnavailablePercent() {
        return this.getNumberAttribute('max_unavailable_percent');
    }
    // minimal_action - computed: true, optional: false, required: false
    get minimalAction() {
        return this.getStringAttribute('minimal_action');
    }
    // most_disruptive_allowed_action - computed: true, optional: false, required: false
    get mostDisruptiveAllowedAction() {
        return this.getStringAttribute('most_disruptive_allowed_action');
    }
    // replacement_method - computed: true, optional: false, required: false
    get replacementMethod() {
        return this.getStringAttribute('replacement_method');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference = DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerUpdatePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerUpdatePolicyList = DataGoogleComputeInstanceGroupManagerUpdatePolicyList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerUpdatePolicyList[_r] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerUpdatePolicyList", version: "5.0.1" };
function dataGoogleComputeInstanceGroupManagerVersionTargetSizeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerVersionTargetSizeToTerraform = dataGoogleComputeInstanceGroupManagerVersionTargetSizeToTerraform;
class DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // fixed - computed: true, optional: false, required: false
    get fixed() {
        return this.getNumberAttribute('fixed');
    }
    // percent - computed: true, optional: false, required: false
    get percent() {
        return this.getNumberAttribute('percent');
    }
}
exports.DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference = DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerVersionTargetSizeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerVersionTargetSizeList = DataGoogleComputeInstanceGroupManagerVersionTargetSizeList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerVersionTargetSizeList[_t] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerVersionTargetSizeList", version: "5.0.1" };
function dataGoogleComputeInstanceGroupManagerVersionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerVersionToTerraform = dataGoogleComputeInstanceGroupManagerVersionToTerraform;
class DataGoogleComputeInstanceGroupManagerVersionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // target_size - computed: true, optional: false, required: false
        this._targetSize = new DataGoogleComputeInstanceGroupManagerVersionTargetSizeList(this, "target_size", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // instance_template - computed: true, optional: false, required: false
    get instanceTemplate() {
        return this.getStringAttribute('instance_template');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get targetSize() {
        return this._targetSize;
    }
}
exports.DataGoogleComputeInstanceGroupManagerVersionOutputReference = DataGoogleComputeInstanceGroupManagerVersionOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerVersionOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerVersionOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGroupManagerVersionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerVersionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerVersionList = DataGoogleComputeInstanceGroupManagerVersionList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerVersionList[_v] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerVersionList", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/compute_instance_group_manager google_compute_instance_group_manager}
*/
class DataGoogleComputeInstanceGroupManager extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/compute_instance_group_manager google_compute_instance_group_manager} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeInstanceGroupManagerConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_compute_instance_group_manager',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // auto_healing_policies - computed: true, optional: false, required: false
        this._autoHealingPolicies = new DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList(this, "auto_healing_policies", false);
        // named_port - computed: true, optional: false, required: false
        this._namedPort = new DataGoogleComputeInstanceGroupManagerNamedPortList(this, "named_port", true);
        // stateful_disk - computed: true, optional: false, required: false
        this._statefulDisk = new DataGoogleComputeInstanceGroupManagerStatefulDiskList(this, "stateful_disk", true);
        // status - computed: true, optional: false, required: false
        this._status = new DataGoogleComputeInstanceGroupManagerStatusList(this, "status", false);
        // update_policy - computed: true, optional: false, required: false
        this._updatePolicy = new DataGoogleComputeInstanceGroupManagerUpdatePolicyList(this, "update_policy", false);
        // version - computed: true, optional: false, required: false
        this._version = new DataGoogleComputeInstanceGroupManagerVersionList(this, "version", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._selfLink = config.selfLink;
        this._zone = config.zone;
    }
    get autoHealingPolicies() {
        return this._autoHealingPolicies;
    }
    // base_instance_name - computed: true, optional: false, required: false
    get baseInstanceName() {
        return this.getStringAttribute('base_instance_name');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // fingerprint - computed: true, optional: false, required: false
    get fingerprint() {
        return this.getStringAttribute('fingerprint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_group - computed: true, optional: false, required: false
    get instanceGroup() {
        return this.getStringAttribute('instance_group');
    }
    // list_managed_instances_results - computed: true, optional: false, required: false
    get listManagedInstancesResults() {
        return this.getStringAttribute('list_managed_instances_results');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namedPort() {
        return this._namedPort;
    }
    // operation - computed: true, optional: false, required: false
    get operation() {
        return this.getStringAttribute('operation');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    set selfLink(value) {
        this._selfLink = value;
    }
    resetSelfLink() {
        this._selfLink = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfLinkInput() {
        return this._selfLink;
    }
    get statefulDisk() {
        return this._statefulDisk;
    }
    get status() {
        return this._status;
    }
    // target_pools - computed: true, optional: false, required: false
    get targetPools() {
        return cdktf.Fn.tolist(this.getListAttribute('target_pools'));
    }
    // target_size - computed: true, optional: false, required: false
    get targetSize() {
        return this.getNumberAttribute('target_size');
    }
    get updatePolicy() {
        return this._updatePolicy;
    }
    get version() {
        return this._version;
    }
    // wait_for_instances - computed: true, optional: false, required: false
    get waitForInstances() {
        return this.getBooleanAttribute('wait_for_instances');
    }
    // wait_for_instances_status - computed: true, optional: false, required: false
    get waitForInstancesStatus() {
        return this.getStringAttribute('wait_for_instances_status');
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            self_link: cdktf.stringToTerraform(this._selfLink),
            zone: cdktf.stringToTerraform(this._zone),
        };
    }
}
exports.DataGoogleComputeInstanceGroupManager = DataGoogleComputeInstanceGroupManager;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManager[_w] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManager", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeInstanceGroupManager.tfResourceType = "google_compute_instance_group_manager";
