// @cdktf/provider-google/lib/data-google-composer-image-versions/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComposerImageVersions = exports.DataGoogleComposerImageVersionsImageVersionsList = exports.DataGoogleComposerImageVersionsImageVersionsOutputReference = exports.dataGoogleComposerImageVersionsImageVersionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComposerImageVersionsImageVersionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComposerImageVersionsImageVersionsToTerraform = dataGoogleComposerImageVersionsImageVersionsToTerraform;
class DataGoogleComposerImageVersionsImageVersionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // image_version_id - computed: true, optional: false, required: false
    get imageVersionId() {
        return this.getStringAttribute('image_version_id');
    }
    // supported_python_versions - computed: true, optional: false, required: false
    get supportedPythonVersions() {
        return this.getListAttribute('supported_python_versions');
    }
}
exports.DataGoogleComposerImageVersionsImageVersionsOutputReference = DataGoogleComposerImageVersionsImageVersionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComposerImageVersionsImageVersionsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComposerImageVersions.DataGoogleComposerImageVersionsImageVersionsOutputReference", version: "5.0.1" };
class DataGoogleComposerImageVersionsImageVersionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComposerImageVersionsImageVersionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComposerImageVersionsImageVersionsList = DataGoogleComposerImageVersionsImageVersionsList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComposerImageVersionsImageVersionsList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComposerImageVersions.DataGoogleComposerImageVersionsImageVersionsList", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/composer_image_versions google_composer_image_versions}
*/
class DataGoogleComposerImageVersions extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/composer_image_versions google_composer_image_versions} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComposerImageVersionsConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_composer_image_versions',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // image_versions - computed: true, optional: false, required: false
        this._imageVersions = new DataGoogleComposerImageVersionsImageVersionsList(this, "image_versions", false);
        this._id = config.id;
        this._project = config.project;
        this._region = config.region;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get imageVersions() {
        return this._imageVersions;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
        };
    }
}
exports.DataGoogleComposerImageVersions = DataGoogleComposerImageVersions;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComposerImageVersions[_c] = { fqn: "@cdktf/provider-google.dataGoogleComposerImageVersions.DataGoogleComposerImageVersions", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComposerImageVersions.tfResourceType = "google_composer_image_versions";
