// @cdktf/provider-google/lib/clouddeploy-delivery-pipeline/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClouddeployDeliveryPipeline = exports.ClouddeployDeliveryPipelineTimeoutsOutputReference = exports.clouddeployDeliveryPipelineTimeoutsToTerraform = exports.ClouddeployDeliveryPipelineSerialPipelineOutputReference = exports.clouddeployDeliveryPipelineSerialPipelineToTerraform = exports.ClouddeployDeliveryPipelineSerialPipelineStagesList = exports.ClouddeployDeliveryPipelineSerialPipelineStagesOutputReference = exports.clouddeployDeliveryPipelineSerialPipelineStagesToTerraform = exports.ClouddeployDeliveryPipelineConditionList = exports.ClouddeployDeliveryPipelineConditionOutputReference = exports.clouddeployDeliveryPipelineConditionToTerraform = exports.ClouddeployDeliveryPipelineConditionTargetsPresentConditionList = exports.ClouddeployDeliveryPipelineConditionTargetsPresentConditionOutputReference = exports.clouddeployDeliveryPipelineConditionTargetsPresentConditionToTerraform = exports.ClouddeployDeliveryPipelineConditionPipelineReadyConditionList = exports.ClouddeployDeliveryPipelineConditionPipelineReadyConditionOutputReference = exports.clouddeployDeliveryPipelineConditionPipelineReadyConditionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function clouddeployDeliveryPipelineConditionPipelineReadyConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.clouddeployDeliveryPipelineConditionPipelineReadyConditionToTerraform = clouddeployDeliveryPipelineConditionPipelineReadyConditionToTerraform;
class ClouddeployDeliveryPipelineConditionPipelineReadyConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getBooleanAttribute('status');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
}
exports.ClouddeployDeliveryPipelineConditionPipelineReadyConditionOutputReference = ClouddeployDeliveryPipelineConditionPipelineReadyConditionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineConditionPipelineReadyConditionOutputReference[_a] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineConditionPipelineReadyConditionOutputReference", version: "4.0.3" };
class ClouddeployDeliveryPipelineConditionPipelineReadyConditionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ClouddeployDeliveryPipelineConditionPipelineReadyConditionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ClouddeployDeliveryPipelineConditionPipelineReadyConditionList = ClouddeployDeliveryPipelineConditionPipelineReadyConditionList;
_b = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineConditionPipelineReadyConditionList[_b] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineConditionPipelineReadyConditionList", version: "4.0.3" };
function clouddeployDeliveryPipelineConditionTargetsPresentConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.clouddeployDeliveryPipelineConditionTargetsPresentConditionToTerraform = clouddeployDeliveryPipelineConditionTargetsPresentConditionToTerraform;
class ClouddeployDeliveryPipelineConditionTargetsPresentConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // missing_targets - computed: true, optional: false, required: false
    get missingTargets() {
        return this.getListAttribute('missing_targets');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getBooleanAttribute('status');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
}
exports.ClouddeployDeliveryPipelineConditionTargetsPresentConditionOutputReference = ClouddeployDeliveryPipelineConditionTargetsPresentConditionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineConditionTargetsPresentConditionOutputReference[_c] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineConditionTargetsPresentConditionOutputReference", version: "4.0.3" };
class ClouddeployDeliveryPipelineConditionTargetsPresentConditionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ClouddeployDeliveryPipelineConditionTargetsPresentConditionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ClouddeployDeliveryPipelineConditionTargetsPresentConditionList = ClouddeployDeliveryPipelineConditionTargetsPresentConditionList;
_d = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineConditionTargetsPresentConditionList[_d] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineConditionTargetsPresentConditionList", version: "4.0.3" };
function clouddeployDeliveryPipelineConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.clouddeployDeliveryPipelineConditionToTerraform = clouddeployDeliveryPipelineConditionToTerraform;
class ClouddeployDeliveryPipelineConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // pipeline_ready_condition - computed: true, optional: false, required: false
        this._pipelineReadyCondition = new ClouddeployDeliveryPipelineConditionPipelineReadyConditionList(this, "pipeline_ready_condition", false);
        // targets_present_condition - computed: true, optional: false, required: false
        this._targetsPresentCondition = new ClouddeployDeliveryPipelineConditionTargetsPresentConditionList(this, "targets_present_condition", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get pipelineReadyCondition() {
        return this._pipelineReadyCondition;
    }
    get targetsPresentCondition() {
        return this._targetsPresentCondition;
    }
}
exports.ClouddeployDeliveryPipelineConditionOutputReference = ClouddeployDeliveryPipelineConditionOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineConditionOutputReference[_e] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineConditionOutputReference", version: "4.0.3" };
class ClouddeployDeliveryPipelineConditionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ClouddeployDeliveryPipelineConditionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ClouddeployDeliveryPipelineConditionList = ClouddeployDeliveryPipelineConditionList;
_f = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineConditionList[_f] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineConditionList", version: "4.0.3" };
function clouddeployDeliveryPipelineSerialPipelineStagesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        profiles: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.profiles),
        target_id: cdktf.stringToTerraform(struct.targetId),
    };
}
exports.clouddeployDeliveryPipelineSerialPipelineStagesToTerraform = clouddeployDeliveryPipelineSerialPipelineStagesToTerraform;
class ClouddeployDeliveryPipelineSerialPipelineStagesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._profiles !== undefined) {
            hasAnyValues = true;
            internalValueResult.profiles = this._profiles;
        }
        if (this._targetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetId = this._targetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._profiles = undefined;
            this._targetId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._profiles = value.profiles;
            this._targetId = value.targetId;
        }
    }
    get profiles() {
        return this.getListAttribute('profiles');
    }
    set profiles(value) {
        this._profiles = value;
    }
    resetProfiles() {
        this._profiles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get profilesInput() {
        return this._profiles;
    }
    get targetId() {
        return this.getStringAttribute('target_id');
    }
    set targetId(value) {
        this._targetId = value;
    }
    resetTargetId() {
        this._targetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetIdInput() {
        return this._targetId;
    }
}
exports.ClouddeployDeliveryPipelineSerialPipelineStagesOutputReference = ClouddeployDeliveryPipelineSerialPipelineStagesOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineSerialPipelineStagesOutputReference[_g] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineSerialPipelineStagesOutputReference", version: "4.0.3" };
class ClouddeployDeliveryPipelineSerialPipelineStagesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ClouddeployDeliveryPipelineSerialPipelineStagesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ClouddeployDeliveryPipelineSerialPipelineStagesList = ClouddeployDeliveryPipelineSerialPipelineStagesList;
_h = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineSerialPipelineStagesList[_h] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineSerialPipelineStagesList", version: "4.0.3" };
function clouddeployDeliveryPipelineSerialPipelineToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        stages: cdktf.listMapper(clouddeployDeliveryPipelineSerialPipelineStagesToTerraform, true)(struct.stages),
    };
}
exports.clouddeployDeliveryPipelineSerialPipelineToTerraform = clouddeployDeliveryPipelineSerialPipelineToTerraform;
class ClouddeployDeliveryPipelineSerialPipelineOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // stages - computed: false, optional: true, required: false
        this._stages = new ClouddeployDeliveryPipelineSerialPipelineStagesList(this, "stages", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._stages?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.stages = this._stages?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._stages.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._stages.internalValue = value.stages;
        }
    }
    get stages() {
        return this._stages;
    }
    putStages(value) {
        this._stages.internalValue = value;
    }
    resetStages() {
        this._stages.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stagesInput() {
        return this._stages.internalValue;
    }
}
exports.ClouddeployDeliveryPipelineSerialPipelineOutputReference = ClouddeployDeliveryPipelineSerialPipelineOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineSerialPipelineOutputReference[_j] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineSerialPipelineOutputReference", version: "4.0.3" };
function clouddeployDeliveryPipelineTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.clouddeployDeliveryPipelineTimeoutsToTerraform = clouddeployDeliveryPipelineTimeoutsToTerraform;
class ClouddeployDeliveryPipelineTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ClouddeployDeliveryPipelineTimeoutsOutputReference = ClouddeployDeliveryPipelineTimeoutsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipelineTimeoutsOutputReference[_k] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipelineTimeoutsOutputReference", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/clouddeploy_delivery_pipeline google_clouddeploy_delivery_pipeline}
*/
class ClouddeployDeliveryPipeline extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/clouddeploy_delivery_pipeline google_clouddeploy_delivery_pipeline} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ClouddeployDeliveryPipelineConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_clouddeploy_delivery_pipeline',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // condition - computed: true, optional: false, required: false
        this._condition = new ClouddeployDeliveryPipelineConditionList(this, "condition", false);
        // serial_pipeline - computed: false, optional: true, required: false
        this._serialPipeline = new ClouddeployDeliveryPipelineSerialPipelineOutputReference(this, "serial_pipeline");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ClouddeployDeliveryPipelineTimeoutsOutputReference(this, "timeouts");
        this._annotations = config.annotations;
        this._description = config.description;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._suspended = config.suspended;
        this._serialPipeline.internalValue = config.serialPipeline;
        this._timeouts.internalValue = config.timeouts;
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    get condition() {
        return this._condition;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get suspended() {
        return this.getBooleanAttribute('suspended');
    }
    set suspended(value) {
        this._suspended = value;
    }
    resetSuspended() {
        this._suspended = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suspendedInput() {
        return this._suspended;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get serialPipeline() {
        return this._serialPipeline;
    }
    putSerialPipeline(value) {
        this._serialPipeline.internalValue = value;
    }
    resetSerialPipeline() {
        this._serialPipeline.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serialPipelineInput() {
        return this._serialPipeline.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            annotations: cdktf.hashMapper(cdktf.stringToTerraform)(this._annotations),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            suspended: cdktf.booleanToTerraform(this._suspended),
            serial_pipeline: clouddeployDeliveryPipelineSerialPipelineToTerraform(this._serialPipeline.internalValue),
            timeouts: clouddeployDeliveryPipelineTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.ClouddeployDeliveryPipeline = ClouddeployDeliveryPipeline;
_l = JSII_RTTI_SYMBOL_1;
ClouddeployDeliveryPipeline[_l] = { fqn: "@cdktf/provider-google.clouddeployDeliveryPipeline.ClouddeployDeliveryPipeline", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
ClouddeployDeliveryPipeline.tfResourceType = "google_clouddeploy_delivery_pipeline";
