// @cdktf/provider-google/lib/data-google-compute-health-check/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeHealthCheck = exports.DataGoogleComputeHealthCheckTcpHealthCheckList = exports.DataGoogleComputeHealthCheckTcpHealthCheckOutputReference = exports.dataGoogleComputeHealthCheckTcpHealthCheckToTerraform = exports.DataGoogleComputeHealthCheckSslHealthCheckList = exports.DataGoogleComputeHealthCheckSslHealthCheckOutputReference = exports.dataGoogleComputeHealthCheckSslHealthCheckToTerraform = exports.DataGoogleComputeHealthCheckLogConfigList = exports.DataGoogleComputeHealthCheckLogConfigOutputReference = exports.dataGoogleComputeHealthCheckLogConfigToTerraform = exports.DataGoogleComputeHealthCheckHttpsHealthCheckList = exports.DataGoogleComputeHealthCheckHttpsHealthCheckOutputReference = exports.dataGoogleComputeHealthCheckHttpsHealthCheckToTerraform = exports.DataGoogleComputeHealthCheckHttpHealthCheckList = exports.DataGoogleComputeHealthCheckHttpHealthCheckOutputReference = exports.dataGoogleComputeHealthCheckHttpHealthCheckToTerraform = exports.DataGoogleComputeHealthCheckHttp2HealthCheckList = exports.DataGoogleComputeHealthCheckHttp2HealthCheckOutputReference = exports.dataGoogleComputeHealthCheckHttp2HealthCheckToTerraform = exports.DataGoogleComputeHealthCheckGrpcHealthCheckList = exports.DataGoogleComputeHealthCheckGrpcHealthCheckOutputReference = exports.dataGoogleComputeHealthCheckGrpcHealthCheckToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeHealthCheckGrpcHealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeHealthCheckGrpcHealthCheckToTerraform = dataGoogleComputeHealthCheckGrpcHealthCheckToTerraform;
class DataGoogleComputeHealthCheckGrpcHealthCheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // grpc_service_name - computed: true, optional: false, required: false
    get grpcServiceName() {
        return this.getStringAttribute('grpc_service_name');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // port_name - computed: true, optional: false, required: false
    get portName() {
        return this.getStringAttribute('port_name');
    }
    // port_specification - computed: true, optional: false, required: false
    get portSpecification() {
        return this.getStringAttribute('port_specification');
    }
}
exports.DataGoogleComputeHealthCheckGrpcHealthCheckOutputReference = DataGoogleComputeHealthCheckGrpcHealthCheckOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckGrpcHealthCheckOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckGrpcHealthCheckOutputReference", version: "4.0.3" };
class DataGoogleComputeHealthCheckGrpcHealthCheckList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeHealthCheckGrpcHealthCheckOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeHealthCheckGrpcHealthCheckList = DataGoogleComputeHealthCheckGrpcHealthCheckList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckGrpcHealthCheckList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckGrpcHealthCheckList", version: "4.0.3" };
function dataGoogleComputeHealthCheckHttp2HealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeHealthCheckHttp2HealthCheckToTerraform = dataGoogleComputeHealthCheckHttp2HealthCheckToTerraform;
class DataGoogleComputeHealthCheckHttp2HealthCheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // host - computed: true, optional: false, required: false
    get host() {
        return this.getStringAttribute('host');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // port_name - computed: true, optional: false, required: false
    get portName() {
        return this.getStringAttribute('port_name');
    }
    // port_specification - computed: true, optional: false, required: false
    get portSpecification() {
        return this.getStringAttribute('port_specification');
    }
    // proxy_header - computed: true, optional: false, required: false
    get proxyHeader() {
        return this.getStringAttribute('proxy_header');
    }
    // request_path - computed: true, optional: false, required: false
    get requestPath() {
        return this.getStringAttribute('request_path');
    }
    // response - computed: true, optional: false, required: false
    get response() {
        return this.getStringAttribute('response');
    }
}
exports.DataGoogleComputeHealthCheckHttp2HealthCheckOutputReference = DataGoogleComputeHealthCheckHttp2HealthCheckOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckHttp2HealthCheckOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckHttp2HealthCheckOutputReference", version: "4.0.3" };
class DataGoogleComputeHealthCheckHttp2HealthCheckList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeHealthCheckHttp2HealthCheckOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeHealthCheckHttp2HealthCheckList = DataGoogleComputeHealthCheckHttp2HealthCheckList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckHttp2HealthCheckList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckHttp2HealthCheckList", version: "4.0.3" };
function dataGoogleComputeHealthCheckHttpHealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeHealthCheckHttpHealthCheckToTerraform = dataGoogleComputeHealthCheckHttpHealthCheckToTerraform;
class DataGoogleComputeHealthCheckHttpHealthCheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // host - computed: true, optional: false, required: false
    get host() {
        return this.getStringAttribute('host');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // port_name - computed: true, optional: false, required: false
    get portName() {
        return this.getStringAttribute('port_name');
    }
    // port_specification - computed: true, optional: false, required: false
    get portSpecification() {
        return this.getStringAttribute('port_specification');
    }
    // proxy_header - computed: true, optional: false, required: false
    get proxyHeader() {
        return this.getStringAttribute('proxy_header');
    }
    // request_path - computed: true, optional: false, required: false
    get requestPath() {
        return this.getStringAttribute('request_path');
    }
    // response - computed: true, optional: false, required: false
    get response() {
        return this.getStringAttribute('response');
    }
}
exports.DataGoogleComputeHealthCheckHttpHealthCheckOutputReference = DataGoogleComputeHealthCheckHttpHealthCheckOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckHttpHealthCheckOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckHttpHealthCheckOutputReference", version: "4.0.3" };
class DataGoogleComputeHealthCheckHttpHealthCheckList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeHealthCheckHttpHealthCheckOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeHealthCheckHttpHealthCheckList = DataGoogleComputeHealthCheckHttpHealthCheckList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckHttpHealthCheckList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckHttpHealthCheckList", version: "4.0.3" };
function dataGoogleComputeHealthCheckHttpsHealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeHealthCheckHttpsHealthCheckToTerraform = dataGoogleComputeHealthCheckHttpsHealthCheckToTerraform;
class DataGoogleComputeHealthCheckHttpsHealthCheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // host - computed: true, optional: false, required: false
    get host() {
        return this.getStringAttribute('host');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // port_name - computed: true, optional: false, required: false
    get portName() {
        return this.getStringAttribute('port_name');
    }
    // port_specification - computed: true, optional: false, required: false
    get portSpecification() {
        return this.getStringAttribute('port_specification');
    }
    // proxy_header - computed: true, optional: false, required: false
    get proxyHeader() {
        return this.getStringAttribute('proxy_header');
    }
    // request_path - computed: true, optional: false, required: false
    get requestPath() {
        return this.getStringAttribute('request_path');
    }
    // response - computed: true, optional: false, required: false
    get response() {
        return this.getStringAttribute('response');
    }
}
exports.DataGoogleComputeHealthCheckHttpsHealthCheckOutputReference = DataGoogleComputeHealthCheckHttpsHealthCheckOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckHttpsHealthCheckOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckHttpsHealthCheckOutputReference", version: "4.0.3" };
class DataGoogleComputeHealthCheckHttpsHealthCheckList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeHealthCheckHttpsHealthCheckOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeHealthCheckHttpsHealthCheckList = DataGoogleComputeHealthCheckHttpsHealthCheckList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckHttpsHealthCheckList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckHttpsHealthCheckList", version: "4.0.3" };
function dataGoogleComputeHealthCheckLogConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeHealthCheckLogConfigToTerraform = dataGoogleComputeHealthCheckLogConfigToTerraform;
class DataGoogleComputeHealthCheckLogConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable - computed: true, optional: false, required: false
    get enable() {
        return this.getBooleanAttribute('enable');
    }
}
exports.DataGoogleComputeHealthCheckLogConfigOutputReference = DataGoogleComputeHealthCheckLogConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckLogConfigOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckLogConfigOutputReference", version: "4.0.3" };
class DataGoogleComputeHealthCheckLogConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeHealthCheckLogConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeHealthCheckLogConfigList = DataGoogleComputeHealthCheckLogConfigList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckLogConfigList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckLogConfigList", version: "4.0.3" };
function dataGoogleComputeHealthCheckSslHealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeHealthCheckSslHealthCheckToTerraform = dataGoogleComputeHealthCheckSslHealthCheckToTerraform;
class DataGoogleComputeHealthCheckSslHealthCheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // port_name - computed: true, optional: false, required: false
    get portName() {
        return this.getStringAttribute('port_name');
    }
    // port_specification - computed: true, optional: false, required: false
    get portSpecification() {
        return this.getStringAttribute('port_specification');
    }
    // proxy_header - computed: true, optional: false, required: false
    get proxyHeader() {
        return this.getStringAttribute('proxy_header');
    }
    // request - computed: true, optional: false, required: false
    get request() {
        return this.getStringAttribute('request');
    }
    // response - computed: true, optional: false, required: false
    get response() {
        return this.getStringAttribute('response');
    }
}
exports.DataGoogleComputeHealthCheckSslHealthCheckOutputReference = DataGoogleComputeHealthCheckSslHealthCheckOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckSslHealthCheckOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckSslHealthCheckOutputReference", version: "4.0.3" };
class DataGoogleComputeHealthCheckSslHealthCheckList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeHealthCheckSslHealthCheckOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeHealthCheckSslHealthCheckList = DataGoogleComputeHealthCheckSslHealthCheckList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckSslHealthCheckList[_m] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckSslHealthCheckList", version: "4.0.3" };
function dataGoogleComputeHealthCheckTcpHealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeHealthCheckTcpHealthCheckToTerraform = dataGoogleComputeHealthCheckTcpHealthCheckToTerraform;
class DataGoogleComputeHealthCheckTcpHealthCheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // port_name - computed: true, optional: false, required: false
    get portName() {
        return this.getStringAttribute('port_name');
    }
    // port_specification - computed: true, optional: false, required: false
    get portSpecification() {
        return this.getStringAttribute('port_specification');
    }
    // proxy_header - computed: true, optional: false, required: false
    get proxyHeader() {
        return this.getStringAttribute('proxy_header');
    }
    // request - computed: true, optional: false, required: false
    get request() {
        return this.getStringAttribute('request');
    }
    // response - computed: true, optional: false, required: false
    get response() {
        return this.getStringAttribute('response');
    }
}
exports.DataGoogleComputeHealthCheckTcpHealthCheckOutputReference = DataGoogleComputeHealthCheckTcpHealthCheckOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckTcpHealthCheckOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckTcpHealthCheckOutputReference", version: "4.0.3" };
class DataGoogleComputeHealthCheckTcpHealthCheckList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeHealthCheckTcpHealthCheckOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeHealthCheckTcpHealthCheckList = DataGoogleComputeHealthCheckTcpHealthCheckList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheckTcpHealthCheckList[_p] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheckTcpHealthCheckList", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/compute_health_check google_compute_health_check}
*/
class DataGoogleComputeHealthCheck extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/compute_health_check google_compute_health_check} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeHealthCheckConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_health_check',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // grpc_health_check - computed: true, optional: false, required: false
        this._grpcHealthCheck = new DataGoogleComputeHealthCheckGrpcHealthCheckList(this, "grpc_health_check", false);
        // http2_health_check - computed: true, optional: false, required: false
        this._http2HealthCheck = new DataGoogleComputeHealthCheckHttp2HealthCheckList(this, "http2_health_check", false);
        // http_health_check - computed: true, optional: false, required: false
        this._httpHealthCheck = new DataGoogleComputeHealthCheckHttpHealthCheckList(this, "http_health_check", false);
        // https_health_check - computed: true, optional: false, required: false
        this._httpsHealthCheck = new DataGoogleComputeHealthCheckHttpsHealthCheckList(this, "https_health_check", false);
        // log_config - computed: true, optional: false, required: false
        this._logConfig = new DataGoogleComputeHealthCheckLogConfigList(this, "log_config", false);
        // ssl_health_check - computed: true, optional: false, required: false
        this._sslHealthCheck = new DataGoogleComputeHealthCheckSslHealthCheckList(this, "ssl_health_check", false);
        // tcp_health_check - computed: true, optional: false, required: false
        this._tcpHealthCheck = new DataGoogleComputeHealthCheckTcpHealthCheckList(this, "tcp_health_check", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // check_interval_sec - computed: true, optional: false, required: false
    get checkIntervalSec() {
        return this.getNumberAttribute('check_interval_sec');
    }
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get grpcHealthCheck() {
        return this._grpcHealthCheck;
    }
    // healthy_threshold - computed: true, optional: false, required: false
    get healthyThreshold() {
        return this.getNumberAttribute('healthy_threshold');
    }
    get http2HealthCheck() {
        return this._http2HealthCheck;
    }
    get httpHealthCheck() {
        return this._httpHealthCheck;
    }
    get httpsHealthCheck() {
        return this._httpsHealthCheck;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get logConfig() {
        return this._logConfig;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get sslHealthCheck() {
        return this._sslHealthCheck;
    }
    get tcpHealthCheck() {
        return this._tcpHealthCheck;
    }
    // timeout_sec - computed: true, optional: false, required: false
    get timeoutSec() {
        return this.getNumberAttribute('timeout_sec');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // unhealthy_threshold - computed: true, optional: false, required: false
    get unhealthyThreshold() {
        return this.getNumberAttribute('unhealthy_threshold');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
}
exports.DataGoogleComputeHealthCheck = DataGoogleComputeHealthCheck;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleComputeHealthCheck[_q] = { fqn: "@cdktf/provider-google.dataGoogleComputeHealthCheck.DataGoogleComputeHealthCheck", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeHealthCheck.tfResourceType = "google_compute_health_check";
