// @cdktf/provider-google/lib/data-google-pubsub-subscription/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGooglePubsubSubscription = exports.DataGooglePubsubSubscriptionRetryPolicyList = exports.DataGooglePubsubSubscriptionRetryPolicyOutputReference = exports.dataGooglePubsubSubscriptionRetryPolicyToTerraform = exports.DataGooglePubsubSubscriptionPushConfigList = exports.DataGooglePubsubSubscriptionPushConfigOutputReference = exports.dataGooglePubsubSubscriptionPushConfigToTerraform = exports.DataGooglePubsubSubscriptionPushConfigOidcTokenList = exports.DataGooglePubsubSubscriptionPushConfigOidcTokenOutputReference = exports.dataGooglePubsubSubscriptionPushConfigOidcTokenToTerraform = exports.DataGooglePubsubSubscriptionExpirationPolicyList = exports.DataGooglePubsubSubscriptionExpirationPolicyOutputReference = exports.dataGooglePubsubSubscriptionExpirationPolicyToTerraform = exports.DataGooglePubsubSubscriptionDeadLetterPolicyList = exports.DataGooglePubsubSubscriptionDeadLetterPolicyOutputReference = exports.dataGooglePubsubSubscriptionDeadLetterPolicyToTerraform = exports.DataGooglePubsubSubscriptionBigqueryConfigList = exports.DataGooglePubsubSubscriptionBigqueryConfigOutputReference = exports.dataGooglePubsubSubscriptionBigqueryConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGooglePubsubSubscriptionBigqueryConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubSubscriptionBigqueryConfigToTerraform = dataGooglePubsubSubscriptionBigqueryConfigToTerraform;
class DataGooglePubsubSubscriptionBigqueryConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // drop_unknown_fields - computed: true, optional: false, required: false
    get dropUnknownFields() {
        return this.getBooleanAttribute('drop_unknown_fields');
    }
    // table - computed: true, optional: false, required: false
    get table() {
        return this.getStringAttribute('table');
    }
    // use_topic_schema - computed: true, optional: false, required: false
    get useTopicSchema() {
        return this.getBooleanAttribute('use_topic_schema');
    }
    // write_metadata - computed: true, optional: false, required: false
    get writeMetadata() {
        return this.getBooleanAttribute('write_metadata');
    }
}
exports.DataGooglePubsubSubscriptionBigqueryConfigOutputReference = DataGooglePubsubSubscriptionBigqueryConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionBigqueryConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionBigqueryConfigOutputReference", version: "5.0.1" };
class DataGooglePubsubSubscriptionBigqueryConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubSubscriptionBigqueryConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubSubscriptionBigqueryConfigList = DataGooglePubsubSubscriptionBigqueryConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionBigqueryConfigList[_b] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionBigqueryConfigList", version: "5.0.1" };
function dataGooglePubsubSubscriptionDeadLetterPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubSubscriptionDeadLetterPolicyToTerraform = dataGooglePubsubSubscriptionDeadLetterPolicyToTerraform;
class DataGooglePubsubSubscriptionDeadLetterPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dead_letter_topic - computed: true, optional: false, required: false
    get deadLetterTopic() {
        return this.getStringAttribute('dead_letter_topic');
    }
    // max_delivery_attempts - computed: true, optional: false, required: false
    get maxDeliveryAttempts() {
        return this.getNumberAttribute('max_delivery_attempts');
    }
}
exports.DataGooglePubsubSubscriptionDeadLetterPolicyOutputReference = DataGooglePubsubSubscriptionDeadLetterPolicyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionDeadLetterPolicyOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionDeadLetterPolicyOutputReference", version: "5.0.1" };
class DataGooglePubsubSubscriptionDeadLetterPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubSubscriptionDeadLetterPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubSubscriptionDeadLetterPolicyList = DataGooglePubsubSubscriptionDeadLetterPolicyList;
_d = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionDeadLetterPolicyList[_d] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionDeadLetterPolicyList", version: "5.0.1" };
function dataGooglePubsubSubscriptionExpirationPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubSubscriptionExpirationPolicyToTerraform = dataGooglePubsubSubscriptionExpirationPolicyToTerraform;
class DataGooglePubsubSubscriptionExpirationPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ttl - computed: true, optional: false, required: false
    get ttl() {
        return this.getStringAttribute('ttl');
    }
}
exports.DataGooglePubsubSubscriptionExpirationPolicyOutputReference = DataGooglePubsubSubscriptionExpirationPolicyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionExpirationPolicyOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionExpirationPolicyOutputReference", version: "5.0.1" };
class DataGooglePubsubSubscriptionExpirationPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubSubscriptionExpirationPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubSubscriptionExpirationPolicyList = DataGooglePubsubSubscriptionExpirationPolicyList;
_f = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionExpirationPolicyList[_f] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionExpirationPolicyList", version: "5.0.1" };
function dataGooglePubsubSubscriptionPushConfigOidcTokenToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubSubscriptionPushConfigOidcTokenToTerraform = dataGooglePubsubSubscriptionPushConfigOidcTokenToTerraform;
class DataGooglePubsubSubscriptionPushConfigOidcTokenOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // audience - computed: true, optional: false, required: false
    get audience() {
        return this.getStringAttribute('audience');
    }
    // service_account_email - computed: true, optional: false, required: false
    get serviceAccountEmail() {
        return this.getStringAttribute('service_account_email');
    }
}
exports.DataGooglePubsubSubscriptionPushConfigOidcTokenOutputReference = DataGooglePubsubSubscriptionPushConfigOidcTokenOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionPushConfigOidcTokenOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionPushConfigOidcTokenOutputReference", version: "5.0.1" };
class DataGooglePubsubSubscriptionPushConfigOidcTokenList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubSubscriptionPushConfigOidcTokenOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubSubscriptionPushConfigOidcTokenList = DataGooglePubsubSubscriptionPushConfigOidcTokenList;
_h = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionPushConfigOidcTokenList[_h] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionPushConfigOidcTokenList", version: "5.0.1" };
function dataGooglePubsubSubscriptionPushConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubSubscriptionPushConfigToTerraform = dataGooglePubsubSubscriptionPushConfigToTerraform;
class DataGooglePubsubSubscriptionPushConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // attributes - computed: true, optional: false, required: false
        this._attributes = new cdktf.StringMap(this, "attributes");
        // oidc_token - computed: true, optional: false, required: false
        this._oidcToken = new DataGooglePubsubSubscriptionPushConfigOidcTokenList(this, "oidc_token", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get attributes() {
        return this._attributes;
    }
    get oidcToken() {
        return this._oidcToken;
    }
    // push_endpoint - computed: true, optional: false, required: false
    get pushEndpoint() {
        return this.getStringAttribute('push_endpoint');
    }
}
exports.DataGooglePubsubSubscriptionPushConfigOutputReference = DataGooglePubsubSubscriptionPushConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionPushConfigOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionPushConfigOutputReference", version: "5.0.1" };
class DataGooglePubsubSubscriptionPushConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubSubscriptionPushConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubSubscriptionPushConfigList = DataGooglePubsubSubscriptionPushConfigList;
_k = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionPushConfigList[_k] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionPushConfigList", version: "5.0.1" };
function dataGooglePubsubSubscriptionRetryPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubSubscriptionRetryPolicyToTerraform = dataGooglePubsubSubscriptionRetryPolicyToTerraform;
class DataGooglePubsubSubscriptionRetryPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // maximum_backoff - computed: true, optional: false, required: false
    get maximumBackoff() {
        return this.getStringAttribute('maximum_backoff');
    }
    // minimum_backoff - computed: true, optional: false, required: false
    get minimumBackoff() {
        return this.getStringAttribute('minimum_backoff');
    }
}
exports.DataGooglePubsubSubscriptionRetryPolicyOutputReference = DataGooglePubsubSubscriptionRetryPolicyOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionRetryPolicyOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionRetryPolicyOutputReference", version: "5.0.1" };
class DataGooglePubsubSubscriptionRetryPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubSubscriptionRetryPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubSubscriptionRetryPolicyList = DataGooglePubsubSubscriptionRetryPolicyList;
_m = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscriptionRetryPolicyList[_m] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscriptionRetryPolicyList", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/pubsub_subscription google_pubsub_subscription}
*/
class DataGooglePubsubSubscription extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/pubsub_subscription google_pubsub_subscription} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGooglePubsubSubscriptionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_pubsub_subscription',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // bigquery_config - computed: true, optional: false, required: false
        this._bigqueryConfig = new DataGooglePubsubSubscriptionBigqueryConfigList(this, "bigquery_config", false);
        // dead_letter_policy - computed: true, optional: false, required: false
        this._deadLetterPolicy = new DataGooglePubsubSubscriptionDeadLetterPolicyList(this, "dead_letter_policy", false);
        // expiration_policy - computed: true, optional: false, required: false
        this._expirationPolicy = new DataGooglePubsubSubscriptionExpirationPolicyList(this, "expiration_policy", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // push_config - computed: true, optional: false, required: false
        this._pushConfig = new DataGooglePubsubSubscriptionPushConfigList(this, "push_config", false);
        // retry_policy - computed: true, optional: false, required: false
        this._retryPolicy = new DataGooglePubsubSubscriptionRetryPolicyList(this, "retry_policy", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // ack_deadline_seconds - computed: true, optional: false, required: false
    get ackDeadlineSeconds() {
        return this.getNumberAttribute('ack_deadline_seconds');
    }
    get bigqueryConfig() {
        return this._bigqueryConfig;
    }
    get deadLetterPolicy() {
        return this._deadLetterPolicy;
    }
    // enable_exactly_once_delivery - computed: true, optional: false, required: false
    get enableExactlyOnceDelivery() {
        return this.getBooleanAttribute('enable_exactly_once_delivery');
    }
    // enable_message_ordering - computed: true, optional: false, required: false
    get enableMessageOrdering() {
        return this.getBooleanAttribute('enable_message_ordering');
    }
    get expirationPolicy() {
        return this._expirationPolicy;
    }
    // filter - computed: true, optional: false, required: false
    get filter() {
        return this.getStringAttribute('filter');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    // message_retention_duration - computed: true, optional: false, required: false
    get messageRetentionDuration() {
        return this.getStringAttribute('message_retention_duration');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get pushConfig() {
        return this._pushConfig;
    }
    // retain_acked_messages - computed: true, optional: false, required: false
    get retainAckedMessages() {
        return this.getBooleanAttribute('retain_acked_messages');
    }
    get retryPolicy() {
        return this._retryPolicy;
    }
    // topic - computed: true, optional: false, required: false
    get topic() {
        return this.getStringAttribute('topic');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
}
exports.DataGooglePubsubSubscription = DataGooglePubsubSubscription;
_o = JSII_RTTI_SYMBOL_1;
DataGooglePubsubSubscription[_o] = { fqn: "@cdktf/provider-google.dataGooglePubsubSubscription.DataGooglePubsubSubscription", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
DataGooglePubsubSubscription.tfResourceType = "google_pubsub_subscription";
