// @cdktf/provider-google/lib/data-catalog-entry/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataCatalogEntry = exports.DataCatalogEntryTimeoutsOutputReference = exports.dataCatalogEntryTimeoutsToTerraform = exports.DataCatalogEntryGcsFilesetSpecOutputReference = exports.dataCatalogEntryGcsFilesetSpecToTerraform = exports.DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsList = exports.DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsOutputReference = exports.dataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsToTerraform = exports.DataCatalogEntryBigqueryTableSpecList = exports.DataCatalogEntryBigqueryTableSpecOutputReference = exports.dataCatalogEntryBigqueryTableSpecToTerraform = exports.DataCatalogEntryBigqueryTableSpecViewSpecList = exports.DataCatalogEntryBigqueryTableSpecViewSpecOutputReference = exports.dataCatalogEntryBigqueryTableSpecViewSpecToTerraform = exports.DataCatalogEntryBigqueryTableSpecTableSpecList = exports.DataCatalogEntryBigqueryTableSpecTableSpecOutputReference = exports.dataCatalogEntryBigqueryTableSpecTableSpecToTerraform = exports.DataCatalogEntryBigqueryDateShardedSpecList = exports.DataCatalogEntryBigqueryDateShardedSpecOutputReference = exports.dataCatalogEntryBigqueryDateShardedSpecToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataCatalogEntryBigqueryDateShardedSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataCatalogEntryBigqueryDateShardedSpecToTerraform = dataCatalogEntryBigqueryDateShardedSpecToTerraform;
class DataCatalogEntryBigqueryDateShardedSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dataset - computed: true, optional: false, required: false
    get dataset() {
        return this.getStringAttribute('dataset');
    }
    // shard_count - computed: true, optional: false, required: false
    get shardCount() {
        return this.getNumberAttribute('shard_count');
    }
    // table_prefix - computed: true, optional: false, required: false
    get tablePrefix() {
        return this.getStringAttribute('table_prefix');
    }
}
exports.DataCatalogEntryBigqueryDateShardedSpecOutputReference = DataCatalogEntryBigqueryDateShardedSpecOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataCatalogEntryBigqueryDateShardedSpecOutputReference[_a] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryBigqueryDateShardedSpecOutputReference", version: "4.0.3" };
class DataCatalogEntryBigqueryDateShardedSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataCatalogEntryBigqueryDateShardedSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataCatalogEntryBigqueryDateShardedSpecList = DataCatalogEntryBigqueryDateShardedSpecList;
_b = JSII_RTTI_SYMBOL_1;
DataCatalogEntryBigqueryDateShardedSpecList[_b] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryBigqueryDateShardedSpecList", version: "4.0.3" };
function dataCatalogEntryBigqueryTableSpecTableSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataCatalogEntryBigqueryTableSpecTableSpecToTerraform = dataCatalogEntryBigqueryTableSpecTableSpecToTerraform;
class DataCatalogEntryBigqueryTableSpecTableSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // grouped_entry - computed: true, optional: false, required: false
    get groupedEntry() {
        return this.getStringAttribute('grouped_entry');
    }
}
exports.DataCatalogEntryBigqueryTableSpecTableSpecOutputReference = DataCatalogEntryBigqueryTableSpecTableSpecOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataCatalogEntryBigqueryTableSpecTableSpecOutputReference[_c] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryBigqueryTableSpecTableSpecOutputReference", version: "4.0.3" };
class DataCatalogEntryBigqueryTableSpecTableSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataCatalogEntryBigqueryTableSpecTableSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataCatalogEntryBigqueryTableSpecTableSpecList = DataCatalogEntryBigqueryTableSpecTableSpecList;
_d = JSII_RTTI_SYMBOL_1;
DataCatalogEntryBigqueryTableSpecTableSpecList[_d] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryBigqueryTableSpecTableSpecList", version: "4.0.3" };
function dataCatalogEntryBigqueryTableSpecViewSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataCatalogEntryBigqueryTableSpecViewSpecToTerraform = dataCatalogEntryBigqueryTableSpecViewSpecToTerraform;
class DataCatalogEntryBigqueryTableSpecViewSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // view_query - computed: true, optional: false, required: false
    get viewQuery() {
        return this.getStringAttribute('view_query');
    }
}
exports.DataCatalogEntryBigqueryTableSpecViewSpecOutputReference = DataCatalogEntryBigqueryTableSpecViewSpecOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataCatalogEntryBigqueryTableSpecViewSpecOutputReference[_e] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryBigqueryTableSpecViewSpecOutputReference", version: "4.0.3" };
class DataCatalogEntryBigqueryTableSpecViewSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataCatalogEntryBigqueryTableSpecViewSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataCatalogEntryBigqueryTableSpecViewSpecList = DataCatalogEntryBigqueryTableSpecViewSpecList;
_f = JSII_RTTI_SYMBOL_1;
DataCatalogEntryBigqueryTableSpecViewSpecList[_f] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryBigqueryTableSpecViewSpecList", version: "4.0.3" };
function dataCatalogEntryBigqueryTableSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataCatalogEntryBigqueryTableSpecToTerraform = dataCatalogEntryBigqueryTableSpecToTerraform;
class DataCatalogEntryBigqueryTableSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // table_spec - computed: true, optional: false, required: false
        this._tableSpec = new DataCatalogEntryBigqueryTableSpecTableSpecList(this, "table_spec", false);
        // view_spec - computed: true, optional: false, required: false
        this._viewSpec = new DataCatalogEntryBigqueryTableSpecViewSpecList(this, "view_spec", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // table_source_type - computed: true, optional: false, required: false
    get tableSourceType() {
        return this.getStringAttribute('table_source_type');
    }
    get tableSpec() {
        return this._tableSpec;
    }
    get viewSpec() {
        return this._viewSpec;
    }
}
exports.DataCatalogEntryBigqueryTableSpecOutputReference = DataCatalogEntryBigqueryTableSpecOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataCatalogEntryBigqueryTableSpecOutputReference[_g] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryBigqueryTableSpecOutputReference", version: "4.0.3" };
class DataCatalogEntryBigqueryTableSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataCatalogEntryBigqueryTableSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataCatalogEntryBigqueryTableSpecList = DataCatalogEntryBigqueryTableSpecList;
_h = JSII_RTTI_SYMBOL_1;
DataCatalogEntryBigqueryTableSpecList[_h] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryBigqueryTableSpecList", version: "4.0.3" };
function dataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsToTerraform = dataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsToTerraform;
class DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // file_path - computed: true, optional: false, required: false
    get filePath() {
        return this.getStringAttribute('file_path');
    }
    // size_bytes - computed: true, optional: false, required: false
    get sizeBytes() {
        return this.getNumberAttribute('size_bytes');
    }
}
exports.DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsOutputReference = DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsOutputReference[_j] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsOutputReference", version: "4.0.3" };
class DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsList = DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsList;
_k = JSII_RTTI_SYMBOL_1;
DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsList[_k] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsList", version: "4.0.3" };
function dataCatalogEntryGcsFilesetSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        file_patterns: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.filePatterns),
    };
}
exports.dataCatalogEntryGcsFilesetSpecToTerraform = dataCatalogEntryGcsFilesetSpecToTerraform;
class DataCatalogEntryGcsFilesetSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // sample_gcs_file_specs - computed: true, optional: false, required: false
        this._sampleGcsFileSpecs = new DataCatalogEntryGcsFilesetSpecSampleGcsFileSpecsList(this, "sample_gcs_file_specs", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filePatterns !== undefined) {
            hasAnyValues = true;
            internalValueResult.filePatterns = this._filePatterns;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._filePatterns = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._filePatterns = value.filePatterns;
        }
    }
    get filePatterns() {
        return this.getListAttribute('file_patterns');
    }
    set filePatterns(value) {
        this._filePatterns = value;
    }
    // Temporarily expose input value. Use with caution.
    get filePatternsInput() {
        return this._filePatterns;
    }
    get sampleGcsFileSpecs() {
        return this._sampleGcsFileSpecs;
    }
}
exports.DataCatalogEntryGcsFilesetSpecOutputReference = DataCatalogEntryGcsFilesetSpecOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataCatalogEntryGcsFilesetSpecOutputReference[_l] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryGcsFilesetSpecOutputReference", version: "4.0.3" };
function dataCatalogEntryTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dataCatalogEntryTimeoutsToTerraform = dataCatalogEntryTimeoutsToTerraform;
class DataCatalogEntryTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DataCatalogEntryTimeoutsOutputReference = DataCatalogEntryTimeoutsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
DataCatalogEntryTimeoutsOutputReference[_m] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntryTimeoutsOutputReference", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/data_catalog_entry google_data_catalog_entry}
*/
class DataCatalogEntry extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/data_catalog_entry google_data_catalog_entry} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataCatalogEntryConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_data_catalog_entry',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // bigquery_date_sharded_spec - computed: true, optional: false, required: false
        this._bigqueryDateShardedSpec = new DataCatalogEntryBigqueryDateShardedSpecList(this, "bigquery_date_sharded_spec", false);
        // bigquery_table_spec - computed: true, optional: false, required: false
        this._bigqueryTableSpec = new DataCatalogEntryBigqueryTableSpecList(this, "bigquery_table_spec", false);
        // gcs_fileset_spec - computed: false, optional: true, required: false
        this._gcsFilesetSpec = new DataCatalogEntryGcsFilesetSpecOutputReference(this, "gcs_fileset_spec");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataCatalogEntryTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._displayName = config.displayName;
        this._entryGroup = config.entryGroup;
        this._entryId = config.entryId;
        this._id = config.id;
        this._linkedResource = config.linkedResource;
        this._schema = config.schema;
        this._type = config.type;
        this._userSpecifiedSystem = config.userSpecifiedSystem;
        this._userSpecifiedType = config.userSpecifiedType;
        this._gcsFilesetSpec.internalValue = config.gcsFilesetSpec;
        this._timeouts.internalValue = config.timeouts;
    }
    get bigqueryDateShardedSpec() {
        return this._bigqueryDateShardedSpec;
    }
    get bigqueryTableSpec() {
        return this._bigqueryTableSpec;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get entryGroup() {
        return this.getStringAttribute('entry_group');
    }
    set entryGroup(value) {
        this._entryGroup = value;
    }
    // Temporarily expose input value. Use with caution.
    get entryGroupInput() {
        return this._entryGroup;
    }
    get entryId() {
        return this.getStringAttribute('entry_id');
    }
    set entryId(value) {
        this._entryId = value;
    }
    // Temporarily expose input value. Use with caution.
    get entryIdInput() {
        return this._entryId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // integrated_system - computed: true, optional: false, required: false
    get integratedSystem() {
        return this.getStringAttribute('integrated_system');
    }
    get linkedResource() {
        return this.getStringAttribute('linked_resource');
    }
    set linkedResource(value) {
        this._linkedResource = value;
    }
    resetLinkedResource() {
        this._linkedResource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get linkedResourceInput() {
        return this._linkedResource;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get schema() {
        return this.getStringAttribute('schema');
    }
    set schema(value) {
        this._schema = value;
    }
    resetSchema() {
        this._schema = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get schemaInput() {
        return this._schema;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get userSpecifiedSystem() {
        return this.getStringAttribute('user_specified_system');
    }
    set userSpecifiedSystem(value) {
        this._userSpecifiedSystem = value;
    }
    resetUserSpecifiedSystem() {
        this._userSpecifiedSystem = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userSpecifiedSystemInput() {
        return this._userSpecifiedSystem;
    }
    get userSpecifiedType() {
        return this.getStringAttribute('user_specified_type');
    }
    set userSpecifiedType(value) {
        this._userSpecifiedType = value;
    }
    resetUserSpecifiedType() {
        this._userSpecifiedType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userSpecifiedTypeInput() {
        return this._userSpecifiedType;
    }
    get gcsFilesetSpec() {
        return this._gcsFilesetSpec;
    }
    putGcsFilesetSpec(value) {
        this._gcsFilesetSpec.internalValue = value;
    }
    resetGcsFilesetSpec() {
        this._gcsFilesetSpec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gcsFilesetSpecInput() {
        return this._gcsFilesetSpec.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            display_name: cdktf.stringToTerraform(this._displayName),
            entry_group: cdktf.stringToTerraform(this._entryGroup),
            entry_id: cdktf.stringToTerraform(this._entryId),
            id: cdktf.stringToTerraform(this._id),
            linked_resource: cdktf.stringToTerraform(this._linkedResource),
            schema: cdktf.stringToTerraform(this._schema),
            type: cdktf.stringToTerraform(this._type),
            user_specified_system: cdktf.stringToTerraform(this._userSpecifiedSystem),
            user_specified_type: cdktf.stringToTerraform(this._userSpecifiedType),
            gcs_fileset_spec: dataCatalogEntryGcsFilesetSpecToTerraform(this._gcsFilesetSpec.internalValue),
            timeouts: dataCatalogEntryTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DataCatalogEntry = DataCatalogEntry;
_o = JSII_RTTI_SYMBOL_1;
DataCatalogEntry[_o] = { fqn: "@cdktf/provider-google.dataCatalogEntry.DataCatalogEntry", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
DataCatalogEntry.tfResourceType = "google_data_catalog_entry";
