// @cdktf/provider-google/lib/dataplex-asset/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataplexAsset = exports.DataplexAssetTimeoutsOutputReference = exports.dataplexAssetTimeoutsToTerraform = exports.DataplexAssetResourceSpecOutputReference = exports.dataplexAssetResourceSpecToTerraform = exports.DataplexAssetDiscoverySpecOutputReference = exports.dataplexAssetDiscoverySpecToTerraform = exports.DataplexAssetDiscoverySpecJsonOptionsOutputReference = exports.dataplexAssetDiscoverySpecJsonOptionsToTerraform = exports.DataplexAssetDiscoverySpecCsvOptionsOutputReference = exports.dataplexAssetDiscoverySpecCsvOptionsToTerraform = exports.DataplexAssetSecurityStatusList = exports.DataplexAssetSecurityStatusOutputReference = exports.dataplexAssetSecurityStatusToTerraform = exports.DataplexAssetResourceStatusList = exports.DataplexAssetResourceStatusOutputReference = exports.dataplexAssetResourceStatusToTerraform = exports.DataplexAssetDiscoveryStatusList = exports.DataplexAssetDiscoveryStatusOutputReference = exports.dataplexAssetDiscoveryStatusToTerraform = exports.DataplexAssetDiscoveryStatusStatsList = exports.DataplexAssetDiscoveryStatusStatsOutputReference = exports.dataplexAssetDiscoveryStatusStatsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataplexAssetDiscoveryStatusStatsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataplexAssetDiscoveryStatusStatsToTerraform = dataplexAssetDiscoveryStatusStatsToTerraform;
class DataplexAssetDiscoveryStatusStatsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data_items - computed: true, optional: false, required: false
    get dataItems() {
        return this.getNumberAttribute('data_items');
    }
    // data_size - computed: true, optional: false, required: false
    get dataSize() {
        return this.getNumberAttribute('data_size');
    }
    // filesets - computed: true, optional: false, required: false
    get filesets() {
        return this.getNumberAttribute('filesets');
    }
    // tables - computed: true, optional: false, required: false
    get tables() {
        return this.getNumberAttribute('tables');
    }
}
exports.DataplexAssetDiscoveryStatusStatsOutputReference = DataplexAssetDiscoveryStatusStatsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataplexAssetDiscoveryStatusStatsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetDiscoveryStatusStatsOutputReference", version: "4.0.3" };
class DataplexAssetDiscoveryStatusStatsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataplexAssetDiscoveryStatusStatsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataplexAssetDiscoveryStatusStatsList = DataplexAssetDiscoveryStatusStatsList;
_b = JSII_RTTI_SYMBOL_1;
DataplexAssetDiscoveryStatusStatsList[_b] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetDiscoveryStatusStatsList", version: "4.0.3" };
function dataplexAssetDiscoveryStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataplexAssetDiscoveryStatusToTerraform = dataplexAssetDiscoveryStatusToTerraform;
class DataplexAssetDiscoveryStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // stats - computed: true, optional: false, required: false
        this._stats = new DataplexAssetDiscoveryStatusStatsList(this, "stats", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // last_run_duration - computed: true, optional: false, required: false
    get lastRunDuration() {
        return this.getStringAttribute('last_run_duration');
    }
    // last_run_time - computed: true, optional: false, required: false
    get lastRunTime() {
        return this.getStringAttribute('last_run_time');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get stats() {
        return this._stats;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
}
exports.DataplexAssetDiscoveryStatusOutputReference = DataplexAssetDiscoveryStatusOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataplexAssetDiscoveryStatusOutputReference[_c] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetDiscoveryStatusOutputReference", version: "4.0.3" };
class DataplexAssetDiscoveryStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataplexAssetDiscoveryStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataplexAssetDiscoveryStatusList = DataplexAssetDiscoveryStatusList;
_d = JSII_RTTI_SYMBOL_1;
DataplexAssetDiscoveryStatusList[_d] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetDiscoveryStatusList", version: "4.0.3" };
function dataplexAssetResourceStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataplexAssetResourceStatusToTerraform = dataplexAssetResourceStatusToTerraform;
class DataplexAssetResourceStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
}
exports.DataplexAssetResourceStatusOutputReference = DataplexAssetResourceStatusOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataplexAssetResourceStatusOutputReference[_e] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetResourceStatusOutputReference", version: "4.0.3" };
class DataplexAssetResourceStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataplexAssetResourceStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataplexAssetResourceStatusList = DataplexAssetResourceStatusList;
_f = JSII_RTTI_SYMBOL_1;
DataplexAssetResourceStatusList[_f] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetResourceStatusList", version: "4.0.3" };
function dataplexAssetSecurityStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataplexAssetSecurityStatusToTerraform = dataplexAssetSecurityStatusToTerraform;
class DataplexAssetSecurityStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
}
exports.DataplexAssetSecurityStatusOutputReference = DataplexAssetSecurityStatusOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataplexAssetSecurityStatusOutputReference[_g] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetSecurityStatusOutputReference", version: "4.0.3" };
class DataplexAssetSecurityStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataplexAssetSecurityStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataplexAssetSecurityStatusList = DataplexAssetSecurityStatusList;
_h = JSII_RTTI_SYMBOL_1;
DataplexAssetSecurityStatusList[_h] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetSecurityStatusList", version: "4.0.3" };
function dataplexAssetDiscoverySpecCsvOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delimiter: cdktf.stringToTerraform(struct.delimiter),
        disable_type_inference: cdktf.booleanToTerraform(struct.disableTypeInference),
        encoding: cdktf.stringToTerraform(struct.encoding),
        header_rows: cdktf.numberToTerraform(struct.headerRows),
    };
}
exports.dataplexAssetDiscoverySpecCsvOptionsToTerraform = dataplexAssetDiscoverySpecCsvOptionsToTerraform;
class DataplexAssetDiscoverySpecCsvOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._delimiter !== undefined) {
            hasAnyValues = true;
            internalValueResult.delimiter = this._delimiter;
        }
        if (this._disableTypeInference !== undefined) {
            hasAnyValues = true;
            internalValueResult.disableTypeInference = this._disableTypeInference;
        }
        if (this._encoding !== undefined) {
            hasAnyValues = true;
            internalValueResult.encoding = this._encoding;
        }
        if (this._headerRows !== undefined) {
            hasAnyValues = true;
            internalValueResult.headerRows = this._headerRows;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._delimiter = undefined;
            this._disableTypeInference = undefined;
            this._encoding = undefined;
            this._headerRows = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._delimiter = value.delimiter;
            this._disableTypeInference = value.disableTypeInference;
            this._encoding = value.encoding;
            this._headerRows = value.headerRows;
        }
    }
    get delimiter() {
        return this.getStringAttribute('delimiter');
    }
    set delimiter(value) {
        this._delimiter = value;
    }
    resetDelimiter() {
        this._delimiter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get delimiterInput() {
        return this._delimiter;
    }
    get disableTypeInference() {
        return this.getBooleanAttribute('disable_type_inference');
    }
    set disableTypeInference(value) {
        this._disableTypeInference = value;
    }
    resetDisableTypeInference() {
        this._disableTypeInference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableTypeInferenceInput() {
        return this._disableTypeInference;
    }
    get encoding() {
        return this.getStringAttribute('encoding');
    }
    set encoding(value) {
        this._encoding = value;
    }
    resetEncoding() {
        this._encoding = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encodingInput() {
        return this._encoding;
    }
    get headerRows() {
        return this.getNumberAttribute('header_rows');
    }
    set headerRows(value) {
        this._headerRows = value;
    }
    resetHeaderRows() {
        this._headerRows = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerRowsInput() {
        return this._headerRows;
    }
}
exports.DataplexAssetDiscoverySpecCsvOptionsOutputReference = DataplexAssetDiscoverySpecCsvOptionsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataplexAssetDiscoverySpecCsvOptionsOutputReference[_j] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetDiscoverySpecCsvOptionsOutputReference", version: "4.0.3" };
function dataplexAssetDiscoverySpecJsonOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disable_type_inference: cdktf.booleanToTerraform(struct.disableTypeInference),
        encoding: cdktf.stringToTerraform(struct.encoding),
    };
}
exports.dataplexAssetDiscoverySpecJsonOptionsToTerraform = dataplexAssetDiscoverySpecJsonOptionsToTerraform;
class DataplexAssetDiscoverySpecJsonOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._disableTypeInference !== undefined) {
            hasAnyValues = true;
            internalValueResult.disableTypeInference = this._disableTypeInference;
        }
        if (this._encoding !== undefined) {
            hasAnyValues = true;
            internalValueResult.encoding = this._encoding;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._disableTypeInference = undefined;
            this._encoding = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._disableTypeInference = value.disableTypeInference;
            this._encoding = value.encoding;
        }
    }
    get disableTypeInference() {
        return this.getBooleanAttribute('disable_type_inference');
    }
    set disableTypeInference(value) {
        this._disableTypeInference = value;
    }
    resetDisableTypeInference() {
        this._disableTypeInference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableTypeInferenceInput() {
        return this._disableTypeInference;
    }
    get encoding() {
        return this.getStringAttribute('encoding');
    }
    set encoding(value) {
        this._encoding = value;
    }
    resetEncoding() {
        this._encoding = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encodingInput() {
        return this._encoding;
    }
}
exports.DataplexAssetDiscoverySpecJsonOptionsOutputReference = DataplexAssetDiscoverySpecJsonOptionsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
DataplexAssetDiscoverySpecJsonOptionsOutputReference[_k] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetDiscoverySpecJsonOptionsOutputReference", version: "4.0.3" };
function dataplexAssetDiscoverySpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        exclude_patterns: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludePatterns),
        include_patterns: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includePatterns),
        schedule: cdktf.stringToTerraform(struct.schedule),
        csv_options: dataplexAssetDiscoverySpecCsvOptionsToTerraform(struct.csvOptions),
        json_options: dataplexAssetDiscoverySpecJsonOptionsToTerraform(struct.jsonOptions),
    };
}
exports.dataplexAssetDiscoverySpecToTerraform = dataplexAssetDiscoverySpecToTerraform;
class DataplexAssetDiscoverySpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // csv_options - computed: false, optional: true, required: false
        this._csvOptions = new DataplexAssetDiscoverySpecCsvOptionsOutputReference(this, "csv_options");
        // json_options - computed: false, optional: true, required: false
        this._jsonOptions = new DataplexAssetDiscoverySpecJsonOptionsOutputReference(this, "json_options");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._excludePatterns !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludePatterns = this._excludePatterns;
        }
        if (this._includePatterns !== undefined) {
            hasAnyValues = true;
            internalValueResult.includePatterns = this._includePatterns;
        }
        if (this._schedule !== undefined) {
            hasAnyValues = true;
            internalValueResult.schedule = this._schedule;
        }
        if (this._csvOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.csvOptions = this._csvOptions?.internalValue;
        }
        if (this._jsonOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.jsonOptions = this._jsonOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._excludePatterns = undefined;
            this._includePatterns = undefined;
            this._schedule = undefined;
            this._csvOptions.internalValue = undefined;
            this._jsonOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._excludePatterns = value.excludePatterns;
            this._includePatterns = value.includePatterns;
            this._schedule = value.schedule;
            this._csvOptions.internalValue = value.csvOptions;
            this._jsonOptions.internalValue = value.jsonOptions;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get excludePatterns() {
        return this.getListAttribute('exclude_patterns');
    }
    set excludePatterns(value) {
        this._excludePatterns = value;
    }
    resetExcludePatterns() {
        this._excludePatterns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludePatternsInput() {
        return this._excludePatterns;
    }
    get includePatterns() {
        return this.getListAttribute('include_patterns');
    }
    set includePatterns(value) {
        this._includePatterns = value;
    }
    resetIncludePatterns() {
        this._includePatterns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includePatternsInput() {
        return this._includePatterns;
    }
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    set schedule(value) {
        this._schedule = value;
    }
    resetSchedule() {
        this._schedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule;
    }
    get csvOptions() {
        return this._csvOptions;
    }
    putCsvOptions(value) {
        this._csvOptions.internalValue = value;
    }
    resetCsvOptions() {
        this._csvOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get csvOptionsInput() {
        return this._csvOptions.internalValue;
    }
    get jsonOptions() {
        return this._jsonOptions;
    }
    putJsonOptions(value) {
        this._jsonOptions.internalValue = value;
    }
    resetJsonOptions() {
        this._jsonOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jsonOptionsInput() {
        return this._jsonOptions.internalValue;
    }
}
exports.DataplexAssetDiscoverySpecOutputReference = DataplexAssetDiscoverySpecOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataplexAssetDiscoverySpecOutputReference[_l] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetDiscoverySpecOutputReference", version: "4.0.3" };
function dataplexAssetResourceSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.dataplexAssetResourceSpecToTerraform = dataplexAssetResourceSpecToTerraform;
class DataplexAssetResourceSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._type = value.type;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.DataplexAssetResourceSpecOutputReference = DataplexAssetResourceSpecOutputReference;
_m = JSII_RTTI_SYMBOL_1;
DataplexAssetResourceSpecOutputReference[_m] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetResourceSpecOutputReference", version: "4.0.3" };
function dataplexAssetTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dataplexAssetTimeoutsToTerraform = dataplexAssetTimeoutsToTerraform;
class DataplexAssetTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DataplexAssetTimeoutsOutputReference = DataplexAssetTimeoutsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataplexAssetTimeoutsOutputReference[_o] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAssetTimeoutsOutputReference", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/dataplex_asset google_dataplex_asset}
*/
class DataplexAsset extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/dataplex_asset google_dataplex_asset} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataplexAssetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_dataplex_asset',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // discovery_status - computed: true, optional: false, required: false
        this._discoveryStatus = new DataplexAssetDiscoveryStatusList(this, "discovery_status", false);
        // resource_status - computed: true, optional: false, required: false
        this._resourceStatus = new DataplexAssetResourceStatusList(this, "resource_status", false);
        // security_status - computed: true, optional: false, required: false
        this._securityStatus = new DataplexAssetSecurityStatusList(this, "security_status", false);
        // discovery_spec - computed: false, optional: false, required: true
        this._discoverySpec = new DataplexAssetDiscoverySpecOutputReference(this, "discovery_spec");
        // resource_spec - computed: false, optional: false, required: true
        this._resourceSpec = new DataplexAssetResourceSpecOutputReference(this, "resource_spec");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataplexAssetTimeoutsOutputReference(this, "timeouts");
        this._dataplexZone = config.dataplexZone;
        this._description = config.description;
        this._displayName = config.displayName;
        this._id = config.id;
        this._labels = config.labels;
        this._lake = config.lake;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._discoverySpec.internalValue = config.discoverySpec;
        this._resourceSpec.internalValue = config.resourceSpec;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get dataplexZone() {
        return this.getStringAttribute('dataplex_zone');
    }
    set dataplexZone(value) {
        this._dataplexZone = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataplexZoneInput() {
        return this._dataplexZone;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get discoveryStatus() {
        return this._discoveryStatus;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get lake() {
        return this.getStringAttribute('lake');
    }
    set lake(value) {
        this._lake = value;
    }
    // Temporarily expose input value. Use with caution.
    get lakeInput() {
        return this._lake;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get resourceStatus() {
        return this._resourceStatus;
    }
    get securityStatus() {
        return this._securityStatus;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get discoverySpec() {
        return this._discoverySpec;
    }
    putDiscoverySpec(value) {
        this._discoverySpec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get discoverySpecInput() {
        return this._discoverySpec.internalValue;
    }
    get resourceSpec() {
        return this._resourceSpec;
    }
    putResourceSpec(value) {
        this._resourceSpec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceSpecInput() {
        return this._resourceSpec.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            dataplex_zone: cdktf.stringToTerraform(this._dataplexZone),
            description: cdktf.stringToTerraform(this._description),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            lake: cdktf.stringToTerraform(this._lake),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            discovery_spec: dataplexAssetDiscoverySpecToTerraform(this._discoverySpec.internalValue),
            resource_spec: dataplexAssetResourceSpecToTerraform(this._resourceSpec.internalValue),
            timeouts: dataplexAssetTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DataplexAsset = DataplexAsset;
_p = JSII_RTTI_SYMBOL_1;
DataplexAsset[_p] = { fqn: "@cdktf/provider-google.dataplexAsset.DataplexAsset", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
DataplexAsset.tfResourceType = "google_dataplex_asset";
