// @cdktf/provider-google/lib/container-node-pool/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerNodePool = exports.ContainerNodePoolUpgradeSettingsOutputReference = exports.containerNodePoolUpgradeSettingsToTerraform = exports.ContainerNodePoolUpgradeSettingsBlueGreenSettingsOutputReference = exports.containerNodePoolUpgradeSettingsBlueGreenSettingsToTerraform = exports.ContainerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyOutputReference = exports.containerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyToTerraform = exports.ContainerNodePoolTimeoutsOutputReference = exports.containerNodePoolTimeoutsToTerraform = exports.ContainerNodePoolPlacementPolicyOutputReference = exports.containerNodePoolPlacementPolicyToTerraform = exports.ContainerNodePoolNodeConfigOutputReference = exports.containerNodePoolNodeConfigToTerraform = exports.ContainerNodePoolNodeConfigWorkloadMetadataConfigOutputReference = exports.containerNodePoolNodeConfigWorkloadMetadataConfigToTerraform = exports.ContainerNodePoolNodeConfigShieldedInstanceConfigOutputReference = exports.containerNodePoolNodeConfigShieldedInstanceConfigToTerraform = exports.ContainerNodePoolNodeConfigReservationAffinityOutputReference = exports.containerNodePoolNodeConfigReservationAffinityToTerraform = exports.ContainerNodePoolNodeConfigLinuxNodeConfigOutputReference = exports.containerNodePoolNodeConfigLinuxNodeConfigToTerraform = exports.ContainerNodePoolNodeConfigKubeletConfigOutputReference = exports.containerNodePoolNodeConfigKubeletConfigToTerraform = exports.ContainerNodePoolNodeConfigGvnicOutputReference = exports.containerNodePoolNodeConfigGvnicToTerraform = exports.ContainerNodePoolNodeConfigGcfsConfigOutputReference = exports.containerNodePoolNodeConfigGcfsConfigToTerraform = exports.ContainerNodePoolNodeConfigTaintList = exports.ContainerNodePoolNodeConfigTaintOutputReference = exports.containerNodePoolNodeConfigTaintToTerraform = exports.ContainerNodePoolNodeConfigGuestAcceleratorList = exports.ContainerNodePoolNodeConfigGuestAcceleratorOutputReference = exports.containerNodePoolNodeConfigGuestAcceleratorToTerraform = exports.ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigList = exports.ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigOutputReference = exports.containerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigToTerraform = exports.ContainerNodePoolNetworkConfigOutputReference = exports.containerNodePoolNetworkConfigToTerraform = exports.ContainerNodePoolManagementOutputReference = exports.containerNodePoolManagementToTerraform = exports.ContainerNodePoolAutoscalingOutputReference = exports.containerNodePoolAutoscalingToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function containerNodePoolAutoscalingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        location_policy: cdktf.stringToTerraform(struct.locationPolicy),
        max_node_count: cdktf.numberToTerraform(struct.maxNodeCount),
        min_node_count: cdktf.numberToTerraform(struct.minNodeCount),
        total_max_node_count: cdktf.numberToTerraform(struct.totalMaxNodeCount),
        total_min_node_count: cdktf.numberToTerraform(struct.totalMinNodeCount),
    };
}
exports.containerNodePoolAutoscalingToTerraform = containerNodePoolAutoscalingToTerraform;
class ContainerNodePoolAutoscalingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._locationPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.locationPolicy = this._locationPolicy;
        }
        if (this._maxNodeCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxNodeCount = this._maxNodeCount;
        }
        if (this._minNodeCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.minNodeCount = this._minNodeCount;
        }
        if (this._totalMaxNodeCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.totalMaxNodeCount = this._totalMaxNodeCount;
        }
        if (this._totalMinNodeCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.totalMinNodeCount = this._totalMinNodeCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._locationPolicy = undefined;
            this._maxNodeCount = undefined;
            this._minNodeCount = undefined;
            this._totalMaxNodeCount = undefined;
            this._totalMinNodeCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._locationPolicy = value.locationPolicy;
            this._maxNodeCount = value.maxNodeCount;
            this._minNodeCount = value.minNodeCount;
            this._totalMaxNodeCount = value.totalMaxNodeCount;
            this._totalMinNodeCount = value.totalMinNodeCount;
        }
    }
    get locationPolicy() {
        return this.getStringAttribute('location_policy');
    }
    set locationPolicy(value) {
        this._locationPolicy = value;
    }
    resetLocationPolicy() {
        this._locationPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationPolicyInput() {
        return this._locationPolicy;
    }
    get maxNodeCount() {
        return this.getNumberAttribute('max_node_count');
    }
    set maxNodeCount(value) {
        this._maxNodeCount = value;
    }
    resetMaxNodeCount() {
        this._maxNodeCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxNodeCountInput() {
        return this._maxNodeCount;
    }
    get minNodeCount() {
        return this.getNumberAttribute('min_node_count');
    }
    set minNodeCount(value) {
        this._minNodeCount = value;
    }
    resetMinNodeCount() {
        this._minNodeCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minNodeCountInput() {
        return this._minNodeCount;
    }
    get totalMaxNodeCount() {
        return this.getNumberAttribute('total_max_node_count');
    }
    set totalMaxNodeCount(value) {
        this._totalMaxNodeCount = value;
    }
    resetTotalMaxNodeCount() {
        this._totalMaxNodeCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get totalMaxNodeCountInput() {
        return this._totalMaxNodeCount;
    }
    get totalMinNodeCount() {
        return this.getNumberAttribute('total_min_node_count');
    }
    set totalMinNodeCount(value) {
        this._totalMinNodeCount = value;
    }
    resetTotalMinNodeCount() {
        this._totalMinNodeCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get totalMinNodeCountInput() {
        return this._totalMinNodeCount;
    }
}
exports.ContainerNodePoolAutoscalingOutputReference = ContainerNodePoolAutoscalingOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ContainerNodePoolAutoscalingOutputReference[_a] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolAutoscalingOutputReference", version: "5.0.1" };
function containerNodePoolManagementToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_repair: cdktf.booleanToTerraform(struct.autoRepair),
        auto_upgrade: cdktf.booleanToTerraform(struct.autoUpgrade),
    };
}
exports.containerNodePoolManagementToTerraform = containerNodePoolManagementToTerraform;
class ContainerNodePoolManagementOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoRepair !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoRepair = this._autoRepair;
        }
        if (this._autoUpgrade !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoUpgrade = this._autoUpgrade;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autoRepair = undefined;
            this._autoUpgrade = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autoRepair = value.autoRepair;
            this._autoUpgrade = value.autoUpgrade;
        }
    }
    get autoRepair() {
        return this.getBooleanAttribute('auto_repair');
    }
    set autoRepair(value) {
        this._autoRepair = value;
    }
    resetAutoRepair() {
        this._autoRepair = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoRepairInput() {
        return this._autoRepair;
    }
    get autoUpgrade() {
        return this.getBooleanAttribute('auto_upgrade');
    }
    set autoUpgrade(value) {
        this._autoUpgrade = value;
    }
    resetAutoUpgrade() {
        this._autoUpgrade = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoUpgradeInput() {
        return this._autoUpgrade;
    }
}
exports.ContainerNodePoolManagementOutputReference = ContainerNodePoolManagementOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ContainerNodePoolManagementOutputReference[_b] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolManagementOutputReference", version: "5.0.1" };
function containerNodePoolNetworkConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create_pod_range: cdktf.booleanToTerraform(struct.createPodRange),
        enable_private_nodes: cdktf.booleanToTerraform(struct.enablePrivateNodes),
        pod_ipv4_cidr_block: cdktf.stringToTerraform(struct.podIpv4CidrBlock),
        pod_range: cdktf.stringToTerraform(struct.podRange),
    };
}
exports.containerNodePoolNetworkConfigToTerraform = containerNodePoolNetworkConfigToTerraform;
class ContainerNodePoolNetworkConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._createPodRange !== undefined) {
            hasAnyValues = true;
            internalValueResult.createPodRange = this._createPodRange;
        }
        if (this._enablePrivateNodes !== undefined) {
            hasAnyValues = true;
            internalValueResult.enablePrivateNodes = this._enablePrivateNodes;
        }
        if (this._podIpv4CidrBlock !== undefined) {
            hasAnyValues = true;
            internalValueResult.podIpv4CidrBlock = this._podIpv4CidrBlock;
        }
        if (this._podRange !== undefined) {
            hasAnyValues = true;
            internalValueResult.podRange = this._podRange;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._createPodRange = undefined;
            this._enablePrivateNodes = undefined;
            this._podIpv4CidrBlock = undefined;
            this._podRange = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._createPodRange = value.createPodRange;
            this._enablePrivateNodes = value.enablePrivateNodes;
            this._podIpv4CidrBlock = value.podIpv4CidrBlock;
            this._podRange = value.podRange;
        }
    }
    get createPodRange() {
        return this.getBooleanAttribute('create_pod_range');
    }
    set createPodRange(value) {
        this._createPodRange = value;
    }
    resetCreatePodRange() {
        this._createPodRange = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createPodRangeInput() {
        return this._createPodRange;
    }
    get enablePrivateNodes() {
        return this.getBooleanAttribute('enable_private_nodes');
    }
    set enablePrivateNodes(value) {
        this._enablePrivateNodes = value;
    }
    resetEnablePrivateNodes() {
        this._enablePrivateNodes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enablePrivateNodesInput() {
        return this._enablePrivateNodes;
    }
    get podIpv4CidrBlock() {
        return this.getStringAttribute('pod_ipv4_cidr_block');
    }
    set podIpv4CidrBlock(value) {
        this._podIpv4CidrBlock = value;
    }
    resetPodIpv4CidrBlock() {
        this._podIpv4CidrBlock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get podIpv4CidrBlockInput() {
        return this._podIpv4CidrBlock;
    }
    get podRange() {
        return this.getStringAttribute('pod_range');
    }
    set podRange(value) {
        this._podRange = value;
    }
    resetPodRange() {
        this._podRange = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get podRangeInput() {
        return this._podRange;
    }
}
exports.ContainerNodePoolNetworkConfigOutputReference = ContainerNodePoolNetworkConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNetworkConfigOutputReference[_c] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNetworkConfigOutputReference", version: "5.0.1" };
function containerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        gpu_sharing_strategy: cdktf.stringToTerraform(struct.gpuSharingStrategy),
        max_shared_clients_per_gpu: cdktf.numberToTerraform(struct.maxSharedClientsPerGpu),
    };
}
exports.containerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigToTerraform = containerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigToTerraform;
class ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._gpuSharingStrategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.gpuSharingStrategy = this._gpuSharingStrategy;
        }
        if (this._maxSharedClientsPerGpu !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxSharedClientsPerGpu = this._maxSharedClientsPerGpu;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._gpuSharingStrategy = undefined;
            this._maxSharedClientsPerGpu = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._gpuSharingStrategy = value.gpuSharingStrategy;
            this._maxSharedClientsPerGpu = value.maxSharedClientsPerGpu;
        }
    }
    get gpuSharingStrategy() {
        return this.getStringAttribute('gpu_sharing_strategy');
    }
    set gpuSharingStrategy(value) {
        this._gpuSharingStrategy = value;
    }
    resetGpuSharingStrategy() {
        this._gpuSharingStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gpuSharingStrategyInput() {
        return this._gpuSharingStrategy;
    }
    get maxSharedClientsPerGpu() {
        return this.getNumberAttribute('max_shared_clients_per_gpu');
    }
    set maxSharedClientsPerGpu(value) {
        this._maxSharedClientsPerGpu = value;
    }
    resetMaxSharedClientsPerGpu() {
        this._maxSharedClientsPerGpu = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxSharedClientsPerGpuInput() {
        return this._maxSharedClientsPerGpu;
    }
}
exports.ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigOutputReference = ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigOutputReference[_d] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigOutputReference", version: "5.0.1" };
class ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigList = ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigList;
_e = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigList[_e] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigList", version: "5.0.1" };
function containerNodePoolNodeConfigGuestAcceleratorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        count: cdktf.numberToTerraform(struct.count),
        gpu_partition_size: cdktf.stringToTerraform(struct.gpuPartitionSize),
        gpu_sharing_config: cdktf.listMapper(containerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigToTerraform, false)(struct.gpuSharingConfig),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.containerNodePoolNodeConfigGuestAcceleratorToTerraform = containerNodePoolNodeConfigGuestAcceleratorToTerraform;
class ContainerNodePoolNodeConfigGuestAcceleratorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // gpu_sharing_config - computed: true, optional: true, required: false
        this._gpuSharingConfig = new ContainerNodePoolNodeConfigGuestAcceleratorGpuSharingConfigList(this, "gpu_sharing_config", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._count !== undefined) {
            hasAnyValues = true;
            internalValueResult.count = this._count;
        }
        if (this._gpuPartitionSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.gpuPartitionSize = this._gpuPartitionSize;
        }
        if (this._gpuSharingConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.gpuSharingConfig = this._gpuSharingConfig?.internalValue;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._count = undefined;
            this._gpuPartitionSize = undefined;
            this._gpuSharingConfig.internalValue = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._count = value.count;
            this._gpuPartitionSize = value.gpuPartitionSize;
            this._gpuSharingConfig.internalValue = value.gpuSharingConfig;
            this._type = value.type;
        }
    }
    get count() {
        return this.getNumberAttribute('count');
    }
    set count(value) {
        this._count = value;
    }
    resetCount() {
        this._count = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countInput() {
        return this._count;
    }
    get gpuPartitionSize() {
        return this.getStringAttribute('gpu_partition_size');
    }
    set gpuPartitionSize(value) {
        this._gpuPartitionSize = value;
    }
    resetGpuPartitionSize() {
        this._gpuPartitionSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gpuPartitionSizeInput() {
        return this._gpuPartitionSize;
    }
    get gpuSharingConfig() {
        return this._gpuSharingConfig;
    }
    putGpuSharingConfig(value) {
        this._gpuSharingConfig.internalValue = value;
    }
    resetGpuSharingConfig() {
        this._gpuSharingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gpuSharingConfigInput() {
        return this._gpuSharingConfig.internalValue;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ContainerNodePoolNodeConfigGuestAcceleratorOutputReference = ContainerNodePoolNodeConfigGuestAcceleratorOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigGuestAcceleratorOutputReference[_f] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigGuestAcceleratorOutputReference", version: "5.0.1" };
class ContainerNodePoolNodeConfigGuestAcceleratorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ContainerNodePoolNodeConfigGuestAcceleratorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ContainerNodePoolNodeConfigGuestAcceleratorList = ContainerNodePoolNodeConfigGuestAcceleratorList;
_g = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigGuestAcceleratorList[_g] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigGuestAcceleratorList", version: "5.0.1" };
function containerNodePoolNodeConfigTaintToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        effect: cdktf.stringToTerraform(struct.effect),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.containerNodePoolNodeConfigTaintToTerraform = containerNodePoolNodeConfigTaintToTerraform;
class ContainerNodePoolNodeConfigTaintOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._effect !== undefined) {
            hasAnyValues = true;
            internalValueResult.effect = this._effect;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._effect = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._effect = value.effect;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get effect() {
        return this.getStringAttribute('effect');
    }
    set effect(value) {
        this._effect = value;
    }
    resetEffect() {
        this._effect = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get effectInput() {
        return this._effect;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ContainerNodePoolNodeConfigTaintOutputReference = ContainerNodePoolNodeConfigTaintOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigTaintOutputReference[_h] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigTaintOutputReference", version: "5.0.1" };
class ContainerNodePoolNodeConfigTaintList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ContainerNodePoolNodeConfigTaintOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ContainerNodePoolNodeConfigTaintList = ContainerNodePoolNodeConfigTaintList;
_j = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigTaintList[_j] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigTaintList", version: "5.0.1" };
function containerNodePoolNodeConfigGcfsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.containerNodePoolNodeConfigGcfsConfigToTerraform = containerNodePoolNodeConfigGcfsConfigToTerraform;
class ContainerNodePoolNodeConfigGcfsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.ContainerNodePoolNodeConfigGcfsConfigOutputReference = ContainerNodePoolNodeConfigGcfsConfigOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigGcfsConfigOutputReference[_k] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigGcfsConfigOutputReference", version: "5.0.1" };
function containerNodePoolNodeConfigGvnicToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.containerNodePoolNodeConfigGvnicToTerraform = containerNodePoolNodeConfigGvnicToTerraform;
class ContainerNodePoolNodeConfigGvnicOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.ContainerNodePoolNodeConfigGvnicOutputReference = ContainerNodePoolNodeConfigGvnicOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigGvnicOutputReference[_l] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigGvnicOutputReference", version: "5.0.1" };
function containerNodePoolNodeConfigKubeletConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cpu_cfs_quota: cdktf.booleanToTerraform(struct.cpuCfsQuota),
        cpu_cfs_quota_period: cdktf.stringToTerraform(struct.cpuCfsQuotaPeriod),
        cpu_manager_policy: cdktf.stringToTerraform(struct.cpuManagerPolicy),
    };
}
exports.containerNodePoolNodeConfigKubeletConfigToTerraform = containerNodePoolNodeConfigKubeletConfigToTerraform;
class ContainerNodePoolNodeConfigKubeletConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cpuCfsQuota !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuCfsQuota = this._cpuCfsQuota;
        }
        if (this._cpuCfsQuotaPeriod !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuCfsQuotaPeriod = this._cpuCfsQuotaPeriod;
        }
        if (this._cpuManagerPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuManagerPolicy = this._cpuManagerPolicy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cpuCfsQuota = undefined;
            this._cpuCfsQuotaPeriod = undefined;
            this._cpuManagerPolicy = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cpuCfsQuota = value.cpuCfsQuota;
            this._cpuCfsQuotaPeriod = value.cpuCfsQuotaPeriod;
            this._cpuManagerPolicy = value.cpuManagerPolicy;
        }
    }
    get cpuCfsQuota() {
        return this.getBooleanAttribute('cpu_cfs_quota');
    }
    set cpuCfsQuota(value) {
        this._cpuCfsQuota = value;
    }
    resetCpuCfsQuota() {
        this._cpuCfsQuota = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuCfsQuotaInput() {
        return this._cpuCfsQuota;
    }
    get cpuCfsQuotaPeriod() {
        return this.getStringAttribute('cpu_cfs_quota_period');
    }
    set cpuCfsQuotaPeriod(value) {
        this._cpuCfsQuotaPeriod = value;
    }
    resetCpuCfsQuotaPeriod() {
        this._cpuCfsQuotaPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuCfsQuotaPeriodInput() {
        return this._cpuCfsQuotaPeriod;
    }
    get cpuManagerPolicy() {
        return this.getStringAttribute('cpu_manager_policy');
    }
    set cpuManagerPolicy(value) {
        this._cpuManagerPolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    get cpuManagerPolicyInput() {
        return this._cpuManagerPolicy;
    }
}
exports.ContainerNodePoolNodeConfigKubeletConfigOutputReference = ContainerNodePoolNodeConfigKubeletConfigOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigKubeletConfigOutputReference[_m] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigKubeletConfigOutputReference", version: "5.0.1" };
function containerNodePoolNodeConfigLinuxNodeConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sysctls: cdktf.hashMapper(cdktf.stringToTerraform)(struct.sysctls),
    };
}
exports.containerNodePoolNodeConfigLinuxNodeConfigToTerraform = containerNodePoolNodeConfigLinuxNodeConfigToTerraform;
class ContainerNodePoolNodeConfigLinuxNodeConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sysctls !== undefined) {
            hasAnyValues = true;
            internalValueResult.sysctls = this._sysctls;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sysctls = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sysctls = value.sysctls;
        }
    }
    get sysctls() {
        return this.getStringMapAttribute('sysctls');
    }
    set sysctls(value) {
        this._sysctls = value;
    }
    // Temporarily expose input value. Use with caution.
    get sysctlsInput() {
        return this._sysctls;
    }
}
exports.ContainerNodePoolNodeConfigLinuxNodeConfigOutputReference = ContainerNodePoolNodeConfigLinuxNodeConfigOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigLinuxNodeConfigOutputReference[_o] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigLinuxNodeConfigOutputReference", version: "5.0.1" };
function containerNodePoolNodeConfigReservationAffinityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        consume_reservation_type: cdktf.stringToTerraform(struct.consumeReservationType),
        key: cdktf.stringToTerraform(struct.key),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.containerNodePoolNodeConfigReservationAffinityToTerraform = containerNodePoolNodeConfigReservationAffinityToTerraform;
class ContainerNodePoolNodeConfigReservationAffinityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._consumeReservationType !== undefined) {
            hasAnyValues = true;
            internalValueResult.consumeReservationType = this._consumeReservationType;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._consumeReservationType = undefined;
            this._key = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._consumeReservationType = value.consumeReservationType;
            this._key = value.key;
            this._values = value.values;
        }
    }
    get consumeReservationType() {
        return this.getStringAttribute('consume_reservation_type');
    }
    set consumeReservationType(value) {
        this._consumeReservationType = value;
    }
    // Temporarily expose input value. Use with caution.
    get consumeReservationTypeInput() {
        return this._consumeReservationType;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.ContainerNodePoolNodeConfigReservationAffinityOutputReference = ContainerNodePoolNodeConfigReservationAffinityOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigReservationAffinityOutputReference[_p] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigReservationAffinityOutputReference", version: "5.0.1" };
function containerNodePoolNodeConfigShieldedInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_integrity_monitoring: cdktf.booleanToTerraform(struct.enableIntegrityMonitoring),
        enable_secure_boot: cdktf.booleanToTerraform(struct.enableSecureBoot),
    };
}
exports.containerNodePoolNodeConfigShieldedInstanceConfigToTerraform = containerNodePoolNodeConfigShieldedInstanceConfigToTerraform;
class ContainerNodePoolNodeConfigShieldedInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableIntegrityMonitoring !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableIntegrityMonitoring = this._enableIntegrityMonitoring;
        }
        if (this._enableSecureBoot !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableSecureBoot = this._enableSecureBoot;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableIntegrityMonitoring = undefined;
            this._enableSecureBoot = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableIntegrityMonitoring = value.enableIntegrityMonitoring;
            this._enableSecureBoot = value.enableSecureBoot;
        }
    }
    get enableIntegrityMonitoring() {
        return this.getBooleanAttribute('enable_integrity_monitoring');
    }
    set enableIntegrityMonitoring(value) {
        this._enableIntegrityMonitoring = value;
    }
    resetEnableIntegrityMonitoring() {
        this._enableIntegrityMonitoring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableIntegrityMonitoringInput() {
        return this._enableIntegrityMonitoring;
    }
    get enableSecureBoot() {
        return this.getBooleanAttribute('enable_secure_boot');
    }
    set enableSecureBoot(value) {
        this._enableSecureBoot = value;
    }
    resetEnableSecureBoot() {
        this._enableSecureBoot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableSecureBootInput() {
        return this._enableSecureBoot;
    }
}
exports.ContainerNodePoolNodeConfigShieldedInstanceConfigOutputReference = ContainerNodePoolNodeConfigShieldedInstanceConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigShieldedInstanceConfigOutputReference[_q] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigShieldedInstanceConfigOutputReference", version: "5.0.1" };
function containerNodePoolNodeConfigWorkloadMetadataConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.stringToTerraform(struct.mode),
    };
}
exports.containerNodePoolNodeConfigWorkloadMetadataConfigToTerraform = containerNodePoolNodeConfigWorkloadMetadataConfigToTerraform;
class ContainerNodePoolNodeConfigWorkloadMetadataConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mode = value.mode;
        }
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
}
exports.ContainerNodePoolNodeConfigWorkloadMetadataConfigOutputReference = ContainerNodePoolNodeConfigWorkloadMetadataConfigOutputReference;
_r = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigWorkloadMetadataConfigOutputReference[_r] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigWorkloadMetadataConfigOutputReference", version: "5.0.1" };
function containerNodePoolNodeConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        boot_disk_kms_key: cdktf.stringToTerraform(struct.bootDiskKmsKey),
        disk_size_gb: cdktf.numberToTerraform(struct.diskSizeGb),
        disk_type: cdktf.stringToTerraform(struct.diskType),
        guest_accelerator: cdktf.listMapper(containerNodePoolNodeConfigGuestAcceleratorToTerraform, false)(struct.guestAccelerator),
        image_type: cdktf.stringToTerraform(struct.imageType),
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        local_ssd_count: cdktf.numberToTerraform(struct.localSsdCount),
        logging_variant: cdktf.stringToTerraform(struct.loggingVariant),
        machine_type: cdktf.stringToTerraform(struct.machineType),
        metadata: cdktf.hashMapper(cdktf.stringToTerraform)(struct.metadata),
        min_cpu_platform: cdktf.stringToTerraform(struct.minCpuPlatform),
        node_group: cdktf.stringToTerraform(struct.nodeGroup),
        oauth_scopes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.oauthScopes),
        preemptible: cdktf.booleanToTerraform(struct.preemptible),
        resource_labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.resourceLabels),
        service_account: cdktf.stringToTerraform(struct.serviceAccount),
        spot: cdktf.booleanToTerraform(struct.spot),
        tags: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tags),
        taint: cdktf.listMapper(containerNodePoolNodeConfigTaintToTerraform, false)(struct.taint),
        gcfs_config: containerNodePoolNodeConfigGcfsConfigToTerraform(struct.gcfsConfig),
        gvnic: containerNodePoolNodeConfigGvnicToTerraform(struct.gvnic),
        kubelet_config: containerNodePoolNodeConfigKubeletConfigToTerraform(struct.kubeletConfig),
        linux_node_config: containerNodePoolNodeConfigLinuxNodeConfigToTerraform(struct.linuxNodeConfig),
        reservation_affinity: containerNodePoolNodeConfigReservationAffinityToTerraform(struct.reservationAffinity),
        shielded_instance_config: containerNodePoolNodeConfigShieldedInstanceConfigToTerraform(struct.shieldedInstanceConfig),
        workload_metadata_config: containerNodePoolNodeConfigWorkloadMetadataConfigToTerraform(struct.workloadMetadataConfig),
    };
}
exports.containerNodePoolNodeConfigToTerraform = containerNodePoolNodeConfigToTerraform;
class ContainerNodePoolNodeConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // guest_accelerator - computed: true, optional: true, required: false
        this._guestAccelerator = new ContainerNodePoolNodeConfigGuestAcceleratorList(this, "guest_accelerator", false);
        // taint - computed: true, optional: true, required: false
        this._taint = new ContainerNodePoolNodeConfigTaintList(this, "taint", false);
        // gcfs_config - computed: false, optional: true, required: false
        this._gcfsConfig = new ContainerNodePoolNodeConfigGcfsConfigOutputReference(this, "gcfs_config");
        // gvnic - computed: false, optional: true, required: false
        this._gvnic = new ContainerNodePoolNodeConfigGvnicOutputReference(this, "gvnic");
        // kubelet_config - computed: false, optional: true, required: false
        this._kubeletConfig = new ContainerNodePoolNodeConfigKubeletConfigOutputReference(this, "kubelet_config");
        // linux_node_config - computed: false, optional: true, required: false
        this._linuxNodeConfig = new ContainerNodePoolNodeConfigLinuxNodeConfigOutputReference(this, "linux_node_config");
        // reservation_affinity - computed: false, optional: true, required: false
        this._reservationAffinity = new ContainerNodePoolNodeConfigReservationAffinityOutputReference(this, "reservation_affinity");
        // shielded_instance_config - computed: false, optional: true, required: false
        this._shieldedInstanceConfig = new ContainerNodePoolNodeConfigShieldedInstanceConfigOutputReference(this, "shielded_instance_config");
        // workload_metadata_config - computed: false, optional: true, required: false
        this._workloadMetadataConfig = new ContainerNodePoolNodeConfigWorkloadMetadataConfigOutputReference(this, "workload_metadata_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bootDiskKmsKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.bootDiskKmsKey = this._bootDiskKmsKey;
        }
        if (this._diskSizeGb !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSizeGb = this._diskSizeGb;
        }
        if (this._diskType !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskType = this._diskType;
        }
        if (this._guestAccelerator?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.guestAccelerator = this._guestAccelerator?.internalValue;
        }
        if (this._imageType !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageType = this._imageType;
        }
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._localSsdCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.localSsdCount = this._localSsdCount;
        }
        if (this._loggingVariant !== undefined) {
            hasAnyValues = true;
            internalValueResult.loggingVariant = this._loggingVariant;
        }
        if (this._machineType !== undefined) {
            hasAnyValues = true;
            internalValueResult.machineType = this._machineType;
        }
        if (this._metadata !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata;
        }
        if (this._minCpuPlatform !== undefined) {
            hasAnyValues = true;
            internalValueResult.minCpuPlatform = this._minCpuPlatform;
        }
        if (this._nodeGroup !== undefined) {
            hasAnyValues = true;
            internalValueResult.nodeGroup = this._nodeGroup;
        }
        if (this._oauthScopes !== undefined) {
            hasAnyValues = true;
            internalValueResult.oauthScopes = this._oauthScopes;
        }
        if (this._preemptible !== undefined) {
            hasAnyValues = true;
            internalValueResult.preemptible = this._preemptible;
        }
        if (this._resourceLabels !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceLabels = this._resourceLabels;
        }
        if (this._serviceAccount !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccount = this._serviceAccount;
        }
        if (this._spot !== undefined) {
            hasAnyValues = true;
            internalValueResult.spot = this._spot;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        if (this._taint?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.taint = this._taint?.internalValue;
        }
        if (this._gcfsConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.gcfsConfig = this._gcfsConfig?.internalValue;
        }
        if (this._gvnic?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.gvnic = this._gvnic?.internalValue;
        }
        if (this._kubeletConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kubeletConfig = this._kubeletConfig?.internalValue;
        }
        if (this._linuxNodeConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.linuxNodeConfig = this._linuxNodeConfig?.internalValue;
        }
        if (this._reservationAffinity?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.reservationAffinity = this._reservationAffinity?.internalValue;
        }
        if (this._shieldedInstanceConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.shieldedInstanceConfig = this._shieldedInstanceConfig?.internalValue;
        }
        if (this._workloadMetadataConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.workloadMetadataConfig = this._workloadMetadataConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bootDiskKmsKey = undefined;
            this._diskSizeGb = undefined;
            this._diskType = undefined;
            this._guestAccelerator.internalValue = undefined;
            this._imageType = undefined;
            this._labels = undefined;
            this._localSsdCount = undefined;
            this._loggingVariant = undefined;
            this._machineType = undefined;
            this._metadata = undefined;
            this._minCpuPlatform = undefined;
            this._nodeGroup = undefined;
            this._oauthScopes = undefined;
            this._preemptible = undefined;
            this._resourceLabels = undefined;
            this._serviceAccount = undefined;
            this._spot = undefined;
            this._tags = undefined;
            this._taint.internalValue = undefined;
            this._gcfsConfig.internalValue = undefined;
            this._gvnic.internalValue = undefined;
            this._kubeletConfig.internalValue = undefined;
            this._linuxNodeConfig.internalValue = undefined;
            this._reservationAffinity.internalValue = undefined;
            this._shieldedInstanceConfig.internalValue = undefined;
            this._workloadMetadataConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bootDiskKmsKey = value.bootDiskKmsKey;
            this._diskSizeGb = value.diskSizeGb;
            this._diskType = value.diskType;
            this._guestAccelerator.internalValue = value.guestAccelerator;
            this._imageType = value.imageType;
            this._labels = value.labels;
            this._localSsdCount = value.localSsdCount;
            this._loggingVariant = value.loggingVariant;
            this._machineType = value.machineType;
            this._metadata = value.metadata;
            this._minCpuPlatform = value.minCpuPlatform;
            this._nodeGroup = value.nodeGroup;
            this._oauthScopes = value.oauthScopes;
            this._preemptible = value.preemptible;
            this._resourceLabels = value.resourceLabels;
            this._serviceAccount = value.serviceAccount;
            this._spot = value.spot;
            this._tags = value.tags;
            this._taint.internalValue = value.taint;
            this._gcfsConfig.internalValue = value.gcfsConfig;
            this._gvnic.internalValue = value.gvnic;
            this._kubeletConfig.internalValue = value.kubeletConfig;
            this._linuxNodeConfig.internalValue = value.linuxNodeConfig;
            this._reservationAffinity.internalValue = value.reservationAffinity;
            this._shieldedInstanceConfig.internalValue = value.shieldedInstanceConfig;
            this._workloadMetadataConfig.internalValue = value.workloadMetadataConfig;
        }
    }
    get bootDiskKmsKey() {
        return this.getStringAttribute('boot_disk_kms_key');
    }
    set bootDiskKmsKey(value) {
        this._bootDiskKmsKey = value;
    }
    resetBootDiskKmsKey() {
        this._bootDiskKmsKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bootDiskKmsKeyInput() {
        return this._bootDiskKmsKey;
    }
    get diskSizeGb() {
        return this.getNumberAttribute('disk_size_gb');
    }
    set diskSizeGb(value) {
        this._diskSizeGb = value;
    }
    resetDiskSizeGb() {
        this._diskSizeGb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeGbInput() {
        return this._diskSizeGb;
    }
    get diskType() {
        return this.getStringAttribute('disk_type');
    }
    set diskType(value) {
        this._diskType = value;
    }
    resetDiskType() {
        this._diskType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeInput() {
        return this._diskType;
    }
    get guestAccelerator() {
        return this._guestAccelerator;
    }
    putGuestAccelerator(value) {
        this._guestAccelerator.internalValue = value;
    }
    resetGuestAccelerator() {
        this._guestAccelerator.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get guestAcceleratorInput() {
        return this._guestAccelerator.internalValue;
    }
    get imageType() {
        return this.getStringAttribute('image_type');
    }
    set imageType(value) {
        this._imageType = value;
    }
    resetImageType() {
        this._imageType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageTypeInput() {
        return this._imageType;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get localSsdCount() {
        return this.getNumberAttribute('local_ssd_count');
    }
    set localSsdCount(value) {
        this._localSsdCount = value;
    }
    resetLocalSsdCount() {
        this._localSsdCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get localSsdCountInput() {
        return this._localSsdCount;
    }
    get loggingVariant() {
        return this.getStringAttribute('logging_variant');
    }
    set loggingVariant(value) {
        this._loggingVariant = value;
    }
    resetLoggingVariant() {
        this._loggingVariant = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loggingVariantInput() {
        return this._loggingVariant;
    }
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    set machineType(value) {
        this._machineType = value;
    }
    resetMachineType() {
        this._machineType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get machineTypeInput() {
        return this._machineType;
    }
    get metadata() {
        return this.getStringMapAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    resetMetadata() {
        this._metadata = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
    get minCpuPlatform() {
        return this.getStringAttribute('min_cpu_platform');
    }
    set minCpuPlatform(value) {
        this._minCpuPlatform = value;
    }
    resetMinCpuPlatform() {
        this._minCpuPlatform = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minCpuPlatformInput() {
        return this._minCpuPlatform;
    }
    get nodeGroup() {
        return this.getStringAttribute('node_group');
    }
    set nodeGroup(value) {
        this._nodeGroup = value;
    }
    resetNodeGroup() {
        this._nodeGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeGroupInput() {
        return this._nodeGroup;
    }
    get oauthScopes() {
        return cdktf.Fn.tolist(this.getListAttribute('oauth_scopes'));
    }
    set oauthScopes(value) {
        this._oauthScopes = value;
    }
    resetOauthScopes() {
        this._oauthScopes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get oauthScopesInput() {
        return this._oauthScopes;
    }
    get preemptible() {
        return this.getBooleanAttribute('preemptible');
    }
    set preemptible(value) {
        this._preemptible = value;
    }
    resetPreemptible() {
        this._preemptible = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preemptibleInput() {
        return this._preemptible;
    }
    get resourceLabels() {
        return this.getStringMapAttribute('resource_labels');
    }
    set resourceLabels(value) {
        this._resourceLabels = value;
    }
    resetResourceLabels() {
        this._resourceLabels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceLabelsInput() {
        return this._resourceLabels;
    }
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    set serviceAccount(value) {
        this._serviceAccount = value;
    }
    resetServiceAccount() {
        this._serviceAccount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountInput() {
        return this._serviceAccount;
    }
    get spot() {
        return this.getBooleanAttribute('spot');
    }
    set spot(value) {
        this._spot = value;
    }
    resetSpot() {
        this._spot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get spotInput() {
        return this._spot;
    }
    get tags() {
        return this.getListAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get taint() {
        return this._taint;
    }
    putTaint(value) {
        this._taint.internalValue = value;
    }
    resetTaint() {
        this._taint.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taintInput() {
        return this._taint.internalValue;
    }
    get gcfsConfig() {
        return this._gcfsConfig;
    }
    putGcfsConfig(value) {
        this._gcfsConfig.internalValue = value;
    }
    resetGcfsConfig() {
        this._gcfsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gcfsConfigInput() {
        return this._gcfsConfig.internalValue;
    }
    get gvnic() {
        return this._gvnic;
    }
    putGvnic(value) {
        this._gvnic.internalValue = value;
    }
    resetGvnic() {
        this._gvnic.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gvnicInput() {
        return this._gvnic.internalValue;
    }
    get kubeletConfig() {
        return this._kubeletConfig;
    }
    putKubeletConfig(value) {
        this._kubeletConfig.internalValue = value;
    }
    resetKubeletConfig() {
        this._kubeletConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kubeletConfigInput() {
        return this._kubeletConfig.internalValue;
    }
    get linuxNodeConfig() {
        return this._linuxNodeConfig;
    }
    putLinuxNodeConfig(value) {
        this._linuxNodeConfig.internalValue = value;
    }
    resetLinuxNodeConfig() {
        this._linuxNodeConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get linuxNodeConfigInput() {
        return this._linuxNodeConfig.internalValue;
    }
    get reservationAffinity() {
        return this._reservationAffinity;
    }
    putReservationAffinity(value) {
        this._reservationAffinity.internalValue = value;
    }
    resetReservationAffinity() {
        this._reservationAffinity.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reservationAffinityInput() {
        return this._reservationAffinity.internalValue;
    }
    get shieldedInstanceConfig() {
        return this._shieldedInstanceConfig;
    }
    putShieldedInstanceConfig(value) {
        this._shieldedInstanceConfig.internalValue = value;
    }
    resetShieldedInstanceConfig() {
        this._shieldedInstanceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shieldedInstanceConfigInput() {
        return this._shieldedInstanceConfig.internalValue;
    }
    get workloadMetadataConfig() {
        return this._workloadMetadataConfig;
    }
    putWorkloadMetadataConfig(value) {
        this._workloadMetadataConfig.internalValue = value;
    }
    resetWorkloadMetadataConfig() {
        this._workloadMetadataConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workloadMetadataConfigInput() {
        return this._workloadMetadataConfig.internalValue;
    }
}
exports.ContainerNodePoolNodeConfigOutputReference = ContainerNodePoolNodeConfigOutputReference;
_s = JSII_RTTI_SYMBOL_1;
ContainerNodePoolNodeConfigOutputReference[_s] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolNodeConfigOutputReference", version: "5.0.1" };
function containerNodePoolPlacementPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.containerNodePoolPlacementPolicyToTerraform = containerNodePoolPlacementPolicyToTerraform;
class ContainerNodePoolPlacementPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ContainerNodePoolPlacementPolicyOutputReference = ContainerNodePoolPlacementPolicyOutputReference;
_t = JSII_RTTI_SYMBOL_1;
ContainerNodePoolPlacementPolicyOutputReference[_t] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolPlacementPolicyOutputReference", version: "5.0.1" };
function containerNodePoolTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.containerNodePoolTimeoutsToTerraform = containerNodePoolTimeoutsToTerraform;
class ContainerNodePoolTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ContainerNodePoolTimeoutsOutputReference = ContainerNodePoolTimeoutsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
ContainerNodePoolTimeoutsOutputReference[_u] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolTimeoutsOutputReference", version: "5.0.1" };
function containerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        batch_node_count: cdktf.numberToTerraform(struct.batchNodeCount),
        batch_percentage: cdktf.numberToTerraform(struct.batchPercentage),
        batch_soak_duration: cdktf.stringToTerraform(struct.batchSoakDuration),
    };
}
exports.containerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyToTerraform = containerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyToTerraform;
class ContainerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._batchNodeCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.batchNodeCount = this._batchNodeCount;
        }
        if (this._batchPercentage !== undefined) {
            hasAnyValues = true;
            internalValueResult.batchPercentage = this._batchPercentage;
        }
        if (this._batchSoakDuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.batchSoakDuration = this._batchSoakDuration;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._batchNodeCount = undefined;
            this._batchPercentage = undefined;
            this._batchSoakDuration = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._batchNodeCount = value.batchNodeCount;
            this._batchPercentage = value.batchPercentage;
            this._batchSoakDuration = value.batchSoakDuration;
        }
    }
    get batchNodeCount() {
        return this.getNumberAttribute('batch_node_count');
    }
    set batchNodeCount(value) {
        this._batchNodeCount = value;
    }
    resetBatchNodeCount() {
        this._batchNodeCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get batchNodeCountInput() {
        return this._batchNodeCount;
    }
    get batchPercentage() {
        return this.getNumberAttribute('batch_percentage');
    }
    set batchPercentage(value) {
        this._batchPercentage = value;
    }
    resetBatchPercentage() {
        this._batchPercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get batchPercentageInput() {
        return this._batchPercentage;
    }
    get batchSoakDuration() {
        return this.getStringAttribute('batch_soak_duration');
    }
    set batchSoakDuration(value) {
        this._batchSoakDuration = value;
    }
    resetBatchSoakDuration() {
        this._batchSoakDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get batchSoakDurationInput() {
        return this._batchSoakDuration;
    }
}
exports.ContainerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyOutputReference = ContainerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyOutputReference;
_v = JSII_RTTI_SYMBOL_1;
ContainerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyOutputReference[_v] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyOutputReference", version: "5.0.1" };
function containerNodePoolUpgradeSettingsBlueGreenSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        node_pool_soak_duration: cdktf.stringToTerraform(struct.nodePoolSoakDuration),
        standard_rollout_policy: containerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyToTerraform(struct.standardRolloutPolicy),
    };
}
exports.containerNodePoolUpgradeSettingsBlueGreenSettingsToTerraform = containerNodePoolUpgradeSettingsBlueGreenSettingsToTerraform;
class ContainerNodePoolUpgradeSettingsBlueGreenSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // standard_rollout_policy - computed: false, optional: false, required: true
        this._standardRolloutPolicy = new ContainerNodePoolUpgradeSettingsBlueGreenSettingsStandardRolloutPolicyOutputReference(this, "standard_rollout_policy");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._nodePoolSoakDuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.nodePoolSoakDuration = this._nodePoolSoakDuration;
        }
        if (this._standardRolloutPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.standardRolloutPolicy = this._standardRolloutPolicy?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._nodePoolSoakDuration = undefined;
            this._standardRolloutPolicy.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._nodePoolSoakDuration = value.nodePoolSoakDuration;
            this._standardRolloutPolicy.internalValue = value.standardRolloutPolicy;
        }
    }
    get nodePoolSoakDuration() {
        return this.getStringAttribute('node_pool_soak_duration');
    }
    set nodePoolSoakDuration(value) {
        this._nodePoolSoakDuration = value;
    }
    resetNodePoolSoakDuration() {
        this._nodePoolSoakDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodePoolSoakDurationInput() {
        return this._nodePoolSoakDuration;
    }
    get standardRolloutPolicy() {
        return this._standardRolloutPolicy;
    }
    putStandardRolloutPolicy(value) {
        this._standardRolloutPolicy.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get standardRolloutPolicyInput() {
        return this._standardRolloutPolicy.internalValue;
    }
}
exports.ContainerNodePoolUpgradeSettingsBlueGreenSettingsOutputReference = ContainerNodePoolUpgradeSettingsBlueGreenSettingsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
ContainerNodePoolUpgradeSettingsBlueGreenSettingsOutputReference[_w] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolUpgradeSettingsBlueGreenSettingsOutputReference", version: "5.0.1" };
function containerNodePoolUpgradeSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_surge: cdktf.numberToTerraform(struct.maxSurge),
        max_unavailable: cdktf.numberToTerraform(struct.maxUnavailable),
        strategy: cdktf.stringToTerraform(struct.strategy),
        blue_green_settings: containerNodePoolUpgradeSettingsBlueGreenSettingsToTerraform(struct.blueGreenSettings),
    };
}
exports.containerNodePoolUpgradeSettingsToTerraform = containerNodePoolUpgradeSettingsToTerraform;
class ContainerNodePoolUpgradeSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // blue_green_settings - computed: false, optional: true, required: false
        this._blueGreenSettings = new ContainerNodePoolUpgradeSettingsBlueGreenSettingsOutputReference(this, "blue_green_settings");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxSurge !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxSurge = this._maxSurge;
        }
        if (this._maxUnavailable !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxUnavailable = this._maxUnavailable;
        }
        if (this._strategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.strategy = this._strategy;
        }
        if (this._blueGreenSettings?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.blueGreenSettings = this._blueGreenSettings?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxSurge = undefined;
            this._maxUnavailable = undefined;
            this._strategy = undefined;
            this._blueGreenSettings.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxSurge = value.maxSurge;
            this._maxUnavailable = value.maxUnavailable;
            this._strategy = value.strategy;
            this._blueGreenSettings.internalValue = value.blueGreenSettings;
        }
    }
    get maxSurge() {
        return this.getNumberAttribute('max_surge');
    }
    set maxSurge(value) {
        this._maxSurge = value;
    }
    resetMaxSurge() {
        this._maxSurge = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxSurgeInput() {
        return this._maxSurge;
    }
    get maxUnavailable() {
        return this.getNumberAttribute('max_unavailable');
    }
    set maxUnavailable(value) {
        this._maxUnavailable = value;
    }
    resetMaxUnavailable() {
        this._maxUnavailable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxUnavailableInput() {
        return this._maxUnavailable;
    }
    get strategy() {
        return this.getStringAttribute('strategy');
    }
    set strategy(value) {
        this._strategy = value;
    }
    resetStrategy() {
        this._strategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get strategyInput() {
        return this._strategy;
    }
    get blueGreenSettings() {
        return this._blueGreenSettings;
    }
    putBlueGreenSettings(value) {
        this._blueGreenSettings.internalValue = value;
    }
    resetBlueGreenSettings() {
        this._blueGreenSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get blueGreenSettingsInput() {
        return this._blueGreenSettings.internalValue;
    }
}
exports.ContainerNodePoolUpgradeSettingsOutputReference = ContainerNodePoolUpgradeSettingsOutputReference;
_x = JSII_RTTI_SYMBOL_1;
ContainerNodePoolUpgradeSettingsOutputReference[_x] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePoolUpgradeSettingsOutputReference", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/container_node_pool google_container_node_pool}
*/
class ContainerNodePool extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/container_node_pool google_container_node_pool} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ContainerNodePoolConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_container_node_pool',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // autoscaling - computed: false, optional: true, required: false
        this._autoscaling = new ContainerNodePoolAutoscalingOutputReference(this, "autoscaling");
        // management - computed: false, optional: true, required: false
        this._management = new ContainerNodePoolManagementOutputReference(this, "management");
        // network_config - computed: false, optional: true, required: false
        this._networkConfig = new ContainerNodePoolNetworkConfigOutputReference(this, "network_config");
        // node_config - computed: false, optional: true, required: false
        this._nodeConfig = new ContainerNodePoolNodeConfigOutputReference(this, "node_config");
        // placement_policy - computed: false, optional: true, required: false
        this._placementPolicy = new ContainerNodePoolPlacementPolicyOutputReference(this, "placement_policy");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ContainerNodePoolTimeoutsOutputReference(this, "timeouts");
        // upgrade_settings - computed: false, optional: true, required: false
        this._upgradeSettings = new ContainerNodePoolUpgradeSettingsOutputReference(this, "upgrade_settings");
        this._cluster = config.cluster;
        this._id = config.id;
        this._initialNodeCount = config.initialNodeCount;
        this._location = config.location;
        this._maxPodsPerNode = config.maxPodsPerNode;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._nodeCount = config.nodeCount;
        this._nodeLocations = config.nodeLocations;
        this._project = config.project;
        this._version = config.version;
        this._autoscaling.internalValue = config.autoscaling;
        this._management.internalValue = config.management;
        this._networkConfig.internalValue = config.networkConfig;
        this._nodeConfig.internalValue = config.nodeConfig;
        this._placementPolicy.internalValue = config.placementPolicy;
        this._timeouts.internalValue = config.timeouts;
        this._upgradeSettings.internalValue = config.upgradeSettings;
    }
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    set cluster(value) {
        this._cluster = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterInput() {
        return this._cluster;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get initialNodeCount() {
        return this.getNumberAttribute('initial_node_count');
    }
    set initialNodeCount(value) {
        this._initialNodeCount = value;
    }
    resetInitialNodeCount() {
        this._initialNodeCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialNodeCountInput() {
        return this._initialNodeCount;
    }
    // instance_group_urls - computed: true, optional: false, required: false
    get instanceGroupUrls() {
        return this.getListAttribute('instance_group_urls');
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // managed_instance_group_urls - computed: true, optional: false, required: false
    get managedInstanceGroupUrls() {
        return this.getListAttribute('managed_instance_group_urls');
    }
    get maxPodsPerNode() {
        return this.getNumberAttribute('max_pods_per_node');
    }
    set maxPodsPerNode(value) {
        this._maxPodsPerNode = value;
    }
    resetMaxPodsPerNode() {
        this._maxPodsPerNode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxPodsPerNodeInput() {
        return this._maxPodsPerNode;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    get nodeCount() {
        return this.getNumberAttribute('node_count');
    }
    set nodeCount(value) {
        this._nodeCount = value;
    }
    resetNodeCount() {
        this._nodeCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeCountInput() {
        return this._nodeCount;
    }
    get nodeLocations() {
        return cdktf.Fn.tolist(this.getListAttribute('node_locations'));
    }
    set nodeLocations(value) {
        this._nodeLocations = value;
    }
    resetNodeLocations() {
        this._nodeLocations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeLocationsInput() {
        return this._nodeLocations;
    }
    // operation - computed: true, optional: false, required: false
    get operation() {
        return this.getStringAttribute('operation');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get autoscaling() {
        return this._autoscaling;
    }
    putAutoscaling(value) {
        this._autoscaling.internalValue = value;
    }
    resetAutoscaling() {
        this._autoscaling.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingInput() {
        return this._autoscaling.internalValue;
    }
    get management() {
        return this._management;
    }
    putManagement(value) {
        this._management.internalValue = value;
    }
    resetManagement() {
        this._management.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get managementInput() {
        return this._management.internalValue;
    }
    get networkConfig() {
        return this._networkConfig;
    }
    putNetworkConfig(value) {
        this._networkConfig.internalValue = value;
    }
    resetNetworkConfig() {
        this._networkConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkConfigInput() {
        return this._networkConfig.internalValue;
    }
    get nodeConfig() {
        return this._nodeConfig;
    }
    putNodeConfig(value) {
        this._nodeConfig.internalValue = value;
    }
    resetNodeConfig() {
        this._nodeConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeConfigInput() {
        return this._nodeConfig.internalValue;
    }
    get placementPolicy() {
        return this._placementPolicy;
    }
    putPlacementPolicy(value) {
        this._placementPolicy.internalValue = value;
    }
    resetPlacementPolicy() {
        this._placementPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementPolicyInput() {
        return this._placementPolicy.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get upgradeSettings() {
        return this._upgradeSettings;
    }
    putUpgradeSettings(value) {
        this._upgradeSettings.internalValue = value;
    }
    resetUpgradeSettings() {
        this._upgradeSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get upgradeSettingsInput() {
        return this._upgradeSettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster: cdktf.stringToTerraform(this._cluster),
            id: cdktf.stringToTerraform(this._id),
            initial_node_count: cdktf.numberToTerraform(this._initialNodeCount),
            location: cdktf.stringToTerraform(this._location),
            max_pods_per_node: cdktf.numberToTerraform(this._maxPodsPerNode),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            node_count: cdktf.numberToTerraform(this._nodeCount),
            node_locations: cdktf.listMapper(cdktf.stringToTerraform, false)(this._nodeLocations),
            project: cdktf.stringToTerraform(this._project),
            version: cdktf.stringToTerraform(this._version),
            autoscaling: containerNodePoolAutoscalingToTerraform(this._autoscaling.internalValue),
            management: containerNodePoolManagementToTerraform(this._management.internalValue),
            network_config: containerNodePoolNetworkConfigToTerraform(this._networkConfig.internalValue),
            node_config: containerNodePoolNodeConfigToTerraform(this._nodeConfig.internalValue),
            placement_policy: containerNodePoolPlacementPolicyToTerraform(this._placementPolicy.internalValue),
            timeouts: containerNodePoolTimeoutsToTerraform(this._timeouts.internalValue),
            upgrade_settings: containerNodePoolUpgradeSettingsToTerraform(this._upgradeSettings.internalValue),
        };
    }
}
exports.ContainerNodePool = ContainerNodePool;
_y = JSII_RTTI_SYMBOL_1;
ContainerNodePool[_y] = { fqn: "@cdktf/provider-google.containerNodePool.ContainerNodePool", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
ContainerNodePool.tfResourceType = "google_container_node_pool";
