// @cdktf/provider-google/lib/data-google-monitoring-uptime-check-ips/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleMonitoringUptimeCheckIps = exports.DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsList = exports.DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsOutputReference = exports.dataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsToTerraform = dataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsToTerraform;
class DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_address - computed: true, optional: false, required: false
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    // region - computed: true, optional: false, required: false
    get region() {
        return this.getStringAttribute('region');
    }
}
exports.DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsOutputReference = DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleMonitoringUptimeCheckIps.DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsOutputReference", version: "4.0.3" };
class DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsList = DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsList[_b] = { fqn: "@cdktf/provider-google.dataGoogleMonitoringUptimeCheckIps.DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsList", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/monitoring_uptime_check_ips google_monitoring_uptime_check_ips}
*/
class DataGoogleMonitoringUptimeCheckIps extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/monitoring_uptime_check_ips google_monitoring_uptime_check_ips} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleMonitoringUptimeCheckIpsConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_monitoring_uptime_check_ips',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // uptime_check_ips - computed: true, optional: false, required: false
        this._uptimeCheckIps = new DataGoogleMonitoringUptimeCheckIpsUptimeCheckIpsList(this, "uptime_check_ips", false);
        this._id = config.id;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get uptimeCheckIps() {
        return this._uptimeCheckIps;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataGoogleMonitoringUptimeCheckIps = DataGoogleMonitoringUptimeCheckIps;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleMonitoringUptimeCheckIps[_c] = { fqn: "@cdktf/provider-google.dataGoogleMonitoringUptimeCheckIps.DataGoogleMonitoringUptimeCheckIps", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleMonitoringUptimeCheckIps.tfResourceType = "google_monitoring_uptime_check_ips";
