// @cdktf/provider-google/lib/cloud-run-v2-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudRunV2Service = exports.CloudRunV2ServiceTrafficList = exports.CloudRunV2ServiceTrafficOutputReference = exports.cloudRunV2ServiceTrafficToTerraform = exports.CloudRunV2ServiceTimeoutsOutputReference = exports.cloudRunV2ServiceTimeoutsToTerraform = exports.CloudRunV2ServiceTemplateOutputReference = exports.cloudRunV2ServiceTemplateToTerraform = exports.CloudRunV2ServiceTemplateVpcAccessOutputReference = exports.cloudRunV2ServiceTemplateVpcAccessToTerraform = exports.CloudRunV2ServiceTemplateVolumesList = exports.CloudRunV2ServiceTemplateVolumesOutputReference = exports.cloudRunV2ServiceTemplateVolumesToTerraform = exports.CloudRunV2ServiceTemplateVolumesSecretOutputReference = exports.cloudRunV2ServiceTemplateVolumesSecretToTerraform = exports.CloudRunV2ServiceTemplateVolumesSecretItemsList = exports.CloudRunV2ServiceTemplateVolumesSecretItemsOutputReference = exports.cloudRunV2ServiceTemplateVolumesSecretItemsToTerraform = exports.CloudRunV2ServiceTemplateVolumesCloudSqlInstanceOutputReference = exports.cloudRunV2ServiceTemplateVolumesCloudSqlInstanceToTerraform = exports.CloudRunV2ServiceTemplateScalingOutputReference = exports.cloudRunV2ServiceTemplateScalingToTerraform = exports.CloudRunV2ServiceTemplateContainersList = exports.CloudRunV2ServiceTemplateContainersOutputReference = exports.cloudRunV2ServiceTemplateContainersToTerraform = exports.CloudRunV2ServiceTemplateContainersVolumeMountsList = exports.CloudRunV2ServiceTemplateContainersVolumeMountsOutputReference = exports.cloudRunV2ServiceTemplateContainersVolumeMountsToTerraform = exports.CloudRunV2ServiceTemplateContainersStartupProbeOutputReference = exports.cloudRunV2ServiceTemplateContainersStartupProbeToTerraform = exports.CloudRunV2ServiceTemplateContainersStartupProbeTcpSocketOutputReference = exports.cloudRunV2ServiceTemplateContainersStartupProbeTcpSocketToTerraform = exports.CloudRunV2ServiceTemplateContainersStartupProbeHttpGetOutputReference = exports.cloudRunV2ServiceTemplateContainersStartupProbeHttpGetToTerraform = exports.CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersList = exports.CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference = exports.cloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform = exports.CloudRunV2ServiceTemplateContainersResourcesOutputReference = exports.cloudRunV2ServiceTemplateContainersResourcesToTerraform = exports.CloudRunV2ServiceTemplateContainersPortsList = exports.CloudRunV2ServiceTemplateContainersPortsOutputReference = exports.cloudRunV2ServiceTemplateContainersPortsToTerraform = exports.CloudRunV2ServiceTemplateContainersLivenessProbeOutputReference = exports.cloudRunV2ServiceTemplateContainersLivenessProbeToTerraform = exports.CloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketOutputReference = exports.cloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketToTerraform = exports.CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetOutputReference = exports.cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetToTerraform = exports.CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersList = exports.CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference = exports.cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform = exports.CloudRunV2ServiceTemplateContainersEnvList = exports.CloudRunV2ServiceTemplateContainersEnvOutputReference = exports.cloudRunV2ServiceTemplateContainersEnvToTerraform = exports.CloudRunV2ServiceTemplateContainersEnvValueSourceOutputReference = exports.cloudRunV2ServiceTemplateContainersEnvValueSourceToTerraform = exports.CloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefOutputReference = exports.cloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefToTerraform = exports.CloudRunV2ServiceBinaryAuthorizationOutputReference = exports.cloudRunV2ServiceBinaryAuthorizationToTerraform = exports.CloudRunV2ServiceTrafficStatusesList = exports.CloudRunV2ServiceTrafficStatusesOutputReference = exports.cloudRunV2ServiceTrafficStatusesToTerraform = exports.CloudRunV2ServiceTerminalConditionList = exports.CloudRunV2ServiceTerminalConditionOutputReference = exports.cloudRunV2ServiceTerminalConditionToTerraform = exports.CloudRunV2ServiceConditionsList = exports.CloudRunV2ServiceConditionsOutputReference = exports.cloudRunV2ServiceConditionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudRunV2ServiceConditionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudRunV2ServiceConditionsToTerraform = cloudRunV2ServiceConditionsToTerraform;
class CloudRunV2ServiceConditionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // execution_reason - computed: true, optional: false, required: false
    get executionReason() {
        return this.getStringAttribute('execution_reason');
    }
    // last_transition_time - computed: true, optional: false, required: false
    get lastTransitionTime() {
        return this.getStringAttribute('last_transition_time');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
    // revision_reason - computed: true, optional: false, required: false
    get revisionReason() {
        return this.getStringAttribute('revision_reason');
    }
    // severity - computed: true, optional: false, required: false
    get severity() {
        return this.getStringAttribute('severity');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.CloudRunV2ServiceConditionsOutputReference = CloudRunV2ServiceConditionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceConditionsOutputReference[_a] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceConditionsOutputReference", version: "5.0.1" };
class CloudRunV2ServiceConditionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceConditionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceConditionsList = CloudRunV2ServiceConditionsList;
_b = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceConditionsList[_b] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceConditionsList", version: "5.0.1" };
function cloudRunV2ServiceTerminalConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudRunV2ServiceTerminalConditionToTerraform = cloudRunV2ServiceTerminalConditionToTerraform;
class CloudRunV2ServiceTerminalConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // execution_reason - computed: true, optional: false, required: false
    get executionReason() {
        return this.getStringAttribute('execution_reason');
    }
    // last_transition_time - computed: true, optional: false, required: false
    get lastTransitionTime() {
        return this.getStringAttribute('last_transition_time');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
    // revision_reason - computed: true, optional: false, required: false
    get revisionReason() {
        return this.getStringAttribute('revision_reason');
    }
    // severity - computed: true, optional: false, required: false
    get severity() {
        return this.getStringAttribute('severity');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.CloudRunV2ServiceTerminalConditionOutputReference = CloudRunV2ServiceTerminalConditionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTerminalConditionOutputReference[_c] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTerminalConditionOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTerminalConditionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTerminalConditionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTerminalConditionList = CloudRunV2ServiceTerminalConditionList;
_d = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTerminalConditionList[_d] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTerminalConditionList", version: "5.0.1" };
function cloudRunV2ServiceTrafficStatusesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudRunV2ServiceTrafficStatusesToTerraform = cloudRunV2ServiceTrafficStatusesToTerraform;
class CloudRunV2ServiceTrafficStatusesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // percent - computed: true, optional: false, required: false
    get percent() {
        return this.getNumberAttribute('percent');
    }
    // revision - computed: true, optional: false, required: false
    get revision() {
        return this.getStringAttribute('revision');
    }
    // tag - computed: true, optional: false, required: false
    get tag() {
        return this.getStringAttribute('tag');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
}
exports.CloudRunV2ServiceTrafficStatusesOutputReference = CloudRunV2ServiceTrafficStatusesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTrafficStatusesOutputReference[_e] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTrafficStatusesOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTrafficStatusesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTrafficStatusesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTrafficStatusesList = CloudRunV2ServiceTrafficStatusesList;
_f = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTrafficStatusesList[_f] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTrafficStatusesList", version: "5.0.1" };
function cloudRunV2ServiceBinaryAuthorizationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        breakglass_justification: cdktf.stringToTerraform(struct.breakglassJustification),
        use_default: cdktf.booleanToTerraform(struct.useDefault),
    };
}
exports.cloudRunV2ServiceBinaryAuthorizationToTerraform = cloudRunV2ServiceBinaryAuthorizationToTerraform;
class CloudRunV2ServiceBinaryAuthorizationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._breakglassJustification !== undefined) {
            hasAnyValues = true;
            internalValueResult.breakglassJustification = this._breakglassJustification;
        }
        if (this._useDefault !== undefined) {
            hasAnyValues = true;
            internalValueResult.useDefault = this._useDefault;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._breakglassJustification = undefined;
            this._useDefault = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._breakglassJustification = value.breakglassJustification;
            this._useDefault = value.useDefault;
        }
    }
    get breakglassJustification() {
        return this.getStringAttribute('breakglass_justification');
    }
    set breakglassJustification(value) {
        this._breakglassJustification = value;
    }
    resetBreakglassJustification() {
        this._breakglassJustification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get breakglassJustificationInput() {
        return this._breakglassJustification;
    }
    get useDefault() {
        return this.getBooleanAttribute('use_default');
    }
    set useDefault(value) {
        this._useDefault = value;
    }
    resetUseDefault() {
        this._useDefault = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useDefaultInput() {
        return this._useDefault;
    }
}
exports.CloudRunV2ServiceBinaryAuthorizationOutputReference = CloudRunV2ServiceBinaryAuthorizationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceBinaryAuthorizationOutputReference[_g] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceBinaryAuthorizationOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret: cdktf.stringToTerraform(struct.secret),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.cloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefToTerraform = cloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefToTerraform;
class CloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secret !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secret = undefined;
            this._version = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secret = value.secret;
            this._version = value.version;
        }
    }
    get secret() {
        return this.getStringAttribute('secret');
    }
    set secret(value) {
        this._secret = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.CloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefOutputReference = CloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefOutputReference;
_h = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefOutputReference[_h] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersEnvValueSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_key_ref: cloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefToTerraform(struct.secretKeyRef),
    };
}
exports.cloudRunV2ServiceTemplateContainersEnvValueSourceToTerraform = cloudRunV2ServiceTemplateContainersEnvValueSourceToTerraform;
class CloudRunV2ServiceTemplateContainersEnvValueSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // secret_key_ref - computed: false, optional: true, required: false
        this._secretKeyRef = new CloudRunV2ServiceTemplateContainersEnvValueSourceSecretKeyRefOutputReference(this, "secret_key_ref");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretKeyRef?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretKeyRef = this._secretKeyRef?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretKeyRef.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretKeyRef.internalValue = value.secretKeyRef;
        }
    }
    get secretKeyRef() {
        return this._secretKeyRef;
    }
    putSecretKeyRef(value) {
        this._secretKeyRef.internalValue = value;
    }
    resetSecretKeyRef() {
        this._secretKeyRef.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretKeyRefInput() {
        return this._secretKeyRef.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateContainersEnvValueSourceOutputReference = CloudRunV2ServiceTemplateContainersEnvValueSourceOutputReference;
_j = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersEnvValueSourceOutputReference[_j] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersEnvValueSourceOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersEnvToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
        value_source: cloudRunV2ServiceTemplateContainersEnvValueSourceToTerraform(struct.valueSource),
    };
}
exports.cloudRunV2ServiceTemplateContainersEnvToTerraform = cloudRunV2ServiceTemplateContainersEnvToTerraform;
class CloudRunV2ServiceTemplateContainersEnvOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // value_source - computed: false, optional: true, required: false
        this._valueSource = new CloudRunV2ServiceTemplateContainersEnvValueSourceOutputReference(this, "value_source");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        if (this._valueSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.valueSource = this._valueSource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
            this._valueSource.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
            this._valueSource.internalValue = value.valueSource;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
    get valueSource() {
        return this._valueSource;
    }
    putValueSource(value) {
        this._valueSource.internalValue = value;
    }
    resetValueSource() {
        this._valueSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueSourceInput() {
        return this._valueSource.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateContainersEnvOutputReference = CloudRunV2ServiceTemplateContainersEnvOutputReference;
_k = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersEnvOutputReference[_k] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersEnvOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTemplateContainersEnvList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTemplateContainersEnvOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTemplateContainersEnvList = CloudRunV2ServiceTemplateContainersEnvList;
_l = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersEnvList[_l] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersEnvList", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform = cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform;
class CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference = CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference;
_m = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference[_m] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersList = CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersList;
_o = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersList[_o] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersList", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        http_headers: cdktf.listMapper(cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersToTerraform, true)(struct.httpHeaders),
    };
}
exports.cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetToTerraform = cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetToTerraform;
class CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // http_headers - computed: false, optional: true, required: false
        this._httpHeaders = new CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetHttpHeadersList(this, "http_headers", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._httpHeaders?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpHeaders = this._httpHeaders?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
            this._httpHeaders.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
            this._httpHeaders.internalValue = value.httpHeaders;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get httpHeaders() {
        return this._httpHeaders;
    }
    putHttpHeaders(value) {
        this._httpHeaders.internalValue = value;
    }
    resetHttpHeaders() {
        this._httpHeaders.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpHeadersInput() {
        return this._httpHeaders.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetOutputReference = CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetOutputReference;
_p = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetOutputReference[_p] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.cloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketToTerraform = cloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketToTerraform;
class CloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._port = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._port = value.port;
        }
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.CloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketOutputReference = CloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketOutputReference[_q] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersLivenessProbeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        failure_threshold: cdktf.numberToTerraform(struct.failureThreshold),
        initial_delay_seconds: cdktf.numberToTerraform(struct.initialDelaySeconds),
        period_seconds: cdktf.numberToTerraform(struct.periodSeconds),
        timeout_seconds: cdktf.numberToTerraform(struct.timeoutSeconds),
        http_get: cloudRunV2ServiceTemplateContainersLivenessProbeHttpGetToTerraform(struct.httpGet),
        tcp_socket: cloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketToTerraform(struct.tcpSocket),
    };
}
exports.cloudRunV2ServiceTemplateContainersLivenessProbeToTerraform = cloudRunV2ServiceTemplateContainersLivenessProbeToTerraform;
class CloudRunV2ServiceTemplateContainersLivenessProbeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // http_get - computed: false, optional: true, required: false
        this._httpGet = new CloudRunV2ServiceTemplateContainersLivenessProbeHttpGetOutputReference(this, "http_get");
        // tcp_socket - computed: false, optional: true, required: false
        this._tcpSocket = new CloudRunV2ServiceTemplateContainersLivenessProbeTcpSocketOutputReference(this, "tcp_socket");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._failureThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.failureThreshold = this._failureThreshold;
        }
        if (this._initialDelaySeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialDelaySeconds = this._initialDelaySeconds;
        }
        if (this._periodSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.periodSeconds = this._periodSeconds;
        }
        if (this._timeoutSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutSeconds = this._timeoutSeconds;
        }
        if (this._httpGet?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpGet = this._httpGet?.internalValue;
        }
        if (this._tcpSocket?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tcpSocket = this._tcpSocket?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._failureThreshold = undefined;
            this._initialDelaySeconds = undefined;
            this._periodSeconds = undefined;
            this._timeoutSeconds = undefined;
            this._httpGet.internalValue = undefined;
            this._tcpSocket.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._failureThreshold = value.failureThreshold;
            this._initialDelaySeconds = value.initialDelaySeconds;
            this._periodSeconds = value.periodSeconds;
            this._timeoutSeconds = value.timeoutSeconds;
            this._httpGet.internalValue = value.httpGet;
            this._tcpSocket.internalValue = value.tcpSocket;
        }
    }
    get failureThreshold() {
        return this.getNumberAttribute('failure_threshold');
    }
    set failureThreshold(value) {
        this._failureThreshold = value;
    }
    resetFailureThreshold() {
        this._failureThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failureThresholdInput() {
        return this._failureThreshold;
    }
    get initialDelaySeconds() {
        return this.getNumberAttribute('initial_delay_seconds');
    }
    set initialDelaySeconds(value) {
        this._initialDelaySeconds = value;
    }
    resetInitialDelaySeconds() {
        this._initialDelaySeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialDelaySecondsInput() {
        return this._initialDelaySeconds;
    }
    get periodSeconds() {
        return this.getNumberAttribute('period_seconds');
    }
    set periodSeconds(value) {
        this._periodSeconds = value;
    }
    resetPeriodSeconds() {
        this._periodSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get periodSecondsInput() {
        return this._periodSeconds;
    }
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    set timeoutSeconds(value) {
        this._timeoutSeconds = value;
    }
    resetTimeoutSeconds() {
        this._timeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutSecondsInput() {
        return this._timeoutSeconds;
    }
    get httpGet() {
        return this._httpGet;
    }
    putHttpGet(value) {
        this._httpGet.internalValue = value;
    }
    resetHttpGet() {
        this._httpGet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpGetInput() {
        return this._httpGet.internalValue;
    }
    get tcpSocket() {
        return this._tcpSocket;
    }
    putTcpSocket(value) {
        this._tcpSocket.internalValue = value;
    }
    resetTcpSocket() {
        this._tcpSocket.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tcpSocketInput() {
        return this._tcpSocket.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateContainersLivenessProbeOutputReference = CloudRunV2ServiceTemplateContainersLivenessProbeOutputReference;
_r = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersLivenessProbeOutputReference[_r] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersLivenessProbeOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersPortsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_port: cdktf.numberToTerraform(struct.containerPort),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.cloudRunV2ServiceTemplateContainersPortsToTerraform = cloudRunV2ServiceTemplateContainersPortsToTerraform;
class CloudRunV2ServiceTemplateContainersPortsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerPort = this._containerPort;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._containerPort = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._containerPort = value.containerPort;
            this._name = value.name;
        }
    }
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    set containerPort(value) {
        this._containerPort = value;
    }
    resetContainerPort() {
        this._containerPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerPortInput() {
        return this._containerPort;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.CloudRunV2ServiceTemplateContainersPortsOutputReference = CloudRunV2ServiceTemplateContainersPortsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersPortsOutputReference[_s] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersPortsOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTemplateContainersPortsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTemplateContainersPortsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTemplateContainersPortsList = CloudRunV2ServiceTemplateContainersPortsList;
_t = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersPortsList[_t] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersPortsList", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cpu_idle: cdktf.booleanToTerraform(struct.cpuIdle),
        limits: cdktf.hashMapper(cdktf.stringToTerraform)(struct.limits),
    };
}
exports.cloudRunV2ServiceTemplateContainersResourcesToTerraform = cloudRunV2ServiceTemplateContainersResourcesToTerraform;
class CloudRunV2ServiceTemplateContainersResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cpuIdle !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuIdle = this._cpuIdle;
        }
        if (this._limits !== undefined) {
            hasAnyValues = true;
            internalValueResult.limits = this._limits;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cpuIdle = undefined;
            this._limits = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cpuIdle = value.cpuIdle;
            this._limits = value.limits;
        }
    }
    get cpuIdle() {
        return this.getBooleanAttribute('cpu_idle');
    }
    set cpuIdle(value) {
        this._cpuIdle = value;
    }
    resetCpuIdle() {
        this._cpuIdle = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuIdleInput() {
        return this._cpuIdle;
    }
    get limits() {
        return this.getStringMapAttribute('limits');
    }
    set limits(value) {
        this._limits = value;
    }
    resetLimits() {
        this._limits = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get limitsInput() {
        return this._limits;
    }
}
exports.CloudRunV2ServiceTemplateContainersResourcesOutputReference = CloudRunV2ServiceTemplateContainersResourcesOutputReference;
_u = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersResourcesOutputReference[_u] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersResourcesOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform = cloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform;
class CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference = CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference;
_v = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference[_v] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersList = CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersList;
_w = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersList[_w] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersList", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersStartupProbeHttpGetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        http_headers: cdktf.listMapper(cloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersToTerraform, true)(struct.httpHeaders),
    };
}
exports.cloudRunV2ServiceTemplateContainersStartupProbeHttpGetToTerraform = cloudRunV2ServiceTemplateContainersStartupProbeHttpGetToTerraform;
class CloudRunV2ServiceTemplateContainersStartupProbeHttpGetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // http_headers - computed: false, optional: true, required: false
        this._httpHeaders = new CloudRunV2ServiceTemplateContainersStartupProbeHttpGetHttpHeadersList(this, "http_headers", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._httpHeaders?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpHeaders = this._httpHeaders?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
            this._httpHeaders.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
            this._httpHeaders.internalValue = value.httpHeaders;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get httpHeaders() {
        return this._httpHeaders;
    }
    putHttpHeaders(value) {
        this._httpHeaders.internalValue = value;
    }
    resetHttpHeaders() {
        this._httpHeaders.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpHeadersInput() {
        return this._httpHeaders.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateContainersStartupProbeHttpGetOutputReference = CloudRunV2ServiceTemplateContainersStartupProbeHttpGetOutputReference;
_x = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersStartupProbeHttpGetOutputReference[_x] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersStartupProbeHttpGetOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersStartupProbeTcpSocketToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.cloudRunV2ServiceTemplateContainersStartupProbeTcpSocketToTerraform = cloudRunV2ServiceTemplateContainersStartupProbeTcpSocketToTerraform;
class CloudRunV2ServiceTemplateContainersStartupProbeTcpSocketOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._port = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._port = value.port;
        }
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.CloudRunV2ServiceTemplateContainersStartupProbeTcpSocketOutputReference = CloudRunV2ServiceTemplateContainersStartupProbeTcpSocketOutputReference;
_y = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersStartupProbeTcpSocketOutputReference[_y] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersStartupProbeTcpSocketOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersStartupProbeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        failure_threshold: cdktf.numberToTerraform(struct.failureThreshold),
        initial_delay_seconds: cdktf.numberToTerraform(struct.initialDelaySeconds),
        period_seconds: cdktf.numberToTerraform(struct.periodSeconds),
        timeout_seconds: cdktf.numberToTerraform(struct.timeoutSeconds),
        http_get: cloudRunV2ServiceTemplateContainersStartupProbeHttpGetToTerraform(struct.httpGet),
        tcp_socket: cloudRunV2ServiceTemplateContainersStartupProbeTcpSocketToTerraform(struct.tcpSocket),
    };
}
exports.cloudRunV2ServiceTemplateContainersStartupProbeToTerraform = cloudRunV2ServiceTemplateContainersStartupProbeToTerraform;
class CloudRunV2ServiceTemplateContainersStartupProbeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // http_get - computed: false, optional: true, required: false
        this._httpGet = new CloudRunV2ServiceTemplateContainersStartupProbeHttpGetOutputReference(this, "http_get");
        // tcp_socket - computed: false, optional: true, required: false
        this._tcpSocket = new CloudRunV2ServiceTemplateContainersStartupProbeTcpSocketOutputReference(this, "tcp_socket");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._failureThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.failureThreshold = this._failureThreshold;
        }
        if (this._initialDelaySeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialDelaySeconds = this._initialDelaySeconds;
        }
        if (this._periodSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.periodSeconds = this._periodSeconds;
        }
        if (this._timeoutSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutSeconds = this._timeoutSeconds;
        }
        if (this._httpGet?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpGet = this._httpGet?.internalValue;
        }
        if (this._tcpSocket?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tcpSocket = this._tcpSocket?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._failureThreshold = undefined;
            this._initialDelaySeconds = undefined;
            this._periodSeconds = undefined;
            this._timeoutSeconds = undefined;
            this._httpGet.internalValue = undefined;
            this._tcpSocket.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._failureThreshold = value.failureThreshold;
            this._initialDelaySeconds = value.initialDelaySeconds;
            this._periodSeconds = value.periodSeconds;
            this._timeoutSeconds = value.timeoutSeconds;
            this._httpGet.internalValue = value.httpGet;
            this._tcpSocket.internalValue = value.tcpSocket;
        }
    }
    get failureThreshold() {
        return this.getNumberAttribute('failure_threshold');
    }
    set failureThreshold(value) {
        this._failureThreshold = value;
    }
    resetFailureThreshold() {
        this._failureThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failureThresholdInput() {
        return this._failureThreshold;
    }
    get initialDelaySeconds() {
        return this.getNumberAttribute('initial_delay_seconds');
    }
    set initialDelaySeconds(value) {
        this._initialDelaySeconds = value;
    }
    resetInitialDelaySeconds() {
        this._initialDelaySeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initialDelaySecondsInput() {
        return this._initialDelaySeconds;
    }
    get periodSeconds() {
        return this.getNumberAttribute('period_seconds');
    }
    set periodSeconds(value) {
        this._periodSeconds = value;
    }
    resetPeriodSeconds() {
        this._periodSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get periodSecondsInput() {
        return this._periodSeconds;
    }
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    set timeoutSeconds(value) {
        this._timeoutSeconds = value;
    }
    resetTimeoutSeconds() {
        this._timeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutSecondsInput() {
        return this._timeoutSeconds;
    }
    get httpGet() {
        return this._httpGet;
    }
    putHttpGet(value) {
        this._httpGet.internalValue = value;
    }
    resetHttpGet() {
        this._httpGet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpGetInput() {
        return this._httpGet.internalValue;
    }
    get tcpSocket() {
        return this._tcpSocket;
    }
    putTcpSocket(value) {
        this._tcpSocket.internalValue = value;
    }
    resetTcpSocket() {
        this._tcpSocket.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tcpSocketInput() {
        return this._tcpSocket.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateContainersStartupProbeOutputReference = CloudRunV2ServiceTemplateContainersStartupProbeOutputReference;
_z = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersStartupProbeOutputReference[_z] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersStartupProbeOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersVolumeMountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mount_path: cdktf.stringToTerraform(struct.mountPath),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.cloudRunV2ServiceTemplateContainersVolumeMountsToTerraform = cloudRunV2ServiceTemplateContainersVolumeMountsToTerraform;
class CloudRunV2ServiceTemplateContainersVolumeMountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mountPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.mountPath = this._mountPath;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._mountPath = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._mountPath = value.mountPath;
            this._name = value.name;
        }
    }
    get mountPath() {
        return this.getStringAttribute('mount_path');
    }
    set mountPath(value) {
        this._mountPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get mountPathInput() {
        return this._mountPath;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.CloudRunV2ServiceTemplateContainersVolumeMountsOutputReference = CloudRunV2ServiceTemplateContainersVolumeMountsOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersVolumeMountsOutputReference[_0] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersVolumeMountsOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTemplateContainersVolumeMountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTemplateContainersVolumeMountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTemplateContainersVolumeMountsList = CloudRunV2ServiceTemplateContainersVolumeMountsList;
_1 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersVolumeMountsList[_1] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersVolumeMountsList", version: "5.0.1" };
function cloudRunV2ServiceTemplateContainersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        args: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.args),
        command: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.command),
        image: cdktf.stringToTerraform(struct.image),
        name: cdktf.stringToTerraform(struct.name),
        working_dir: cdktf.stringToTerraform(struct.workingDir),
        env: cdktf.listMapper(cloudRunV2ServiceTemplateContainersEnvToTerraform, true)(struct.env),
        liveness_probe: cloudRunV2ServiceTemplateContainersLivenessProbeToTerraform(struct.livenessProbe),
        ports: cdktf.listMapper(cloudRunV2ServiceTemplateContainersPortsToTerraform, true)(struct.ports),
        resources: cloudRunV2ServiceTemplateContainersResourcesToTerraform(struct.resources),
        startup_probe: cloudRunV2ServiceTemplateContainersStartupProbeToTerraform(struct.startupProbe),
        volume_mounts: cdktf.listMapper(cloudRunV2ServiceTemplateContainersVolumeMountsToTerraform, true)(struct.volumeMounts),
    };
}
exports.cloudRunV2ServiceTemplateContainersToTerraform = cloudRunV2ServiceTemplateContainersToTerraform;
class CloudRunV2ServiceTemplateContainersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // env - computed: false, optional: true, required: false
        this._env = new CloudRunV2ServiceTemplateContainersEnvList(this, "env", false);
        // liveness_probe - computed: false, optional: true, required: false
        this._livenessProbe = new CloudRunV2ServiceTemplateContainersLivenessProbeOutputReference(this, "liveness_probe");
        // ports - computed: false, optional: true, required: false
        this._ports = new CloudRunV2ServiceTemplateContainersPortsList(this, "ports", false);
        // resources - computed: false, optional: true, required: false
        this._resources = new CloudRunV2ServiceTemplateContainersResourcesOutputReference(this, "resources");
        // startup_probe - computed: false, optional: true, required: false
        this._startupProbe = new CloudRunV2ServiceTemplateContainersStartupProbeOutputReference(this, "startup_probe");
        // volume_mounts - computed: false, optional: true, required: false
        this._volumeMounts = new CloudRunV2ServiceTemplateContainersVolumeMountsList(this, "volume_mounts", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._args !== undefined) {
            hasAnyValues = true;
            internalValueResult.args = this._args;
        }
        if (this._command !== undefined) {
            hasAnyValues = true;
            internalValueResult.command = this._command;
        }
        if (this._image !== undefined) {
            hasAnyValues = true;
            internalValueResult.image = this._image;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._workingDir !== undefined) {
            hasAnyValues = true;
            internalValueResult.workingDir = this._workingDir;
        }
        if (this._env?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.env = this._env?.internalValue;
        }
        if (this._livenessProbe?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.livenessProbe = this._livenessProbe?.internalValue;
        }
        if (this._ports?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ports = this._ports?.internalValue;
        }
        if (this._resources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources?.internalValue;
        }
        if (this._startupProbe?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.startupProbe = this._startupProbe?.internalValue;
        }
        if (this._volumeMounts?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeMounts = this._volumeMounts?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._args = undefined;
            this._command = undefined;
            this._image = undefined;
            this._name = undefined;
            this._workingDir = undefined;
            this._env.internalValue = undefined;
            this._livenessProbe.internalValue = undefined;
            this._ports.internalValue = undefined;
            this._resources.internalValue = undefined;
            this._startupProbe.internalValue = undefined;
            this._volumeMounts.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._args = value.args;
            this._command = value.command;
            this._image = value.image;
            this._name = value.name;
            this._workingDir = value.workingDir;
            this._env.internalValue = value.env;
            this._livenessProbe.internalValue = value.livenessProbe;
            this._ports.internalValue = value.ports;
            this._resources.internalValue = value.resources;
            this._startupProbe.internalValue = value.startupProbe;
            this._volumeMounts.internalValue = value.volumeMounts;
        }
    }
    get args() {
        return this.getListAttribute('args');
    }
    set args(value) {
        this._args = value;
    }
    resetArgs() {
        this._args = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argsInput() {
        return this._args;
    }
    get command() {
        return this.getListAttribute('command');
    }
    set command(value) {
        this._command = value;
    }
    resetCommand() {
        this._command = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commandInput() {
        return this._command;
    }
    get image() {
        return this.getStringAttribute('image');
    }
    set image(value) {
        this._image = value;
    }
    // Temporarily expose input value. Use with caution.
    get imageInput() {
        return this._image;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get workingDir() {
        return this.getStringAttribute('working_dir');
    }
    set workingDir(value) {
        this._workingDir = value;
    }
    resetWorkingDir() {
        this._workingDir = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workingDirInput() {
        return this._workingDir;
    }
    get env() {
        return this._env;
    }
    putEnv(value) {
        this._env.internalValue = value;
    }
    resetEnv() {
        this._env.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get envInput() {
        return this._env.internalValue;
    }
    get livenessProbe() {
        return this._livenessProbe;
    }
    putLivenessProbe(value) {
        this._livenessProbe.internalValue = value;
    }
    resetLivenessProbe() {
        this._livenessProbe.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get livenessProbeInput() {
        return this._livenessProbe.internalValue;
    }
    get ports() {
        return this._ports;
    }
    putPorts(value) {
        this._ports.internalValue = value;
    }
    resetPorts() {
        this._ports.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portsInput() {
        return this._ports.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    resetResources() {
        this._resources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
    get startupProbe() {
        return this._startupProbe;
    }
    putStartupProbe(value) {
        this._startupProbe.internalValue = value;
    }
    resetStartupProbe() {
        this._startupProbe.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startupProbeInput() {
        return this._startupProbe.internalValue;
    }
    get volumeMounts() {
        return this._volumeMounts;
    }
    putVolumeMounts(value) {
        this._volumeMounts.internalValue = value;
    }
    resetVolumeMounts() {
        this._volumeMounts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeMountsInput() {
        return this._volumeMounts.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateContainersOutputReference = CloudRunV2ServiceTemplateContainersOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersOutputReference[_2] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTemplateContainersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTemplateContainersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTemplateContainersList = CloudRunV2ServiceTemplateContainersList;
_3 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateContainersList[_3] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateContainersList", version: "5.0.1" };
function cloudRunV2ServiceTemplateScalingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_instance_count: cdktf.numberToTerraform(struct.maxInstanceCount),
        min_instance_count: cdktf.numberToTerraform(struct.minInstanceCount),
    };
}
exports.cloudRunV2ServiceTemplateScalingToTerraform = cloudRunV2ServiceTemplateScalingToTerraform;
class CloudRunV2ServiceTemplateScalingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxInstanceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxInstanceCount = this._maxInstanceCount;
        }
        if (this._minInstanceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.minInstanceCount = this._minInstanceCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxInstanceCount = undefined;
            this._minInstanceCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxInstanceCount = value.maxInstanceCount;
            this._minInstanceCount = value.minInstanceCount;
        }
    }
    get maxInstanceCount() {
        return this.getNumberAttribute('max_instance_count');
    }
    set maxInstanceCount(value) {
        this._maxInstanceCount = value;
    }
    resetMaxInstanceCount() {
        this._maxInstanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxInstanceCountInput() {
        return this._maxInstanceCount;
    }
    get minInstanceCount() {
        return this.getNumberAttribute('min_instance_count');
    }
    set minInstanceCount(value) {
        this._minInstanceCount = value;
    }
    resetMinInstanceCount() {
        this._minInstanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minInstanceCountInput() {
        return this._minInstanceCount;
    }
}
exports.CloudRunV2ServiceTemplateScalingOutputReference = CloudRunV2ServiceTemplateScalingOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateScalingOutputReference[_4] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateScalingOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateVolumesCloudSqlInstanceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instances: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.instances),
    };
}
exports.cloudRunV2ServiceTemplateVolumesCloudSqlInstanceToTerraform = cloudRunV2ServiceTemplateVolumesCloudSqlInstanceToTerraform;
class CloudRunV2ServiceTemplateVolumesCloudSqlInstanceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instances !== undefined) {
            hasAnyValues = true;
            internalValueResult.instances = this._instances;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._instances = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._instances = value.instances;
        }
    }
    get instances() {
        return this.getListAttribute('instances');
    }
    set instances(value) {
        this._instances = value;
    }
    resetInstances() {
        this._instances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instancesInput() {
        return this._instances;
    }
}
exports.CloudRunV2ServiceTemplateVolumesCloudSqlInstanceOutputReference = CloudRunV2ServiceTemplateVolumesCloudSqlInstanceOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateVolumesCloudSqlInstanceOutputReference[_5] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateVolumesCloudSqlInstanceOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateVolumesSecretItemsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.numberToTerraform(struct.mode),
        path: cdktf.stringToTerraform(struct.path),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.cloudRunV2ServiceTemplateVolumesSecretItemsToTerraform = cloudRunV2ServiceTemplateVolumesSecretItemsToTerraform;
class CloudRunV2ServiceTemplateVolumesSecretItemsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._mode = undefined;
            this._path = undefined;
            this._version = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._mode = value.mode;
            this._path = value.path;
            this._version = value.version;
        }
    }
    get mode() {
        return this.getNumberAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.CloudRunV2ServiceTemplateVolumesSecretItemsOutputReference = CloudRunV2ServiceTemplateVolumesSecretItemsOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateVolumesSecretItemsOutputReference[_6] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateVolumesSecretItemsOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTemplateVolumesSecretItemsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTemplateVolumesSecretItemsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTemplateVolumesSecretItemsList = CloudRunV2ServiceTemplateVolumesSecretItemsList;
_7 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateVolumesSecretItemsList[_7] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateVolumesSecretItemsList", version: "5.0.1" };
function cloudRunV2ServiceTemplateVolumesSecretToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_mode: cdktf.numberToTerraform(struct.defaultMode),
        secret: cdktf.stringToTerraform(struct.secret),
        items: cdktf.listMapper(cloudRunV2ServiceTemplateVolumesSecretItemsToTerraform, true)(struct.items),
    };
}
exports.cloudRunV2ServiceTemplateVolumesSecretToTerraform = cloudRunV2ServiceTemplateVolumesSecretToTerraform;
class CloudRunV2ServiceTemplateVolumesSecretOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // items - computed: false, optional: true, required: false
        this._items = new CloudRunV2ServiceTemplateVolumesSecretItemsList(this, "items", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultMode = this._defaultMode;
        }
        if (this._secret !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret;
        }
        if (this._items?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.items = this._items?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultMode = undefined;
            this._secret = undefined;
            this._items.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultMode = value.defaultMode;
            this._secret = value.secret;
            this._items.internalValue = value.items;
        }
    }
    get defaultMode() {
        return this.getNumberAttribute('default_mode');
    }
    set defaultMode(value) {
        this._defaultMode = value;
    }
    resetDefaultMode() {
        this._defaultMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultModeInput() {
        return this._defaultMode;
    }
    get secret() {
        return this.getStringAttribute('secret');
    }
    set secret(value) {
        this._secret = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret;
    }
    get items() {
        return this._items;
    }
    putItems(value) {
        this._items.internalValue = value;
    }
    resetItems() {
        this._items.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get itemsInput() {
        return this._items.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateVolumesSecretOutputReference = CloudRunV2ServiceTemplateVolumesSecretOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateVolumesSecretOutputReference[_8] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateVolumesSecretOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        cloud_sql_instance: cloudRunV2ServiceTemplateVolumesCloudSqlInstanceToTerraform(struct.cloudSqlInstance),
        secret: cloudRunV2ServiceTemplateVolumesSecretToTerraform(struct.secret),
    };
}
exports.cloudRunV2ServiceTemplateVolumesToTerraform = cloudRunV2ServiceTemplateVolumesToTerraform;
class CloudRunV2ServiceTemplateVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cloud_sql_instance - computed: false, optional: true, required: false
        this._cloudSqlInstance = new CloudRunV2ServiceTemplateVolumesCloudSqlInstanceOutputReference(this, "cloud_sql_instance");
        // secret - computed: false, optional: true, required: false
        this._secret = new CloudRunV2ServiceTemplateVolumesSecretOutputReference(this, "secret");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._cloudSqlInstance?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudSqlInstance = this._cloudSqlInstance?.internalValue;
        }
        if (this._secret?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._cloudSqlInstance.internalValue = undefined;
            this._secret.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._cloudSqlInstance.internalValue = value.cloudSqlInstance;
            this._secret.internalValue = value.secret;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get cloudSqlInstance() {
        return this._cloudSqlInstance;
    }
    putCloudSqlInstance(value) {
        this._cloudSqlInstance.internalValue = value;
    }
    resetCloudSqlInstance() {
        this._cloudSqlInstance.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudSqlInstanceInput() {
        return this._cloudSqlInstance.internalValue;
    }
    get secret() {
        return this._secret;
    }
    putSecret(value) {
        this._secret.internalValue = value;
    }
    resetSecret() {
        this._secret.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateVolumesOutputReference = CloudRunV2ServiceTemplateVolumesOutputReference;
_9 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateVolumesOutputReference[_9] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateVolumesOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTemplateVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTemplateVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTemplateVolumesList = CloudRunV2ServiceTemplateVolumesList;
_10 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateVolumesList[_10] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateVolumesList", version: "5.0.1" };
function cloudRunV2ServiceTemplateVpcAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connector: cdktf.stringToTerraform(struct.connector),
        egress: cdktf.stringToTerraform(struct.egress),
    };
}
exports.cloudRunV2ServiceTemplateVpcAccessToTerraform = cloudRunV2ServiceTemplateVpcAccessToTerraform;
class CloudRunV2ServiceTemplateVpcAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connector !== undefined) {
            hasAnyValues = true;
            internalValueResult.connector = this._connector;
        }
        if (this._egress !== undefined) {
            hasAnyValues = true;
            internalValueResult.egress = this._egress;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._connector = undefined;
            this._egress = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._connector = value.connector;
            this._egress = value.egress;
        }
    }
    get connector() {
        return this.getStringAttribute('connector');
    }
    set connector(value) {
        this._connector = value;
    }
    resetConnector() {
        this._connector = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectorInput() {
        return this._connector;
    }
    get egress() {
        return this.getStringAttribute('egress');
    }
    set egress(value) {
        this._egress = value;
    }
    resetEgress() {
        this._egress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get egressInput() {
        return this._egress;
    }
}
exports.CloudRunV2ServiceTemplateVpcAccessOutputReference = CloudRunV2ServiceTemplateVpcAccessOutputReference;
_11 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateVpcAccessOutputReference[_11] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateVpcAccessOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_key: cdktf.stringToTerraform(struct.encryptionKey),
        execution_environment: cdktf.stringToTerraform(struct.executionEnvironment),
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        max_instance_request_concurrency: cdktf.numberToTerraform(struct.maxInstanceRequestConcurrency),
        revision: cdktf.stringToTerraform(struct.revision),
        service_account: cdktf.stringToTerraform(struct.serviceAccount),
        timeout: cdktf.stringToTerraform(struct.timeout),
        containers: cdktf.listMapper(cloudRunV2ServiceTemplateContainersToTerraform, true)(struct.containers),
        scaling: cloudRunV2ServiceTemplateScalingToTerraform(struct.scaling),
        volumes: cdktf.listMapper(cloudRunV2ServiceTemplateVolumesToTerraform, true)(struct.volumes),
        vpc_access: cloudRunV2ServiceTemplateVpcAccessToTerraform(struct.vpcAccess),
    };
}
exports.cloudRunV2ServiceTemplateToTerraform = cloudRunV2ServiceTemplateToTerraform;
class CloudRunV2ServiceTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // containers - computed: false, optional: true, required: false
        this._containers = new CloudRunV2ServiceTemplateContainersList(this, "containers", false);
        // scaling - computed: false, optional: true, required: false
        this._scaling = new CloudRunV2ServiceTemplateScalingOutputReference(this, "scaling");
        // volumes - computed: false, optional: true, required: false
        this._volumes = new CloudRunV2ServiceTemplateVolumesList(this, "volumes", false);
        // vpc_access - computed: false, optional: true, required: false
        this._vpcAccess = new CloudRunV2ServiceTemplateVpcAccessOutputReference(this, "vpc_access");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionKey = this._encryptionKey;
        }
        if (this._executionEnvironment !== undefined) {
            hasAnyValues = true;
            internalValueResult.executionEnvironment = this._executionEnvironment;
        }
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._maxInstanceRequestConcurrency !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxInstanceRequestConcurrency = this._maxInstanceRequestConcurrency;
        }
        if (this._revision !== undefined) {
            hasAnyValues = true;
            internalValueResult.revision = this._revision;
        }
        if (this._serviceAccount !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccount = this._serviceAccount;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        if (this._containers?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.containers = this._containers?.internalValue;
        }
        if (this._scaling?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.scaling = this._scaling?.internalValue;
        }
        if (this._volumes?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumes = this._volumes?.internalValue;
        }
        if (this._vpcAccess?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpcAccess = this._vpcAccess?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._encryptionKey = undefined;
            this._executionEnvironment = undefined;
            this._labels = undefined;
            this._maxInstanceRequestConcurrency = undefined;
            this._revision = undefined;
            this._serviceAccount = undefined;
            this._timeout = undefined;
            this._containers.internalValue = undefined;
            this._scaling.internalValue = undefined;
            this._volumes.internalValue = undefined;
            this._vpcAccess.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._encryptionKey = value.encryptionKey;
            this._executionEnvironment = value.executionEnvironment;
            this._labels = value.labels;
            this._maxInstanceRequestConcurrency = value.maxInstanceRequestConcurrency;
            this._revision = value.revision;
            this._serviceAccount = value.serviceAccount;
            this._timeout = value.timeout;
            this._containers.internalValue = value.containers;
            this._scaling.internalValue = value.scaling;
            this._volumes.internalValue = value.volumes;
            this._vpcAccess.internalValue = value.vpcAccess;
        }
    }
    get encryptionKey() {
        return this.getStringAttribute('encryption_key');
    }
    set encryptionKey(value) {
        this._encryptionKey = value;
    }
    resetEncryptionKey() {
        this._encryptionKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionKeyInput() {
        return this._encryptionKey;
    }
    get executionEnvironment() {
        return this.getStringAttribute('execution_environment');
    }
    set executionEnvironment(value) {
        this._executionEnvironment = value;
    }
    resetExecutionEnvironment() {
        this._executionEnvironment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get executionEnvironmentInput() {
        return this._executionEnvironment;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get maxInstanceRequestConcurrency() {
        return this.getNumberAttribute('max_instance_request_concurrency');
    }
    set maxInstanceRequestConcurrency(value) {
        this._maxInstanceRequestConcurrency = value;
    }
    resetMaxInstanceRequestConcurrency() {
        this._maxInstanceRequestConcurrency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxInstanceRequestConcurrencyInput() {
        return this._maxInstanceRequestConcurrency;
    }
    get revision() {
        return this.getStringAttribute('revision');
    }
    set revision(value) {
        this._revision = value;
    }
    resetRevision() {
        this._revision = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get revisionInput() {
        return this._revision;
    }
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    set serviceAccount(value) {
        this._serviceAccount = value;
    }
    resetServiceAccount() {
        this._serviceAccount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountInput() {
        return this._serviceAccount;
    }
    get timeout() {
        return this.getStringAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get containers() {
        return this._containers;
    }
    putContainers(value) {
        this._containers.internalValue = value;
    }
    resetContainers() {
        this._containers.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containersInput() {
        return this._containers.internalValue;
    }
    get scaling() {
        return this._scaling;
    }
    putScaling(value) {
        this._scaling.internalValue = value;
    }
    resetScaling() {
        this._scaling.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scalingInput() {
        return this._scaling.internalValue;
    }
    get volumes() {
        return this._volumes;
    }
    putVolumes(value) {
        this._volumes.internalValue = value;
    }
    resetVolumes() {
        this._volumes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesInput() {
        return this._volumes.internalValue;
    }
    get vpcAccess() {
        return this._vpcAccess;
    }
    putVpcAccess(value) {
        this._vpcAccess.internalValue = value;
    }
    resetVpcAccess() {
        this._vpcAccess.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcAccessInput() {
        return this._vpcAccess.internalValue;
    }
}
exports.CloudRunV2ServiceTemplateOutputReference = CloudRunV2ServiceTemplateOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTemplateOutputReference[_12] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTemplateOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudRunV2ServiceTimeoutsToTerraform = cloudRunV2ServiceTimeoutsToTerraform;
class CloudRunV2ServiceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CloudRunV2ServiceTimeoutsOutputReference = CloudRunV2ServiceTimeoutsOutputReference;
_13 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTimeoutsOutputReference[_13] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTimeoutsOutputReference", version: "5.0.1" };
function cloudRunV2ServiceTrafficToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        percent: cdktf.numberToTerraform(struct.percent),
        revision: cdktf.stringToTerraform(struct.revision),
        tag: cdktf.stringToTerraform(struct.tag),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.cloudRunV2ServiceTrafficToTerraform = cloudRunV2ServiceTrafficToTerraform;
class CloudRunV2ServiceTrafficOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._percent !== undefined) {
            hasAnyValues = true;
            internalValueResult.percent = this._percent;
        }
        if (this._revision !== undefined) {
            hasAnyValues = true;
            internalValueResult.revision = this._revision;
        }
        if (this._tag !== undefined) {
            hasAnyValues = true;
            internalValueResult.tag = this._tag;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._percent = undefined;
            this._revision = undefined;
            this._tag = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._percent = value.percent;
            this._revision = value.revision;
            this._tag = value.tag;
            this._type = value.type;
        }
    }
    get percent() {
        return this.getNumberAttribute('percent');
    }
    set percent(value) {
        this._percent = value;
    }
    resetPercent() {
        this._percent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get percentInput() {
        return this._percent;
    }
    get revision() {
        return this.getStringAttribute('revision');
    }
    set revision(value) {
        this._revision = value;
    }
    resetRevision() {
        this._revision = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get revisionInput() {
        return this._revision;
    }
    get tag() {
        return this.getStringAttribute('tag');
    }
    set tag(value) {
        this._tag = value;
    }
    resetTag() {
        this._tag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagInput() {
        return this._tag;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.CloudRunV2ServiceTrafficOutputReference = CloudRunV2ServiceTrafficOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTrafficOutputReference[_14] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTrafficOutputReference", version: "5.0.1" };
class CloudRunV2ServiceTrafficList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2ServiceTrafficOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2ServiceTrafficList = CloudRunV2ServiceTrafficList;
_15 = JSII_RTTI_SYMBOL_1;
CloudRunV2ServiceTrafficList[_15] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2ServiceTrafficList", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/cloud_run_v2_service google_cloud_run_v2_service}
*/
class CloudRunV2Service extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/cloud_run_v2_service google_cloud_run_v2_service} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudRunV2ServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloud_run_v2_service',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // conditions - computed: true, optional: false, required: false
        this._conditions = new CloudRunV2ServiceConditionsList(this, "conditions", false);
        // terminal_condition - computed: true, optional: false, required: false
        this._terminalCondition = new CloudRunV2ServiceTerminalConditionList(this, "terminal_condition", false);
        // traffic_statuses - computed: true, optional: false, required: false
        this._trafficStatuses = new CloudRunV2ServiceTrafficStatusesList(this, "traffic_statuses", false);
        // binary_authorization - computed: false, optional: true, required: false
        this._binaryAuthorization = new CloudRunV2ServiceBinaryAuthorizationOutputReference(this, "binary_authorization");
        // template - computed: false, optional: false, required: true
        this._template = new CloudRunV2ServiceTemplateOutputReference(this, "template");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudRunV2ServiceTimeoutsOutputReference(this, "timeouts");
        // traffic - computed: false, optional: true, required: false
        this._traffic = new CloudRunV2ServiceTrafficList(this, "traffic", false);
        this._client = config.client;
        this._clientVersion = config.clientVersion;
        this._description = config.description;
        this._id = config.id;
        this._ingress = config.ingress;
        this._labels = config.labels;
        this._launchStage = config.launchStage;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._binaryAuthorization.internalValue = config.binaryAuthorization;
        this._template.internalValue = config.template;
        this._timeouts.internalValue = config.timeouts;
        this._traffic.internalValue = config.traffic;
    }
    get client() {
        return this.getStringAttribute('client');
    }
    set client(value) {
        this._client = value;
    }
    resetClient() {
        this._client = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientInput() {
        return this._client;
    }
    get clientVersion() {
        return this.getStringAttribute('client_version');
    }
    set clientVersion(value) {
        this._clientVersion = value;
    }
    resetClientVersion() {
        this._clientVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientVersionInput() {
        return this._clientVersion;
    }
    get conditions() {
        return this._conditions;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    // generation - computed: true, optional: false, required: false
    get generation() {
        return this.getStringAttribute('generation');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ingress() {
        return this.getStringAttribute('ingress');
    }
    set ingress(value) {
        this._ingress = value;
    }
    resetIngress() {
        this._ingress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ingressInput() {
        return this._ingress;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    // latest_created_revision - computed: true, optional: false, required: false
    get latestCreatedRevision() {
        return this.getStringAttribute('latest_created_revision');
    }
    // latest_ready_revision - computed: true, optional: false, required: false
    get latestReadyRevision() {
        return this.getStringAttribute('latest_ready_revision');
    }
    get launchStage() {
        return this.getStringAttribute('launch_stage');
    }
    set launchStage(value) {
        this._launchStage = value;
    }
    resetLaunchStage() {
        this._launchStage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchStageInput() {
        return this._launchStage;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // observed_generation - computed: true, optional: false, required: false
    get observedGeneration() {
        return this.getStringAttribute('observed_generation');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // reconciling - computed: true, optional: false, required: false
    get reconciling() {
        return this.getBooleanAttribute('reconciling');
    }
    get terminalCondition() {
        return this._terminalCondition;
    }
    get trafficStatuses() {
        return this._trafficStatuses;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
    get binaryAuthorization() {
        return this._binaryAuthorization;
    }
    putBinaryAuthorization(value) {
        this._binaryAuthorization.internalValue = value;
    }
    resetBinaryAuthorization() {
        this._binaryAuthorization.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get binaryAuthorizationInput() {
        return this._binaryAuthorization.internalValue;
    }
    get template() {
        return this._template;
    }
    putTemplate(value) {
        this._template.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get templateInput() {
        return this._template.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get traffic() {
        return this._traffic;
    }
    putTraffic(value) {
        this._traffic.internalValue = value;
    }
    resetTraffic() {
        this._traffic.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get trafficInput() {
        return this._traffic.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            client: cdktf.stringToTerraform(this._client),
            client_version: cdktf.stringToTerraform(this._clientVersion),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            ingress: cdktf.stringToTerraform(this._ingress),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            launch_stage: cdktf.stringToTerraform(this._launchStage),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            binary_authorization: cloudRunV2ServiceBinaryAuthorizationToTerraform(this._binaryAuthorization.internalValue),
            template: cloudRunV2ServiceTemplateToTerraform(this._template.internalValue),
            timeouts: cloudRunV2ServiceTimeoutsToTerraform(this._timeouts.internalValue),
            traffic: cdktf.listMapper(cloudRunV2ServiceTrafficToTerraform, true)(this._traffic.internalValue),
        };
    }
}
exports.CloudRunV2Service = CloudRunV2Service;
_16 = JSII_RTTI_SYMBOL_1;
CloudRunV2Service[_16] = { fqn: "@cdktf/provider-google.cloudRunV2Service.CloudRunV2Service", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
CloudRunV2Service.tfResourceType = "google_cloud_run_v2_service";
