// @cdktf/provider-google/lib/data-google-pubsub-topic/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGooglePubsubTopic = exports.DataGooglePubsubTopicSchemaSettingsList = exports.DataGooglePubsubTopicSchemaSettingsOutputReference = exports.dataGooglePubsubTopicSchemaSettingsToTerraform = exports.DataGooglePubsubTopicMessageStoragePolicyList = exports.DataGooglePubsubTopicMessageStoragePolicyOutputReference = exports.dataGooglePubsubTopicMessageStoragePolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGooglePubsubTopicMessageStoragePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicMessageStoragePolicyToTerraform = dataGooglePubsubTopicMessageStoragePolicyToTerraform;
class DataGooglePubsubTopicMessageStoragePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allowed_persistence_regions - computed: true, optional: false, required: false
    get allowedPersistenceRegions() {
        return this.getListAttribute('allowed_persistence_regions');
    }
}
exports.DataGooglePubsubTopicMessageStoragePolicyOutputReference = DataGooglePubsubTopicMessageStoragePolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicMessageStoragePolicyOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicMessageStoragePolicyOutputReference", version: "5.0.1" };
class DataGooglePubsubTopicMessageStoragePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicMessageStoragePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicMessageStoragePolicyList = DataGooglePubsubTopicMessageStoragePolicyList;
_b = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicMessageStoragePolicyList[_b] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicMessageStoragePolicyList", version: "5.0.1" };
function dataGooglePubsubTopicSchemaSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicSchemaSettingsToTerraform = dataGooglePubsubTopicSchemaSettingsToTerraform;
class DataGooglePubsubTopicSchemaSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // encoding - computed: true, optional: false, required: false
    get encoding() {
        return this.getStringAttribute('encoding');
    }
    // schema - computed: true, optional: false, required: false
    get schema() {
        return this.getStringAttribute('schema');
    }
}
exports.DataGooglePubsubTopicSchemaSettingsOutputReference = DataGooglePubsubTopicSchemaSettingsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicSchemaSettingsOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicSchemaSettingsOutputReference", version: "5.0.1" };
class DataGooglePubsubTopicSchemaSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicSchemaSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicSchemaSettingsList = DataGooglePubsubTopicSchemaSettingsList;
_d = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicSchemaSettingsList[_d] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicSchemaSettingsList", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/pubsub_topic google_pubsub_topic}
*/
class DataGooglePubsubTopic extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/pubsub_topic google_pubsub_topic} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGooglePubsubTopicConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_pubsub_topic',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // message_storage_policy - computed: true, optional: false, required: false
        this._messageStoragePolicy = new DataGooglePubsubTopicMessageStoragePolicyList(this, "message_storage_policy", false);
        // schema_settings - computed: true, optional: false, required: false
        this._schemaSettings = new DataGooglePubsubTopicSchemaSettingsList(this, "schema_settings", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kms_key_name - computed: true, optional: false, required: false
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
    get labels() {
        return this._labels;
    }
    // message_retention_duration - computed: true, optional: false, required: false
    get messageRetentionDuration() {
        return this.getStringAttribute('message_retention_duration');
    }
    get messageStoragePolicy() {
        return this._messageStoragePolicy;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get schemaSettings() {
        return this._schemaSettings;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
}
exports.DataGooglePubsubTopic = DataGooglePubsubTopic;
_e = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopic[_e] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopic", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
DataGooglePubsubTopic.tfResourceType = "google_pubsub_topic";
