// @cdktf/provider-google/lib/data-google-compute-instance/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeInstance = exports.DataGoogleComputeInstanceShieldedInstanceConfigList = exports.DataGoogleComputeInstanceShieldedInstanceConfigOutputReference = exports.dataGoogleComputeInstanceShieldedInstanceConfigToTerraform = exports.DataGoogleComputeInstanceServiceAccountList = exports.DataGoogleComputeInstanceServiceAccountOutputReference = exports.dataGoogleComputeInstanceServiceAccountToTerraform = exports.DataGoogleComputeInstanceScratchDiskList = exports.DataGoogleComputeInstanceScratchDiskOutputReference = exports.dataGoogleComputeInstanceScratchDiskToTerraform = exports.DataGoogleComputeInstanceSchedulingList = exports.DataGoogleComputeInstanceSchedulingOutputReference = exports.dataGoogleComputeInstanceSchedulingToTerraform = exports.DataGoogleComputeInstanceSchedulingNodeAffinitiesList = exports.DataGoogleComputeInstanceSchedulingNodeAffinitiesOutputReference = exports.dataGoogleComputeInstanceSchedulingNodeAffinitiesToTerraform = exports.DataGoogleComputeInstanceReservationAffinityList = exports.DataGoogleComputeInstanceReservationAffinityOutputReference = exports.dataGoogleComputeInstanceReservationAffinityToTerraform = exports.DataGoogleComputeInstanceReservationAffinitySpecificReservationList = exports.DataGoogleComputeInstanceReservationAffinitySpecificReservationOutputReference = exports.dataGoogleComputeInstanceReservationAffinitySpecificReservationToTerraform = exports.DataGoogleComputeInstanceNetworkInterfaceList = exports.DataGoogleComputeInstanceNetworkInterfaceOutputReference = exports.dataGoogleComputeInstanceNetworkInterfaceToTerraform = exports.DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigList = exports.DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigOutputReference = exports.dataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigToTerraform = exports.DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeList = exports.DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeOutputReference = exports.dataGoogleComputeInstanceNetworkInterfaceAliasIpRangeToTerraform = exports.DataGoogleComputeInstanceNetworkInterfaceAccessConfigList = exports.DataGoogleComputeInstanceNetworkInterfaceAccessConfigOutputReference = exports.dataGoogleComputeInstanceNetworkInterfaceAccessConfigToTerraform = exports.DataGoogleComputeInstanceGuestAcceleratorList = exports.DataGoogleComputeInstanceGuestAcceleratorOutputReference = exports.dataGoogleComputeInstanceGuestAcceleratorToTerraform = exports.DataGoogleComputeInstanceConfidentialInstanceConfigList = exports.DataGoogleComputeInstanceConfidentialInstanceConfigOutputReference = exports.dataGoogleComputeInstanceConfidentialInstanceConfigToTerraform = exports.DataGoogleComputeInstanceBootDiskList = exports.DataGoogleComputeInstanceBootDiskOutputReference = exports.dataGoogleComputeInstanceBootDiskToTerraform = exports.DataGoogleComputeInstanceBootDiskInitializeParamsList = exports.DataGoogleComputeInstanceBootDiskInitializeParamsOutputReference = exports.dataGoogleComputeInstanceBootDiskInitializeParamsToTerraform = exports.DataGoogleComputeInstanceAttachedDiskList = exports.DataGoogleComputeInstanceAttachedDiskOutputReference = exports.dataGoogleComputeInstanceAttachedDiskToTerraform = exports.DataGoogleComputeInstanceAdvancedMachineFeaturesList = exports.DataGoogleComputeInstanceAdvancedMachineFeaturesOutputReference = exports.dataGoogleComputeInstanceAdvancedMachineFeaturesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeInstanceAdvancedMachineFeaturesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceAdvancedMachineFeaturesToTerraform = dataGoogleComputeInstanceAdvancedMachineFeaturesToTerraform;
class DataGoogleComputeInstanceAdvancedMachineFeaturesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_nested_virtualization - computed: true, optional: false, required: false
    get enableNestedVirtualization() {
        return this.getBooleanAttribute('enable_nested_virtualization');
    }
    // threads_per_core - computed: true, optional: false, required: false
    get threadsPerCore() {
        return this.getNumberAttribute('threads_per_core');
    }
    // visible_core_count - computed: true, optional: false, required: false
    get visibleCoreCount() {
        return this.getNumberAttribute('visible_core_count');
    }
}
exports.DataGoogleComputeInstanceAdvancedMachineFeaturesOutputReference = DataGoogleComputeInstanceAdvancedMachineFeaturesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceAdvancedMachineFeaturesOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceAdvancedMachineFeaturesOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceAdvancedMachineFeaturesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceAdvancedMachineFeaturesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceAdvancedMachineFeaturesList = DataGoogleComputeInstanceAdvancedMachineFeaturesList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceAdvancedMachineFeaturesList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceAdvancedMachineFeaturesList", version: "5.0.1" };
function dataGoogleComputeInstanceAttachedDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceAttachedDiskToTerraform = dataGoogleComputeInstanceAttachedDiskToTerraform;
class DataGoogleComputeInstanceAttachedDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // disk_encryption_key_raw - computed: true, optional: false, required: false
    get diskEncryptionKeyRaw() {
        return this.getStringAttribute('disk_encryption_key_raw');
    }
    // disk_encryption_key_sha256 - computed: true, optional: false, required: false
    get diskEncryptionKeySha256() {
        return this.getStringAttribute('disk_encryption_key_sha256');
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
    // source - computed: true, optional: false, required: false
    get source() {
        return this.getStringAttribute('source');
    }
}
exports.DataGoogleComputeInstanceAttachedDiskOutputReference = DataGoogleComputeInstanceAttachedDiskOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceAttachedDiskOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceAttachedDiskOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceAttachedDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceAttachedDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceAttachedDiskList = DataGoogleComputeInstanceAttachedDiskList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceAttachedDiskList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceAttachedDiskList", version: "5.0.1" };
function dataGoogleComputeInstanceBootDiskInitializeParamsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceBootDiskInitializeParamsToTerraform = dataGoogleComputeInstanceBootDiskInitializeParamsToTerraform;
class DataGoogleComputeInstanceBootDiskInitializeParamsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // image - computed: true, optional: false, required: false
    get image() {
        return this.getStringAttribute('image');
    }
    get labels() {
        return this._labels;
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceBootDiskInitializeParamsOutputReference = DataGoogleComputeInstanceBootDiskInitializeParamsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceBootDiskInitializeParamsOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceBootDiskInitializeParamsOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceBootDiskInitializeParamsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceBootDiskInitializeParamsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceBootDiskInitializeParamsList = DataGoogleComputeInstanceBootDiskInitializeParamsList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceBootDiskInitializeParamsList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceBootDiskInitializeParamsList", version: "5.0.1" };
function dataGoogleComputeInstanceBootDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceBootDiskToTerraform = dataGoogleComputeInstanceBootDiskToTerraform;
class DataGoogleComputeInstanceBootDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // initialize_params - computed: true, optional: false, required: false
        this._initializeParams = new DataGoogleComputeInstanceBootDiskInitializeParamsList(this, "initialize_params", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // auto_delete - computed: true, optional: false, required: false
    get autoDelete() {
        return this.getBooleanAttribute('auto_delete');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // disk_encryption_key_raw - computed: true, optional: false, required: false
    get diskEncryptionKeyRaw() {
        return this.getStringAttribute('disk_encryption_key_raw');
    }
    // disk_encryption_key_sha256 - computed: true, optional: false, required: false
    get diskEncryptionKeySha256() {
        return this.getStringAttribute('disk_encryption_key_sha256');
    }
    get initializeParams() {
        return this._initializeParams;
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
    // source - computed: true, optional: false, required: false
    get source() {
        return this.getStringAttribute('source');
    }
}
exports.DataGoogleComputeInstanceBootDiskOutputReference = DataGoogleComputeInstanceBootDiskOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceBootDiskOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceBootDiskOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceBootDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceBootDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceBootDiskList = DataGoogleComputeInstanceBootDiskList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceBootDiskList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceBootDiskList", version: "5.0.1" };
function dataGoogleComputeInstanceConfidentialInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceConfidentialInstanceConfigToTerraform = dataGoogleComputeInstanceConfidentialInstanceConfigToTerraform;
class DataGoogleComputeInstanceConfidentialInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_confidential_compute - computed: true, optional: false, required: false
    get enableConfidentialCompute() {
        return this.getBooleanAttribute('enable_confidential_compute');
    }
}
exports.DataGoogleComputeInstanceConfidentialInstanceConfigOutputReference = DataGoogleComputeInstanceConfidentialInstanceConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceConfidentialInstanceConfigOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceConfidentialInstanceConfigOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceConfidentialInstanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceConfidentialInstanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceConfidentialInstanceConfigList = DataGoogleComputeInstanceConfidentialInstanceConfigList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceConfidentialInstanceConfigList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceConfidentialInstanceConfigList", version: "5.0.1" };
function dataGoogleComputeInstanceGuestAcceleratorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGuestAcceleratorToTerraform = dataGoogleComputeInstanceGuestAcceleratorToTerraform;
class DataGoogleComputeInstanceGuestAcceleratorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // count - computed: true, optional: false, required: false
    get count() {
        return this.getNumberAttribute('count');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceGuestAcceleratorOutputReference = DataGoogleComputeInstanceGuestAcceleratorOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGuestAcceleratorOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceGuestAcceleratorOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceGuestAcceleratorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGuestAcceleratorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGuestAcceleratorList = DataGoogleComputeInstanceGuestAcceleratorList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGuestAcceleratorList[_m] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceGuestAcceleratorList", version: "5.0.1" };
function dataGoogleComputeInstanceNetworkInterfaceAccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceNetworkInterfaceAccessConfigToTerraform = dataGoogleComputeInstanceNetworkInterfaceAccessConfigToTerraform;
class DataGoogleComputeInstanceNetworkInterfaceAccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // nat_ip - computed: true, optional: false, required: false
    get natIp() {
        return this.getStringAttribute('nat_ip');
    }
    // network_tier - computed: true, optional: false, required: false
    get networkTier() {
        return this.getStringAttribute('network_tier');
    }
    // public_ptr_domain_name - computed: true, optional: false, required: false
    get publicPtrDomainName() {
        return this.getStringAttribute('public_ptr_domain_name');
    }
}
exports.DataGoogleComputeInstanceNetworkInterfaceAccessConfigOutputReference = DataGoogleComputeInstanceNetworkInterfaceAccessConfigOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceNetworkInterfaceAccessConfigOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceNetworkInterfaceAccessConfigOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceNetworkInterfaceAccessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceNetworkInterfaceAccessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceNetworkInterfaceAccessConfigList = DataGoogleComputeInstanceNetworkInterfaceAccessConfigList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceNetworkInterfaceAccessConfigList[_p] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceNetworkInterfaceAccessConfigList", version: "5.0.1" };
function dataGoogleComputeInstanceNetworkInterfaceAliasIpRangeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceNetworkInterfaceAliasIpRangeToTerraform = dataGoogleComputeInstanceNetworkInterfaceAliasIpRangeToTerraform;
class DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_cidr_range - computed: true, optional: false, required: false
    get ipCidrRange() {
        return this.getStringAttribute('ip_cidr_range');
    }
    // subnetwork_range_name - computed: true, optional: false, required: false
    get subnetworkRangeName() {
        return this.getStringAttribute('subnetwork_range_name');
    }
}
exports.DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeOutputReference = DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeList = DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeList[_r] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeList", version: "5.0.1" };
function dataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigToTerraform = dataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigToTerraform;
class DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // external_ipv6 - computed: true, optional: false, required: false
    get externalIpv6() {
        return this.getStringAttribute('external_ipv6');
    }
    // external_ipv6_prefix_length - computed: true, optional: false, required: false
    get externalIpv6PrefixLength() {
        return this.getStringAttribute('external_ipv6_prefix_length');
    }
    // network_tier - computed: true, optional: false, required: false
    get networkTier() {
        return this.getStringAttribute('network_tier');
    }
    // public_ptr_domain_name - computed: true, optional: false, required: false
    get publicPtrDomainName() {
        return this.getStringAttribute('public_ptr_domain_name');
    }
}
exports.DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigOutputReference = DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigList = DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigList[_t] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigList", version: "5.0.1" };
function dataGoogleComputeInstanceNetworkInterfaceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceNetworkInterfaceToTerraform = dataGoogleComputeInstanceNetworkInterfaceToTerraform;
class DataGoogleComputeInstanceNetworkInterfaceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // access_config - computed: true, optional: false, required: false
        this._accessConfig = new DataGoogleComputeInstanceNetworkInterfaceAccessConfigList(this, "access_config", false);
        // alias_ip_range - computed: true, optional: false, required: false
        this._aliasIpRange = new DataGoogleComputeInstanceNetworkInterfaceAliasIpRangeList(this, "alias_ip_range", false);
        // ipv6_access_config - computed: true, optional: false, required: false
        this._ipv6AccessConfig = new DataGoogleComputeInstanceNetworkInterfaceIpv6AccessConfigList(this, "ipv6_access_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get accessConfig() {
        return this._accessConfig;
    }
    get aliasIpRange() {
        return this._aliasIpRange;
    }
    get ipv6AccessConfig() {
        return this._ipv6AccessConfig;
    }
    // ipv6_access_type - computed: true, optional: false, required: false
    get ipv6AccessType() {
        return this.getStringAttribute('ipv6_access_type');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    // network_ip - computed: true, optional: false, required: false
    get networkIp() {
        return this.getStringAttribute('network_ip');
    }
    // nic_type - computed: true, optional: false, required: false
    get nicType() {
        return this.getStringAttribute('nic_type');
    }
    // queue_count - computed: true, optional: false, required: false
    get queueCount() {
        return this.getNumberAttribute('queue_count');
    }
    // stack_type - computed: true, optional: false, required: false
    get stackType() {
        return this.getStringAttribute('stack_type');
    }
    // subnetwork - computed: true, optional: false, required: false
    get subnetwork() {
        return this.getStringAttribute('subnetwork');
    }
    // subnetwork_project - computed: true, optional: false, required: false
    get subnetworkProject() {
        return this.getStringAttribute('subnetwork_project');
    }
}
exports.DataGoogleComputeInstanceNetworkInterfaceOutputReference = DataGoogleComputeInstanceNetworkInterfaceOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceNetworkInterfaceOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceNetworkInterfaceOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceNetworkInterfaceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceNetworkInterfaceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceNetworkInterfaceList = DataGoogleComputeInstanceNetworkInterfaceList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceNetworkInterfaceList[_v] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceNetworkInterfaceList", version: "5.0.1" };
function dataGoogleComputeInstanceReservationAffinitySpecificReservationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceReservationAffinitySpecificReservationToTerraform = dataGoogleComputeInstanceReservationAffinitySpecificReservationToTerraform;
class DataGoogleComputeInstanceReservationAffinitySpecificReservationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return this.getListAttribute('values');
    }
}
exports.DataGoogleComputeInstanceReservationAffinitySpecificReservationOutputReference = DataGoogleComputeInstanceReservationAffinitySpecificReservationOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceReservationAffinitySpecificReservationOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceReservationAffinitySpecificReservationOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceReservationAffinitySpecificReservationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceReservationAffinitySpecificReservationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceReservationAffinitySpecificReservationList = DataGoogleComputeInstanceReservationAffinitySpecificReservationList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceReservationAffinitySpecificReservationList[_x] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceReservationAffinitySpecificReservationList", version: "5.0.1" };
function dataGoogleComputeInstanceReservationAffinityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceReservationAffinityToTerraform = dataGoogleComputeInstanceReservationAffinityToTerraform;
class DataGoogleComputeInstanceReservationAffinityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // specific_reservation - computed: true, optional: false, required: false
        this._specificReservation = new DataGoogleComputeInstanceReservationAffinitySpecificReservationList(this, "specific_reservation", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get specificReservation() {
        return this._specificReservation;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceReservationAffinityOutputReference = DataGoogleComputeInstanceReservationAffinityOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceReservationAffinityOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceReservationAffinityOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceReservationAffinityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceReservationAffinityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceReservationAffinityList = DataGoogleComputeInstanceReservationAffinityList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceReservationAffinityList[_z] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceReservationAffinityList", version: "5.0.1" };
function dataGoogleComputeInstanceSchedulingNodeAffinitiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceSchedulingNodeAffinitiesToTerraform = dataGoogleComputeInstanceSchedulingNodeAffinitiesToTerraform;
class DataGoogleComputeInstanceSchedulingNodeAffinitiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // operator - computed: true, optional: false, required: false
    get operator() {
        return this.getStringAttribute('operator');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataGoogleComputeInstanceSchedulingNodeAffinitiesOutputReference = DataGoogleComputeInstanceSchedulingNodeAffinitiesOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceSchedulingNodeAffinitiesOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceSchedulingNodeAffinitiesOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceSchedulingNodeAffinitiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceSchedulingNodeAffinitiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceSchedulingNodeAffinitiesList = DataGoogleComputeInstanceSchedulingNodeAffinitiesList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceSchedulingNodeAffinitiesList[_1] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceSchedulingNodeAffinitiesList", version: "5.0.1" };
function dataGoogleComputeInstanceSchedulingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceSchedulingToTerraform = dataGoogleComputeInstanceSchedulingToTerraform;
class DataGoogleComputeInstanceSchedulingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // node_affinities - computed: true, optional: false, required: false
        this._nodeAffinities = new DataGoogleComputeInstanceSchedulingNodeAffinitiesList(this, "node_affinities", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // automatic_restart - computed: true, optional: false, required: false
    get automaticRestart() {
        return this.getBooleanAttribute('automatic_restart');
    }
    // instance_termination_action - computed: true, optional: false, required: false
    get instanceTerminationAction() {
        return this.getStringAttribute('instance_termination_action');
    }
    // min_node_cpus - computed: true, optional: false, required: false
    get minNodeCpus() {
        return this.getNumberAttribute('min_node_cpus');
    }
    get nodeAffinities() {
        return this._nodeAffinities;
    }
    // on_host_maintenance - computed: true, optional: false, required: false
    get onHostMaintenance() {
        return this.getStringAttribute('on_host_maintenance');
    }
    // preemptible - computed: true, optional: false, required: false
    get preemptible() {
        return this.getBooleanAttribute('preemptible');
    }
    // provisioning_model - computed: true, optional: false, required: false
    get provisioningModel() {
        return this.getStringAttribute('provisioning_model');
    }
}
exports.DataGoogleComputeInstanceSchedulingOutputReference = DataGoogleComputeInstanceSchedulingOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceSchedulingOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceSchedulingOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceSchedulingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceSchedulingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceSchedulingList = DataGoogleComputeInstanceSchedulingList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceSchedulingList[_3] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceSchedulingList", version: "5.0.1" };
function dataGoogleComputeInstanceScratchDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceScratchDiskToTerraform = dataGoogleComputeInstanceScratchDiskToTerraform;
class DataGoogleComputeInstanceScratchDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // interface - computed: true, optional: false, required: false
    get interface() {
        return this.getStringAttribute('interface');
    }
}
exports.DataGoogleComputeInstanceScratchDiskOutputReference = DataGoogleComputeInstanceScratchDiskOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceScratchDiskOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceScratchDiskOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceScratchDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceScratchDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceScratchDiskList = DataGoogleComputeInstanceScratchDiskList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceScratchDiskList[_5] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceScratchDiskList", version: "5.0.1" };
function dataGoogleComputeInstanceServiceAccountToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceServiceAccountToTerraform = dataGoogleComputeInstanceServiceAccountToTerraform;
class DataGoogleComputeInstanceServiceAccountOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
    // scopes - computed: true, optional: false, required: false
    get scopes() {
        return cdktf.Fn.tolist(this.getListAttribute('scopes'));
    }
}
exports.DataGoogleComputeInstanceServiceAccountOutputReference = DataGoogleComputeInstanceServiceAccountOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceServiceAccountOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceServiceAccountOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceServiceAccountList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceServiceAccountOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceServiceAccountList = DataGoogleComputeInstanceServiceAccountList;
_7 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceServiceAccountList[_7] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceServiceAccountList", version: "5.0.1" };
function dataGoogleComputeInstanceShieldedInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceShieldedInstanceConfigToTerraform = dataGoogleComputeInstanceShieldedInstanceConfigToTerraform;
class DataGoogleComputeInstanceShieldedInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_integrity_monitoring - computed: true, optional: false, required: false
    get enableIntegrityMonitoring() {
        return this.getBooleanAttribute('enable_integrity_monitoring');
    }
    // enable_secure_boot - computed: true, optional: false, required: false
    get enableSecureBoot() {
        return this.getBooleanAttribute('enable_secure_boot');
    }
    // enable_vtpm - computed: true, optional: false, required: false
    get enableVtpm() {
        return this.getBooleanAttribute('enable_vtpm');
    }
}
exports.DataGoogleComputeInstanceShieldedInstanceConfigOutputReference = DataGoogleComputeInstanceShieldedInstanceConfigOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceShieldedInstanceConfigOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceShieldedInstanceConfigOutputReference", version: "5.0.1" };
class DataGoogleComputeInstanceShieldedInstanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceShieldedInstanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceShieldedInstanceConfigList = DataGoogleComputeInstanceShieldedInstanceConfigList;
_9 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceShieldedInstanceConfigList[_9] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstanceShieldedInstanceConfigList", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/compute_instance google_compute_instance}
*/
class DataGoogleComputeInstance extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/compute_instance google_compute_instance} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeInstanceConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_compute_instance',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // advanced_machine_features - computed: true, optional: false, required: false
        this._advancedMachineFeatures = new DataGoogleComputeInstanceAdvancedMachineFeaturesList(this, "advanced_machine_features", false);
        // attached_disk - computed: true, optional: false, required: false
        this._attachedDisk = new DataGoogleComputeInstanceAttachedDiskList(this, "attached_disk", false);
        // boot_disk - computed: true, optional: false, required: false
        this._bootDisk = new DataGoogleComputeInstanceBootDiskList(this, "boot_disk", false);
        // confidential_instance_config - computed: true, optional: false, required: false
        this._confidentialInstanceConfig = new DataGoogleComputeInstanceConfidentialInstanceConfigList(this, "confidential_instance_config", false);
        // guest_accelerator - computed: true, optional: false, required: false
        this._guestAccelerator = new DataGoogleComputeInstanceGuestAcceleratorList(this, "guest_accelerator", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // metadata - computed: true, optional: false, required: false
        this._metadata = new cdktf.StringMap(this, "metadata");
        // network_interface - computed: true, optional: false, required: false
        this._networkInterface = new DataGoogleComputeInstanceNetworkInterfaceList(this, "network_interface", false);
        // reservation_affinity - computed: true, optional: false, required: false
        this._reservationAffinity = new DataGoogleComputeInstanceReservationAffinityList(this, "reservation_affinity", false);
        // scheduling - computed: true, optional: false, required: false
        this._scheduling = new DataGoogleComputeInstanceSchedulingList(this, "scheduling", false);
        // scratch_disk - computed: true, optional: false, required: false
        this._scratchDisk = new DataGoogleComputeInstanceScratchDiskList(this, "scratch_disk", false);
        // service_account - computed: true, optional: false, required: false
        this._serviceAccount = new DataGoogleComputeInstanceServiceAccountList(this, "service_account", false);
        // shielded_instance_config - computed: true, optional: false, required: false
        this._shieldedInstanceConfig = new DataGoogleComputeInstanceShieldedInstanceConfigList(this, "shielded_instance_config", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._selfLink = config.selfLink;
        this._zone = config.zone;
    }
    get advancedMachineFeatures() {
        return this._advancedMachineFeatures;
    }
    // allow_stopping_for_update - computed: true, optional: false, required: false
    get allowStoppingForUpdate() {
        return this.getBooleanAttribute('allow_stopping_for_update');
    }
    get attachedDisk() {
        return this._attachedDisk;
    }
    get bootDisk() {
        return this._bootDisk;
    }
    // can_ip_forward - computed: true, optional: false, required: false
    get canIpForward() {
        return this.getBooleanAttribute('can_ip_forward');
    }
    get confidentialInstanceConfig() {
        return this._confidentialInstanceConfig;
    }
    // cpu_platform - computed: true, optional: false, required: false
    get cpuPlatform() {
        return this.getStringAttribute('cpu_platform');
    }
    // current_status - computed: true, optional: false, required: false
    get currentStatus() {
        return this.getStringAttribute('current_status');
    }
    // deletion_protection - computed: true, optional: false, required: false
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // desired_status - computed: true, optional: false, required: false
    get desiredStatus() {
        return this.getStringAttribute('desired_status');
    }
    // enable_display - computed: true, optional: false, required: false
    get enableDisplay() {
        return this.getBooleanAttribute('enable_display');
    }
    get guestAccelerator() {
        return this._guestAccelerator;
    }
    // hostname - computed: true, optional: false, required: false
    get hostname() {
        return this.getStringAttribute('hostname');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_id - computed: true, optional: false, required: false
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    // label_fingerprint - computed: true, optional: false, required: false
    get labelFingerprint() {
        return this.getStringAttribute('label_fingerprint');
    }
    get labels() {
        return this._labels;
    }
    // machine_type - computed: true, optional: false, required: false
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    get metadata() {
        return this._metadata;
    }
    // metadata_fingerprint - computed: true, optional: false, required: false
    get metadataFingerprint() {
        return this.getStringAttribute('metadata_fingerprint');
    }
    // metadata_startup_script - computed: true, optional: false, required: false
    get metadataStartupScript() {
        return this.getStringAttribute('metadata_startup_script');
    }
    // min_cpu_platform - computed: true, optional: false, required: false
    get minCpuPlatform() {
        return this.getStringAttribute('min_cpu_platform');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get networkInterface() {
        return this._networkInterface;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get reservationAffinity() {
        return this._reservationAffinity;
    }
    // resource_policies - computed: true, optional: false, required: false
    get resourcePolicies() {
        return this.getListAttribute('resource_policies');
    }
    get scheduling() {
        return this._scheduling;
    }
    get scratchDisk() {
        return this._scratchDisk;
    }
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    set selfLink(value) {
        this._selfLink = value;
    }
    resetSelfLink() {
        this._selfLink = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfLinkInput() {
        return this._selfLink;
    }
    get serviceAccount() {
        return this._serviceAccount;
    }
    get shieldedInstanceConfig() {
        return this._shieldedInstanceConfig;
    }
    // tags - computed: true, optional: false, required: false
    get tags() {
        return cdktf.Fn.tolist(this.getListAttribute('tags'));
    }
    // tags_fingerprint - computed: true, optional: false, required: false
    get tagsFingerprint() {
        return this.getStringAttribute('tags_fingerprint');
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            self_link: cdktf.stringToTerraform(this._selfLink),
            zone: cdktf.stringToTerraform(this._zone),
        };
    }
}
exports.DataGoogleComputeInstance = DataGoogleComputeInstance;
_10 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstance[_10] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstance.DataGoogleComputeInstance", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeInstance.tfResourceType = "google_compute_instance";
