// cdktf/lib/backends/manta-backend.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTerraformRemoteStateManta = exports.MantaBackend = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_backend_1 = require("../terraform-backend");
const util_1 = require("../util");
const terraform_remote_state_1 = require("../terraform-remote-state");
/**
 * @deprecated CDK for Terraform no longer supports the manta backend. Terraform deprecated manta in v1.2.3 and removed it in v1.3.
 */
class MantaBackend extends terraform_backend_1.TerraformBackend {
    constructor(scope, props) {
        super(scope, "backend", "manta");
        this.props = props;
    }
    synthesizeAttributes() {
        return util_1.keysToSnakeCase({ ...this.props });
    }
    getRemoteStateDataSource(scope, name, _fromStack) {
        return new DataTerraformRemoteStateManta(scope, name, {
            ...this.props,
            workspace: "${terraform.workspace}",
        });
    }
}
exports.MantaBackend = MantaBackend;
_a = JSII_RTTI_SYMBOL_1;
MantaBackend[_a] = { fqn: "cdktf.MantaBackend", version: "0.14.3" };
/**
 * @deprecated CDK for Terraform no longer supports the manta backend. Terraform deprecated manta in v1.2.3 and removed it in v1.3.
 */
class DataTerraformRemoteStateManta extends terraform_remote_state_1.TerraformRemoteState {
    constructor(scope, id, config) {
        super(scope, id, "manta", config);
    }
}
exports.DataTerraformRemoteStateManta = DataTerraformRemoteStateManta;
_b = JSII_RTTI_SYMBOL_1;
DataTerraformRemoteStateManta[_b] = { fqn: "cdktf.DataTerraformRemoteStateManta", version: "0.14.3" };
