// @cdktf/provider-google/lib/certificate-manager-certificate/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CertificateManagerCertificate = exports.CertificateManagerCertificateTimeoutsOutputReference = exports.certificateManagerCertificateTimeoutsToTerraform = exports.CertificateManagerCertificateSelfManagedOutputReference = exports.certificateManagerCertificateSelfManagedToTerraform = exports.CertificateManagerCertificateManagedOutputReference = exports.certificateManagerCertificateManagedToTerraform = exports.CertificateManagerCertificateManagedProvisioningIssueList = exports.CertificateManagerCertificateManagedProvisioningIssueOutputReference = exports.certificateManagerCertificateManagedProvisioningIssueToTerraform = exports.CertificateManagerCertificateManagedAuthorizationAttemptInfoList = exports.CertificateManagerCertificateManagedAuthorizationAttemptInfoOutputReference = exports.certificateManagerCertificateManagedAuthorizationAttemptInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function certificateManagerCertificateManagedAuthorizationAttemptInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.certificateManagerCertificateManagedAuthorizationAttemptInfoToTerraform = certificateManagerCertificateManagedAuthorizationAttemptInfoToTerraform;
class CertificateManagerCertificateManagedAuthorizationAttemptInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // details - computed: true, optional: false, required: false
    get details() {
        return this.getStringAttribute('details');
    }
    // domain - computed: true, optional: false, required: false
    get domain() {
        return this.getStringAttribute('domain');
    }
    // failure_reason - computed: true, optional: false, required: false
    get failureReason() {
        return this.getStringAttribute('failure_reason');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.CertificateManagerCertificateManagedAuthorizationAttemptInfoOutputReference = CertificateManagerCertificateManagedAuthorizationAttemptInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CertificateManagerCertificateManagedAuthorizationAttemptInfoOutputReference[_a] = { fqn: "@cdktf/provider-google.certificateManagerCertificate.CertificateManagerCertificateManagedAuthorizationAttemptInfoOutputReference", version: "5.0.1" };
class CertificateManagerCertificateManagedAuthorizationAttemptInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CertificateManagerCertificateManagedAuthorizationAttemptInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CertificateManagerCertificateManagedAuthorizationAttemptInfoList = CertificateManagerCertificateManagedAuthorizationAttemptInfoList;
_b = JSII_RTTI_SYMBOL_1;
CertificateManagerCertificateManagedAuthorizationAttemptInfoList[_b] = { fqn: "@cdktf/provider-google.certificateManagerCertificate.CertificateManagerCertificateManagedAuthorizationAttemptInfoList", version: "5.0.1" };
function certificateManagerCertificateManagedProvisioningIssueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.certificateManagerCertificateManagedProvisioningIssueToTerraform = certificateManagerCertificateManagedProvisioningIssueToTerraform;
class CertificateManagerCertificateManagedProvisioningIssueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // details - computed: true, optional: false, required: false
    get details() {
        return this.getStringAttribute('details');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
}
exports.CertificateManagerCertificateManagedProvisioningIssueOutputReference = CertificateManagerCertificateManagedProvisioningIssueOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CertificateManagerCertificateManagedProvisioningIssueOutputReference[_c] = { fqn: "@cdktf/provider-google.certificateManagerCertificate.CertificateManagerCertificateManagedProvisioningIssueOutputReference", version: "5.0.1" };
class CertificateManagerCertificateManagedProvisioningIssueList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CertificateManagerCertificateManagedProvisioningIssueOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CertificateManagerCertificateManagedProvisioningIssueList = CertificateManagerCertificateManagedProvisioningIssueList;
_d = JSII_RTTI_SYMBOL_1;
CertificateManagerCertificateManagedProvisioningIssueList[_d] = { fqn: "@cdktf/provider-google.certificateManagerCertificate.CertificateManagerCertificateManagedProvisioningIssueList", version: "5.0.1" };
function certificateManagerCertificateManagedToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dns_authorizations: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.dnsAuthorizations),
        domains: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.domains),
    };
}
exports.certificateManagerCertificateManagedToTerraform = certificateManagerCertificateManagedToTerraform;
class CertificateManagerCertificateManagedOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // authorization_attempt_info - computed: true, optional: false, required: false
        this._authorizationAttemptInfo = new CertificateManagerCertificateManagedAuthorizationAttemptInfoList(this, "authorization_attempt_info", false);
        // provisioning_issue - computed: true, optional: false, required: false
        this._provisioningIssue = new CertificateManagerCertificateManagedProvisioningIssueList(this, "provisioning_issue", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dnsAuthorizations !== undefined) {
            hasAnyValues = true;
            internalValueResult.dnsAuthorizations = this._dnsAuthorizations;
        }
        if (this._domains !== undefined) {
            hasAnyValues = true;
            internalValueResult.domains = this._domains;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dnsAuthorizations = undefined;
            this._domains = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dnsAuthorizations = value.dnsAuthorizations;
            this._domains = value.domains;
        }
    }
    get authorizationAttemptInfo() {
        return this._authorizationAttemptInfo;
    }
    get dnsAuthorizations() {
        return this.getListAttribute('dns_authorizations');
    }
    set dnsAuthorizations(value) {
        this._dnsAuthorizations = value;
    }
    resetDnsAuthorizations() {
        this._dnsAuthorizations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dnsAuthorizationsInput() {
        return this._dnsAuthorizations;
    }
    get domains() {
        return this.getListAttribute('domains');
    }
    set domains(value) {
        this._domains = value;
    }
    resetDomains() {
        this._domains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainsInput() {
        return this._domains;
    }
    get provisioningIssue() {
        return this._provisioningIssue;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.CertificateManagerCertificateManagedOutputReference = CertificateManagerCertificateManagedOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CertificateManagerCertificateManagedOutputReference[_e] = { fqn: "@cdktf/provider-google.certificateManagerCertificate.CertificateManagerCertificateManagedOutputReference", version: "5.0.1" };
function certificateManagerCertificateSelfManagedToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate_pem: cdktf.stringToTerraform(struct.certificatePem),
        pem_certificate: cdktf.stringToTerraform(struct.pemCertificate),
        pem_private_key: cdktf.stringToTerraform(struct.pemPrivateKey),
        private_key_pem: cdktf.stringToTerraform(struct.privateKeyPem),
    };
}
exports.certificateManagerCertificateSelfManagedToTerraform = certificateManagerCertificateSelfManagedToTerraform;
class CertificateManagerCertificateSelfManagedOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificatePem !== undefined) {
            hasAnyValues = true;
            internalValueResult.certificatePem = this._certificatePem;
        }
        if (this._pemCertificate !== undefined) {
            hasAnyValues = true;
            internalValueResult.pemCertificate = this._pemCertificate;
        }
        if (this._pemPrivateKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.pemPrivateKey = this._pemPrivateKey;
        }
        if (this._privateKeyPem !== undefined) {
            hasAnyValues = true;
            internalValueResult.privateKeyPem = this._privateKeyPem;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificatePem = undefined;
            this._pemCertificate = undefined;
            this._pemPrivateKey = undefined;
            this._privateKeyPem = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificatePem = value.certificatePem;
            this._pemCertificate = value.pemCertificate;
            this._pemPrivateKey = value.pemPrivateKey;
            this._privateKeyPem = value.privateKeyPem;
        }
    }
    get certificatePem() {
        return this.getStringAttribute('certificate_pem');
    }
    set certificatePem(value) {
        this._certificatePem = value;
    }
    resetCertificatePem() {
        this._certificatePem = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get certificatePemInput() {
        return this._certificatePem;
    }
    get pemCertificate() {
        return this.getStringAttribute('pem_certificate');
    }
    set pemCertificate(value) {
        this._pemCertificate = value;
    }
    resetPemCertificate() {
        this._pemCertificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pemCertificateInput() {
        return this._pemCertificate;
    }
    get pemPrivateKey() {
        return this.getStringAttribute('pem_private_key');
    }
    set pemPrivateKey(value) {
        this._pemPrivateKey = value;
    }
    resetPemPrivateKey() {
        this._pemPrivateKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pemPrivateKeyInput() {
        return this._pemPrivateKey;
    }
    get privateKeyPem() {
        return this.getStringAttribute('private_key_pem');
    }
    set privateKeyPem(value) {
        this._privateKeyPem = value;
    }
    resetPrivateKeyPem() {
        this._privateKeyPem = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privateKeyPemInput() {
        return this._privateKeyPem;
    }
}
exports.CertificateManagerCertificateSelfManagedOutputReference = CertificateManagerCertificateSelfManagedOutputReference;
_f = JSII_RTTI_SYMBOL_1;
CertificateManagerCertificateSelfManagedOutputReference[_f] = { fqn: "@cdktf/provider-google.certificateManagerCertificate.CertificateManagerCertificateSelfManagedOutputReference", version: "5.0.1" };
function certificateManagerCertificateTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.certificateManagerCertificateTimeoutsToTerraform = certificateManagerCertificateTimeoutsToTerraform;
class CertificateManagerCertificateTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CertificateManagerCertificateTimeoutsOutputReference = CertificateManagerCertificateTimeoutsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CertificateManagerCertificateTimeoutsOutputReference[_g] = { fqn: "@cdktf/provider-google.certificateManagerCertificate.CertificateManagerCertificateTimeoutsOutputReference", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/certificate_manager_certificate google_certificate_manager_certificate}
*/
class CertificateManagerCertificate extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/certificate_manager_certificate google_certificate_manager_certificate} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CertificateManagerCertificateConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_certificate_manager_certificate',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // managed - computed: false, optional: true, required: false
        this._managed = new CertificateManagerCertificateManagedOutputReference(this, "managed");
        // self_managed - computed: false, optional: true, required: false
        this._selfManaged = new CertificateManagerCertificateSelfManagedOutputReference(this, "self_managed");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CertificateManagerCertificateTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._project = config.project;
        this._scope = config.scope;
        this._managed.internalValue = config.managed;
        this._selfManaged.internalValue = config.selfManaged;
        this._timeouts.internalValue = config.timeouts;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get scope() {
        return this.getStringAttribute('scope');
    }
    set scope(value) {
        this._scope = value;
    }
    resetScope() {
        this._scope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scopeInput() {
        return this._scope;
    }
    get managed() {
        return this._managed;
    }
    putManaged(value) {
        this._managed.internalValue = value;
    }
    resetManaged() {
        this._managed.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get managedInput() {
        return this._managed.internalValue;
    }
    get selfManaged() {
        return this._selfManaged;
    }
    putSelfManaged(value) {
        this._selfManaged.internalValue = value;
    }
    resetSelfManaged() {
        this._selfManaged.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfManagedInput() {
        return this._selfManaged.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            scope: cdktf.stringToTerraform(this._scope),
            managed: certificateManagerCertificateManagedToTerraform(this._managed.internalValue),
            self_managed: certificateManagerCertificateSelfManagedToTerraform(this._selfManaged.internalValue),
            timeouts: certificateManagerCertificateTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.CertificateManagerCertificate = CertificateManagerCertificate;
_h = JSII_RTTI_SYMBOL_1;
CertificateManagerCertificate[_h] = { fqn: "@cdktf/provider-google.certificateManagerCertificate.CertificateManagerCertificate", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
CertificateManagerCertificate.tfResourceType = "google_certificate_manager_certificate";
