// @cdktf/provider-google/lib/game-services-game-server-deployment-rollout/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameServicesGameServerDeploymentRollout = exports.GameServicesGameServerDeploymentRolloutTimeoutsOutputReference = exports.gameServicesGameServerDeploymentRolloutTimeoutsToTerraform = exports.GameServicesGameServerDeploymentRolloutGameServerConfigOverridesList = exports.GameServicesGameServerDeploymentRolloutGameServerConfigOverridesOutputReference = exports.gameServicesGameServerDeploymentRolloutGameServerConfigOverridesToTerraform = exports.GameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorOutputReference = exports.gameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function gameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        realms: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.realms),
    };
}
exports.gameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorToTerraform = gameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorToTerraform;
class GameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._realms !== undefined) {
            hasAnyValues = true;
            internalValueResult.realms = this._realms;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._realms = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._realms = value.realms;
        }
    }
    get realms() {
        return this.getListAttribute('realms');
    }
    set realms(value) {
        this._realms = value;
    }
    resetRealms() {
        this._realms = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get realmsInput() {
        return this._realms;
    }
}
exports.GameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorOutputReference = GameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorOutputReference[_a] = { fqn: "@cdktf/provider-google.gameServicesGameServerDeploymentRollout.GameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorOutputReference", version: "5.0.1" };
function gameServicesGameServerDeploymentRolloutGameServerConfigOverridesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        config_version: cdktf.stringToTerraform(struct.configVersion),
        realms_selector: gameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorToTerraform(struct.realmsSelector),
    };
}
exports.gameServicesGameServerDeploymentRolloutGameServerConfigOverridesToTerraform = gameServicesGameServerDeploymentRolloutGameServerConfigOverridesToTerraform;
class GameServicesGameServerDeploymentRolloutGameServerConfigOverridesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // realms_selector - computed: false, optional: true, required: false
        this._realmsSelector = new GameServicesGameServerDeploymentRolloutGameServerConfigOverridesRealmsSelectorOutputReference(this, "realms_selector");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._configVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.configVersion = this._configVersion;
        }
        if (this._realmsSelector?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.realmsSelector = this._realmsSelector?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._configVersion = undefined;
            this._realmsSelector.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._configVersion = value.configVersion;
            this._realmsSelector.internalValue = value.realmsSelector;
        }
    }
    get configVersion() {
        return this.getStringAttribute('config_version');
    }
    set configVersion(value) {
        this._configVersion = value;
    }
    resetConfigVersion() {
        this._configVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configVersionInput() {
        return this._configVersion;
    }
    get realmsSelector() {
        return this._realmsSelector;
    }
    putRealmsSelector(value) {
        this._realmsSelector.internalValue = value;
    }
    resetRealmsSelector() {
        this._realmsSelector.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get realmsSelectorInput() {
        return this._realmsSelector.internalValue;
    }
}
exports.GameServicesGameServerDeploymentRolloutGameServerConfigOverridesOutputReference = GameServicesGameServerDeploymentRolloutGameServerConfigOverridesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GameServicesGameServerDeploymentRolloutGameServerConfigOverridesOutputReference[_b] = { fqn: "@cdktf/provider-google.gameServicesGameServerDeploymentRollout.GameServicesGameServerDeploymentRolloutGameServerConfigOverridesOutputReference", version: "5.0.1" };
class GameServicesGameServerDeploymentRolloutGameServerConfigOverridesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GameServicesGameServerDeploymentRolloutGameServerConfigOverridesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GameServicesGameServerDeploymentRolloutGameServerConfigOverridesList = GameServicesGameServerDeploymentRolloutGameServerConfigOverridesList;
_c = JSII_RTTI_SYMBOL_1;
GameServicesGameServerDeploymentRolloutGameServerConfigOverridesList[_c] = { fqn: "@cdktf/provider-google.gameServicesGameServerDeploymentRollout.GameServicesGameServerDeploymentRolloutGameServerConfigOverridesList", version: "5.0.1" };
function gameServicesGameServerDeploymentRolloutTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.gameServicesGameServerDeploymentRolloutTimeoutsToTerraform = gameServicesGameServerDeploymentRolloutTimeoutsToTerraform;
class GameServicesGameServerDeploymentRolloutTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.GameServicesGameServerDeploymentRolloutTimeoutsOutputReference = GameServicesGameServerDeploymentRolloutTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
GameServicesGameServerDeploymentRolloutTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-google.gameServicesGameServerDeploymentRollout.GameServicesGameServerDeploymentRolloutTimeoutsOutputReference", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/game_services_game_server_deployment_rollout google_game_services_game_server_deployment_rollout}
*/
class GameServicesGameServerDeploymentRollout extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/game_services_game_server_deployment_rollout google_game_services_game_server_deployment_rollout} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GameServicesGameServerDeploymentRolloutConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_game_services_game_server_deployment_rollout',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // game_server_config_overrides - computed: false, optional: true, required: false
        this._gameServerConfigOverrides = new GameServicesGameServerDeploymentRolloutGameServerConfigOverridesList(this, "game_server_config_overrides", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GameServicesGameServerDeploymentRolloutTimeoutsOutputReference(this, "timeouts");
        this._defaultGameServerConfig = config.defaultGameServerConfig;
        this._deploymentId = config.deploymentId;
        this._id = config.id;
        this._project = config.project;
        this._gameServerConfigOverrides.internalValue = config.gameServerConfigOverrides;
        this._timeouts.internalValue = config.timeouts;
    }
    get defaultGameServerConfig() {
        return this.getStringAttribute('default_game_server_config');
    }
    set defaultGameServerConfig(value) {
        this._defaultGameServerConfig = value;
    }
    // Temporarily expose input value. Use with caution.
    get defaultGameServerConfigInput() {
        return this._defaultGameServerConfig;
    }
    get deploymentId() {
        return this.getStringAttribute('deployment_id');
    }
    set deploymentId(value) {
        this._deploymentId = value;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentIdInput() {
        return this._deploymentId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get gameServerConfigOverrides() {
        return this._gameServerConfigOverrides;
    }
    putGameServerConfigOverrides(value) {
        this._gameServerConfigOverrides.internalValue = value;
    }
    resetGameServerConfigOverrides() {
        this._gameServerConfigOverrides.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gameServerConfigOverridesInput() {
        return this._gameServerConfigOverrides.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            default_game_server_config: cdktf.stringToTerraform(this._defaultGameServerConfig),
            deployment_id: cdktf.stringToTerraform(this._deploymentId),
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
            game_server_config_overrides: cdktf.listMapper(gameServicesGameServerDeploymentRolloutGameServerConfigOverridesToTerraform, true)(this._gameServerConfigOverrides.internalValue),
            timeouts: gameServicesGameServerDeploymentRolloutTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.GameServicesGameServerDeploymentRollout = GameServicesGameServerDeploymentRollout;
_e = JSII_RTTI_SYMBOL_1;
GameServicesGameServerDeploymentRollout[_e] = { fqn: "@cdktf/provider-google.gameServicesGameServerDeploymentRollout.GameServicesGameServerDeploymentRollout", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
GameServicesGameServerDeploymentRollout.tfResourceType = "google_game_services_game_server_deployment_rollout";
