// @cdktf/provider-google/lib/data-google-compute-lb-ip-ranges/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeLbIpRanges = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/compute_lb_ip_ranges google_compute_lb_ip_ranges}
*/
class DataGoogleComputeLbIpRanges extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/compute_lb_ip_ranges google_compute_lb_ip_ranges} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeLbIpRangesConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_compute_lb_ip_ranges',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // http_ssl_tcp_internal - computed: true, optional: false, required: false
    get httpSslTcpInternal() {
        return this.getListAttribute('http_ssl_tcp_internal');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getListAttribute('network');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataGoogleComputeLbIpRanges = DataGoogleComputeLbIpRanges;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeLbIpRanges[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeLbIpRanges.DataGoogleComputeLbIpRanges", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeLbIpRanges.tfResourceType = "google_compute_lb_ip_ranges";
