// @cdktf/provider-google/lib/container-attached-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerAttachedCluster = exports.ContainerAttachedClusterTimeoutsOutputReference = exports.containerAttachedClusterTimeoutsToTerraform = exports.ContainerAttachedClusterOidcConfigOutputReference = exports.containerAttachedClusterOidcConfigToTerraform = exports.ContainerAttachedClusterMonitoringConfigOutputReference = exports.containerAttachedClusterMonitoringConfigToTerraform = exports.ContainerAttachedClusterMonitoringConfigManagedPrometheusConfigOutputReference = exports.containerAttachedClusterMonitoringConfigManagedPrometheusConfigToTerraform = exports.ContainerAttachedClusterLoggingConfigOutputReference = exports.containerAttachedClusterLoggingConfigToTerraform = exports.ContainerAttachedClusterLoggingConfigComponentConfigOutputReference = exports.containerAttachedClusterLoggingConfigComponentConfigToTerraform = exports.ContainerAttachedClusterFleetOutputReference = exports.containerAttachedClusterFleetToTerraform = exports.ContainerAttachedClusterAuthorizationOutputReference = exports.containerAttachedClusterAuthorizationToTerraform = exports.ContainerAttachedClusterWorkloadIdentityConfigList = exports.ContainerAttachedClusterWorkloadIdentityConfigOutputReference = exports.containerAttachedClusterWorkloadIdentityConfigToTerraform = exports.ContainerAttachedClusterErrorsList = exports.ContainerAttachedClusterErrorsOutputReference = exports.containerAttachedClusterErrorsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function containerAttachedClusterErrorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.containerAttachedClusterErrorsToTerraform = containerAttachedClusterErrorsToTerraform;
class ContainerAttachedClusterErrorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
}
exports.ContainerAttachedClusterErrorsOutputReference = ContainerAttachedClusterErrorsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterErrorsOutputReference[_a] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterErrorsOutputReference", version: "5.0.1" };
class ContainerAttachedClusterErrorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ContainerAttachedClusterErrorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ContainerAttachedClusterErrorsList = ContainerAttachedClusterErrorsList;
_b = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterErrorsList[_b] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterErrorsList", version: "5.0.1" };
function containerAttachedClusterWorkloadIdentityConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.containerAttachedClusterWorkloadIdentityConfigToTerraform = containerAttachedClusterWorkloadIdentityConfigToTerraform;
class ContainerAttachedClusterWorkloadIdentityConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // identity_provider - computed: true, optional: false, required: false
    get identityProvider() {
        return this.getStringAttribute('identity_provider');
    }
    // issuer_uri - computed: true, optional: false, required: false
    get issuerUri() {
        return this.getStringAttribute('issuer_uri');
    }
    // workload_pool - computed: true, optional: false, required: false
    get workloadPool() {
        return this.getStringAttribute('workload_pool');
    }
}
exports.ContainerAttachedClusterWorkloadIdentityConfigOutputReference = ContainerAttachedClusterWorkloadIdentityConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterWorkloadIdentityConfigOutputReference[_c] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterWorkloadIdentityConfigOutputReference", version: "5.0.1" };
class ContainerAttachedClusterWorkloadIdentityConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ContainerAttachedClusterWorkloadIdentityConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ContainerAttachedClusterWorkloadIdentityConfigList = ContainerAttachedClusterWorkloadIdentityConfigList;
_d = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterWorkloadIdentityConfigList[_d] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterWorkloadIdentityConfigList", version: "5.0.1" };
function containerAttachedClusterAuthorizationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        admin_users: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.adminUsers),
    };
}
exports.containerAttachedClusterAuthorizationToTerraform = containerAttachedClusterAuthorizationToTerraform;
class ContainerAttachedClusterAuthorizationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._adminUsers !== undefined) {
            hasAnyValues = true;
            internalValueResult.adminUsers = this._adminUsers;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._adminUsers = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._adminUsers = value.adminUsers;
        }
    }
    get adminUsers() {
        return this.getListAttribute('admin_users');
    }
    set adminUsers(value) {
        this._adminUsers = value;
    }
    resetAdminUsers() {
        this._adminUsers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminUsersInput() {
        return this._adminUsers;
    }
}
exports.ContainerAttachedClusterAuthorizationOutputReference = ContainerAttachedClusterAuthorizationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterAuthorizationOutputReference[_e] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterAuthorizationOutputReference", version: "5.0.1" };
function containerAttachedClusterFleetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        project: cdktf.stringToTerraform(struct.project),
    };
}
exports.containerAttachedClusterFleetToTerraform = containerAttachedClusterFleetToTerraform;
class ContainerAttachedClusterFleetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._project !== undefined) {
            hasAnyValues = true;
            internalValueResult.project = this._project;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._project = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._project = value.project;
        }
    }
    // membership - computed: true, optional: false, required: false
    get membership() {
        return this.getStringAttribute('membership');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
}
exports.ContainerAttachedClusterFleetOutputReference = ContainerAttachedClusterFleetOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterFleetOutputReference[_f] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterFleetOutputReference", version: "5.0.1" };
function containerAttachedClusterLoggingConfigComponentConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_components: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.enableComponents),
    };
}
exports.containerAttachedClusterLoggingConfigComponentConfigToTerraform = containerAttachedClusterLoggingConfigComponentConfigToTerraform;
class ContainerAttachedClusterLoggingConfigComponentConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableComponents !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableComponents = this._enableComponents;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableComponents = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableComponents = value.enableComponents;
        }
    }
    get enableComponents() {
        return this.getListAttribute('enable_components');
    }
    set enableComponents(value) {
        this._enableComponents = value;
    }
    resetEnableComponents() {
        this._enableComponents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableComponentsInput() {
        return this._enableComponents;
    }
}
exports.ContainerAttachedClusterLoggingConfigComponentConfigOutputReference = ContainerAttachedClusterLoggingConfigComponentConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterLoggingConfigComponentConfigOutputReference[_g] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterLoggingConfigComponentConfigOutputReference", version: "5.0.1" };
function containerAttachedClusterLoggingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        component_config: containerAttachedClusterLoggingConfigComponentConfigToTerraform(struct.componentConfig),
    };
}
exports.containerAttachedClusterLoggingConfigToTerraform = containerAttachedClusterLoggingConfigToTerraform;
class ContainerAttachedClusterLoggingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // component_config - computed: false, optional: true, required: false
        this._componentConfig = new ContainerAttachedClusterLoggingConfigComponentConfigOutputReference(this, "component_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._componentConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.componentConfig = this._componentConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._componentConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._componentConfig.internalValue = value.componentConfig;
        }
    }
    get componentConfig() {
        return this._componentConfig;
    }
    putComponentConfig(value) {
        this._componentConfig.internalValue = value;
    }
    resetComponentConfig() {
        this._componentConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get componentConfigInput() {
        return this._componentConfig.internalValue;
    }
}
exports.ContainerAttachedClusterLoggingConfigOutputReference = ContainerAttachedClusterLoggingConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterLoggingConfigOutputReference[_h] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterLoggingConfigOutputReference", version: "5.0.1" };
function containerAttachedClusterMonitoringConfigManagedPrometheusConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.containerAttachedClusterMonitoringConfigManagedPrometheusConfigToTerraform = containerAttachedClusterMonitoringConfigManagedPrometheusConfigToTerraform;
class ContainerAttachedClusterMonitoringConfigManagedPrometheusConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.ContainerAttachedClusterMonitoringConfigManagedPrometheusConfigOutputReference = ContainerAttachedClusterMonitoringConfigManagedPrometheusConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterMonitoringConfigManagedPrometheusConfigOutputReference[_j] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterMonitoringConfigManagedPrometheusConfigOutputReference", version: "5.0.1" };
function containerAttachedClusterMonitoringConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        managed_prometheus_config: containerAttachedClusterMonitoringConfigManagedPrometheusConfigToTerraform(struct.managedPrometheusConfig),
    };
}
exports.containerAttachedClusterMonitoringConfigToTerraform = containerAttachedClusterMonitoringConfigToTerraform;
class ContainerAttachedClusterMonitoringConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // managed_prometheus_config - computed: false, optional: true, required: false
        this._managedPrometheusConfig = new ContainerAttachedClusterMonitoringConfigManagedPrometheusConfigOutputReference(this, "managed_prometheus_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._managedPrometheusConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.managedPrometheusConfig = this._managedPrometheusConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._managedPrometheusConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._managedPrometheusConfig.internalValue = value.managedPrometheusConfig;
        }
    }
    get managedPrometheusConfig() {
        return this._managedPrometheusConfig;
    }
    putManagedPrometheusConfig(value) {
        this._managedPrometheusConfig.internalValue = value;
    }
    resetManagedPrometheusConfig() {
        this._managedPrometheusConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get managedPrometheusConfigInput() {
        return this._managedPrometheusConfig.internalValue;
    }
}
exports.ContainerAttachedClusterMonitoringConfigOutputReference = ContainerAttachedClusterMonitoringConfigOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterMonitoringConfigOutputReference[_k] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterMonitoringConfigOutputReference", version: "5.0.1" };
function containerAttachedClusterOidcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        issuer_url: cdktf.stringToTerraform(struct.issuerUrl),
        jwks: cdktf.stringToTerraform(struct.jwks),
    };
}
exports.containerAttachedClusterOidcConfigToTerraform = containerAttachedClusterOidcConfigToTerraform;
class ContainerAttachedClusterOidcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._issuerUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.issuerUrl = this._issuerUrl;
        }
        if (this._jwks !== undefined) {
            hasAnyValues = true;
            internalValueResult.jwks = this._jwks;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._issuerUrl = undefined;
            this._jwks = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._issuerUrl = value.issuerUrl;
            this._jwks = value.jwks;
        }
    }
    get issuerUrl() {
        return this.getStringAttribute('issuer_url');
    }
    set issuerUrl(value) {
        this._issuerUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    get issuerUrlInput() {
        return this._issuerUrl;
    }
    get jwks() {
        return this.getStringAttribute('jwks');
    }
    set jwks(value) {
        this._jwks = value;
    }
    resetJwks() {
        this._jwks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jwksInput() {
        return this._jwks;
    }
}
exports.ContainerAttachedClusterOidcConfigOutputReference = ContainerAttachedClusterOidcConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterOidcConfigOutputReference[_l] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterOidcConfigOutputReference", version: "5.0.1" };
function containerAttachedClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.containerAttachedClusterTimeoutsToTerraform = containerAttachedClusterTimeoutsToTerraform;
class ContainerAttachedClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ContainerAttachedClusterTimeoutsOutputReference = ContainerAttachedClusterTimeoutsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ContainerAttachedClusterTimeoutsOutputReference[_m] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedClusterTimeoutsOutputReference", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/container_attached_cluster google_container_attached_cluster}
*/
class ContainerAttachedCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/container_attached_cluster google_container_attached_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ContainerAttachedClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_container_attached_cluster',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // errors - computed: true, optional: false, required: false
        this._errors = new ContainerAttachedClusterErrorsList(this, "errors", false);
        // workload_identity_config - computed: true, optional: false, required: false
        this._workloadIdentityConfig = new ContainerAttachedClusterWorkloadIdentityConfigList(this, "workload_identity_config", false);
        // authorization - computed: false, optional: true, required: false
        this._authorization = new ContainerAttachedClusterAuthorizationOutputReference(this, "authorization");
        // fleet - computed: false, optional: false, required: true
        this._fleet = new ContainerAttachedClusterFleetOutputReference(this, "fleet");
        // logging_config - computed: false, optional: true, required: false
        this._loggingConfig = new ContainerAttachedClusterLoggingConfigOutputReference(this, "logging_config");
        // monitoring_config - computed: false, optional: true, required: false
        this._monitoringConfig = new ContainerAttachedClusterMonitoringConfigOutputReference(this, "monitoring_config");
        // oidc_config - computed: false, optional: false, required: true
        this._oidcConfig = new ContainerAttachedClusterOidcConfigOutputReference(this, "oidc_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ContainerAttachedClusterTimeoutsOutputReference(this, "timeouts");
        this._annotations = config.annotations;
        this._description = config.description;
        this._distribution = config.distribution;
        this._id = config.id;
        this._location = config.location;
        this._name = config.name;
        this._platformVersion = config.platformVersion;
        this._project = config.project;
        this._authorization.internalValue = config.authorization;
        this._fleet.internalValue = config.fleet;
        this._loggingConfig.internalValue = config.loggingConfig;
        this._monitoringConfig.internalValue = config.monitoringConfig;
        this._oidcConfig.internalValue = config.oidcConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    // cluster_region - computed: true, optional: false, required: false
    get clusterRegion() {
        return this.getStringAttribute('cluster_region');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get distribution() {
        return this.getStringAttribute('distribution');
    }
    set distribution(value) {
        this._distribution = value;
    }
    // Temporarily expose input value. Use with caution.
    get distributionInput() {
        return this._distribution;
    }
    get errors() {
        return this._errors;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kubernetes_version - computed: true, optional: false, required: false
    get kubernetesVersion() {
        return this.getStringAttribute('kubernetes_version');
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    set platformVersion(value) {
        this._platformVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get platformVersionInput() {
        return this._platformVersion;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // reconciling - computed: true, optional: false, required: false
    get reconciling() {
        return this.getBooleanAttribute('reconciling');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get workloadIdentityConfig() {
        return this._workloadIdentityConfig;
    }
    get authorization() {
        return this._authorization;
    }
    putAuthorization(value) {
        this._authorization.internalValue = value;
    }
    resetAuthorization() {
        this._authorization.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationInput() {
        return this._authorization.internalValue;
    }
    get fleet() {
        return this._fleet;
    }
    putFleet(value) {
        this._fleet.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get fleetInput() {
        return this._fleet.internalValue;
    }
    get loggingConfig() {
        return this._loggingConfig;
    }
    putLoggingConfig(value) {
        this._loggingConfig.internalValue = value;
    }
    resetLoggingConfig() {
        this._loggingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loggingConfigInput() {
        return this._loggingConfig.internalValue;
    }
    get monitoringConfig() {
        return this._monitoringConfig;
    }
    putMonitoringConfig(value) {
        this._monitoringConfig.internalValue = value;
    }
    resetMonitoringConfig() {
        this._monitoringConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get monitoringConfigInput() {
        return this._monitoringConfig.internalValue;
    }
    get oidcConfig() {
        return this._oidcConfig;
    }
    putOidcConfig(value) {
        this._oidcConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get oidcConfigInput() {
        return this._oidcConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            annotations: cdktf.hashMapper(cdktf.stringToTerraform)(this._annotations),
            description: cdktf.stringToTerraform(this._description),
            distribution: cdktf.stringToTerraform(this._distribution),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            platform_version: cdktf.stringToTerraform(this._platformVersion),
            project: cdktf.stringToTerraform(this._project),
            authorization: containerAttachedClusterAuthorizationToTerraform(this._authorization.internalValue),
            fleet: containerAttachedClusterFleetToTerraform(this._fleet.internalValue),
            logging_config: containerAttachedClusterLoggingConfigToTerraform(this._loggingConfig.internalValue),
            monitoring_config: containerAttachedClusterMonitoringConfigToTerraform(this._monitoringConfig.internalValue),
            oidc_config: containerAttachedClusterOidcConfigToTerraform(this._oidcConfig.internalValue),
            timeouts: containerAttachedClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.ContainerAttachedCluster = ContainerAttachedCluster;
_o = JSII_RTTI_SYMBOL_1;
ContainerAttachedCluster[_o] = { fqn: "@cdktf/provider-google.containerAttachedCluster.ContainerAttachedCluster", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
ContainerAttachedCluster.tfResourceType = "google_container_attached_cluster";
