// @cdktf/provider-google/lib/data-google-beyondcorp-app-connection/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleBeyondcorpAppConnection = exports.DataGoogleBeyondcorpAppConnectionGatewayList = exports.DataGoogleBeyondcorpAppConnectionGatewayOutputReference = exports.dataGoogleBeyondcorpAppConnectionGatewayToTerraform = exports.DataGoogleBeyondcorpAppConnectionApplicationEndpointList = exports.DataGoogleBeyondcorpAppConnectionApplicationEndpointOutputReference = exports.dataGoogleBeyondcorpAppConnectionApplicationEndpointToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleBeyondcorpAppConnectionApplicationEndpointToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleBeyondcorpAppConnectionApplicationEndpointToTerraform = dataGoogleBeyondcorpAppConnectionApplicationEndpointToTerraform;
class DataGoogleBeyondcorpAppConnectionApplicationEndpointOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // host - computed: true, optional: false, required: false
    get host() {
        return this.getStringAttribute('host');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataGoogleBeyondcorpAppConnectionApplicationEndpointOutputReference = DataGoogleBeyondcorpAppConnectionApplicationEndpointOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnectionApplicationEndpointOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnection.DataGoogleBeyondcorpAppConnectionApplicationEndpointOutputReference", version: "5.0.1" };
class DataGoogleBeyondcorpAppConnectionApplicationEndpointList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleBeyondcorpAppConnectionApplicationEndpointOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleBeyondcorpAppConnectionApplicationEndpointList = DataGoogleBeyondcorpAppConnectionApplicationEndpointList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnectionApplicationEndpointList[_b] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnection.DataGoogleBeyondcorpAppConnectionApplicationEndpointList", version: "5.0.1" };
function dataGoogleBeyondcorpAppConnectionGatewayToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleBeyondcorpAppConnectionGatewayToTerraform = dataGoogleBeyondcorpAppConnectionGatewayToTerraform;
class DataGoogleBeyondcorpAppConnectionGatewayOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // app_gateway - computed: true, optional: false, required: false
    get appGateway() {
        return this.getStringAttribute('app_gateway');
    }
    // ingress_port - computed: true, optional: false, required: false
    get ingressPort() {
        return this.getNumberAttribute('ingress_port');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
}
exports.DataGoogleBeyondcorpAppConnectionGatewayOutputReference = DataGoogleBeyondcorpAppConnectionGatewayOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnectionGatewayOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnection.DataGoogleBeyondcorpAppConnectionGatewayOutputReference", version: "5.0.1" };
class DataGoogleBeyondcorpAppConnectionGatewayList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleBeyondcorpAppConnectionGatewayOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleBeyondcorpAppConnectionGatewayList = DataGoogleBeyondcorpAppConnectionGatewayList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnectionGatewayList[_d] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnection.DataGoogleBeyondcorpAppConnectionGatewayList", version: "5.0.1" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/beyondcorp_app_connection google_beyondcorp_app_connection}
*/
class DataGoogleBeyondcorpAppConnection extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/beyondcorp_app_connection google_beyondcorp_app_connection} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleBeyondcorpAppConnectionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_beyondcorp_app_connection',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.49.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // application_endpoint - computed: true, optional: false, required: false
        this._applicationEndpoint = new DataGoogleBeyondcorpAppConnectionApplicationEndpointList(this, "application_endpoint", false);
        // gateway - computed: true, optional: false, required: false
        this._gateway = new DataGoogleBeyondcorpAppConnectionGatewayList(this, "gateway", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
    }
    get applicationEndpoint() {
        return this._applicationEndpoint;
    }
    // connectors - computed: true, optional: false, required: false
    get connectors() {
        return this.getListAttribute('connectors');
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    get gateway() {
        return this._gateway;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
        };
    }
}
exports.DataGoogleBeyondcorpAppConnection = DataGoogleBeyondcorpAppConnection;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleBeyondcorpAppConnection[_e] = { fqn: "@cdktf/provider-google.dataGoogleBeyondcorpAppConnection.DataGoogleBeyondcorpAppConnection", version: "5.0.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleBeyondcorpAppConnection.tfResourceType = "google_beyondcorp_app_connection";
