// @cdktf/provider-google/lib/data-loss-prevention-stored-info-type/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataLossPreventionStoredInfoType = exports.DataLossPreventionStoredInfoTypeTimeoutsOutputReference = exports.dataLossPreventionStoredInfoTypeTimeoutsToTerraform = exports.DataLossPreventionStoredInfoTypeRegexOutputReference = exports.dataLossPreventionStoredInfoTypeRegexToTerraform = exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputReference = exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryToTerraform = exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathOutputReference = exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathToTerraform = exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetOutputReference = exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetToTerraform = exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldOutputReference = exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldToTerraform = exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableOutputReference = exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableToTerraform = exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldOutputReference = exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldToTerraform = exports.DataLossPreventionStoredInfoTypeDictionaryOutputReference = exports.dataLossPreventionStoredInfoTypeDictionaryToTerraform = exports.DataLossPreventionStoredInfoTypeDictionaryWordListOutputReference = exports.dataLossPreventionStoredInfoTypeDictionaryWordListToTerraform = exports.DataLossPreventionStoredInfoTypeDictionaryCloudStoragePathOutputReference = exports.dataLossPreventionStoredInfoTypeDictionaryCloudStoragePathToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataLossPreventionStoredInfoTypeDictionaryCloudStoragePathToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.dataLossPreventionStoredInfoTypeDictionaryCloudStoragePathToTerraform = dataLossPreventionStoredInfoTypeDictionaryCloudStoragePathToTerraform;
class DataLossPreventionStoredInfoTypeDictionaryCloudStoragePathOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.DataLossPreventionStoredInfoTypeDictionaryCloudStoragePathOutputReference = DataLossPreventionStoredInfoTypeDictionaryCloudStoragePathOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeDictionaryCloudStoragePathOutputReference[_a] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeDictionaryCloudStoragePathOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeDictionaryWordListToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        words: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.words),
    };
}
exports.dataLossPreventionStoredInfoTypeDictionaryWordListToTerraform = dataLossPreventionStoredInfoTypeDictionaryWordListToTerraform;
class DataLossPreventionStoredInfoTypeDictionaryWordListOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._words !== undefined) {
            hasAnyValues = true;
            internalValueResult.words = this._words;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._words = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._words = value.words;
        }
    }
    get words() {
        return this.getListAttribute('words');
    }
    set words(value) {
        this._words = value;
    }
    // Temporarily expose input value. Use with caution.
    get wordsInput() {
        return this._words;
    }
}
exports.DataLossPreventionStoredInfoTypeDictionaryWordListOutputReference = DataLossPreventionStoredInfoTypeDictionaryWordListOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeDictionaryWordListOutputReference[_b] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeDictionaryWordListOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeDictionaryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloud_storage_path: dataLossPreventionStoredInfoTypeDictionaryCloudStoragePathToTerraform(struct.cloudStoragePath),
        word_list: dataLossPreventionStoredInfoTypeDictionaryWordListToTerraform(struct.wordList),
    };
}
exports.dataLossPreventionStoredInfoTypeDictionaryToTerraform = dataLossPreventionStoredInfoTypeDictionaryToTerraform;
class DataLossPreventionStoredInfoTypeDictionaryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cloud_storage_path - computed: false, optional: true, required: false
        this._cloudStoragePath = new DataLossPreventionStoredInfoTypeDictionaryCloudStoragePathOutputReference(this, "cloud_storage_path");
        // word_list - computed: false, optional: true, required: false
        this._wordList = new DataLossPreventionStoredInfoTypeDictionaryWordListOutputReference(this, "word_list");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cloudStoragePath?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudStoragePath = this._cloudStoragePath?.internalValue;
        }
        if (this._wordList?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.wordList = this._wordList?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cloudStoragePath.internalValue = undefined;
            this._wordList.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cloudStoragePath.internalValue = value.cloudStoragePath;
            this._wordList.internalValue = value.wordList;
        }
    }
    get cloudStoragePath() {
        return this._cloudStoragePath;
    }
    putCloudStoragePath(value) {
        this._cloudStoragePath.internalValue = value;
    }
    resetCloudStoragePath() {
        this._cloudStoragePath.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudStoragePathInput() {
        return this._cloudStoragePath.internalValue;
    }
    get wordList() {
        return this._wordList;
    }
    putWordList(value) {
        this._wordList.internalValue = value;
    }
    resetWordList() {
        this._wordList.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get wordListInput() {
        return this._wordList.internalValue;
    }
}
exports.DataLossPreventionStoredInfoTypeDictionaryOutputReference = DataLossPreventionStoredInfoTypeDictionaryOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeDictionaryOutputReference[_c] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeDictionaryOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldToTerraform = dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldToTerraform;
class DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldOutputReference = DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldOutputReference[_d] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dataset_id: cdktf.stringToTerraform(struct.datasetId),
        project_id: cdktf.stringToTerraform(struct.projectId),
        table_id: cdktf.stringToTerraform(struct.tableId),
    };
}
exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableToTerraform = dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableToTerraform;
class DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._datasetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.datasetId = this._datasetId;
        }
        if (this._projectId !== undefined) {
            hasAnyValues = true;
            internalValueResult.projectId = this._projectId;
        }
        if (this._tableId !== undefined) {
            hasAnyValues = true;
            internalValueResult.tableId = this._tableId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._datasetId = undefined;
            this._projectId = undefined;
            this._tableId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._datasetId = value.datasetId;
            this._projectId = value.projectId;
            this._tableId = value.tableId;
        }
    }
    get datasetId() {
        return this.getStringAttribute('dataset_id');
    }
    set datasetId(value) {
        this._datasetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get datasetIdInput() {
        return this._datasetId;
    }
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    set projectId(value) {
        this._projectId = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectIdInput() {
        return this._projectId;
    }
    get tableId() {
        return this.getStringAttribute('table_id');
    }
    set tableId(value) {
        this._tableId = value;
    }
    // Temporarily expose input value. Use with caution.
    get tableIdInput() {
        return this._tableId;
    }
}
exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableOutputReference = DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableOutputReference[_e] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field: dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldToTerraform(struct.field),
        table: dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableToTerraform(struct.table),
    };
}
exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldToTerraform = dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldToTerraform;
class DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // field - computed: false, optional: false, required: true
        this._field = new DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldFieldOutputReference(this, "field");
        // table - computed: false, optional: false, required: true
        this._table = new DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldTableOutputReference(this, "table");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._field?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field?.internalValue;
        }
        if (this._table?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.table = this._table?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._field.internalValue = undefined;
            this._table.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._field.internalValue = value.field;
            this._table.internalValue = value.table;
        }
    }
    get field() {
        return this._field;
    }
    putField(value) {
        this._field.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field.internalValue;
    }
    get table() {
        return this._table;
    }
    putTable(value) {
        this._table.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get tableInput() {
        return this._table.internalValue;
    }
}
exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldOutputReference = DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldOutputReference[_f] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        url: cdktf.stringToTerraform(struct.url),
    };
}
exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetToTerraform = dataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetToTerraform;
class DataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._url !== undefined) {
            hasAnyValues = true;
            internalValueResult.url = this._url;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._url = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._url = value.url;
        }
    }
    get url() {
        return this.getStringAttribute('url');
    }
    set url(value) {
        this._url = value;
    }
    // Temporarily expose input value. Use with caution.
    get urlInput() {
        return this._url;
    }
}
exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetOutputReference = DataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetOutputReference[_g] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathToTerraform = dataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathToTerraform;
class DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathOutputReference = DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathOutputReference;
_h = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathOutputReference[_h] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeLargeCustomDictionaryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        big_query_field: dataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldToTerraform(struct.bigQueryField),
        cloud_storage_file_set: dataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetToTerraform(struct.cloudStorageFileSet),
        output_path: dataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathToTerraform(struct.outputPath),
    };
}
exports.dataLossPreventionStoredInfoTypeLargeCustomDictionaryToTerraform = dataLossPreventionStoredInfoTypeLargeCustomDictionaryToTerraform;
class DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // big_query_field - computed: false, optional: true, required: false
        this._bigQueryField = new DataLossPreventionStoredInfoTypeLargeCustomDictionaryBigQueryFieldOutputReference(this, "big_query_field");
        // cloud_storage_file_set - computed: false, optional: true, required: false
        this._cloudStorageFileSet = new DataLossPreventionStoredInfoTypeLargeCustomDictionaryCloudStorageFileSetOutputReference(this, "cloud_storage_file_set");
        // output_path - computed: false, optional: false, required: true
        this._outputPath = new DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputPathOutputReference(this, "output_path");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bigQueryField?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.bigQueryField = this._bigQueryField?.internalValue;
        }
        if (this._cloudStorageFileSet?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudStorageFileSet = this._cloudStorageFileSet?.internalValue;
        }
        if (this._outputPath?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.outputPath = this._outputPath?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bigQueryField.internalValue = undefined;
            this._cloudStorageFileSet.internalValue = undefined;
            this._outputPath.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bigQueryField.internalValue = value.bigQueryField;
            this._cloudStorageFileSet.internalValue = value.cloudStorageFileSet;
            this._outputPath.internalValue = value.outputPath;
        }
    }
    get bigQueryField() {
        return this._bigQueryField;
    }
    putBigQueryField(value) {
        this._bigQueryField.internalValue = value;
    }
    resetBigQueryField() {
        this._bigQueryField.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigQueryFieldInput() {
        return this._bigQueryField.internalValue;
    }
    get cloudStorageFileSet() {
        return this._cloudStorageFileSet;
    }
    putCloudStorageFileSet(value) {
        this._cloudStorageFileSet.internalValue = value;
    }
    resetCloudStorageFileSet() {
        this._cloudStorageFileSet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudStorageFileSetInput() {
        return this._cloudStorageFileSet.internalValue;
    }
    get outputPath() {
        return this._outputPath;
    }
    putOutputPath(value) {
        this._outputPath.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get outputPathInput() {
        return this._outputPath.internalValue;
    }
}
exports.DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputReference = DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputReference[_j] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeRegexToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        group_indexes: cdktf.listMapper(cdktf.numberToTerraform, false)(struct.groupIndexes),
        pattern: cdktf.stringToTerraform(struct.pattern),
    };
}
exports.dataLossPreventionStoredInfoTypeRegexToTerraform = dataLossPreventionStoredInfoTypeRegexToTerraform;
class DataLossPreventionStoredInfoTypeRegexOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._groupIndexes !== undefined) {
            hasAnyValues = true;
            internalValueResult.groupIndexes = this._groupIndexes;
        }
        if (this._pattern !== undefined) {
            hasAnyValues = true;
            internalValueResult.pattern = this._pattern;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._groupIndexes = undefined;
            this._pattern = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._groupIndexes = value.groupIndexes;
            this._pattern = value.pattern;
        }
    }
    get groupIndexes() {
        return this.getNumberListAttribute('group_indexes');
    }
    set groupIndexes(value) {
        this._groupIndexes = value;
    }
    resetGroupIndexes() {
        this._groupIndexes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupIndexesInput() {
        return this._groupIndexes;
    }
    get pattern() {
        return this.getStringAttribute('pattern');
    }
    set pattern(value) {
        this._pattern = value;
    }
    // Temporarily expose input value. Use with caution.
    get patternInput() {
        return this._pattern;
    }
}
exports.DataLossPreventionStoredInfoTypeRegexOutputReference = DataLossPreventionStoredInfoTypeRegexOutputReference;
_k = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeRegexOutputReference[_k] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeRegexOutputReference", version: "4.0.3" };
function dataLossPreventionStoredInfoTypeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dataLossPreventionStoredInfoTypeTimeoutsToTerraform = dataLossPreventionStoredInfoTypeTimeoutsToTerraform;
class DataLossPreventionStoredInfoTypeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DataLossPreventionStoredInfoTypeTimeoutsOutputReference = DataLossPreventionStoredInfoTypeTimeoutsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoTypeTimeoutsOutputReference[_l] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoTypeTimeoutsOutputReference", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/data_loss_prevention_stored_info_type google_data_loss_prevention_stored_info_type}
*/
class DataLossPreventionStoredInfoType extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/data_loss_prevention_stored_info_type google_data_loss_prevention_stored_info_type} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataLossPreventionStoredInfoTypeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_data_loss_prevention_stored_info_type',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dictionary - computed: false, optional: true, required: false
        this._dictionary = new DataLossPreventionStoredInfoTypeDictionaryOutputReference(this, "dictionary");
        // large_custom_dictionary - computed: false, optional: true, required: false
        this._largeCustomDictionary = new DataLossPreventionStoredInfoTypeLargeCustomDictionaryOutputReference(this, "large_custom_dictionary");
        // regex - computed: false, optional: true, required: false
        this._regex = new DataLossPreventionStoredInfoTypeRegexOutputReference(this, "regex");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataLossPreventionStoredInfoTypeTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._displayName = config.displayName;
        this._id = config.id;
        this._parent = config.parent;
        this._dictionary.internalValue = config.dictionary;
        this._largeCustomDictionary.internalValue = config.largeCustomDictionary;
        this._regex.internalValue = config.regex;
        this._timeouts.internalValue = config.timeouts;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get parent() {
        return this.getStringAttribute('parent');
    }
    set parent(value) {
        this._parent = value;
    }
    // Temporarily expose input value. Use with caution.
    get parentInput() {
        return this._parent;
    }
    get dictionary() {
        return this._dictionary;
    }
    putDictionary(value) {
        this._dictionary.internalValue = value;
    }
    resetDictionary() {
        this._dictionary.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dictionaryInput() {
        return this._dictionary.internalValue;
    }
    get largeCustomDictionary() {
        return this._largeCustomDictionary;
    }
    putLargeCustomDictionary(value) {
        this._largeCustomDictionary.internalValue = value;
    }
    resetLargeCustomDictionary() {
        this._largeCustomDictionary.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get largeCustomDictionaryInput() {
        return this._largeCustomDictionary.internalValue;
    }
    get regex() {
        return this._regex;
    }
    putRegex(value) {
        this._regex.internalValue = value;
    }
    resetRegex() {
        this._regex.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regexInput() {
        return this._regex.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            parent: cdktf.stringToTerraform(this._parent),
            dictionary: dataLossPreventionStoredInfoTypeDictionaryToTerraform(this._dictionary.internalValue),
            large_custom_dictionary: dataLossPreventionStoredInfoTypeLargeCustomDictionaryToTerraform(this._largeCustomDictionary.internalValue),
            regex: dataLossPreventionStoredInfoTypeRegexToTerraform(this._regex.internalValue),
            timeouts: dataLossPreventionStoredInfoTypeTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DataLossPreventionStoredInfoType = DataLossPreventionStoredInfoType;
_m = JSII_RTTI_SYMBOL_1;
DataLossPreventionStoredInfoType[_m] = { fqn: "@cdktf/provider-google.dataLossPreventionStoredInfoType.DataLossPreventionStoredInfoType", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
DataLossPreventionStoredInfoType.tfResourceType = "google_data_loss_prevention_stored_info_type";
