// @cdktf/provider-google/lib/compute-region-instance-group-manager/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComputeRegionInstanceGroupManager = exports.ComputeRegionInstanceGroupManagerVersionList = exports.ComputeRegionInstanceGroupManagerVersionOutputReference = exports.computeRegionInstanceGroupManagerVersionToTerraform = exports.ComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference = exports.computeRegionInstanceGroupManagerVersionTargetSizeToTerraform = exports.ComputeRegionInstanceGroupManagerUpdatePolicyOutputReference = exports.computeRegionInstanceGroupManagerUpdatePolicyToTerraform = exports.ComputeRegionInstanceGroupManagerTimeoutsOutputReference = exports.computeRegionInstanceGroupManagerTimeoutsToTerraform = exports.ComputeRegionInstanceGroupManagerStatefulDiskList = exports.ComputeRegionInstanceGroupManagerStatefulDiskOutputReference = exports.computeRegionInstanceGroupManagerStatefulDiskToTerraform = exports.ComputeRegionInstanceGroupManagerNamedPortList = exports.ComputeRegionInstanceGroupManagerNamedPortOutputReference = exports.computeRegionInstanceGroupManagerNamedPortToTerraform = exports.ComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference = exports.computeRegionInstanceGroupManagerAutoHealingPoliciesToTerraform = exports.ComputeRegionInstanceGroupManagerStatusList = exports.ComputeRegionInstanceGroupManagerStatusOutputReference = exports.computeRegionInstanceGroupManagerStatusToTerraform = exports.ComputeRegionInstanceGroupManagerStatusVersionTargetList = exports.ComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference = exports.computeRegionInstanceGroupManagerStatusVersionTargetToTerraform = exports.ComputeRegionInstanceGroupManagerStatusStatefulList = exports.ComputeRegionInstanceGroupManagerStatusStatefulOutputReference = exports.computeRegionInstanceGroupManagerStatusStatefulToTerraform = exports.ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList = exports.ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = exports.computeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function computeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = computeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform;
class ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all_effective - computed: true, optional: false, required: false
    get allEffective() {
        return this.getBooleanAttribute('all_effective');
    }
}
exports.ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference[_a] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference", version: "4.0.3" };
class ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList = ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList;
_b = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList[_b] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList", version: "4.0.3" };
function computeRegionInstanceGroupManagerStatusStatefulToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeRegionInstanceGroupManagerStatusStatefulToTerraform = computeRegionInstanceGroupManagerStatusStatefulToTerraform;
class ComputeRegionInstanceGroupManagerStatusStatefulOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // per_instance_configs - computed: true, optional: false, required: false
        this._perInstanceConfigs = new ComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList(this, "per_instance_configs", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // has_stateful_config - computed: true, optional: false, required: false
    get hasStatefulConfig() {
        return this.getBooleanAttribute('has_stateful_config');
    }
    get perInstanceConfigs() {
        return this._perInstanceConfigs;
    }
}
exports.ComputeRegionInstanceGroupManagerStatusStatefulOutputReference = ComputeRegionInstanceGroupManagerStatusStatefulOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatusStatefulOutputReference[_c] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatusStatefulOutputReference", version: "4.0.3" };
class ComputeRegionInstanceGroupManagerStatusStatefulList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeRegionInstanceGroupManagerStatusStatefulOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeRegionInstanceGroupManagerStatusStatefulList = ComputeRegionInstanceGroupManagerStatusStatefulList;
_d = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatusStatefulList[_d] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatusStatefulList", version: "4.0.3" };
function computeRegionInstanceGroupManagerStatusVersionTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeRegionInstanceGroupManagerStatusVersionTargetToTerraform = computeRegionInstanceGroupManagerStatusVersionTargetToTerraform;
class ComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_reached - computed: true, optional: false, required: false
    get isReached() {
        return this.getBooleanAttribute('is_reached');
    }
}
exports.ComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference = ComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference[_e] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference", version: "4.0.3" };
class ComputeRegionInstanceGroupManagerStatusVersionTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeRegionInstanceGroupManagerStatusVersionTargetList = ComputeRegionInstanceGroupManagerStatusVersionTargetList;
_f = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatusVersionTargetList[_f] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatusVersionTargetList", version: "4.0.3" };
function computeRegionInstanceGroupManagerStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeRegionInstanceGroupManagerStatusToTerraform = computeRegionInstanceGroupManagerStatusToTerraform;
class ComputeRegionInstanceGroupManagerStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // stateful - computed: true, optional: false, required: false
        this._stateful = new ComputeRegionInstanceGroupManagerStatusStatefulList(this, "stateful", false);
        // version_target - computed: true, optional: false, required: false
        this._versionTarget = new ComputeRegionInstanceGroupManagerStatusVersionTargetList(this, "version_target", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_stable - computed: true, optional: false, required: false
    get isStable() {
        return this.getBooleanAttribute('is_stable');
    }
    get stateful() {
        return this._stateful;
    }
    get versionTarget() {
        return this._versionTarget;
    }
}
exports.ComputeRegionInstanceGroupManagerStatusOutputReference = ComputeRegionInstanceGroupManagerStatusOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatusOutputReference[_g] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatusOutputReference", version: "4.0.3" };
class ComputeRegionInstanceGroupManagerStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeRegionInstanceGroupManagerStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeRegionInstanceGroupManagerStatusList = ComputeRegionInstanceGroupManagerStatusList;
_h = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatusList[_h] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatusList", version: "4.0.3" };
function computeRegionInstanceGroupManagerAutoHealingPoliciesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        health_check: cdktf.stringToTerraform(struct.healthCheck),
        initial_delay_sec: cdktf.numberToTerraform(struct.initialDelaySec),
    };
}
exports.computeRegionInstanceGroupManagerAutoHealingPoliciesToTerraform = computeRegionInstanceGroupManagerAutoHealingPoliciesToTerraform;
class ComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._healthCheck !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck;
        }
        if (this._initialDelaySec !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialDelaySec = this._initialDelaySec;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._healthCheck = undefined;
            this._initialDelaySec = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._healthCheck = value.healthCheck;
            this._initialDelaySec = value.initialDelaySec;
        }
    }
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    set healthCheck(value) {
        this._healthCheck = value;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck;
    }
    get initialDelaySec() {
        return this.getNumberAttribute('initial_delay_sec');
    }
    set initialDelaySec(value) {
        this._initialDelaySec = value;
    }
    // Temporarily expose input value. Use with caution.
    get initialDelaySecInput() {
        return this._initialDelaySec;
    }
}
exports.ComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference = ComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference[_j] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference", version: "4.0.3" };
function computeRegionInstanceGroupManagerNamedPortToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.computeRegionInstanceGroupManagerNamedPortToTerraform = computeRegionInstanceGroupManagerNamedPortToTerraform;
class ComputeRegionInstanceGroupManagerNamedPortOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._port = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._port = value.port;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.ComputeRegionInstanceGroupManagerNamedPortOutputReference = ComputeRegionInstanceGroupManagerNamedPortOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerNamedPortOutputReference[_k] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerNamedPortOutputReference", version: "4.0.3" };
class ComputeRegionInstanceGroupManagerNamedPortList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeRegionInstanceGroupManagerNamedPortOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeRegionInstanceGroupManagerNamedPortList = ComputeRegionInstanceGroupManagerNamedPortList;
_l = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerNamedPortList[_l] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerNamedPortList", version: "4.0.3" };
function computeRegionInstanceGroupManagerStatefulDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_rule: cdktf.stringToTerraform(struct.deleteRule),
        device_name: cdktf.stringToTerraform(struct.deviceName),
    };
}
exports.computeRegionInstanceGroupManagerStatefulDiskToTerraform = computeRegionInstanceGroupManagerStatefulDiskToTerraform;
class ComputeRegionInstanceGroupManagerStatefulDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteRule !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteRule = this._deleteRule;
        }
        if (this._deviceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.deviceName = this._deviceName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._deleteRule = undefined;
            this._deviceName = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._deleteRule = value.deleteRule;
            this._deviceName = value.deviceName;
        }
    }
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    set deleteRule(value) {
        this._deleteRule = value;
    }
    resetDeleteRule() {
        this._deleteRule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteRuleInput() {
        return this._deleteRule;
    }
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    set deviceName(value) {
        this._deviceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get deviceNameInput() {
        return this._deviceName;
    }
}
exports.ComputeRegionInstanceGroupManagerStatefulDiskOutputReference = ComputeRegionInstanceGroupManagerStatefulDiskOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatefulDiskOutputReference[_m] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatefulDiskOutputReference", version: "4.0.3" };
class ComputeRegionInstanceGroupManagerStatefulDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeRegionInstanceGroupManagerStatefulDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeRegionInstanceGroupManagerStatefulDiskList = ComputeRegionInstanceGroupManagerStatefulDiskList;
_o = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerStatefulDiskList[_o] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerStatefulDiskList", version: "4.0.3" };
function computeRegionInstanceGroupManagerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.computeRegionInstanceGroupManagerTimeoutsToTerraform = computeRegionInstanceGroupManagerTimeoutsToTerraform;
class ComputeRegionInstanceGroupManagerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ComputeRegionInstanceGroupManagerTimeoutsOutputReference = ComputeRegionInstanceGroupManagerTimeoutsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerTimeoutsOutputReference[_p] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerTimeoutsOutputReference", version: "4.0.3" };
function computeRegionInstanceGroupManagerUpdatePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instance_redistribution_type: cdktf.stringToTerraform(struct.instanceRedistributionType),
        max_surge_fixed: cdktf.numberToTerraform(struct.maxSurgeFixed),
        max_surge_percent: cdktf.numberToTerraform(struct.maxSurgePercent),
        max_unavailable_fixed: cdktf.numberToTerraform(struct.maxUnavailableFixed),
        max_unavailable_percent: cdktf.numberToTerraform(struct.maxUnavailablePercent),
        minimal_action: cdktf.stringToTerraform(struct.minimalAction),
        most_disruptive_allowed_action: cdktf.stringToTerraform(struct.mostDisruptiveAllowedAction),
        replacement_method: cdktf.stringToTerraform(struct.replacementMethod),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.computeRegionInstanceGroupManagerUpdatePolicyToTerraform = computeRegionInstanceGroupManagerUpdatePolicyToTerraform;
class ComputeRegionInstanceGroupManagerUpdatePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instanceRedistributionType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceRedistributionType = this._instanceRedistributionType;
        }
        if (this._maxSurgeFixed !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxSurgeFixed = this._maxSurgeFixed;
        }
        if (this._maxSurgePercent !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxSurgePercent = this._maxSurgePercent;
        }
        if (this._maxUnavailableFixed !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxUnavailableFixed = this._maxUnavailableFixed;
        }
        if (this._maxUnavailablePercent !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxUnavailablePercent = this._maxUnavailablePercent;
        }
        if (this._minimalAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.minimalAction = this._minimalAction;
        }
        if (this._mostDisruptiveAllowedAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.mostDisruptiveAllowedAction = this._mostDisruptiveAllowedAction;
        }
        if (this._replacementMethod !== undefined) {
            hasAnyValues = true;
            internalValueResult.replacementMethod = this._replacementMethod;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._instanceRedistributionType = undefined;
            this._maxSurgeFixed = undefined;
            this._maxSurgePercent = undefined;
            this._maxUnavailableFixed = undefined;
            this._maxUnavailablePercent = undefined;
            this._minimalAction = undefined;
            this._mostDisruptiveAllowedAction = undefined;
            this._replacementMethod = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._instanceRedistributionType = value.instanceRedistributionType;
            this._maxSurgeFixed = value.maxSurgeFixed;
            this._maxSurgePercent = value.maxSurgePercent;
            this._maxUnavailableFixed = value.maxUnavailableFixed;
            this._maxUnavailablePercent = value.maxUnavailablePercent;
            this._minimalAction = value.minimalAction;
            this._mostDisruptiveAllowedAction = value.mostDisruptiveAllowedAction;
            this._replacementMethod = value.replacementMethod;
            this._type = value.type;
        }
    }
    get instanceRedistributionType() {
        return this.getStringAttribute('instance_redistribution_type');
    }
    set instanceRedistributionType(value) {
        this._instanceRedistributionType = value;
    }
    resetInstanceRedistributionType() {
        this._instanceRedistributionType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceRedistributionTypeInput() {
        return this._instanceRedistributionType;
    }
    get maxSurgeFixed() {
        return this.getNumberAttribute('max_surge_fixed');
    }
    set maxSurgeFixed(value) {
        this._maxSurgeFixed = value;
    }
    resetMaxSurgeFixed() {
        this._maxSurgeFixed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxSurgeFixedInput() {
        return this._maxSurgeFixed;
    }
    get maxSurgePercent() {
        return this.getNumberAttribute('max_surge_percent');
    }
    set maxSurgePercent(value) {
        this._maxSurgePercent = value;
    }
    resetMaxSurgePercent() {
        this._maxSurgePercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxSurgePercentInput() {
        return this._maxSurgePercent;
    }
    get maxUnavailableFixed() {
        return this.getNumberAttribute('max_unavailable_fixed');
    }
    set maxUnavailableFixed(value) {
        this._maxUnavailableFixed = value;
    }
    resetMaxUnavailableFixed() {
        this._maxUnavailableFixed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxUnavailableFixedInput() {
        return this._maxUnavailableFixed;
    }
    get maxUnavailablePercent() {
        return this.getNumberAttribute('max_unavailable_percent');
    }
    set maxUnavailablePercent(value) {
        this._maxUnavailablePercent = value;
    }
    resetMaxUnavailablePercent() {
        this._maxUnavailablePercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxUnavailablePercentInput() {
        return this._maxUnavailablePercent;
    }
    get minimalAction() {
        return this.getStringAttribute('minimal_action');
    }
    set minimalAction(value) {
        this._minimalAction = value;
    }
    // Temporarily expose input value. Use with caution.
    get minimalActionInput() {
        return this._minimalAction;
    }
    get mostDisruptiveAllowedAction() {
        return this.getStringAttribute('most_disruptive_allowed_action');
    }
    set mostDisruptiveAllowedAction(value) {
        this._mostDisruptiveAllowedAction = value;
    }
    resetMostDisruptiveAllowedAction() {
        this._mostDisruptiveAllowedAction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mostDisruptiveAllowedActionInput() {
        return this._mostDisruptiveAllowedAction;
    }
    get replacementMethod() {
        return this.getStringAttribute('replacement_method');
    }
    set replacementMethod(value) {
        this._replacementMethod = value;
    }
    resetReplacementMethod() {
        this._replacementMethod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replacementMethodInput() {
        return this._replacementMethod;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ComputeRegionInstanceGroupManagerUpdatePolicyOutputReference = ComputeRegionInstanceGroupManagerUpdatePolicyOutputReference;
_q = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerUpdatePolicyOutputReference[_q] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerUpdatePolicyOutputReference", version: "4.0.3" };
function computeRegionInstanceGroupManagerVersionTargetSizeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        fixed: cdktf.numberToTerraform(struct.fixed),
        percent: cdktf.numberToTerraform(struct.percent),
    };
}
exports.computeRegionInstanceGroupManagerVersionTargetSizeToTerraform = computeRegionInstanceGroupManagerVersionTargetSizeToTerraform;
class ComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fixed !== undefined) {
            hasAnyValues = true;
            internalValueResult.fixed = this._fixed;
        }
        if (this._percent !== undefined) {
            hasAnyValues = true;
            internalValueResult.percent = this._percent;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fixed = undefined;
            this._percent = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fixed = value.fixed;
            this._percent = value.percent;
        }
    }
    get fixed() {
        return this.getNumberAttribute('fixed');
    }
    set fixed(value) {
        this._fixed = value;
    }
    resetFixed() {
        this._fixed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fixedInput() {
        return this._fixed;
    }
    get percent() {
        return this.getNumberAttribute('percent');
    }
    set percent(value) {
        this._percent = value;
    }
    resetPercent() {
        this._percent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get percentInput() {
        return this._percent;
    }
}
exports.ComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference = ComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference;
_r = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference[_r] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference", version: "4.0.3" };
function computeRegionInstanceGroupManagerVersionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instance_template: cdktf.stringToTerraform(struct.instanceTemplate),
        name: cdktf.stringToTerraform(struct.name),
        target_size: computeRegionInstanceGroupManagerVersionTargetSizeToTerraform(struct.targetSize),
    };
}
exports.computeRegionInstanceGroupManagerVersionToTerraform = computeRegionInstanceGroupManagerVersionToTerraform;
class ComputeRegionInstanceGroupManagerVersionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // target_size - computed: false, optional: true, required: false
        this._targetSize = new ComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference(this, "target_size");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instanceTemplate !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceTemplate = this._instanceTemplate;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._targetSize?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetSize = this._targetSize?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._instanceTemplate = undefined;
            this._name = undefined;
            this._targetSize.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._instanceTemplate = value.instanceTemplate;
            this._name = value.name;
            this._targetSize.internalValue = value.targetSize;
        }
    }
    get instanceTemplate() {
        return this.getStringAttribute('instance_template');
    }
    set instanceTemplate(value) {
        this._instanceTemplate = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTemplateInput() {
        return this._instanceTemplate;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get targetSize() {
        return this._targetSize;
    }
    putTargetSize(value) {
        this._targetSize.internalValue = value;
    }
    resetTargetSize() {
        this._targetSize.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetSizeInput() {
        return this._targetSize.internalValue;
    }
}
exports.ComputeRegionInstanceGroupManagerVersionOutputReference = ComputeRegionInstanceGroupManagerVersionOutputReference;
_s = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerVersionOutputReference[_s] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerVersionOutputReference", version: "4.0.3" };
class ComputeRegionInstanceGroupManagerVersionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeRegionInstanceGroupManagerVersionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeRegionInstanceGroupManagerVersionList = ComputeRegionInstanceGroupManagerVersionList;
_t = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManagerVersionList[_t] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManagerVersionList", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/r/compute_region_instance_group_manager google_compute_region_instance_group_manager}
*/
class ComputeRegionInstanceGroupManager extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/r/compute_region_instance_group_manager google_compute_region_instance_group_manager} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComputeRegionInstanceGroupManagerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_region_instance_group_manager',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // status - computed: true, optional: false, required: false
        this._status = new ComputeRegionInstanceGroupManagerStatusList(this, "status", false);
        // auto_healing_policies - computed: false, optional: true, required: false
        this._autoHealingPolicies = new ComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference(this, "auto_healing_policies");
        // named_port - computed: false, optional: true, required: false
        this._namedPort = new ComputeRegionInstanceGroupManagerNamedPortList(this, "named_port", true);
        // stateful_disk - computed: false, optional: true, required: false
        this._statefulDisk = new ComputeRegionInstanceGroupManagerStatefulDiskList(this, "stateful_disk", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComputeRegionInstanceGroupManagerTimeoutsOutputReference(this, "timeouts");
        // update_policy - computed: false, optional: true, required: false
        this._updatePolicy = new ComputeRegionInstanceGroupManagerUpdatePolicyOutputReference(this, "update_policy");
        // version - computed: false, optional: false, required: true
        this._version = new ComputeRegionInstanceGroupManagerVersionList(this, "version", false);
        this._baseInstanceName = config.baseInstanceName;
        this._description = config.description;
        this._distributionPolicyTargetShape = config.distributionPolicyTargetShape;
        this._distributionPolicyZones = config.distributionPolicyZones;
        this._id = config.id;
        this._listManagedInstancesResults = config.listManagedInstancesResults;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
        this._targetPools = config.targetPools;
        this._targetSize = config.targetSize;
        this._waitForInstances = config.waitForInstances;
        this._waitForInstancesStatus = config.waitForInstancesStatus;
        this._autoHealingPolicies.internalValue = config.autoHealingPolicies;
        this._namedPort.internalValue = config.namedPort;
        this._statefulDisk.internalValue = config.statefulDisk;
        this._timeouts.internalValue = config.timeouts;
        this._updatePolicy.internalValue = config.updatePolicy;
        this._version.internalValue = config.version;
    }
    get baseInstanceName() {
        return this.getStringAttribute('base_instance_name');
    }
    set baseInstanceName(value) {
        this._baseInstanceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get baseInstanceNameInput() {
        return this._baseInstanceName;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get distributionPolicyTargetShape() {
        return this.getStringAttribute('distribution_policy_target_shape');
    }
    set distributionPolicyTargetShape(value) {
        this._distributionPolicyTargetShape = value;
    }
    resetDistributionPolicyTargetShape() {
        this._distributionPolicyTargetShape = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get distributionPolicyTargetShapeInput() {
        return this._distributionPolicyTargetShape;
    }
    get distributionPolicyZones() {
        return cdktf.Fn.tolist(this.getListAttribute('distribution_policy_zones'));
    }
    set distributionPolicyZones(value) {
        this._distributionPolicyZones = value;
    }
    resetDistributionPolicyZones() {
        this._distributionPolicyZones = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get distributionPolicyZonesInput() {
        return this._distributionPolicyZones;
    }
    // fingerprint - computed: true, optional: false, required: false
    get fingerprint() {
        return this.getStringAttribute('fingerprint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_group - computed: true, optional: false, required: false
    get instanceGroup() {
        return this.getStringAttribute('instance_group');
    }
    get listManagedInstancesResults() {
        return this.getStringAttribute('list_managed_instances_results');
    }
    set listManagedInstancesResults(value) {
        this._listManagedInstancesResults = value;
    }
    resetListManagedInstancesResults() {
        this._listManagedInstancesResults = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get listManagedInstancesResultsInput() {
        return this._listManagedInstancesResults;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get status() {
        return this._status;
    }
    get targetPools() {
        return cdktf.Fn.tolist(this.getListAttribute('target_pools'));
    }
    set targetPools(value) {
        this._targetPools = value;
    }
    resetTargetPools() {
        this._targetPools = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetPoolsInput() {
        return this._targetPools;
    }
    get targetSize() {
        return this.getNumberAttribute('target_size');
    }
    set targetSize(value) {
        this._targetSize = value;
    }
    resetTargetSize() {
        this._targetSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetSizeInput() {
        return this._targetSize;
    }
    get waitForInstances() {
        return this.getBooleanAttribute('wait_for_instances');
    }
    set waitForInstances(value) {
        this._waitForInstances = value;
    }
    resetWaitForInstances() {
        this._waitForInstances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForInstancesInput() {
        return this._waitForInstances;
    }
    get waitForInstancesStatus() {
        return this.getStringAttribute('wait_for_instances_status');
    }
    set waitForInstancesStatus(value) {
        this._waitForInstancesStatus = value;
    }
    resetWaitForInstancesStatus() {
        this._waitForInstancesStatus = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForInstancesStatusInput() {
        return this._waitForInstancesStatus;
    }
    get autoHealingPolicies() {
        return this._autoHealingPolicies;
    }
    putAutoHealingPolicies(value) {
        this._autoHealingPolicies.internalValue = value;
    }
    resetAutoHealingPolicies() {
        this._autoHealingPolicies.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoHealingPoliciesInput() {
        return this._autoHealingPolicies.internalValue;
    }
    get namedPort() {
        return this._namedPort;
    }
    putNamedPort(value) {
        this._namedPort.internalValue = value;
    }
    resetNamedPort() {
        this._namedPort.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namedPortInput() {
        return this._namedPort.internalValue;
    }
    get statefulDisk() {
        return this._statefulDisk;
    }
    putStatefulDisk(value) {
        this._statefulDisk.internalValue = value;
    }
    resetStatefulDisk() {
        this._statefulDisk.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statefulDiskInput() {
        return this._statefulDisk.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get updatePolicy() {
        return this._updatePolicy;
    }
    putUpdatePolicy(value) {
        this._updatePolicy.internalValue = value;
    }
    resetUpdatePolicy() {
        this._updatePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updatePolicyInput() {
        return this._updatePolicy.internalValue;
    }
    get version() {
        return this._version;
    }
    putVersion(value) {
        this._version.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            base_instance_name: cdktf.stringToTerraform(this._baseInstanceName),
            description: cdktf.stringToTerraform(this._description),
            distribution_policy_target_shape: cdktf.stringToTerraform(this._distributionPolicyTargetShape),
            distribution_policy_zones: cdktf.listMapper(cdktf.stringToTerraform, false)(this._distributionPolicyZones),
            id: cdktf.stringToTerraform(this._id),
            list_managed_instances_results: cdktf.stringToTerraform(this._listManagedInstancesResults),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
            target_pools: cdktf.listMapper(cdktf.stringToTerraform, false)(this._targetPools),
            target_size: cdktf.numberToTerraform(this._targetSize),
            wait_for_instances: cdktf.booleanToTerraform(this._waitForInstances),
            wait_for_instances_status: cdktf.stringToTerraform(this._waitForInstancesStatus),
            auto_healing_policies: computeRegionInstanceGroupManagerAutoHealingPoliciesToTerraform(this._autoHealingPolicies.internalValue),
            named_port: cdktf.listMapper(computeRegionInstanceGroupManagerNamedPortToTerraform, true)(this._namedPort.internalValue),
            stateful_disk: cdktf.listMapper(computeRegionInstanceGroupManagerStatefulDiskToTerraform, true)(this._statefulDisk.internalValue),
            timeouts: computeRegionInstanceGroupManagerTimeoutsToTerraform(this._timeouts.internalValue),
            update_policy: computeRegionInstanceGroupManagerUpdatePolicyToTerraform(this._updatePolicy.internalValue),
            version: cdktf.listMapper(computeRegionInstanceGroupManagerVersionToTerraform, true)(this._version.internalValue),
        };
    }
}
exports.ComputeRegionInstanceGroupManager = ComputeRegionInstanceGroupManager;
_u = JSII_RTTI_SYMBOL_1;
ComputeRegionInstanceGroupManager[_u] = { fqn: "@cdktf/provider-google.computeRegionInstanceGroupManager.ComputeRegionInstanceGroupManager", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
ComputeRegionInstanceGroupManager.tfResourceType = "google_compute_region_instance_group_manager";
