// @cdktf/provider-google/lib/data-google-storage-bucket/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleStorageBucket = exports.DataGoogleStorageBucketWebsiteList = exports.DataGoogleStorageBucketWebsiteOutputReference = exports.dataGoogleStorageBucketWebsiteToTerraform = exports.DataGoogleStorageBucketVersioningList = exports.DataGoogleStorageBucketVersioningOutputReference = exports.dataGoogleStorageBucketVersioningToTerraform = exports.DataGoogleStorageBucketRetentionPolicyList = exports.DataGoogleStorageBucketRetentionPolicyOutputReference = exports.dataGoogleStorageBucketRetentionPolicyToTerraform = exports.DataGoogleStorageBucketLoggingList = exports.DataGoogleStorageBucketLoggingOutputReference = exports.dataGoogleStorageBucketLoggingToTerraform = exports.DataGoogleStorageBucketLifecycleRuleList = exports.DataGoogleStorageBucketLifecycleRuleOutputReference = exports.dataGoogleStorageBucketLifecycleRuleToTerraform = exports.DataGoogleStorageBucketLifecycleRuleConditionList = exports.DataGoogleStorageBucketLifecycleRuleConditionOutputReference = exports.dataGoogleStorageBucketLifecycleRuleConditionToTerraform = exports.DataGoogleStorageBucketLifecycleRuleActionList = exports.DataGoogleStorageBucketLifecycleRuleActionOutputReference = exports.dataGoogleStorageBucketLifecycleRuleActionToTerraform = exports.DataGoogleStorageBucketEncryptionList = exports.DataGoogleStorageBucketEncryptionOutputReference = exports.dataGoogleStorageBucketEncryptionToTerraform = exports.DataGoogleStorageBucketCustomPlacementConfigList = exports.DataGoogleStorageBucketCustomPlacementConfigOutputReference = exports.dataGoogleStorageBucketCustomPlacementConfigToTerraform = exports.DataGoogleStorageBucketCorsList = exports.DataGoogleStorageBucketCorsOutputReference = exports.dataGoogleStorageBucketCorsToTerraform = exports.DataGoogleStorageBucketAutoclassList = exports.DataGoogleStorageBucketAutoclassOutputReference = exports.dataGoogleStorageBucketAutoclassToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleStorageBucketAutoclassToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketAutoclassToTerraform = dataGoogleStorageBucketAutoclassToTerraform;
class DataGoogleStorageBucketAutoclassOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataGoogleStorageBucketAutoclassOutputReference = DataGoogleStorageBucketAutoclassOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketAutoclassOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketAutoclassOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketAutoclassList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketAutoclassOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketAutoclassList = DataGoogleStorageBucketAutoclassList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketAutoclassList[_b] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketAutoclassList", version: "4.0.3" };
function dataGoogleStorageBucketCorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketCorsToTerraform = dataGoogleStorageBucketCorsToTerraform;
class DataGoogleStorageBucketCorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_age_seconds - computed: true, optional: false, required: false
    get maxAgeSeconds() {
        return this.getNumberAttribute('max_age_seconds');
    }
    // method - computed: true, optional: false, required: false
    get method() {
        return this.getListAttribute('method');
    }
    // origin - computed: true, optional: false, required: false
    get origin() {
        return this.getListAttribute('origin');
    }
    // response_header - computed: true, optional: false, required: false
    get responseHeader() {
        return this.getListAttribute('response_header');
    }
}
exports.DataGoogleStorageBucketCorsOutputReference = DataGoogleStorageBucketCorsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketCorsOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketCorsOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketCorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketCorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketCorsList = DataGoogleStorageBucketCorsList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketCorsList[_d] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketCorsList", version: "4.0.3" };
function dataGoogleStorageBucketCustomPlacementConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketCustomPlacementConfigToTerraform = dataGoogleStorageBucketCustomPlacementConfigToTerraform;
class DataGoogleStorageBucketCustomPlacementConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data_locations - computed: true, optional: false, required: false
    get dataLocations() {
        return cdktf.Fn.tolist(this.getListAttribute('data_locations'));
    }
}
exports.DataGoogleStorageBucketCustomPlacementConfigOutputReference = DataGoogleStorageBucketCustomPlacementConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketCustomPlacementConfigOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketCustomPlacementConfigOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketCustomPlacementConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketCustomPlacementConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketCustomPlacementConfigList = DataGoogleStorageBucketCustomPlacementConfigList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketCustomPlacementConfigList[_f] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketCustomPlacementConfigList", version: "4.0.3" };
function dataGoogleStorageBucketEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketEncryptionToTerraform = dataGoogleStorageBucketEncryptionToTerraform;
class DataGoogleStorageBucketEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // default_kms_key_name - computed: true, optional: false, required: false
    get defaultKmsKeyName() {
        return this.getStringAttribute('default_kms_key_name');
    }
}
exports.DataGoogleStorageBucketEncryptionOutputReference = DataGoogleStorageBucketEncryptionOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketEncryptionOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketEncryptionOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketEncryptionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketEncryptionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketEncryptionList = DataGoogleStorageBucketEncryptionList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketEncryptionList[_h] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketEncryptionList", version: "4.0.3" };
function dataGoogleStorageBucketLifecycleRuleActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketLifecycleRuleActionToTerraform = dataGoogleStorageBucketLifecycleRuleActionToTerraform;
class DataGoogleStorageBucketLifecycleRuleActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // storage_class - computed: true, optional: false, required: false
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleStorageBucketLifecycleRuleActionOutputReference = DataGoogleStorageBucketLifecycleRuleActionOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleActionOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleActionOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketLifecycleRuleActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketLifecycleRuleActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketLifecycleRuleActionList = DataGoogleStorageBucketLifecycleRuleActionList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleActionList[_k] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleActionList", version: "4.0.3" };
function dataGoogleStorageBucketLifecycleRuleConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketLifecycleRuleConditionToTerraform = dataGoogleStorageBucketLifecycleRuleConditionToTerraform;
class DataGoogleStorageBucketLifecycleRuleConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // age - computed: true, optional: false, required: false
    get age() {
        return this.getNumberAttribute('age');
    }
    // created_before - computed: true, optional: false, required: false
    get createdBefore() {
        return this.getStringAttribute('created_before');
    }
    // custom_time_before - computed: true, optional: false, required: false
    get customTimeBefore() {
        return this.getStringAttribute('custom_time_before');
    }
    // days_since_custom_time - computed: true, optional: false, required: false
    get daysSinceCustomTime() {
        return this.getNumberAttribute('days_since_custom_time');
    }
    // days_since_noncurrent_time - computed: true, optional: false, required: false
    get daysSinceNoncurrentTime() {
        return this.getNumberAttribute('days_since_noncurrent_time');
    }
    // matches_prefix - computed: true, optional: false, required: false
    get matchesPrefix() {
        return this.getListAttribute('matches_prefix');
    }
    // matches_storage_class - computed: true, optional: false, required: false
    get matchesStorageClass() {
        return this.getListAttribute('matches_storage_class');
    }
    // matches_suffix - computed: true, optional: false, required: false
    get matchesSuffix() {
        return this.getListAttribute('matches_suffix');
    }
    // noncurrent_time_before - computed: true, optional: false, required: false
    get noncurrentTimeBefore() {
        return this.getStringAttribute('noncurrent_time_before');
    }
    // num_newer_versions - computed: true, optional: false, required: false
    get numNewerVersions() {
        return this.getNumberAttribute('num_newer_versions');
    }
    // with_state - computed: true, optional: false, required: false
    get withState() {
        return this.getStringAttribute('with_state');
    }
}
exports.DataGoogleStorageBucketLifecycleRuleConditionOutputReference = DataGoogleStorageBucketLifecycleRuleConditionOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleConditionOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleConditionOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketLifecycleRuleConditionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketLifecycleRuleConditionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketLifecycleRuleConditionList = DataGoogleStorageBucketLifecycleRuleConditionList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleConditionList[_m] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleConditionList", version: "4.0.3" };
function dataGoogleStorageBucketLifecycleRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketLifecycleRuleToTerraform = dataGoogleStorageBucketLifecycleRuleToTerraform;
class DataGoogleStorageBucketLifecycleRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // action - computed: true, optional: false, required: false
        this._action = new DataGoogleStorageBucketLifecycleRuleActionList(this, "action", true);
        // condition - computed: true, optional: false, required: false
        this._condition = new DataGoogleStorageBucketLifecycleRuleConditionList(this, "condition", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get action() {
        return this._action;
    }
    get condition() {
        return this._condition;
    }
}
exports.DataGoogleStorageBucketLifecycleRuleOutputReference = DataGoogleStorageBucketLifecycleRuleOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketLifecycleRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketLifecycleRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketLifecycleRuleList = DataGoogleStorageBucketLifecycleRuleList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleList[_p] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleList", version: "4.0.3" };
function dataGoogleStorageBucketLoggingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketLoggingToTerraform = dataGoogleStorageBucketLoggingToTerraform;
class DataGoogleStorageBucketLoggingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // log_bucket - computed: true, optional: false, required: false
    get logBucket() {
        return this.getStringAttribute('log_bucket');
    }
    // log_object_prefix - computed: true, optional: false, required: false
    get logObjectPrefix() {
        return this.getStringAttribute('log_object_prefix');
    }
}
exports.DataGoogleStorageBucketLoggingOutputReference = DataGoogleStorageBucketLoggingOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLoggingOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLoggingOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketLoggingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketLoggingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketLoggingList = DataGoogleStorageBucketLoggingList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLoggingList[_r] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLoggingList", version: "4.0.3" };
function dataGoogleStorageBucketRetentionPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketRetentionPolicyToTerraform = dataGoogleStorageBucketRetentionPolicyToTerraform;
class DataGoogleStorageBucketRetentionPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_locked - computed: true, optional: false, required: false
    get isLocked() {
        return this.getBooleanAttribute('is_locked');
    }
    // retention_period - computed: true, optional: false, required: false
    get retentionPeriod() {
        return this.getNumberAttribute('retention_period');
    }
}
exports.DataGoogleStorageBucketRetentionPolicyOutputReference = DataGoogleStorageBucketRetentionPolicyOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketRetentionPolicyOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketRetentionPolicyOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketRetentionPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketRetentionPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketRetentionPolicyList = DataGoogleStorageBucketRetentionPolicyList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketRetentionPolicyList[_t] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketRetentionPolicyList", version: "4.0.3" };
function dataGoogleStorageBucketVersioningToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketVersioningToTerraform = dataGoogleStorageBucketVersioningToTerraform;
class DataGoogleStorageBucketVersioningOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataGoogleStorageBucketVersioningOutputReference = DataGoogleStorageBucketVersioningOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketVersioningOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketVersioningOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketVersioningList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketVersioningOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketVersioningList = DataGoogleStorageBucketVersioningList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketVersioningList[_v] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketVersioningList", version: "4.0.3" };
function dataGoogleStorageBucketWebsiteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketWebsiteToTerraform = dataGoogleStorageBucketWebsiteToTerraform;
class DataGoogleStorageBucketWebsiteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // main_page_suffix - computed: true, optional: false, required: false
    get mainPageSuffix() {
        return this.getStringAttribute('main_page_suffix');
    }
    // not_found_page - computed: true, optional: false, required: false
    get notFoundPage() {
        return this.getStringAttribute('not_found_page');
    }
}
exports.DataGoogleStorageBucketWebsiteOutputReference = DataGoogleStorageBucketWebsiteOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketWebsiteOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketWebsiteOutputReference", version: "4.0.3" };
class DataGoogleStorageBucketWebsiteList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketWebsiteOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketWebsiteList = DataGoogleStorageBucketWebsiteList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketWebsiteList[_x] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketWebsiteList", version: "4.0.3" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/google/d/storage_bucket google_storage_bucket}
*/
class DataGoogleStorageBucket extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/google/d/storage_bucket google_storage_bucket} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleStorageBucketConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_storage_bucket',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '4.46.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // autoclass - computed: true, optional: false, required: false
        this._autoclass = new DataGoogleStorageBucketAutoclassList(this, "autoclass", false);
        // cors - computed: true, optional: false, required: false
        this._cors = new DataGoogleStorageBucketCorsList(this, "cors", false);
        // custom_placement_config - computed: true, optional: false, required: false
        this._customPlacementConfig = new DataGoogleStorageBucketCustomPlacementConfigList(this, "custom_placement_config", false);
        // encryption - computed: true, optional: false, required: false
        this._encryption = new DataGoogleStorageBucketEncryptionList(this, "encryption", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // lifecycle_rule - computed: true, optional: false, required: false
        this._lifecycleRule = new DataGoogleStorageBucketLifecycleRuleList(this, "lifecycle_rule", false);
        // logging - computed: true, optional: false, required: false
        this._logging = new DataGoogleStorageBucketLoggingList(this, "logging", false);
        // retention_policy - computed: true, optional: false, required: false
        this._retentionPolicy = new DataGoogleStorageBucketRetentionPolicyList(this, "retention_policy", false);
        // versioning - computed: true, optional: false, required: false
        this._versioning = new DataGoogleStorageBucketVersioningList(this, "versioning", false);
        // website - computed: true, optional: false, required: false
        this._website = new DataGoogleStorageBucketWebsiteList(this, "website", false);
        this._id = config.id;
        this._name = config.name;
    }
    get autoclass() {
        return this._autoclass;
    }
    get cors() {
        return this._cors;
    }
    get customPlacementConfig() {
        return this._customPlacementConfig;
    }
    // default_event_based_hold - computed: true, optional: false, required: false
    get defaultEventBasedHold() {
        return this.getBooleanAttribute('default_event_based_hold');
    }
    get encryption() {
        return this._encryption;
    }
    // force_destroy - computed: true, optional: false, required: false
    get forceDestroy() {
        return this.getBooleanAttribute('force_destroy');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get lifecycleRule() {
        return this._lifecycleRule;
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    get logging() {
        return this._logging;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // project - computed: true, optional: false, required: false
    get project() {
        return this.getStringAttribute('project');
    }
    // public_access_prevention - computed: true, optional: false, required: false
    get publicAccessPrevention() {
        return this.getStringAttribute('public_access_prevention');
    }
    // requester_pays - computed: true, optional: false, required: false
    get requesterPays() {
        return this.getBooleanAttribute('requester_pays');
    }
    get retentionPolicy() {
        return this._retentionPolicy;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    // storage_class - computed: true, optional: false, required: false
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    // uniform_bucket_level_access - computed: true, optional: false, required: false
    get uniformBucketLevelAccess() {
        return this.getBooleanAttribute('uniform_bucket_level_access');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    get versioning() {
        return this._versioning;
    }
    get website() {
        return this._website;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataGoogleStorageBucket = DataGoogleStorageBucket;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucket[_y] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucket", version: "4.0.3" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleStorageBucket.tfResourceType = "google_storage_bucket";
